// @progress/kendo-angular-dialog/dist/npm/window/actions/window-minimize-action.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var drag_resize_service_1 = require("./../drag-resize.service");
var kendo_angular_buttons_1 = require("@progress/kendo-angular-buttons");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
var WindowMinimizeActionDirective = /** @class */ (function (_super) {
    tslib_1.__extends(WindowMinimizeActionDirective, _super);
    function WindowMinimizeActionDirective(el, renderer, _service, localization, ngZone) {
        var _this = _super.call(this, el, renderer, null, localization, ngZone) || this;
        _this.buttonType = 'button';
        _this.window = _service;
        _this.fillMode = 'flat';
        _this.icon = 'window-minimize';
        return _this;
    }
    /**
     * @hidden
     */
    WindowMinimizeActionDirective.prototype.onClick = function () {
        if (!this.isDisabled) {
            this.window.minimizeAction();
        }
    };
    Object.defineProperty(WindowMinimizeActionDirective.prototype, "visible", {
        get: function () {
            return this.window.options.state === 'default' ? 'inline-flex' : 'none';
        },
        enumerable: true,
        configurable: true
    });
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], WindowMinimizeActionDirective.prototype, "window", void 0);
    tslib_1.__decorate([
        core_1.HostBinding('attr.type'),
        tslib_1.__metadata("design:type", String)
    ], WindowMinimizeActionDirective.prototype, "buttonType", void 0);
    tslib_1.__decorate([
        core_1.HostListener('click'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], WindowMinimizeActionDirective.prototype, "onClick", null);
    tslib_1.__decorate([
        core_1.HostBinding('style.display'),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [])
    ], WindowMinimizeActionDirective.prototype, "visible", null);
    WindowMinimizeActionDirective = tslib_1.__decorate([
        core_1.Directive({
            exportAs: 'kendoWindowMinimizeAction',
            providers: [
                kendo_angular_l10n_1.LocalizationService,
                {
                    provide: kendo_angular_l10n_1.L10N_PREFIX,
                    useValue: 'kendo.button'
                }
            ],
            selector: 'button[kendoWindowMinimizeAction]' // tslint:disable-line
        }),
        tslib_1.__param(2, core_1.Optional()),
        tslib_1.__metadata("design:paramtypes", [core_1.ElementRef,
            core_1.Renderer2,
            drag_resize_service_1.DragResizeService,
            kendo_angular_l10n_1.LocalizationService,
            core_1.NgZone])
    ], WindowMinimizeActionDirective);
    return WindowMinimizeActionDirective;
}(kendo_angular_buttons_1.Button));
exports.WindowMinimizeActionDirective = WindowMinimizeActionDirective;
