// @progress/kendo-angular-pager/dist/es/pager/pager.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { L10N_PREFIX } from '@progress/kendo-angular-l10n';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NumericTextBoxModule } from '@progress/kendo-angular-inputs';
import { DropDownListModule } from '@progress/kendo-angular-dropdowns';
import { PagerComponent } from './pager.component';
import { PagerPrevButtonsComponent } from './pager-prev-buttons.component';
import { PagerNextButtonsComponent } from './pager-next-buttons.component';
import { PagerNumericButtonsComponent } from './pager-numeric-buttons.component';
import { PagerInputComponent } from './pager-input.component';
import { PagerInfoComponent } from './pager-info.component';
import { PagerPageSizesComponent } from './pager-page-sizes.component';
import { PagerTemplateDirective } from './pager-template.directive';
import { EventsModule } from '@progress/kendo-angular-common';
import { LocalizedMessagesDirective } from './localization/localized-messages.directive';
import { CustomMessagesComponent } from './localization/custom-messages.component';
import { ResizeSensorModule } from '@progress/kendo-angular-common';
var importedModules = [
    CommonModule,
    NumericTextBoxModule,
    DropDownListModule,
    EventsModule,
    ResizeSensorModule
];
var INTERNAL_COMPONENTS = [
    PagerComponent,
    PagerPrevButtonsComponent,
    PagerNextButtonsComponent,
    PagerNumericButtonsComponent,
    PagerInputComponent,
    PagerInfoComponent,
    PagerPageSizesComponent,
    PagerTemplateDirective,
    LocalizedMessagesDirective,
    CustomMessagesComponent
];
var providers = [
    {
        provide: L10N_PREFIX,
        useValue: 'kendo.pager'
    }
];
/**
 * @hidden
 */
var PagerModule = /** @class */ (function () {
    function PagerModule() {
    }
    PagerModule.exports = function () {
        return [
            PagerComponent,
            PagerPrevButtonsComponent,
            PagerNextButtonsComponent,
            PagerNumericButtonsComponent,
            PagerInputComponent,
            PagerInfoComponent,
            PagerPageSizesComponent,
            PagerTemplateDirective
        ];
    };
    PagerModule = tslib_1.__decorate([
        NgModule({
            declarations: [INTERNAL_COMPONENTS],
            exports: [INTERNAL_COMPONENTS],
            imports: importedModules.slice(),
            providers: [providers]
        })
    ], PagerModule);
    return PagerModule;
}());
export { PagerModule };
