// @safe-fleet/sf-menu/esm2015/lib/components/menu-toggle-btn/menu-toggle-btn.component.js
import { Component } from '@angular/core';
import { SfMenuService } from '../../sf-menu.service';
export class MenuToggleBtnComponent {
    constructor(svc) {
        this.svc = svc;
        this.collapsed = false;
    }
    ngOnInit() {
        this.color = this.svc.colors.gray700;
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            this.menuConfig = cfg;
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
        });
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
    toggleMenu() {
        this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
    }
}
MenuToggleBtnComponent.decorators = [
    { type: Component, args: [{
                selector: 'sf-menu-toggle-btn',
                template: "<button class=\"sf-menu-toggle-btn\" [class.collapsed]=\"collapsed\" (click)=\"toggleMenu()\" [style.color]=\"color\">\n  <div>\n    <em class=\"fas fa-chevron-left\"></em>\n    <em class=\"fas fa-chevron-left\"></em>\n  </div>\n</button>\n",
                styles: [".sf-menu-toggle-btn{display:none;position:absolute;z-index:10000;top:8px;right:8px;background-color:transparent;border:none;width:1.8rem;height:1.8rem;transition:transform .15s ease-in-out;cursor:pointer}@media screen and (min-width:768px){.sf-menu-toggle-btn{display:block}}.sf-menu-toggle-btn div{display:flex;justify-content:center;align-items:center}.sf-menu-toggle-btn .fas{font-size:16px;opacity:.64}.sf-menu-toggle-btn .fas:nth-child(2){margin-left:-4px}.sf-menu-toggle-btn.collapsed .fas{transform:rotate(180deg)}.sf-menu-toggle-btn:hover .fas{opacity:1}"]
            },] }
];
MenuToggleBtnComponent.ctorParameters = () => [
    { type: SfMenuService }
];
