// @progress/kendo-angular-menu/dist/npm/rendering/arrow.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var items_service_1 = require("../services/items.service");
/**
 * Represents a directive that can be used in the [`linkTemplate`]({% slug api_menu_itemlinktemplatedirective %})
 * of the items to render the default expand arrow.
 */
var ExpandArrowDirective = /** @class */ (function () {
    function ExpandArrowDirective(itemsService) {
        this.itemsService = itemsService;
        this.hostClasses = true;
        this.role = 'presentation';
    }
    Object.defineProperty(ExpandArrowDirective.prototype, "arrowDown", {
        get: function () {
            return !this.item.horizontal;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpandArrowDirective.prototype, "arrowRight", {
        get: function () {
            return this.item.horizontal && !this.item.rtl;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpandArrowDirective.prototype, "arrowLeft", {
        get: function () {
            return this.item.horizontal && this.item.rtl;
        },
        enumerable: true,
        configurable: true
    });
    ExpandArrowDirective.prototype.ngOnInit = function () {
        if (core_1.isDevMode() && !this.index) {
            throw new Error('The kendoMenuExpandArrow directive requires the item index to be set.');
        }
        this.item = this.itemsService.get(this.index) || {};
    };
    tslib_1.__decorate([
        core_1.Input('kendoMenuExpandArrow'),
        tslib_1.__metadata("design:type", String)
    ], ExpandArrowDirective.prototype, "index", void 0);
    tslib_1.__decorate([
        core_1.HostBinding('class.k-icon'),
        core_1.HostBinding('class.k-menu-expand-arrow'),
        tslib_1.__metadata("design:type", Boolean)
    ], ExpandArrowDirective.prototype, "hostClasses", void 0);
    tslib_1.__decorate([
        core_1.HostBinding('attr.role'),
        tslib_1.__metadata("design:type", String)
    ], ExpandArrowDirective.prototype, "role", void 0);
    tslib_1.__decorate([
        core_1.HostBinding('class.k-i-arrow-60-down'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [])
    ], ExpandArrowDirective.prototype, "arrowDown", null);
    tslib_1.__decorate([
        core_1.HostBinding('class.k-i-arrow-60-right'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [])
    ], ExpandArrowDirective.prototype, "arrowRight", null);
    tslib_1.__decorate([
        core_1.HostBinding('class.k-i-arrow-60-left'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [])
    ], ExpandArrowDirective.prototype, "arrowLeft", null);
    ExpandArrowDirective = tslib_1.__decorate([
        core_1.Directive({
            selector: '[kendoMenuExpandArrow]'
        }),
        tslib_1.__metadata("design:paramtypes", [items_service_1.ItemsService])
    ], ExpandArrowDirective);
    return ExpandArrowDirective;
}());
exports.ExpandArrowDirective = ExpandArrowDirective;
