// @progress/kendo-angular-layout/dist/npm/tilelayout/constants.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @hidden
 */
exports.RESIZE_HANDLE_UNIDIRECTIONAL_SIZE = 9;
/**
 * @hidden
 */
exports.RESIZE_HANDLE_BIDIRECTIONAL_SIZE = 25;
/**
 * @hidden
 */
exports.RESIZE_HANDLE_UNIDIRECTIONAL_OVERLAP = 5;
/**
 * @hidden
 */
exports.RESIZE_HANDLE_BIDIRECTIONAL_OVERLAP = 18;
/**
 * @hidden
 */
exports.DRAGGED_ZINDEX = 10;
/**
 * @hidden
 */
exports.RESIZE_DIRECTIONS = ['ew', 'ns', 'nwse'];
/**
 * @hidden
 */
exports.RTL_RESIZE_DIRECTIONS = ['ew', 'ns', 'nesw'];
/**
 * @hidden
 */
exports.RESIZE_HANDLE_DIMENSIONS = {
    // numeric values represent pixels
    ew: {
        width: exports.RESIZE_HANDLE_UNIDIRECTIONAL_SIZE,
        overlapX: exports.RESIZE_HANDLE_UNIDIRECTIONAL_OVERLAP
    },
    ns: {
        height: exports.RESIZE_HANDLE_UNIDIRECTIONAL_SIZE,
        overlapY: exports.RESIZE_HANDLE_UNIDIRECTIONAL_OVERLAP
    },
    nwse: {
        width: exports.RESIZE_HANDLE_BIDIRECTIONAL_SIZE,
        height: exports.RESIZE_HANDLE_BIDIRECTIONAL_SIZE,
        overlapX: exports.RESIZE_HANDLE_BIDIRECTIONAL_OVERLAP,
        overlapY: exports.RESIZE_HANDLE_BIDIRECTIONAL_OVERLAP
    },
    nesw: {
        width: exports.RESIZE_HANDLE_BIDIRECTIONAL_SIZE,
        height: exports.RESIZE_HANDLE_BIDIRECTIONAL_SIZE,
        overlapX: exports.RESIZE_HANDLE_BIDIRECTIONAL_OVERLAP,
        overlapY: exports.RESIZE_HANDLE_BIDIRECTIONAL_OVERLAP
    }
};
/**
 * @hidden
 */
exports.OVERLAP_THRESHOLD = 0.3;
/**
 * @hidden
 */
exports.REVERSE_OVERLAP_THRESHOLD = -0.7;
/**
 * @hidden
 */
exports.HINT_BORDERS_HEIGHT = 2;
