// @progress/kendo-angular-buttons/dist/es2015/floatingactionbutton/dial-list.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, HostBinding, Input, TemplateRef } from "@angular/core";
import { Subscription } from 'rxjs';
import { FocusService } from '../focusable/focus.service';
/**
 * @hidden
 */
let DialListComponent = class DialListComponent {
    constructor(focusService, cdr) {
        this.focusService = focusService;
        this.cdr = cdr;
        this.hostClass = true;
        this.subscriptions = new Subscription();
        this.subscriptions.add(this.focusService.onFocus.subscribe(() => this.cdr.detectChanges()));
    }
    get bottomClass() {
        return this.align.vertical === 'top' || this.align.vertical === 'middle';
    }
    get topClass() {
        return this.align.vertical === 'bottom';
    }
    isFocused(index) {
        return this.focusService.isFocused(index);
    }
    ngOnDestroy() {
        this.subscriptions.unsubscribe();
    }
};
tslib_1.__decorate([
    HostBinding('class.k-fab-items'),
    tslib_1.__metadata("design:type", Boolean)
], DialListComponent.prototype, "hostClass", void 0);
tslib_1.__decorate([
    HostBinding('class.k-fab-items-bottom'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], DialListComponent.prototype, "bottomClass", null);
tslib_1.__decorate([
    HostBinding('class.k-fab-items-top'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], DialListComponent.prototype, "topClass", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Array)
], DialListComponent.prototype, "dialItems", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", TemplateRef)
], DialListComponent.prototype, "dialItemTemplate", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], DialListComponent.prototype, "align", void 0);
DialListComponent = tslib_1.__decorate([
    Component({
        selector: '[kendoDialList]',
        template: `
        <ng-container *ngFor='let item of dialItems; let idx = index'>
            <li
                kendoButtonFocusable
                kendoDialItem
                [item]="dialItems[idx]"
                [index]="idx"
                [dialItemTemplate]="dialItemTemplate"
                [isFocused]="isFocused(idx)"
                [ngClass]='item.cssClass'
                [ngStyle]='item.cssStyle'
                [align]="align"
            >
            </li>
        </ng-container>
    `
    }),
    tslib_1.__metadata("design:paramtypes", [FocusService, ChangeDetectorRef])
], DialListComponent);
export { DialListComponent };
