// @safe-fleet/sf-menu/esm2015/lib/components/menu/sf-menu.component.js
import { Component, Input, ViewChild } from '@angular/core';
import { trigger, transition, style, animate } from '@angular/animations';
import { MenuConfig } from '../../models/menu-config';
import { SfMenuService } from '../../sf-menu.service';
import { TooltipDirective } from '@progress/kendo-angular-tooltip';
export class SfMenuComponent {
    constructor(svc) {
        this.svc = svc;
        this.collapsed = false;
    }
    set config(value) {
        this.svc.setMenuConfig(value);
    }
    ngOnInit() {
        const clr = this.svc.colors;
        this.gradient = `linear-gradient(180deg, ${clr.gray200} 0%, ${clr.gray100} 100%)`;
        this.color = clr.gray900;
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            this.menuConfig = cfg;
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
            if (this.tooltipDir) {
                this.tooltipDir.hide();
            }
        });
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
    toggleMenu() {
        this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
    }
}
SfMenuComponent.decorators = [
    { type: Component, args: [{
                selector: 'sf-menu',
                template: "<div class=\"sf-menu-container\" *ngIf=\"!menuConfig.hideMenu\">\n  <sf-top-menu [ngStyle]=\"{ 'border-bottom-color': menuConfig?.borderColor }\"></sf-top-menu>\n  <div\n    class=\"sf-menu-content\"\n    [class.collapsed]=\"collapsed\"\n    [ngStyle]=\"{ 'border-top-color': menuConfig?.borderColor }\"\n    [style.background]=\"gradient\"\n    kendoTooltip\n    showOn=\"{{ menuConfig.tooltips || collapsed ? 'hover' : 'none' }}\"\n    position=\"bottom\"\n  >\n    <sf-customer-logo [class.collapsed]=\"collapsed\" *ngIf=\"menuConfig.logoUrl || menuConfig.logoClass\"></sf-customer-logo>\n    <sf-menu-products [class.collapsed]=\"collapsed\"></sf-menu-products>\n    <div class=\"fm-menu-scrollable-content\">\n      <div class=\"scroll\">\n        <sf-menu-shortcuts *ngIf=\"menuConfig?.shortcuts && menuConfig.shortcuts.length > 0\"></sf-menu-shortcuts>\n        <sf-menu-routes *ngIf=\"menuConfig?.routes\"></sf-menu-routes>\n      </div>\n    </div>\n    <sf-menu-user></sf-menu-user>\n    <sf-menu-footer-logo></sf-menu-footer-logo>\n  </div>\n  <sf-menu-toggle-btn></sf-menu-toggle-btn>\n  <div class=\"sf-menu-backdrop\" *ngIf=\"!collapsed\" (click)=\"toggleMenu()\" [@appear]></div>\n</div>\n<div class=\"application-content\">\n  <ng-content></ng-content>\n</div>\n",
                animations: [
                    trigger('appear', [
                        transition(':enter', [style({ opacity: 0 }), animate(180, style({ opacity: 1 }))]),
                        transition(':leave', [style({ opacity: 1 }), animate(180, style({ opacity: 0 }))])
                    ])
                ],
                styles: [":host{width:100%;height:100%;display:grid;grid-template-areas:\"menu\" \"content\";grid-template-rows:auto 1fr;grid-template-columns:1fr}@media screen and (min-width:768px){:host{grid-template-areas:\"menu content\";grid-template-rows:1fr;grid-template-columns:auto 1fr}}.sf-menu-container{grid-area:menu;position:relative}.sf-menu-content{height:auto;max-height:calc(100% - 70px);width:250px;border-bottom:none;transition:transform .12s ease-out,width .12s ease-out;overflow-y:auto;position:fixed;z-index:10000}.sf-menu-content::-webkit-scrollbar{width:6px;background-color:transparent}.sf-menu-content::-webkit-scrollbar-thumb{border-radius:10px;background-color:var(--clr-gray-400)}.sf-menu-content.collapsed{transform:translateX(-100%)}@media screen and (min-width:768px){.sf-menu-content{display:flex;flex-direction:column;justify-content:space-between;border-top:5px solid #00a2fe;position:relative;height:100%;max-height:100%;width:250px;overflow:hidden}.sf-menu-content.collapsed{transform:translateX(0);width:75px}}@media screen and (max-width:380px){.sf-menu-content{max-height:calc(100% - 48px)}}.sf-menu-backdrop{position:fixed;width:100%;height:100%;background-color:rgba(0,0,0,.71);z-index:9999}.application-content{grid-area:content;overflow:auto;height:100%;position:relative}.sf-menu-logo-bg{background:50%/contain no-repeat}.fm-menu-scrollable-content{display:flex;flex-direction:column;width:100%;height:auto;align-items:flex-start;justify-content:flex-start;overflow:hidden;position:relative;flex:1 0 120px;padding:0}.fm-menu-scrollable-content:before{content:\" \";width:100%;height:1px;background-color:#fff;opacity:.15}.fm-menu-scrollable-content ul{list-style:none;padding-left:0;width:100%}.fm-menu-scrollable-content :before{display:none}.fm-menu-scrollable-content .scroll{overflow-y:auto;overflow-x:hidden;width:100%}.fm-menu-scrollable-content .scroll::-webkit-scrollbar{width:6px;background-color:transparent}.fm-menu-scrollable-content .scroll::-webkit-scrollbar-thumb{border-radius:10px;background-color:#5b616a}@media screen and (min-width:768px){.fm-menu-scrollable-content .scroll{max-height:65vh}}"]
            },] }
];
SfMenuComponent.ctorParameters = () => [
    { type: SfMenuService }
];
SfMenuComponent.propDecorators = {
    borderColor: [{ type: Input }],
    config: [{ type: Input, args: ['config',] }],
    tooltipDir: [{ type: ViewChild, args: [TooltipDirective,] }]
};
