// angular-auth-oidc-client/esm2020/lib/config/provider/config.provider.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class ConfigurationProvider {
    constructor() {
        this.configsInternal = {};
    }
    hasAtLeastOneConfig() {
        return Object.keys(this.configsInternal).length > 0;
    }
    hasManyConfigs() {
        return Object.keys(this.configsInternal).length > 1;
    }
    setConfig(readyConfig) {
        const { configId } = readyConfig;
        this.configsInternal[configId] = readyConfig;
    }
    getOpenIDConfiguration(configId) {
        if (!!configId) {
            return this.configsInternal[configId] || null;
        }
        const [, value] = Object.entries(this.configsInternal)[0] || [[null, null]];
        return value || null;
    }
    getAllConfigurations() {
        return Object.values(this.configsInternal);
    }
}
ConfigurationProvider.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: ConfigurationProvider, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
ConfigurationProvider.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: ConfigurationProvider });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: ConfigurationProvider, decorators: [{
            type: Injectable
        }] });
