// @progress/kendo-angular-pager/dist/npm/main.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var pagesizechange_event_1 = require("./pager/pagesizechange-event");
exports.PageSizeChangeEvent = pagesizechange_event_1.PageSizeChangeEvent;
var custom_messages_component_1 = require("./pager/localization/custom-messages.component");
exports.CustomMessagesComponent = custom_messages_component_1.CustomMessagesComponent;
var pager_prev_buttons_component_1 = require("./pager/pager-prev-buttons.component");
exports.PagerPrevButtonsComponent = pager_prev_buttons_component_1.PagerPrevButtonsComponent;
var pager_page_sizes_component_1 = require("./pager/pager-page-sizes.component");
exports.PagerPageSizesComponent = pager_page_sizes_component_1.PagerPageSizesComponent;
var pager_template_directive_1 = require("./pager/pager-template.directive");
exports.PagerTemplateDirective = pager_template_directive_1.PagerTemplateDirective;
var pager_numeric_buttons_component_1 = require("./pager/pager-numeric-buttons.component");
exports.PagerNumericButtonsComponent = pager_numeric_buttons_component_1.PagerNumericButtonsComponent;
var pager_next_buttons_component_1 = require("./pager/pager-next-buttons.component");
exports.PagerNextButtonsComponent = pager_next_buttons_component_1.PagerNextButtonsComponent;
var pager_input_component_1 = require("./pager/pager-input.component");
exports.PagerInputComponent = pager_input_component_1.PagerInputComponent;
var pager_info_component_1 = require("./pager/pager-info.component");
exports.PagerInfoComponent = pager_info_component_1.PagerInfoComponent;
var pager_component_1 = require("./pager/pager.component");
exports.PagerComponent = pager_component_1.PagerComponent;
var pager_module_1 = require("./pager/pager.module");
exports.PagerModule = pager_module_1.PagerModule;
