// @progress/kendo-angular-treeview/dist/es/drag-and-drop/editing-services/drag-and-drop-asset.service.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { isPresent } from '../../utils';
/**
 * @hidden
 */
var DragAndDropAssetService = /** @class */ (function () {
    function DragAndDropAssetService() {
    }
    Object.defineProperty(DragAndDropAssetService.prototype, "componentRef", {
        get: function () {
            if (!isPresent(this._componentRef)) {
                throw new Error('The `initalize` method must be called before calling other service methods.');
            }
            return this._componentRef;
        },
        set: function (componentRef) {
            this._componentRef = componentRef;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DragAndDropAssetService.prototype, "element", {
        get: function () {
            return this.componentRef.location.nativeElement;
        },
        enumerable: true,
        configurable: true
    });
    DragAndDropAssetService.prototype.ngOnDestroy = function () {
        if (!isPresent(this._componentRef)) {
            return;
        }
        this.element.parentElement.removeChild(this.element);
        this.componentRef.destroy();
        this.componentRef = null;
    };
    DragAndDropAssetService.prototype.show = function () {
        this.element.style.display = '';
    };
    DragAndDropAssetService.prototype.hide = function () {
        this.element.style.display = 'none';
    };
    DragAndDropAssetService.prototype.move = function (left, top, offset) {
        if (offset === void 0) { offset = 0; }
        this.element.style.left = left + offset + "px";
        this.element.style.top = top + offset + "px";
    };
    return DragAndDropAssetService;
}());
export { DragAndDropAssetService };
