// @progress/kendo-angular-dateinputs/dist/npm/timepicker/services/dom.service.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var kendo_angular_common_1 = require("@progress/kendo-angular-common");
var util_1 = require("../../util");
var div = util_1.domContainerFactory('div');
var ul = util_1.domContainerFactory('ul');
var li = util_1.domContainerFactory('li');
var span = util_1.domContainerFactory('span');
var listTitle = function () { return span('hour', 'k-title k-timeselector-title'); };
var ɵ0 = listTitle;
exports.ɵ0 = ɵ0;
var listItem = function () { return li('<span>02</span>', 'k-item'); };
var ɵ1 = listItem;
exports.ɵ1 = ɵ1;
var list = function () { return ul([listItem()], 'k-reset'); };
var ɵ2 = list;
exports.ɵ2 = ɵ2;
var scrollable = function () { return (div([list()], 'k-time-container k-flex k-content k-calendar-content k-scrollable')); };
var ɵ3 = scrollable;
exports.ɵ3 = ɵ3;
var timeListWrapper = function () {
    if (!kendo_angular_common_1.isDocumentAvailable()) {
        return null;
    }
    return div([listTitle(), div([scrollable()], 'k-time-list')], 'k-time-list-wrapper', { left: '-10000px', position: 'absolute' });
};
var ɵ4 = timeListWrapper;
exports.ɵ4 = ɵ4;
var TIMELIST_WRAPPER = timeListWrapper();
/**
 * @hidden
 */
var TimePickerDOMService = /** @class */ (function () {
    function TimePickerDOMService() {
    }
    TimePickerDOMService.prototype.ensureHeights = function () {
        if (this.timeListHeight !== undefined) {
            return;
        }
        this.calculateHeights();
    };
    TimePickerDOMService.prototype.calculateHeights = function (container) {
        if (!kendo_angular_common_1.isDocumentAvailable()) {
            return;
        }
        var listContainer = container && container.querySelector('.k-time-list-container');
        var hostContainer = listContainer || document.body;
        var wrapper = hostContainer.appendChild(TIMELIST_WRAPPER);
        this.timeListHeight = wrapper.querySelector('.k-scrollable').getBoundingClientRect().height;
        this.itemHeight = wrapper.querySelector('li').getBoundingClientRect().height;
        hostContainer.removeChild(wrapper);
    };
    TimePickerDOMService.prototype.isActive = function (element) {
        if (!kendo_angular_common_1.isDocumentAvailable() || !element) {
            return false;
        }
        return (element.nativeElement || element) === document.activeElement;
    };
    TimePickerDOMService = tslib_1.__decorate([
        core_1.Injectable()
    ], TimePickerDOMService);
    return TimePickerDOMService;
}());
exports.TimePickerDOMService = TimePickerDOMService;
