// @progress/kendo-angular-progressbar/dist/npm/common/util.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var constants_1 = require("./constants");
var core_1 = require("@angular/core");
/**
 * @hidden
 */
exports.reverseChunks = function (orientation, reverse) { return (orientation === 'vertical' && !reverse || orientation === 'horizontal' && reverse); };
/**
 * @hidden
 */
exports.formatValue = function (value, min, max, label) {
    var defaultFormattedValue = truncateNumber(value);
    if (typeof label !== 'boolean') {
        if (typeof label.format === 'string') {
            switch (label.format) {
                case 'value':
                    return defaultFormattedValue;
                case 'percent':
                    return Math.floor(calculatePercentage(value, min, max)) + "%";
                default:
                    return defaultFormattedValue;
            }
        }
        else if (typeof label.format === 'function') {
            return label.format(value);
        }
        else {
            return defaultFormattedValue;
        }
    }
    return defaultFormattedValue;
};
/**
 * @hidden
 */
exports.validateRange = function (min, max) {
    if (core_1.isDevMode && min > max) {
        throw new Error(constants_1.MIN_MAX_ERROR_MESSAGE);
    }
};
/**
 * @hidden
 */
exports.adjustValueToRange = function (min, max, value) { return Math.max(Math.min(value, max), min); };
/**
 * @hidden
 */
var calculatePercentage = function (value, min, max) {
    var onePercent = Math.abs((max - min) / 100);
    return Math.abs((value - min) / onePercent);
};
var ɵ0 = calculatePercentage;
exports.ɵ0 = ɵ0;
/**
 * @hidden
 */
var truncateNumber = function (value) {
    var numberParts = value.toString().split('.');
    return numberParts.length === 1 ? "" + numberParts[0] : numberParts[0] + "." + numberParts[1].substr(0, constants_1.LABEL_DECIMALS);
};
var ɵ1 = truncateNumber;
exports.ɵ1 = ɵ1;
/**
 * @hidden
 */
exports.calculateRatio = function (min, max, value) { return Math.max((value - min) / (max - min), constants_1.MIN_RATIO); };
/**
 * @hidden
 */
exports.extractValueFromChanges = function (changes, type, value) {
    return changes[type] && changes[type].currentValue !== undefined ? changes[type].currentValue : value;
};
/**
 * @hidden
 */
exports.runAnimation = function (changes, animation, previousValue, displayValue) { return animation && typeof requestAnimationFrame !== 'undefined' && changes.value && previousValue !== displayValue; };
/**
 * @hidden
 */
exports.stopCurrentAnimation = function (changes) {
    var isAnimationChanged = Boolean(changes.animation);
    var hasAnimation = isAnimationChanged && changes.animation.currentValue;
    return isAnimationChanged && !hasAnimation;
};
