// @progress/kendo-angular-inputs/dist/es/colorpicker/services/flatcolorpicker.service.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Injectable } from "@angular/core";
import { DEFAULT_ACCESSIBLE_PRESET, DEFAULT_PRESET } from "../constants";
import { PALETTEPRESETS } from "../models";
/**
 * @hidden
 */
var FlatColorPickerService = /** @class */ (function () {
    function FlatColorPickerService() {
    }
    FlatColorPickerService.prototype.getPaletteSettings = function (settings, format) {
        var defaultPreset = (format !== 'name') ? DEFAULT_PRESET : DEFAULT_ACCESSIBLE_PRESET;
        var settingsPalette = settings.palette;
        var presetColumns = typeof settingsPalette === 'string' && PALETTEPRESETS[settingsPalette] ?
            PALETTEPRESETS[settingsPalette].columns :
            undefined;
        return {
            palette: settingsPalette || defaultPreset,
            tileSize: settings.tileSize || 24,
            columns: settings.columns || presetColumns || 10
        };
    };
    FlatColorPickerService.prototype.paletteTileLayout = function (tileSize) {
        if (typeof tileSize === 'number') {
            return { width: tileSize, height: tileSize };
        }
        return {
            width: tileSize.width ? tileSize.width : tileSize.height,
            height: tileSize.height ? tileSize.height : tileSize.width
        };
    };
    FlatColorPickerService = tslib_1.__decorate([
        Injectable()
    ], FlatColorPickerService);
    return FlatColorPickerService;
}());
export { FlatColorPickerService };
