// @progress/kendo-angular-menu/dist/npm/data-binding/hierachy-binding.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var menu_base_1 = require("../menu-base");
var binding_directive_base_1 = require("./binding-directive-base");
var utils_1 = require("./utils");
var getField = function (field, level) { return Array.isArray(field) ? field[level] || utils_1.last(field) : field; };
var ɵ0 = getField;
exports.ɵ0 = ɵ0;
/* tslint:disable:no-input-rename */
/**
 * A directive that converts the provided hierarchical data to [MenuItems]({% slug api_menu_menuitem %}) and binds them to the Menu.
 */
var HierarchyBindingDirective = /** @class */ (function (_super) {
    tslib_1.__extends(HierarchyBindingDirective, _super);
    function HierarchyBindingDirective(menu) {
        return _super.call(this, menu) || this;
    }
    HierarchyBindingDirective.prototype.mapItems = function (items, level) {
        var _this = this;
        if (level === void 0) { level = 0; }
        return items.map(function (item) {
            var menuItem = _this.createItem(item, level);
            var children = _this.getChildren(item, level);
            if (children) {
                menuItem.items = _this.mapItems(children, level + 1);
            }
            return menuItem;
        });
    };
    HierarchyBindingDirective.prototype.createItem = function (item, level) {
        var result = { data: item };
        var fields = this.fields;
        for (var idx = 0; idx < fields.length; idx++) {
            var _a = fields[idx], target = _a.target, source = _a.source;
            result[target] = utils_1.getter(getField(source, level))(item);
        }
        return result;
    };
    HierarchyBindingDirective.prototype.getChildren = function (item, level) {
        if (this.childrenField) {
            var field = getField(this.childrenField, level);
            return item[field];
        }
    };
    tslib_1.__decorate([
        core_1.Input("kendoMenuHierarchyBinding"),
        tslib_1.__metadata("design:type", Array)
    ], HierarchyBindingDirective.prototype, "data", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], HierarchyBindingDirective.prototype, "textField", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], HierarchyBindingDirective.prototype, "urlField", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], HierarchyBindingDirective.prototype, "iconField", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], HierarchyBindingDirective.prototype, "disabledField", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], HierarchyBindingDirective.prototype, "cssClassField", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], HierarchyBindingDirective.prototype, "cssStyleField", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], HierarchyBindingDirective.prototype, "separatorField", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], HierarchyBindingDirective.prototype, "childrenField", void 0);
    HierarchyBindingDirective = tslib_1.__decorate([
        core_1.Directive({
            exportAs: 'kendoMenuHierarchyBinding',
            selector: '[kendoMenuHierarchyBinding]'
        }),
        tslib_1.__metadata("design:paramtypes", [menu_base_1.MenuBase])
    ], HierarchyBindingDirective);
    return HierarchyBindingDirective;
}(binding_directive_base_1.BindingDirectiveBase));
exports.HierarchyBindingDirective = HierarchyBindingDirective;
