// @progress/kendo-angular-buttons/dist/es/main.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
export { ButtonDirective, ButtonDirective as Button } from './button/button.directive';
export { ButtonGroupComponent, ButtonGroupComponent as ButtonGroup } from './buttongroup/buttongroup.component';
export { ButtonGroupModule } from './buttongroup/buttongroup.module';
export { ButtonModule } from './button/button.module';
export { ButtonsModule } from './buttons.module';
export { SplitButtonComponent, SplitButtonComponent as SplitButton } from './splitbutton/splitbutton.component';
export { SplitButtonModule } from './splitbutton/splitbutton.module';
export { SplitButtonCustomMessagesComponent } from './splitbutton/localization/custom-messages.component';
export { DropDownButtonComponent, DropDownButtonComponent as DropDownButton } from './dropdownbutton/dropdownbutton.component';
export { DropDownButtonModule } from './dropdownbutton/dropdownbutton.module';
export { ChipComponent } from './chip/chip.component';
export { ChipListComponent } from './chip/chip-list.component';
export { ChipModule } from './chip/chip.module';
// Floating Action Button
export { FloatingActionButtonModule } from './floatingactionbutton/floatingactionbutton.module';
export { FloatingActionButtonComponent } from './floatingactionbutton/floatingactionbutton.component';
export { FloatingActionButtonTemplateDirective } from './floatingactionbutton/templates/fab-template.directive';
export { DialItemTemplateDirective } from './floatingactionbutton/templates/dial-item-template.directive';
