// @progress/kendo-angular-buttons/dist/es/util.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
/* tslint:disable:no-null-keyword */
/* tslint:disable:no-bitwise */
import { detectDesktopBrowser, detectMobileOS } from "@progress/kendo-common";
var resolvedPromise = Promise.resolve(null);
/**
 * @hidden
 */
export var isPresent = function (value) { return value !== null && value !== undefined; };
/**
 * @hidden
 */
export var tick = function (f) { return (resolvedPromise.then(f)); };
/**
 * @hidden
 */
function isDocumentNode(container) {
    return container.nodeType === 9;
}
/**
 * @hidden
 */
export function closest(element, selector) {
    if (element.closest) {
        return element.closest(selector);
    }
    var matches = Element.prototype.matches ?
        function (el, sel) { return el.matches(sel); }
        : function (el, sel) { return el.msMatchesSelector(sel); };
    var node = element;
    while (node && !isDocumentNode(node)) {
        if (matches(node, selector)) {
            return node;
        }
        node = node.parentNode;
    }
}
/**
 * @hidden
 */
export var replaceMessagePlaceholder = function (message, name, value) {
    return message.replace(new RegExp("{\\s*" + name + "\\s*}", 'g'), value);
};
/**
 * @hidden
 */
export var SIZES = {
    small: 'sm',
    medium: 'md',
    large: 'lg'
};
var ROUNDNESS = {
    small: 'sm',
    medium: 'md',
    large: 'lg',
    full: 'full'
};
/**
 * @hidden
 *
 * Returns the styling classes to be added and removed
 */
export var getStylingClasses = function (componentType, stylingOption, previousValue, newValue) {
    switch (stylingOption) {
        case 'size':
            return {
                toRemove: "k-" + componentType + "-" + SIZES[previousValue],
                toAdd: newValue ? "k-" + componentType + "-" + SIZES[newValue] : null
            };
        case 'rounded':
            return {
                toRemove: "k-rounded-" + ROUNDNESS[previousValue],
                toAdd: newValue ? "k-rounded-" + ROUNDNESS[newValue] : null
            };
        case 'fillMode':
        case 'shape':
            return {
                toRemove: "k-" + componentType + "-" + previousValue,
                toAdd: newValue ? "k-" + componentType + "-" + newValue : null
            };
        default:
            break;
    }
};
/**
 * @hidden
 *
 * Returns the themeColor classes to be added and removed
 */
export var getThemeColorClasses = function (componentType, prevFillMode, fillMode, previousValue, newValue) {
    return {
        toRemove: "k-" + componentType + "-" + prevFillMode + "-" + previousValue,
        toAdd: newValue ? "k-" + componentType + "-" + fillMode + "-" + newValue : null
    };
};
/**
 * @hidden
 *
 * Returns true if the used browser is Firefox.
 */
export var isFirefox = function (userAgent) {
    var desktopBrowser = detectDesktopBrowser(userAgent);
    var mobileOS = detectMobileOS(userAgent);
    return (desktopBrowser && desktopBrowser.mozilla) || (mobileOS && mobileOS.browser === 'firefox');
};
