// @progress/kendo-angular-tooltip/dist/npm/popover.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var common_1 = require("@angular/common");
var core_1 = require("@angular/core");
var kendo_angular_popup_1 = require("@progress/kendo-angular-popup");
var anchor_directive_1 = require("./popover/anchor.directive");
var container_directive_1 = require("./popover/container.directive");
var popover_component_1 = require("./popover/popover.component");
var title_template_directive_1 = require("./popover/template-directives/title-template.directive");
var body_template_directive_1 = require("./popover/template-directives/body-template.directive");
var actions_template_directive_1 = require("./popover/template-directives/actions-template.directive");
var DIRECTIVES = [
    actions_template_directive_1.PopoverActionsTemplateDirective,
    body_template_directive_1.PopoverBodyTemplateDirective,
    title_template_directive_1.PopoverTitleTemplateDirective,
    anchor_directive_1.PopoverAnchorDirective,
    container_directive_1.PopoverContainerDirective
];
/**
 * Represents the [NgModule]({{ site.data.urls.angular['ngmoduleapi'] }})
 * definition for the Popover component.
 *
 * @example
 *
 * ```ts-no-run
 * import { PopoverModule } from '@progress/kendo-angular-tooltip';
 *
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 * import { NgModule } from '@angular/core';
 *
 * import { AppComponent } from './app.component';
 *
 * _@NgModule({
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, PopoverModule],
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
var PopoverModule = /** @class */ (function () {
    function PopoverModule() {
    }
    PopoverModule = tslib_1.__decorate([
        core_1.NgModule({
            declarations: DIRECTIVES.concat([
                popover_component_1.PopoverComponent
            ]),
            entryComponents: [popover_component_1.PopoverComponent],
            exports: DIRECTIVES.concat([popover_component_1.PopoverComponent]),
            imports: [
                common_1.CommonModule,
                kendo_angular_popup_1.PopupModule
            ]
        })
    ], PopoverModule);
    return PopoverModule;
}());
exports.PopoverModule = PopoverModule;
