// @progress/kendo-angular-tooltip/dist/npm/tooltip.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var tooltip_directive_1 = require("./tooltip/tooltip.directive");
var kendo_angular_popup_1 = require("@progress/kendo-angular-popup");
var tooltip_content_component_1 = require("./tooltip/tooltip.content.component");
var localized_messages_directive_1 = require("./localization/localized-messages.directive");
var COMPONENT_DIRECTIVES = [tooltip_directive_1.TooltipDirective, tooltip_content_component_1.TooltipContentComponent, localized_messages_directive_1.LocalizedMessagesDirective];
var COMPONENT_MODULES = [kendo_angular_popup_1.PopupModule];
/**
 * Represents the [NgModule]({{ site.data.urls.angular['ngmoduleapi'] }})
 * definition for the Tooltip component.
 *
 * The package exports:
 * - `KendoTooltipDirective`&mdash;The Tooltip directive class.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the Tooltip module
 * import { TooltipModule } from '@progress/kendo-angular-tooltip';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * _@NgModule{{
 *    declarations: [AppComponent], // declare app component
 *    imports:      [BrowserModule, TooltipModule], // import TooltipModule module
 *    bootstrap:    [AppComponent]
 * }}
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 * ```
 */
var TooltipModule = /** @class */ (function () {
    function TooltipModule() {
    }
    TooltipModule = tslib_1.__decorate([
        core_1.NgModule({
            declarations: [COMPONENT_DIRECTIVES],
            entryComponents: [tooltip_content_component_1.TooltipContentComponent],
            imports: [common_1.CommonModule].concat(COMPONENT_MODULES),
            exports: [COMPONENT_DIRECTIVES]
        })
    ], TooltipModule);
    return TooltipModule;
}());
exports.TooltipModule = TooltipModule;
