// @progress/kendo-angular-menu/dist/es2015/rendering/arrow.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Directive, Input, HostBinding, isDevMode } from '@angular/core';
import { ItemsService } from '../services/items.service';
/**
 * Represents a directive that can be used in the [`linkTemplate`]({% slug api_menu_itemlinktemplatedirective %})
 * of the items to render the default expand arrow.
 */
let ExpandArrowDirective = class ExpandArrowDirective {
    constructor(itemsService) {
        this.itemsService = itemsService;
        this.hostClasses = true;
        this.role = 'presentation';
    }
    get arrowDown() {
        return !this.item.horizontal;
    }
    get arrowRight() {
        return this.item.horizontal && !this.item.rtl;
    }
    get arrowLeft() {
        return this.item.horizontal && this.item.rtl;
    }
    ngOnInit() {
        if (isDevMode() && !this.index) {
            throw new Error('The kendoMenuExpandArrow directive requires the item index to be set.');
        }
        this.item = this.itemsService.get(this.index) || {};
    }
};
tslib_1.__decorate([
    Input('kendoMenuExpandArrow'),
    tslib_1.__metadata("design:type", String)
], ExpandArrowDirective.prototype, "index", void 0);
tslib_1.__decorate([
    HostBinding('class.k-icon'),
    HostBinding('class.k-menu-expand-arrow'),
    tslib_1.__metadata("design:type", Boolean)
], ExpandArrowDirective.prototype, "hostClasses", void 0);
tslib_1.__decorate([
    HostBinding('attr.role'),
    tslib_1.__metadata("design:type", String)
], ExpandArrowDirective.prototype, "role", void 0);
tslib_1.__decorate([
    HostBinding('class.k-i-arrow-60-down'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], ExpandArrowDirective.prototype, "arrowDown", null);
tslib_1.__decorate([
    HostBinding('class.k-i-arrow-60-right'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], ExpandArrowDirective.prototype, "arrowRight", null);
tslib_1.__decorate([
    HostBinding('class.k-i-arrow-60-left'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], ExpandArrowDirective.prototype, "arrowLeft", null);
ExpandArrowDirective = tslib_1.__decorate([
    Directive({
        selector: '[kendoMenuExpandArrow]'
    }),
    tslib_1.__metadata("design:paramtypes", [ItemsService])
], ExpandArrowDirective);
export { ExpandArrowDirective };
