// @safe-fleet/sf-security-menu/esm2015/lib/services/security/security.service.js
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { catchError, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class SecurityService {
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    setSecurityInformation(securityEndpoint, settingsEndpoint, headers) {
        this.securityHeaders = headers;
        this.securityEndpoint = securityEndpoint;
        this.settingsEndpoint = settingsEndpoint;
    }
    getApplications() {
        const url = `${this.securityEndpoint}/applications`;
        return this.httpClient.get(url, {
            headers: this.securityHeaders,
        });
    }
    getPermissions() {
        return this.permissions;
    }
    getInfo() {
        return this.applicationInfo;
    }
    getSecurityPermissions(resourceName) {
        const url = `${this.securityEndpoint}/permissions?resourceName=${resourceName}`;
        return this.httpClient
            .get(url, {
            headers: this.securityHeaders,
        })
            .pipe(map((x) => (this.permissions = x)), catchError((error) => {
            throw error;
        }));
    }
    getApplicationInfo() {
        const url = `${this.settingsEndpoint}/info`;
        return this.httpClient
            .get(url, {
            headers: this.securityHeaders,
        })
            .pipe(map((x) => (this.applicationInfo = x)), catchError((error) => {
            throw error;
        }));
    }
}
SecurityService.ɵprov = i0.ɵɵdefineInjectable({ factory: function SecurityService_Factory() { return new SecurityService(i0.ɵɵinject(i1.HttpClient)); }, token: SecurityService, providedIn: "root" });
SecurityService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
SecurityService.ctorParameters = () => [
    { type: HttpClient }
];
