// @progress/kendo-angular-layout/dist/es/stepper/localization/custom-messages.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { StepperMessages } from './messages';
/**
 * Custom component messages override default component messages
 * ([see example]({% slug rtl_layout %})).
 */
var StepperCustomMessagesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(StepperCustomMessagesComponent, _super);
    function StepperCustomMessagesComponent(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    StepperCustomMessagesComponent_1 = StepperCustomMessagesComponent;
    Object.defineProperty(StepperCustomMessagesComponent.prototype, "override", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    var StepperCustomMessagesComponent_1;
    StepperCustomMessagesComponent = StepperCustomMessagesComponent_1 = tslib_1.__decorate([
        Component({
            providers: [
                {
                    provide: StepperMessages,
                    useExisting: forwardRef(function () { return StepperCustomMessagesComponent_1; })
                }
            ],
            selector: 'kendo-stepper-messages',
            template: ""
        }),
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], StepperCustomMessagesComponent);
    return StepperCustomMessagesComponent;
}(StepperMessages));
export { StepperCustomMessagesComponent };
