// @safe-fleet/sf-menu/__ivy_ngcc__/fesm2015/safe-fleet-sf-menu.js
import { BehaviorSubject, Subject } from 'rxjs';
import { Injectable, Component, Input, ViewChild, EventEmitter, Directive, ElementRef, Output, HostListener, HostBinding, NgModule } from '@angular/core';
import { NavigationEnd, Router, RouterModule } from '@angular/router';
import { filter } from 'rxjs/operators';
import { trigger, transition, style, animate } from '@angular/animations';
import { TooltipDirective, TooltipModule } from '@progress/kendo-angular-tooltip';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { InputsModule } from '@progress/kendo-angular-inputs';
import { MenusModule } from '@progress/kendo-angular-menu';
import { PopupModule } from '@progress/kendo-angular-popup';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '@progress/kendo-angular-tooltip';
import * as ɵngcc4 from '@progress/kendo-angular-menu';
import * as ɵngcc5 from '@progress/kendo-angular-popup';
import * as ɵngcc6 from '@progress/kendo-angular-inputs';
import * as ɵngcc7 from '@angular/forms';

function SfMenuComponent_div_0_sf_customer_logo_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "sf-customer-logo");
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("collapsed", ctx_r1.collapsed);
} }
function SfMenuComponent_div_0_sf_menu_shortcuts_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "sf-menu-shortcuts");
} }
function SfMenuComponent_div_0_sf_menu_routes_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "sf-menu-routes");
} }
function SfMenuComponent_div_0_div_12_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵlistener("click", function SfMenuComponent_div_0_div_12_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(2); return ctx_r5.toggleMenu(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("@appear", undefined);
} }
const _c0 = function (a0) { return { "border-bottom-color": a0 }; };
const _c1 = function (a0) { return { "border-top-color": a0 }; };
function SfMenuComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelement(1, "sf-top-menu", 3);
    ɵngcc0.ɵɵelementStart(2, "div", 4);
    ɵngcc0.ɵɵtemplate(3, SfMenuComponent_div_0_sf_customer_logo_3_Template, 1, 2, "sf-customer-logo", 5);
    ɵngcc0.ɵɵelement(4, "sf-menu-products");
    ɵngcc0.ɵɵelementStart(5, "div", 6)(6, "div", 7);
    ɵngcc0.ɵɵtemplate(7, SfMenuComponent_div_0_sf_menu_shortcuts_7_Template, 1, 0, "sf-menu-shortcuts", 8);
    ɵngcc0.ɵɵtemplate(8, SfMenuComponent_div_0_sf_menu_routes_8_Template, 1, 0, "sf-menu-routes", 8);
    ɵngcc0.ɵɵelementEnd()();
    ɵngcc0.ɵɵelement(9, "sf-menu-user")(10, "sf-menu-footer-logo");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(11, "sf-menu-toggle-btn");
    ɵngcc0.ɵɵtemplate(12, SfMenuComponent_div_0_div_12_Template, 1, 1, "div", 9);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(13, _c0, ctx_r0.menuConfig == null ? null : ctx_r0.menuConfig.borderColor));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵstyleProp("background", ctx_r0.gradient);
    ɵngcc0.ɵɵclassProp("collapsed", ctx_r0.collapsed);
    ɵngcc0.ɵɵpropertyInterpolate("showOn", ctx_r0.menuConfig.tooltips || ctx_r0.collapsed ? "hover" : "none");
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(15, _c1, ctx_r0.menuConfig == null ? null : ctx_r0.menuConfig.borderColor));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.menuConfig.logoUrl || ctx_r0.menuConfig.logoClass);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassProp("collapsed", ctx_r0.collapsed);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", (ctx_r0.menuConfig == null ? null : ctx_r0.menuConfig.shortcuts) && ctx_r0.menuConfig.shortcuts.length > 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.menuConfig == null ? null : ctx_r0.menuConfig.routes);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.collapsed);
} }
const _c2 = ["*"];
const _c3 = function (a0) { return { "background-image": a0 }; };
function CustomerLogoComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 2);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(1, _c3, "url(" + ctx_r0.menuConfig.logoUrl + ")"));
} }
function CustomerLogoComponent_em_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "em", 3);
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r1.menuConfig.logoClass);
} }
function MenuFooterLogoComponent__svg_svg_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(0, "svg", 3);
    ɵngcc0.ɵɵelement(1, "path", 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("@appear-fast", undefined);
    ɵngcc0.ɵɵattribute("fill", ctx_r0.color);
} }
function MenuFooterLogoComponent__svg_svg_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(0, "svg", 5);
    ɵngcc0.ɵɵelement(1, "path", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("@appear-fast", undefined);
    ɵngcc0.ɵɵattribute("fill", ctx_r1.color);
} }
function MenuFooterLogoComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 7);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵstyleProp("color", ctx_r2.text);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r2.menuConfig.versionText, "\n");
} }
const _c4 = ["sf-menu-item", ""];
function MenuItemComponent_div_2_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 8);
    ɵngcc0.ɵɵelement(1, "em", 9);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("collapsed", ctx_r2.collapsed);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r2.item.iconClass);
} }
function MenuItemComponent_div_2_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 10);
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(1, _c3, "url(" + ctx_r3.item.iconPath + ")"));
} }
function MenuItemComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 5);
    ɵngcc0.ɵɵtemplate(1, MenuItemComponent_div_2_div_1_Template, 2, 3, "div", 6);
    ɵngcc0.ɵɵtemplate(2, MenuItemComponent_div_2_div_2_Template, 1, 3, "div", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.item.iconClass);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.item.iconClass && ctx_r0.item.iconPath);
} }
function MenuItemComponent_ul_6_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "li", 13);
} if (rf & 2) {
    const child_r5 = ctx.$implicit;
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("show-arrow", child_r5.items)("active-item", child_r5.active);
    ɵngcc0.ɵɵproperty("item", child_r5)("level", ctx_r4.level + 1)("itemType", ctx_r4.itemType);
} }
function MenuItemComponent_ul_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul", 11);
    ɵngcc0.ɵɵtemplate(1, MenuItemComponent_ul_6_li_1_Template, 1, 7, "li", 12);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.item.items);
} }
const _c5 = function (a0, a1) { return { "border-left-color": a0, "padding-left": a1 }; };
function MenuProductsComponent_div_0_em_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "em", 7);
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r3.menuConfig.currentProduct.iconClass);
} }
function MenuProductsComponent_div_0_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 8);
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(1, _c3, "url(" + ctx_r4.menuConfig.currentProduct.iconPath + ")"));
} }
function MenuProductsComponent_div_0_em_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "em", 9);
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵstyleProp("color", ctx_r5.chevron);
    ɵngcc0.ɵɵclassProp("expanded", ctx_r5.productsExpanded)("collapsed", ctx_r5.collapsed);
} }
function MenuProductsComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵlistener("click", function MenuProductsComponent_div_0_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.currentProductClicked(); });
    ɵngcc0.ɵɵtemplate(1, MenuProductsComponent_div_0_em_1_Template, 1, 1, "em", 4);
    ɵngcc0.ɵɵtemplate(2, MenuProductsComponent_div_0_div_2_Template, 1, 3, "div", 5);
    ɵngcc0.ɵɵtemplate(3, MenuProductsComponent_div_0_em_3_Template, 1, 6, "em", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("collapsed", ctx_r0.collapsed);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.menuConfig.currentProduct.iconClass);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.menuConfig.currentProduct.iconClass && ctx_r0.menuConfig.currentProduct.iconPath);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.menuConfig.products && ctx_r0.menuConfig.products.length > 0);
} }
function MenuProductsComponent_div_1_h4_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "h4");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r8.menuConfig.currentProduct.text, " ");
} }
function MenuProductsComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵtemplate(1, MenuProductsComponent_div_1_h4_1_Template, 2, 1, "h4", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵstyleProp("color", ctx_r1.text);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.menuConfig.currentProduct == null ? null : ctx_r1.menuConfig.currentProduct.text);
} }
function MenuProductsComponent_div_2_a_1_em_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "em", 7);
} if (rf & 2) {
    const option_r10 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngClass", option_r10.iconClass);
} }
function MenuProductsComponent_div_2_a_1_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 8);
} if (rf & 2) {
    const option_r10 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(1, _c3, "url(" + option_r10.iconPath + ")"));
} }
function MenuProductsComponent_div_2_a_1_Template(rf, ctx) { if (rf & 1) {
    const _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 14);
    ɵngcc0.ɵɵlistener("click", function MenuProductsComponent_div_2_a_1_Template_a_click_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r16); const option_r10 = restoredCtx.$implicit; const ctx_r15 = ɵngcc0.ɵɵnextContext(2); return ctx_r15.productClicked(option_r10, $event); });
    ɵngcc0.ɵɵtemplate(1, MenuProductsComponent_div_2_a_1_em_1_Template, 1, 1, "em", 4);
    ɵngcc0.ɵɵtemplate(2, MenuProductsComponent_div_2_a_1_div_2_Template, 1, 3, "div", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const option_r10 = ctx.$implicit;
    const ctx_r9 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("collapsed", ctx_r9.collapsed)("two-columns", ctx_r9.menuConfig.products.length > 3);
    ɵngcc0.ɵɵattribute("href", option_r10.routeHref, ɵngcc0.ɵɵsanitizeUrl);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", option_r10.iconClass);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !option_r10.iconClass && option_r10.iconPath);
} }
function MenuProductsComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵtemplate(1, MenuProductsComponent_div_2_a_1_Template, 3, 7, "a", 13);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵstyleProp("background-color", ctx_r2.productsBg);
    ɵngcc0.ɵɵclassProp("expanded", ctx_r2.productsExpanded);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r2.menuConfig.products);
} }
function MenuRoutesComponent_ul_0_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "li", 3);
} if (rf & 2) {
    const item_r3 = ctx.$implicit;
    ɵngcc0.ɵɵclassProp("show-arrow", item_r3.items)("active-item", item_r3.active);
    ɵngcc0.ɵɵproperty("item", item_r3)("level", 1)("itemType", 2);
} }
function MenuRoutesComponent_ul_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul");
    ɵngcc0.ɵɵtemplate(1, MenuRoutesComponent_ul_0_li_1_Template, 1, 7, "li", 2);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("@appear-fast", undefined);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.menuConfig.routes);
} }
function MenuRoutesComponent_kendo_menu_1_ng_template_1_em_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "em", 11);
} if (rf & 2) {
    const item_r5 = ɵngcc0.ɵɵnextContext().item;
    ɵngcc0.ɵɵproperty("ngClass", item_r5.iconClass);
} }
function MenuRoutesComponent_kendo_menu_1_ng_template_1_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 12);
} if (rf & 2) {
    const item_r5 = ɵngcc0.ɵɵnextContext().item;
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(1, _c3, "url(" + item_r5.iconPath + ")"));
} }
function MenuRoutesComponent_kendo_menu_1_ng_template_1_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r5 = ɵngcc0.ɵɵnextContext().item;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(item_r5.text);
} }
function MenuRoutesComponent_kendo_menu_1_ng_template_1_em_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "em", 13);
} }
const _c6 = function (a0) { return { "border-left-color": a0 }; };
function MenuRoutesComponent_kendo_menu_1_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    const _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 6);
    ɵngcc0.ɵɵlistener("click", function MenuRoutesComponent_kendo_menu_1_ng_template_1_Template_div_click_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r15); const item_r5 = restoredCtx.item; const ctx_r14 = ɵngcc0.ɵɵnextContext(2); return ctx_r14.routeClicked(item_r5, $event); });
    ɵngcc0.ɵɵelementStart(1, "div", 7);
    ɵngcc0.ɵɵtemplate(2, MenuRoutesComponent_kendo_menu_1_ng_template_1_em_2_Template, 1, 1, "em", 8);
    ɵngcc0.ɵɵtemplate(3, MenuRoutesComponent_kendo_menu_1_ng_template_1_div_3_Template, 1, 3, "div", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, MenuRoutesComponent_kendo_menu_1_ng_template_1_span_4_Template, 2, 1, "span", 0);
    ɵngcc0.ɵɵtemplate(5, MenuRoutesComponent_kendo_menu_1_ng_template_1_em_5_Template, 1, 0, "em", 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r5 = ctx.item;
    const index_r6 = ctx.index;
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("active-item", item_r5.active);
    ɵngcc0.ɵɵpropertyInterpolate("title", item_r5.items ? undefined : item_r5.text);
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(11, _c6, ctx_r4.menuConfig == null ? null : ctx_r4.menuConfig.borderColor))("@appear-fast", undefined);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassProp("collapsed", ctx_r4.collapsed);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r5.iconClass);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !item_r5.iconClass && item_r5.iconPath);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.showText(index_r6));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r5.items);
} }
function MenuRoutesComponent_kendo_menu_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "kendo-menu", 4);
    ɵngcc0.ɵɵtemplate(1, MenuRoutesComponent_kendo_menu_1_ng_template_1_Template, 6, 13, "ng-template", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("items", ctx_r1.menuConfig.routes)("vertical", true);
} }
function MenuShortcutsComponent_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "li", 1);
} if (rf & 2) {
    const item_r1 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("item", item_r1)("level", 1)("itemType", 1);
} }
function MenuUserComponent_div_7_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div")(1, "span", 11);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 12);
    ɵngcc0.ɵɵlistener("click", function MenuUserComponent_div_7_Template_span_click_3_listener() { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.showMultipleTenantMenu(); });
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1("", ctx_r0.menuConfig.userName, " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", (ctx_r0.menuConfig == null ? null : ctx_r0.menuConfig.multipleTenants == null ? null : ctx_r0.menuConfig.multipleTenants.length) >= 2 ? "" : ctx_r0.menuConfig.fleetName, " ");
} }
function MenuUserComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵlistener("click", function MenuUserComponent_div_8_Template_div_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.onSsoManageAccount($event, ctx_r8.menuConfig.ssoManageAccount); });
    ɵngcc0.ɵɵelementStart(1, "div", 14);
    ɵngcc0.ɵɵelement(2, "em", 15);
    ɵngcc0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassProp("disable", ctx_r1.menuConfig.ssoManageAccount.disable);
    ɵngcc0.ɵɵpropertyInterpolate("showOn", !ctx_r1.menuConfig.tooltips ? "hover" : "none");
    ɵngcc0.ɵɵpropertyInterpolate("title", ctx_r1.menuConfig.ssoManageAccount.title);
} }
function MenuUserComponent_li_9_span_7_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 23);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r14 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r14.menuConfig.fleetTitle, " ");
} }
function MenuUserComponent_li_9_span_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtemplate(1, MenuUserComponent_li_9_span_7_div_1_Template, 2, 1, "div", 22);
    ɵngcc0.ɵɵelementStart(2, "div");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r11.menuConfig.fleetTitle);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r11.menuConfig.fleetName, " ");
} }
function MenuUserComponent_li_9_em_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "em", 24);
} }
function MenuUserComponent_li_9_kendo_popup_9_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r18 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 28);
    ɵngcc0.ɵɵlistener("click", function MenuUserComponent_li_9_kendo_popup_9_div_2_Template_div_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r18); const tenant_r16 = restoredCtx.$implicit; const ctx_r17 = ɵngcc0.ɵɵnextContext(3); return ctx_r17.changeTenant(tenant_r16); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const tenant_r16 = ctx.$implicit;
    ɵngcc0.ɵɵclassProp("active", tenant_r16.active);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", tenant_r16.id, " ");
} }
function MenuUserComponent_li_9_kendo_popup_9_Template(rf, ctx) { if (rf & 1) {
    const _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "kendo-popup", 25);
    ɵngcc0.ɵɵlistener("clickOutside", function MenuUserComponent_li_9_kendo_popup_9_Template_kendo_popup_clickOutside_0_listener() { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r19 = ɵngcc0.ɵɵnextContext(2); return ctx_r19.hideMultipleTenantMenu(); });
    ɵngcc0.ɵɵelementStart(1, "div", 26);
    ɵngcc0.ɵɵtemplate(2, MenuUserComponent_li_9_kendo_popup_9_div_2_Template, 2, 3, "div", 27);
    ɵngcc0.ɵɵelementEnd()();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    const _r10 = ɵngcc0.ɵɵreference(1);
    const ctx_r13 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("anchor", _r10)("anchorAlign", ctx_r13.anchorAlign)("popupAlign", ctx_r13.popupAlign);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵstyleProp("color", ctx_r13.color)("background-color", ctx_r13.tenantBg);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r13.tenants);
} }
function MenuUserComponent_li_9_Template(rf, ctx) { if (rf & 1) {
    const _r22 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", null, 16)(2, "div", 17)(3, "div", 18);
    ɵngcc0.ɵɵlistener("click", function MenuUserComponent_li_9_Template_div_click_3_listener() { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r21 = ɵngcc0.ɵɵnextContext(); return ctx_r21.showMultipleTenantMenu(); });
    ɵngcc0.ɵɵelementStart(4, "div", 3)(5, "div", 4);
    ɵngcc0.ɵɵelement(6, "em", 19);
    ɵngcc0.ɵɵelementEnd()();
    ɵngcc0.ɵɵtemplate(7, MenuUserComponent_li_9_span_7_Template, 4, 2, "span", 6);
    ɵngcc0.ɵɵtemplate(8, MenuUserComponent_li_9_em_8_Template, 1, 0, "em", 20);
    ɵngcc0.ɵɵelementEnd()();
    ɵngcc0.ɵɵtemplate(9, MenuUserComponent_li_9_kendo_popup_9_Template, 3, 8, "kendo-popup", 21);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵclassProp("show-arrow", ctx_r2.menuConfig == null ? null : ctx_r2.menuConfig.multipleTenants)("collapsed", ctx_r2.collapsed);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassProp("expanded-item", ctx_r2.multipleTenantMenu);
    ɵngcc0.ɵɵpropertyInterpolate("title", ctx_r2.menuConfig.tooltips || ctx_r2.collapsed ? ctx_r2.menuConfig.fleetName : "");
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(11, _c6, ctx_r2.menuConfig == null ? null : ctx_r2.menuConfig.borderColor));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.collapsed);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.collapsed);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.multipleTenantMenu);
} }
function MenuUserComponent_li_10_span_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r23 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r23.menuConfig == null ? null : ctx_r23.menuConfig.themeTitle);
} }
function MenuUserComponent_li_10_kendo_switch_7_Template(rf, ctx) { if (rf & 1) {
    const _r26 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "kendo-switch", 33);
    ɵngcc0.ɵɵlistener("valueChange", function MenuUserComponent_li_10_kendo_switch_7_Template_kendo_switch_valueChange_0_listener() { ɵngcc0.ɵɵrestoreView(_r26); const ctx_r25 = ɵngcc0.ɵɵnextContext(2); return ctx_r25.setDarkMode(); })("ngModelChange", function MenuUserComponent_li_10_kendo_switch_7_Template_kendo_switch_ngModelChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r26); const ctx_r27 = ɵngcc0.ɵɵnextContext(2); return ctx_r27.darkMode = $event; });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r24 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r24.darkMode);
} }
function MenuUserComponent_li_10_Template(rf, ctx) { if (rf & 1) {
    const _r29 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 29);
    ɵngcc0.ɵɵlistener("click", function MenuUserComponent_li_10_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r29); const ctx_r28 = ɵngcc0.ɵɵnextContext(); return ctx_r28.themeIconClicked(); });
    ɵngcc0.ɵɵelementStart(1, "div", 17)(2, "div", 30)(3, "div", 3)(4, "div", 4);
    ɵngcc0.ɵɵelement(5, "em", 31);
    ɵngcc0.ɵɵelementEnd()();
    ɵngcc0.ɵɵtemplate(6, MenuUserComponent_li_10_span_6_Template, 2, 1, "span", 6);
    ɵngcc0.ɵɵtemplate(7, MenuUserComponent_li_10_kendo_switch_7_Template, 1, 1, "kendo-switch", 32);
    ɵngcc0.ɵɵelementEnd()()();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassProp("collapsed", ctx_r3.collapsed);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ctx_r3.menuConfig.tooltips || ctx_r3.collapsed ? ctx_r3.menuConfig.themeTitle : "");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r3.collapsed);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r3.collapsed);
} }
function MenuUserComponent_li_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "li", 34);
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("show-arrow", ctx_r4.userActions.help.items)("active-item", ctx_r4.userActions.help.active);
    ɵngcc0.ɵɵproperty("item", ctx_r4.userActions.help)("itemType", 3);
} }
function MenuUserComponent_li_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "li", 35);
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("show-arrow", ctx_r5.userActions.logout.items)("active-item", ctx_r5.userActions.logout.active);
    ɵngcc0.ɵɵproperty("item", ctx_r5.userActions.logout)("itemType", 3);
} }
function TopMenuComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 8);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(1, _c3, "url(" + ctx_r0.menuConfig.logoUrl + ")"));
} }
function TopMenuComponent_em_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "em", 9);
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r1.menuConfig.logoClass);
} }
function TopMenuComponent__svg_svg_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(0, "svg", 10);
    ɵngcc0.ɵɵelement(1, "path", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("@appear-fast", undefined);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("fill", ctx_r2.iconColor);
} }
function TopMenuComponent_ng_container_7_em_1_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "em", 14);
    ɵngcc0.ɵɵlistener("click", function TopMenuComponent_ng_container_7_em_1_Template_em_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r9); const shortcut_r4 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.shortcutClicked(shortcut_r4); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const shortcut_r4 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵstyleProp("background-color", shortcut_r4.active ? ctx_r5.activeLink : "")("color", ctx_r5.iconColor);
    ɵngcc0.ɵɵproperty("ngClass", shortcut_r4.iconClass);
} }
function TopMenuComponent_ng_container_7_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 15)(1, "div", 16);
    ɵngcc0.ɵɵlistener("click", function TopMenuComponent_ng_container_7_div_2_Template_div_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r13); const shortcut_r4 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.shortcutClicked(shortcut_r4); });
    ɵngcc0.ɵɵelementEnd()();
} if (rf & 2) {
    const shortcut_r4 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵstyleProp("background-color", shortcut_r4.active ? ctx_r6.activeLink : "");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(3, _c3, "url(" + shortcut_r4.iconPath + ")"));
} }
function TopMenuComponent_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, TopMenuComponent_ng_container_7_em_1_Template, 1, 5, "em", 12);
    ɵngcc0.ɵɵtemplate(2, TopMenuComponent_ng_container_7_div_2_Template, 2, 5, "div", 13);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const shortcut_r4 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", shortcut_r4.iconClass);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !shortcut_r4.iconClass && shortcut_r4.iconPath);
} }
class MenuConfig {
    constructor() {
        this.helpText = 'Help';
        this.logoutText = 'Logout';
        // optional menu
        this.hideMenu = false;
        this.tooltips = false;
        this.denyChangeTheme = false;
        this.menuIgnoresTheme = true;
        this.hideProductName = false;
        this.routeClickPreventCollapse = false;
        this.disableUserNameClick = false;
    }
}
class SsoManageAccountConfig {
    constructor() {
        this.disable = false;
    }
}

var MenuEventType;
(function (MenuEventType) {
    MenuEventType[MenuEventType["ProductClick"] = 0] = "ProductClick";
    MenuEventType[MenuEventType["ShortcutClick"] = 1] = "ShortcutClick";
    MenuEventType[MenuEventType["RouteClick"] = 2] = "RouteClick";
    MenuEventType[MenuEventType["UserActionClick"] = 3] = "UserActionClick";
    MenuEventType[MenuEventType["ChangeTenantClick"] = 4] = "ChangeTenantClick";
    MenuEventType[MenuEventType["OpenTenantListClick"] = 5] = "OpenTenantListClick";
})(MenuEventType || (MenuEventType = {}));

class SfMenuService {
    constructor(router) {
        this.router = router;
        this.SMALL_RESOLUTION = 768;
        this.menuCollapse$ = new BehaviorSubject(false);
        this.events$ = new Subject();
        this.menuConfig$ = new BehaviorSubject(undefined);
        // store nodes to expand on navigation end
        this.nodesToExpand = [];
        this.themes = ['light', 'dark'];
        // theming
        this.setBaseColors();
        this.currentTheme = new BehaviorSubject(this.themes[0]);
        this.storedTheme = localStorage.getItem('theme');
        if (!this.storedTheme) {
            this.storedTheme = window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches ? this.themes[1] : this.themes[0];
        }
        this.setMenuConfig(new MenuConfig());
        this.router.events.pipe(filter((e) => e instanceof NavigationEnd)).subscribe((e) => {
            var _a, _b;
            this.nodesToExpand.length = 0;
            (_b = (_a = this.menuConfig) === null || _a === void 0 ? void 0 : _a.routes) === null || _b === void 0 ? void 0 : _b.forEach((r) => this.setCurrentRouteActive(r, e.urlAfterRedirects));
        });
    }
    get colors() {
        return this.menuConfig.menuIgnoresTheme || this.menuConfig.denyChangeTheme ? this.noThemeColors : this.themeColors;
    }
    setMenuConfig(cfg) {
        this.menuConfig = cfg || new MenuConfig();
        this.menuConfig$.next(this.menuConfig);
        this.setTheme(this.storedTheme);
    }
    setBaseColors() {
        const notheme = {};
        const theme = {};
        const root = getComputedStyle(document.querySelector(':root'));
        ['blue', 'gray', 'green', 'red', 'orange', 'purple', 'teal', 'yellow'].forEach((col) => {
            for (let i = 1; i < 10; i++) {
                notheme[`${col}${i * 100}`] = root.getPropertyValue(`--clr-${col}-${1000 - i * 100}`);
                theme[`${col}${i * 100}`] = `var(--clr-${col}-${i * 100})`;
            }
        });
        this.noThemeColors = notheme;
        this.themeColors = theme;
    }
    routeClicked(item, evt) {
        evt.stopPropagation();
        this.events$.next({
            type: MenuEventType.RouteClick,
            value: item
        });
        if (item.items) {
            item.expanded = !item.expanded;
            return;
        }
        if (item.routePath) {
            if (!this.menuConfig.routeClickPreventCollapse) {
                this.collapseAllItems(this.menuConfig.routes);
            }
            this.router.navigate([item.routePath]);
            return;
        }
        if (item.routeHref) {
            if (!this.menuConfig.routeClickPreventCollapse) {
                this.collapseAllItems(this.menuConfig.routes);
            }
            if (item.newTab) {
                window.open(item.routeHref);
                return;
            }
            if (item.newWindow) {
                let top;
                let left;
                let width;
                let height;
                if (item.newWindowWidth > 200 && item.newWindowHeight > 100) {
                    width = item.newWindowWidth;
                    height = item.newWindowHeight;
                    left = window.screenLeft + window.outerWidth / 2 - width / 2;
                    top = window.screenTop + window.outerHeight / 2 - height / 2;
                }
                else {
                    height = window.outerHeight;
                    width = window.outerWidth;
                    left = window.screenLeft;
                    top = window.screenTop;
                }
                window.open(item.routeHref, item.id, `top=${top},left=${left},height=${height},width=${width},resizable=1,scrollbars=1`);
            }
        }
        if (window.innerWidth <= this.SMALL_RESOLUTION && !this.menuCollapse$.value) {
            this.menuCollapse$.next(!this.menuCollapse$.value);
        }
    }
    setCurrentRouteActive(route, url) {
        var _a, _b;
        route.active = !((_a = route.items) === null || _a === void 0 ? void 0 : _a.length) && `/${route.routePath}` === url;
        if (route.active) {
            this.nodesToExpand.forEach((n) => (n.expanded = true));
        }
        this.nodesToExpand.push(route);
        (_b = route.items) === null || _b === void 0 ? void 0 : _b.forEach((ch) => this.setCurrentRouteActive(ch, url));
        this.nodesToExpand.pop();
    }
    // theming
    setTheme(theme) {
        if (this.menuConfig.denyChangeTheme || this.themes.indexOf(theme) < 0) {
            theme = this.themes[0];
        }
        this.currentTheme.next(theme);
        localStorage.setItem('theme', theme);
        this.themes.forEach((t) => document.querySelector('html').classList.toggle(t, t === theme));
    }
    collapseAllItems(routes) {
        if (routes) {
            routes.forEach((route) => {
                route.expanded = false;
                if (route.items) {
                    this.collapseAllItems(route.items);
                }
            });
        }
    }
}
SfMenuService.ɵfac = function SfMenuService_Factory(t) { return new (t || SfMenuService)(ɵngcc0.ɵɵinject(ɵngcc1.Router)); };
SfMenuService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: SfMenuService, factory: SfMenuService.ɵfac });
SfMenuService.ctorParameters = () => [
    { type: Router }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SfMenuService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.Router }]; }, null); })();

class SfMenuComponent {
    constructor(svc) {
        this.svc = svc;
        this.collapsed = false;
    }
    set config(value) {
        this.svc.setMenuConfig(value);
    }
    ngOnInit() {
        const clr = this.svc.colors;
        this.gradient = `linear-gradient(180deg, ${clr.gray200} 0%, ${clr.gray100} 100%)`;
        this.color = clr.gray900;
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            this.menuConfig = cfg;
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
            if (this.tooltipDir) {
                this.tooltipDir.hide();
            }
        });
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
    toggleMenu() {
        this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
    }
}
SfMenuComponent.ɵfac = function SfMenuComponent_Factory(t) { return new (t || SfMenuComponent)(ɵngcc0.ɵɵdirectiveInject(SfMenuService)); };
SfMenuComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SfMenuComponent, selectors: [["sf-menu"]], viewQuery: function SfMenuComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(TooltipDirective, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.tooltipDir = _t.first);
    } }, inputs: { config: "config", borderColor: "borderColor" }, ngContentSelectors: _c2, decls: 3, vars: 1, consts: [["class", "sf-menu-container", 4, "ngIf"], [1, "application-content"], [1, "sf-menu-container"], [3, "ngStyle"], ["kendoTooltip", "", "position", "bottom", 1, "sf-menu-content", 3, "ngStyle", "showOn"], [3, "collapsed", 4, "ngIf"], [1, "fm-menu-scrollable-content"], [1, "scroll"], [4, "ngIf"], ["class", "sf-menu-backdrop", 3, "click", 4, "ngIf"], [1, "sf-menu-backdrop", 3, "click"]], template: function SfMenuComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, SfMenuComponent_div_0_Template, 13, 17, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵprojection(2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.menuConfig.hideMenu);
    } }, directives: function () { return [ɵngcc2.NgIf, TopMenuComponent, ɵngcc2.NgStyle, ɵngcc3.TooltipDirective, CustomerLogoComponent, MenuProductsComponent, MenuShortcutsComponent, MenuRoutesComponent, MenuUserComponent, MenuFooterLogoComponent, MenuToggleBtnComponent]; }, styles: ["[_nghost-%COMP%]{width:100%;height:100%;display:grid;grid-template-areas:\"menu\" \"content\";grid-template-rows:auto 1fr;grid-template-columns:1fr}@media screen and (min-width:768px){[_nghost-%COMP%]{grid-template-areas:\"menu content\";grid-template-rows:1fr;grid-template-columns:auto 1fr}}.sf-menu-container[_ngcontent-%COMP%]{grid-area:menu;position:relative}.sf-menu-content[_ngcontent-%COMP%]{height:auto;max-height:calc(100% - 70px);width:250px;border-bottom:none;transition:transform .12s ease-out,width .12s ease-out;overflow-y:auto;position:fixed;z-index:10000}.sf-menu-content[_ngcontent-%COMP%]::-webkit-scrollbar{width:6px;background-color:transparent}.sf-menu-content[_ngcontent-%COMP%]::-webkit-scrollbar-thumb{border-radius:10px;background-color:var(--clr-gray-400)}.sf-menu-content.collapsed[_ngcontent-%COMP%]{transform:translateX(-100%)}@media screen and (min-width:768px){.sf-menu-content[_ngcontent-%COMP%]{display:flex;flex-direction:column;justify-content:space-between;border-top:5px solid #00a2fe;position:relative;height:100%;max-height:100%;width:250px;overflow:hidden}.sf-menu-content.collapsed[_ngcontent-%COMP%]{transform:translateX(0);width:75px}}@media screen and (max-width:380px){.sf-menu-content[_ngcontent-%COMP%]{max-height:calc(100% - 48px)}}.sf-menu-backdrop[_ngcontent-%COMP%]{position:fixed;width:100%;height:100%;background-color:rgba(0,0,0,.71);z-index:9999}.application-content[_ngcontent-%COMP%]{grid-area:content;overflow:auto;height:100%;position:relative}.sf-menu-logo-bg[_ngcontent-%COMP%]{background:50%/contain no-repeat}.fm-menu-scrollable-content[_ngcontent-%COMP%]{display:flex;flex-direction:column;width:100%;height:auto;align-items:flex-start;justify-content:flex-start;overflow:hidden;position:relative;flex:1 0 120px;padding:0}.fm-menu-scrollable-content[_ngcontent-%COMP%]:before{content:\" \";width:100%;height:1px;background-color:#fff;opacity:.15}.fm-menu-scrollable-content[_ngcontent-%COMP%]   ul[_ngcontent-%COMP%]{list-style:none;padding-left:0;width:100%}.fm-menu-scrollable-content[_ngcontent-%COMP%]   [_ngcontent-%COMP%]:before{display:none}.fm-menu-scrollable-content[_ngcontent-%COMP%]   .scroll[_ngcontent-%COMP%]{overflow-y:auto;overflow-x:hidden;width:100%}.fm-menu-scrollable-content[_ngcontent-%COMP%]   .scroll[_ngcontent-%COMP%]::-webkit-scrollbar{width:6px;background-color:transparent}.fm-menu-scrollable-content[_ngcontent-%COMP%]   .scroll[_ngcontent-%COMP%]::-webkit-scrollbar-thumb{border-radius:10px;background-color:#5b616a}@media screen and (min-width:768px){.fm-menu-scrollable-content[_ngcontent-%COMP%]   .scroll[_ngcontent-%COMP%]{max-height:65vh}}"], data: { animation: [
            trigger('appear', [
                transition(':enter', [style({ opacity: 0 }), animate(180, style({ opacity: 1 }))]),
                transition(':leave', [style({ opacity: 1 }), animate(180, style({ opacity: 0 }))])
            ])
        ] } });
SfMenuComponent.ctorParameters = () => [
    { type: SfMenuService }
];
SfMenuComponent.propDecorators = {
    borderColor: [{ type: Input }],
    config: [{ type: Input, args: ['config',] }],
    tooltipDir: [{ type: ViewChild, args: [TooltipDirective,] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SfMenuComponent, [{
        type: Component,
        args: [{
                selector: 'sf-menu',
                template: "<div class=\"sf-menu-container\" *ngIf=\"!menuConfig.hideMenu\">\n  <sf-top-menu [ngStyle]=\"{ 'border-bottom-color': menuConfig?.borderColor }\"></sf-top-menu>\n  <div\n    class=\"sf-menu-content\"\n    [class.collapsed]=\"collapsed\"\n    [ngStyle]=\"{ 'border-top-color': menuConfig?.borderColor }\"\n    [style.background]=\"gradient\"\n    kendoTooltip\n    showOn=\"{{ menuConfig.tooltips || collapsed ? 'hover' : 'none' }}\"\n    position=\"bottom\"\n  >\n    <sf-customer-logo [class.collapsed]=\"collapsed\" *ngIf=\"menuConfig.logoUrl || menuConfig.logoClass\"></sf-customer-logo>\n    <sf-menu-products [class.collapsed]=\"collapsed\"></sf-menu-products>\n    <div class=\"fm-menu-scrollable-content\">\n      <div class=\"scroll\">\n        <sf-menu-shortcuts *ngIf=\"menuConfig?.shortcuts && menuConfig.shortcuts.length > 0\"></sf-menu-shortcuts>\n        <sf-menu-routes *ngIf=\"menuConfig?.routes\"></sf-menu-routes>\n      </div>\n    </div>\n    <sf-menu-user></sf-menu-user>\n    <sf-menu-footer-logo></sf-menu-footer-logo>\n  </div>\n  <sf-menu-toggle-btn></sf-menu-toggle-btn>\n  <div class=\"sf-menu-backdrop\" *ngIf=\"!collapsed\" (click)=\"toggleMenu()\" [@appear]></div>\n</div>\n<div class=\"application-content\">\n  <ng-content></ng-content>\n</div>\n",
                animations: [
                    trigger('appear', [
                        transition(':enter', [style({ opacity: 0 }), animate(180, style({ opacity: 1 }))]),
                        transition(':leave', [style({ opacity: 1 }), animate(180, style({ opacity: 0 }))])
                    ])
                ],
                styles: [":host{width:100%;height:100%;display:grid;grid-template-areas:\"menu\" \"content\";grid-template-rows:auto 1fr;grid-template-columns:1fr}@media screen and (min-width:768px){:host{grid-template-areas:\"menu content\";grid-template-rows:1fr;grid-template-columns:auto 1fr}}.sf-menu-container{grid-area:menu;position:relative}.sf-menu-content{height:auto;max-height:calc(100% - 70px);width:250px;border-bottom:none;transition:transform .12s ease-out,width .12s ease-out;overflow-y:auto;position:fixed;z-index:10000}.sf-menu-content::-webkit-scrollbar{width:6px;background-color:transparent}.sf-menu-content::-webkit-scrollbar-thumb{border-radius:10px;background-color:var(--clr-gray-400)}.sf-menu-content.collapsed{transform:translateX(-100%)}@media screen and (min-width:768px){.sf-menu-content{display:flex;flex-direction:column;justify-content:space-between;border-top:5px solid #00a2fe;position:relative;height:100%;max-height:100%;width:250px;overflow:hidden}.sf-menu-content.collapsed{transform:translateX(0);width:75px}}@media screen and (max-width:380px){.sf-menu-content{max-height:calc(100% - 48px)}}.sf-menu-backdrop{position:fixed;width:100%;height:100%;background-color:rgba(0,0,0,.71);z-index:9999}.application-content{grid-area:content;overflow:auto;height:100%;position:relative}.sf-menu-logo-bg{background:50%/contain no-repeat}.fm-menu-scrollable-content{display:flex;flex-direction:column;width:100%;height:auto;align-items:flex-start;justify-content:flex-start;overflow:hidden;position:relative;flex:1 0 120px;padding:0}.fm-menu-scrollable-content:before{content:\" \";width:100%;height:1px;background-color:#fff;opacity:.15}.fm-menu-scrollable-content ul{list-style:none;padding-left:0;width:100%}.fm-menu-scrollable-content :before{display:none}.fm-menu-scrollable-content .scroll{overflow-y:auto;overflow-x:hidden;width:100%}.fm-menu-scrollable-content .scroll::-webkit-scrollbar{width:6px;background-color:transparent}.fm-menu-scrollable-content .scroll::-webkit-scrollbar-thumb{border-radius:10px;background-color:#5b616a}@media screen and (min-width:768px){.fm-menu-scrollable-content .scroll{max-height:65vh}}"]
            }]
    }], function () { return [{ type: SfMenuService }]; }, { config: [{
            type: Input,
            args: ['config']
        }], borderColor: [{
            type: Input
        }], tooltipDir: [{
            type: ViewChild,
            args: [TooltipDirective]
        }] }); })();

class ClickOutsideDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.clickOutside = new EventEmitter();
        this.ready = false;
        setTimeout(() => (this.ready = true), 200);
    }
    onClick(event, targetElement) {
        if (!this.ready || !this.elementRef || !targetElement) {
            return;
        }
        const clickedInside = this.elementRef.nativeElement.contains(targetElement);
        if (!clickedInside) {
            this.clickOutside.emit(event);
        }
    }
}
ClickOutsideDirective.ɵfac = function ClickOutsideDirective_Factory(t) { return new (t || ClickOutsideDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ClickOutsideDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: ClickOutsideDirective, selectors: [["", "clickOutside", ""]], hostBindings: function ClickOutsideDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ClickOutsideDirective_click_HostBindingHandler($event) { return ctx.onClick($event, $event.target); }, false, ɵngcc0.ɵɵresolveDocument);
    } }, outputs: { clickOutside: "clickOutside" } });
ClickOutsideDirective.ctorParameters = () => [
    { type: ElementRef }
];
ClickOutsideDirective.propDecorators = {
    clickOutside: [{ type: Output }],
    onClick: [{ type: HostListener, args: ['document:click', ['$event', '$event.target'],] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClickOutsideDirective, [{
        type: Directive,
        args: [{
                selector: '[clickOutside]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { clickOutside: [{
            type: Output
        }], onClick: [{
            type: HostListener,
            args: ['document:click', ['$event', '$event.target']]
        }] }); })();

class CustomerLogoComponent {
    constructor(svc) {
        this.svc = svc;
        this.collapsed = false;
    }
    get clr() {
        return this.background;
    }
    ngOnInit() {
        this.background = this.svc.colors.gray100;
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            this.menuConfig = cfg;
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
        });
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
}
CustomerLogoComponent.ɵfac = function CustomerLogoComponent_Factory(t) { return new (t || CustomerLogoComponent)(ɵngcc0.ɵɵdirectiveInject(SfMenuService)); };
CustomerLogoComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: CustomerLogoComponent, selectors: [["sf-customer-logo"]], hostVars: 2, hostBindings: function CustomerLogoComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("background-color", ctx.clr);
    } }, decls: 2, vars: 2, consts: [["data-testid", "menu-logo-bg", "class", "sf-menu-logo-bg", 3, "ngStyle", 4, "ngIf"], [3, "ngClass", 4, "ngIf"], ["data-testid", "menu-logo-bg", 1, "sf-menu-logo-bg", 3, "ngStyle"], [3, "ngClass"]], template: function CustomerLogoComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, CustomerLogoComponent_div_0_Template, 1, 3, "div", 0);
        ɵngcc0.ɵɵtemplate(1, CustomerLogoComponent_em_1_Template, 1, 1, "em", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.menuConfig.logoUrl && !ctx.menuConfig.logoClass);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.menuConfig.logoClass && !ctx.menuConfig.logoUrl);
    } }, directives: [ɵngcc2.NgIf, ɵngcc2.NgStyle, ɵngcc2.NgClass], styles: ["[_nghost-%COMP%]{display:flex;align-items:center;justify-content:center;width:100%;flex-direction:column;height:auto;padding:5px;position:relative}[_nghost-%COMP%]   .sf-menu-logo-bg[_ngcontent-%COMP%]{background:50%/contain no-repeat;height:2.0625rem;width:4.8125rem}.collapsed[_nghost-%COMP%]   .sf-menu-logo-bg[_ngcontent-%COMP%]{width:80%}"] });
CustomerLogoComponent.ctorParameters = () => [
    { type: SfMenuService }
];
CustomerLogoComponent.propDecorators = {
    clr: [{ type: HostBinding, args: ['style.background-color',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CustomerLogoComponent, [{
        type: Component,
        args: [{
                selector: 'sf-customer-logo',
                template: "<div\n  data-testid=\"menu-logo-bg\"\n  class=\"sf-menu-logo-bg\"\n  *ngIf=\"menuConfig.logoUrl && !menuConfig.logoClass\"\n  [ngStyle]=\"{\n    'background-image': 'url(' + menuConfig.logoUrl + ')'\n  }\"\n></div>\n<em [ngClass]=\"menuConfig.logoClass\" *ngIf=\"menuConfig.logoClass && !menuConfig.logoUrl\"></em>\n",
                styles: [":host{display:flex;align-items:center;justify-content:center;width:100%;flex-direction:column;height:auto;padding:5px;position:relative}:host .sf-menu-logo-bg{background:50%/contain no-repeat;height:2.0625rem;width:4.8125rem}:host.collapsed .sf-menu-logo-bg{width:80%}"]
            }]
    }], function () { return [{ type: SfMenuService }]; }, { clr: [{
            type: HostBinding,
            args: ['style.background-color']
        }] }); })();

class MenuFooterLogoComponent {
    constructor(svc) {
        this.svc = svc;
        this.collapsed = false;
    }
    ngOnInit() {
        this.color = this.svc.colors.gray900;
        this.text = this.svc.colors.gray800;
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            this.menuConfig = cfg;
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
        });
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
}
MenuFooterLogoComponent.ɵfac = function MenuFooterLogoComponent_Factory(t) { return new (t || MenuFooterLogoComponent)(ɵngcc0.ɵɵdirectiveInject(SfMenuService)); };
MenuFooterLogoComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: MenuFooterLogoComponent, selectors: [["sf-menu-footer-logo"]], decls: 3, vars: 3, consts: [["class", "sf-menu-logo", "xmlns", "http://www.w3.org/2000/svg", "viewBox", "0 0 1182 263", 4, "ngIf"], ["class", "sf-menu-logo", "xmlns", "http://www.w3.org/2000/svg", "viewBox", "0 0 256 256", 4, "ngIf"], ["class", "sf-version-text", 3, "color", 4, "ngIf"], ["xmlns", "http://www.w3.org/2000/svg", "viewBox", "0 0 1182 263", 1, "sf-menu-logo"], ["d", "M55.2,145.2 C74.8,147.2 94.4,150.1 94.4,162.9 C94.4,179.3 70.4,183.8 49.4,183.8 C22,183.8 2.1,178 0.4,160 L11.7,160 C13.7,174 28.8,177.4 49.5,177.4 C65.7,177.4 82.5,174.7 82.5,164.1 C82.5,153.9 62.8,152.7 43.2,150.7 C17.5,148.2 3.7,143.7 3.7,132.9 C3.7,118.3 25.4,113.5 46.3,113.5 C68.3,113.5 90.6,118.4 91.3,134.3 L80.1,134.3 C78.6,122.9 63.6,119.9 46.2,119.9 C28.9,119.9 15.7,124.1 15.7,131.7 C15.7,140.4 28.9,142.5 55.2,145.2 Z M128.3,155.4 L176,155.4 L152.3,122.1 L128.3,155.4 Z M146,115.2 L159,115.2 L208,182.1 L195,182.1 L180.6,161.8 L123.8,161.8 L109.1,182.1 L97,182.1 L146,115.2 Z M217,182.1 L217,115.2 L297.9,115.2 L297.9,121.6 L228.2,121.6 L228.2,144.3 L294.4,144.3 L294.4,150.6 L228.2,150.6 L228.2,182.1 L217,182.1 Z M322.4,150.6 L322.4,175.7 L398.3,175.7 L398.3,182.1 L311.2,182.1 L311.2,115.2 L397.1,115.2 L397.1,121.6 L322.4,121.6 L322.4,144.2 L394.4,144.2 L394.4,150.6 L322.4,150.6 Z M707.1,182.1 L707.1,115.2 L788,115.2 L788,121.6 L718.3,121.6 L718.3,144.3 L784.5,144.3 L784.5,150.6 L718.3,150.6 L718.3,182.1 L707.1,182.1 Z M810.7,115.2 L810.7,175.7 L879.1,175.7 L879.1,182.1 L799.5,182.1 L799.5,115.2 L810.7,115.2 Z M902.1,150.6 L902.1,175.7 L978,175.7 L978,182.1 L890.9,182.1 L890.9,115.2 L976.8,115.2 L976.8,121.6 L902.1,121.6 L902.1,144.2 L974.1,144.2 L974.1,150.6 L902.1,150.6 Z M1005.7,150.6 L1005.7,175.7 L1081.5,175.7 L1081.5,182.1 L994.5,182.1 L994.5,115.2 L1080.3,115.2 L1080.3,121.6 L1005.7,121.6 L1005.7,144.2 L1077.6,144.2 L1077.6,150.6 L1005.7,150.6 Z M1085.9,115.2 L1181.8,115.2 L1181.8,121.6 L1139.4,121.6 L1139.4,182.1 L1128.2,182.1 L1128.2,121.6 L1085.9,121.6 L1085.9,115.2 Z M645.7,172.8 C668.1,124.1 664.4,42.1 664.3,38.6 L664.1,34.3 L659.8,34.3 C595.3,34.3 555.2,9.6 554.8,9.3 L552.4,7.8 L550,9.3 C549.7,9.6 509.7,34.3 445.2,34.3 L440.9,34.3 L440.7,38.6 C440.5,42 436.8,124.1 459.2,172.8 C476.4,210 508.1,237.8 550.9,253.4 L552.4,254 L553.9,253.4 C596.8,237.9 628.5,210 645.7,172.8 Z M433.1,28.5 C507.3,28.5 552.3,0 552.3,0 C552.3,0 597.5,28.5 671.6,28.5 C671.6,28.5 676.1,121.7 651.4,175.3 C634.1,212.6 602.2,244.2 552.2,262.3 C502.4,244.2 470.5,212.6 453.2,175.3 C428.5,121.7 433.1,28.5 433.1,28.5 Z M629.9,76.8 C630.1,70.1 630.1,64.5 630,60.5 L523.9,60.5 C486.9,60.5 479.1,77.1 479.1,91.1 C479.1,101.6 484.9,111 496.3,118.9 L523.5,137.5 C529.9,141.8 532.8,144.4 534.1,145.8 C536.2,147.9 537.2,150 537.2,151.8 C537.2,154.8 537.2,161.2 523.6,161.2 L487,161 C490.3,166.8 494.3,172.4 498.8,177.6 L520.3,177.6 C532.7,177.6 544.6,175.9 551.9,172.5 L551.9,211.1 C562.6,207.2 569.7,202.4 578,196.9 L578,122.7 L625.3,122.7 C626.4,117.1 627.2,111.4 627.9,105.8 L578,105.8 L578,81 C578,78.8 578.4,77.7 578.9,77.4 C578.9,77.3 580.2,76.8 585.1,76.8 L629.9,76.8 Z M552.3,18.4 C561.9,23.8 598.9,42.4 655.4,43.3 C655.8,60.8 656.2,128.5 637.5,169.1 C621.5,203.6 592.1,229.7 552.3,244.5 C512.6,229.6 483.2,203.6 467.3,169.1 C448.6,128.5 449,60.8 449.4,43.3 C505.9,42.4 542.8,23.8 552.3,18.4 Z M507.6,86.1 C507.6,83.1 507.6,76.8 521.2,76.8 L551.9,76.8 L551.9,120 C551.4,119.7 551,119.4 550.5,119.1 L521.1,100.9 C516,97.4 512.5,94.5 510.5,92.4 C508.5,90.3 507.6,88.2 507.6,86.1 Z"], ["xmlns", "http://www.w3.org/2000/svg", "viewBox", "0 0 256 256", 1, "sf-menu-logo"], ["d", "M219.012551,168.99085 C240.789143,121.646207 237.192116,41.9283263 237.0949,38.5257339 L236.900466,34.345406 L232.720138,34.345406 C170.01522,34.345406 131.031232,10.332825 130.642364,10.0411742 L128.309158,8.58292032 L125.975952,10.0411742 C125.684301,10.332825 86.7975301,34.345406 24.0926121,34.345406 L19.9122842,34.345406 L19.7178504,38.5257339 C19.5234165,41.8311094 15.9263902,121.646207 37.7029819,168.99085 C54.4242934,205.155547 85.2420593,232.181853 126.850904,247.347693 L128.309158,247.930995 L129.767412,247.347693 C171.473474,232.27907 202.29124,205.155547 219.012551,168.99085 Z M128.211941,1 C128.211941,1 172.153992,28.7068242 244.191735,28.7068242 C244.191735,28.7068242 248.566497,119.313 224.553916,171.421273 C207.735388,207.683187 176.723188,238.403736 128.114724,256 C79.7006944,238.403736 48.6884947,207.683187 31.8699663,171.421273 C7.85738524,119.313 12.3293639,28.7068242 12.3293639,28.7068242 C84.4643239,28.7068242 128.211941,1 128.211941,1 Z M203.652277,75.6626001 C203.84671,69.149066 203.84671,63.704918 203.749494,59.8162409 L100.602334,59.8162409 C64.6320707,59.8162409 57.0491504,75.9542509 57.0491504,89.5646207 C57.0491504,99.772398 62.6877322,108.910789 73.7704619,116.590926 L100.213466,134.673275 C106.435349,138.853603 109.25464,141.381243 110.51846,142.74228 C112.560016,144.783835 113.532185,146.825391 113.532185,148.575295 C113.532185,151.491803 113.532185,157.713687 100.310683,157.713687 L64.7292876,157.519253 C67.9374462,163.157835 71.8261233,168.601982 76.200885,173.657263 L97.1025244,173.657263 C109.157423,173.657263 120.726238,172.004575 127.823073,168.699199 L127.823073,206.224933 C138.225285,202.433473 145.127686,197.767061 153.196691,192.42013 L153.196691,120.28517 L199.180298,120.28517 C200.249684,114.841022 201.02742,109.299657 201.707938,103.855509 L153.196691,103.855509 L153.196691,79.745711 C153.196691,77.6069386 153.585559,76.5375524 154.071644,76.2459016 C154.071644,76.1486847 155.335464,75.6626001 160.099093,75.6626001 L203.652277,75.6626001 Z M128.211941,18.8879146 C137.544766,24.1376287 173.515029,42.2199771 228.442593,43.0949295 C228.831461,60.1078917 229.220328,125.923751 211.040763,165.393824 C195.486055,198.933664 166.904278,224.307282 128.211941,238.695387 C89.616821,224.210065 61.0350444,198.933664 45.577553,165.393824 C27.3979876,125.923751 27.7868553,60.1078917 28.175723,43.0949295 C83.1032869,42.2199771 118.976333,24.1376287 128.211941,18.8879146 Z M84.7559746,84.7037743 C84.7559746,81.7872665 84.7559746,75.6626001 97.9774767,75.6626001 L127.823073,75.6626001 L127.823073,117.660313 C127.336989,117.368662 126.948121,117.077011 126.462036,116.78536 L97.8802598,99.0918795 C92.9221965,95.6892871 89.5196041,92.8699962 87.5752655,90.8284407 C85.630927,88.7868852 84.7559746,86.7453298 84.7559746,84.7037743 Z"], [1, "sf-version-text"]], template: function MenuFooterLogoComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, MenuFooterLogoComponent__svg_svg_0_Template, 2, 2, "svg", 0);
        ɵngcc0.ɵɵtemplate(1, MenuFooterLogoComponent__svg_svg_1_Template, 2, 2, "svg", 1);
        ɵngcc0.ɵɵtemplate(2, MenuFooterLogoComponent_div_2_Template, 2, 3, "div", 2);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.collapsed);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.collapsed);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.collapsed && ctx.menuConfig.versionText);
    } }, directives: [ɵngcc2.NgIf], styles: ["[_nghost-%COMP%]{display:flex;flex-direction:column;justify-content:center;align-items:center;padding:15px 5px}[_nghost-%COMP%]   .sf-menu-logo[_ngcontent-%COMP%]{-o-object-fit:contain;object-fit:contain;text-align:center;align-self:center;justify-self:center;height:auto;width:50%;max-width:90%;height:2em}[_nghost-%COMP%]   .sf-menu-logo-bg[_ngcontent-%COMP%]{width:100%;height:4em}@media screen and (min-width:768px){[_nghost-%COMP%]   .sf-menu-logo-bg[_ngcontent-%COMP%]{width:65%}}[_nghost-%COMP%]   .sf-version-text[_ngcontent-%COMP%]{font-size:.7em;padding:10px 10px 0;text-align:center}@media screen and (min-width:768px){.collapsed[_nghost-%COMP%]   .sf-menu-logo[_ngcontent-%COMP%]{max-width:100%;width:100%}}"], data: { animation: [
            trigger('appear-fast', [transition(':enter', [style({ opacity: 0 }), animate('80ms 120ms ease-in', style({ opacity: 1 }))])])
        ] } });
MenuFooterLogoComponent.ctorParameters = () => [
    { type: SfMenuService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MenuFooterLogoComponent, [{
        type: Component,
        args: [{
                selector: 'sf-menu-footer-logo',
                template: "<svg class=\"sf-menu-logo\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 1182 263\" [attr.fill]=\"color\" *ngIf=\"!collapsed\" [@appear-fast]>\n  <path\n    d=\"M55.2,145.2 C74.8,147.2 94.4,150.1 94.4,162.9 C94.4,179.3 70.4,183.8 49.4,183.8 C22,183.8 2.1,178 0.4,160 L11.7,160 C13.7,174 28.8,177.4 49.5,177.4 C65.7,177.4 82.5,174.7 82.5,164.1 C82.5,153.9 62.8,152.7 43.2,150.7 C17.5,148.2 3.7,143.7 3.7,132.9 C3.7,118.3 25.4,113.5 46.3,113.5 C68.3,113.5 90.6,118.4 91.3,134.3 L80.1,134.3 C78.6,122.9 63.6,119.9 46.2,119.9 C28.9,119.9 15.7,124.1 15.7,131.7 C15.7,140.4 28.9,142.5 55.2,145.2 Z M128.3,155.4 L176,155.4 L152.3,122.1 L128.3,155.4 Z M146,115.2 L159,115.2 L208,182.1 L195,182.1 L180.6,161.8 L123.8,161.8 L109.1,182.1 L97,182.1 L146,115.2 Z M217,182.1 L217,115.2 L297.9,115.2 L297.9,121.6 L228.2,121.6 L228.2,144.3 L294.4,144.3 L294.4,150.6 L228.2,150.6 L228.2,182.1 L217,182.1 Z M322.4,150.6 L322.4,175.7 L398.3,175.7 L398.3,182.1 L311.2,182.1 L311.2,115.2 L397.1,115.2 L397.1,121.6 L322.4,121.6 L322.4,144.2 L394.4,144.2 L394.4,150.6 L322.4,150.6 Z M707.1,182.1 L707.1,115.2 L788,115.2 L788,121.6 L718.3,121.6 L718.3,144.3 L784.5,144.3 L784.5,150.6 L718.3,150.6 L718.3,182.1 L707.1,182.1 Z M810.7,115.2 L810.7,175.7 L879.1,175.7 L879.1,182.1 L799.5,182.1 L799.5,115.2 L810.7,115.2 Z M902.1,150.6 L902.1,175.7 L978,175.7 L978,182.1 L890.9,182.1 L890.9,115.2 L976.8,115.2 L976.8,121.6 L902.1,121.6 L902.1,144.2 L974.1,144.2 L974.1,150.6 L902.1,150.6 Z M1005.7,150.6 L1005.7,175.7 L1081.5,175.7 L1081.5,182.1 L994.5,182.1 L994.5,115.2 L1080.3,115.2 L1080.3,121.6 L1005.7,121.6 L1005.7,144.2 L1077.6,144.2 L1077.6,150.6 L1005.7,150.6 Z M1085.9,115.2 L1181.8,115.2 L1181.8,121.6 L1139.4,121.6 L1139.4,182.1 L1128.2,182.1 L1128.2,121.6 L1085.9,121.6 L1085.9,115.2 Z M645.7,172.8 C668.1,124.1 664.4,42.1 664.3,38.6 L664.1,34.3 L659.8,34.3 C595.3,34.3 555.2,9.6 554.8,9.3 L552.4,7.8 L550,9.3 C549.7,9.6 509.7,34.3 445.2,34.3 L440.9,34.3 L440.7,38.6 C440.5,42 436.8,124.1 459.2,172.8 C476.4,210 508.1,237.8 550.9,253.4 L552.4,254 L553.9,253.4 C596.8,237.9 628.5,210 645.7,172.8 Z M433.1,28.5 C507.3,28.5 552.3,0 552.3,0 C552.3,0 597.5,28.5 671.6,28.5 C671.6,28.5 676.1,121.7 651.4,175.3 C634.1,212.6 602.2,244.2 552.2,262.3 C502.4,244.2 470.5,212.6 453.2,175.3 C428.5,121.7 433.1,28.5 433.1,28.5 Z M629.9,76.8 C630.1,70.1 630.1,64.5 630,60.5 L523.9,60.5 C486.9,60.5 479.1,77.1 479.1,91.1 C479.1,101.6 484.9,111 496.3,118.9 L523.5,137.5 C529.9,141.8 532.8,144.4 534.1,145.8 C536.2,147.9 537.2,150 537.2,151.8 C537.2,154.8 537.2,161.2 523.6,161.2 L487,161 C490.3,166.8 494.3,172.4 498.8,177.6 L520.3,177.6 C532.7,177.6 544.6,175.9 551.9,172.5 L551.9,211.1 C562.6,207.2 569.7,202.4 578,196.9 L578,122.7 L625.3,122.7 C626.4,117.1 627.2,111.4 627.9,105.8 L578,105.8 L578,81 C578,78.8 578.4,77.7 578.9,77.4 C578.9,77.3 580.2,76.8 585.1,76.8 L629.9,76.8 Z M552.3,18.4 C561.9,23.8 598.9,42.4 655.4,43.3 C655.8,60.8 656.2,128.5 637.5,169.1 C621.5,203.6 592.1,229.7 552.3,244.5 C512.6,229.6 483.2,203.6 467.3,169.1 C448.6,128.5 449,60.8 449.4,43.3 C505.9,42.4 542.8,23.8 552.3,18.4 Z M507.6,86.1 C507.6,83.1 507.6,76.8 521.2,76.8 L551.9,76.8 L551.9,120 C551.4,119.7 551,119.4 550.5,119.1 L521.1,100.9 C516,97.4 512.5,94.5 510.5,92.4 C508.5,90.3 507.6,88.2 507.6,86.1 Z\"\n  />\n</svg>\n\n<svg class=\"sf-menu-logo\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 256 256\" [attr.fill]=\"color\" *ngIf=\"collapsed\" [@appear-fast]>\n  <path\n    d=\"M219.012551,168.99085 C240.789143,121.646207 237.192116,41.9283263 237.0949,38.5257339 L236.900466,34.345406 L232.720138,34.345406 C170.01522,34.345406 131.031232,10.332825 130.642364,10.0411742 L128.309158,8.58292032 L125.975952,10.0411742 C125.684301,10.332825 86.7975301,34.345406 24.0926121,34.345406 L19.9122842,34.345406 L19.7178504,38.5257339 C19.5234165,41.8311094 15.9263902,121.646207 37.7029819,168.99085 C54.4242934,205.155547 85.2420593,232.181853 126.850904,247.347693 L128.309158,247.930995 L129.767412,247.347693 C171.473474,232.27907 202.29124,205.155547 219.012551,168.99085 Z M128.211941,1 C128.211941,1 172.153992,28.7068242 244.191735,28.7068242 C244.191735,28.7068242 248.566497,119.313 224.553916,171.421273 C207.735388,207.683187 176.723188,238.403736 128.114724,256 C79.7006944,238.403736 48.6884947,207.683187 31.8699663,171.421273 C7.85738524,119.313 12.3293639,28.7068242 12.3293639,28.7068242 C84.4643239,28.7068242 128.211941,1 128.211941,1 Z M203.652277,75.6626001 C203.84671,69.149066 203.84671,63.704918 203.749494,59.8162409 L100.602334,59.8162409 C64.6320707,59.8162409 57.0491504,75.9542509 57.0491504,89.5646207 C57.0491504,99.772398 62.6877322,108.910789 73.7704619,116.590926 L100.213466,134.673275 C106.435349,138.853603 109.25464,141.381243 110.51846,142.74228 C112.560016,144.783835 113.532185,146.825391 113.532185,148.575295 C113.532185,151.491803 113.532185,157.713687 100.310683,157.713687 L64.7292876,157.519253 C67.9374462,163.157835 71.8261233,168.601982 76.200885,173.657263 L97.1025244,173.657263 C109.157423,173.657263 120.726238,172.004575 127.823073,168.699199 L127.823073,206.224933 C138.225285,202.433473 145.127686,197.767061 153.196691,192.42013 L153.196691,120.28517 L199.180298,120.28517 C200.249684,114.841022 201.02742,109.299657 201.707938,103.855509 L153.196691,103.855509 L153.196691,79.745711 C153.196691,77.6069386 153.585559,76.5375524 154.071644,76.2459016 C154.071644,76.1486847 155.335464,75.6626001 160.099093,75.6626001 L203.652277,75.6626001 Z M128.211941,18.8879146 C137.544766,24.1376287 173.515029,42.2199771 228.442593,43.0949295 C228.831461,60.1078917 229.220328,125.923751 211.040763,165.393824 C195.486055,198.933664 166.904278,224.307282 128.211941,238.695387 C89.616821,224.210065 61.0350444,198.933664 45.577553,165.393824 C27.3979876,125.923751 27.7868553,60.1078917 28.175723,43.0949295 C83.1032869,42.2199771 118.976333,24.1376287 128.211941,18.8879146 Z M84.7559746,84.7037743 C84.7559746,81.7872665 84.7559746,75.6626001 97.9774767,75.6626001 L127.823073,75.6626001 L127.823073,117.660313 C127.336989,117.368662 126.948121,117.077011 126.462036,116.78536 L97.8802598,99.0918795 C92.9221965,95.6892871 89.5196041,92.8699962 87.5752655,90.8284407 C85.630927,88.7868852 84.7559746,86.7453298 84.7559746,84.7037743 Z\"\n  />\n</svg>\n\n<div class=\"sf-version-text\" *ngIf=\"!collapsed && menuConfig.versionText\" [style.color]=\"text\">\n  {{ menuConfig.versionText }}\n</div>\n",
                animations: [
                    trigger('appear-fast', [transition(':enter', [style({ opacity: 0 }), animate('80ms 120ms ease-in', style({ opacity: 1 }))])])
                ],
                styles: [":host{display:flex;flex-direction:column;justify-content:center;align-items:center;padding:15px 5px}:host .sf-menu-logo{-o-object-fit:contain;object-fit:contain;text-align:center;align-self:center;justify-self:center;height:auto;width:50%;max-width:90%;height:2em}:host .sf-menu-logo-bg{width:100%;height:4em}@media screen and (min-width:768px){:host .sf-menu-logo-bg{width:65%}}:host .sf-version-text{font-size:.7em;padding:10px 10px 0;text-align:center}@media screen and (min-width:768px){:host.collapsed .sf-menu-logo{max-width:100%;width:100%}}"]
            }]
    }], function () { return [{ type: SfMenuService }]; }, null); })();

class MenuItem {
    constructor() {
        this.newTab = false;
        this.newWindow = false;
        this.expanded = false;
        this.active = false;
    }
}

class MenuItemComponent {
    constructor(svc) {
        this.svc = svc;
        this.collapsed = false;
        this.level = 1;
    }
    ngOnInit() {
        this.text = this.svc.colors.gray900;
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            this.menuConfig = cfg;
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
        });
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
    click(evt) {
        switch (this.itemType) {
            case MenuEventType.ShortcutClick:
                this.shortcutClicked();
                break;
            case MenuEventType.UserActionClick:
                this.userOptionClicked();
                break;
            case MenuEventType.RouteClick:
                this.svc.routeClicked(this.item, evt);
                break;
            default:
                console.error('event not found');
        }
    }
    shortcutClicked() {
        this.svc.events$.next({
            type: MenuEventType.ShortcutClick,
            value: this.item
        });
        if (window.innerWidth <= this.svc.SMALL_RESOLUTION && !this.collapsed) {
            this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
        }
    }
    userOptionClicked() {
        this.svc.events$.next({
            type: MenuEventType.UserActionClick,
            value: this.item
        });
        if (window.innerWidth <= this.svc.SMALL_RESOLUTION && !this.collapsed) {
            this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
        }
    }
}
MenuItemComponent.ɵfac = function MenuItemComponent_Factory(t) { return new (t || MenuItemComponent)(ɵngcc0.ɵɵdirectiveInject(SfMenuService)); };
MenuItemComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: MenuItemComponent, selectors: [["", "sf-menu-item", ""]], inputs: { level: "level", item: "item", itemType: "itemType" }, attrs: _c4, decls: 7, vars: 18, consts: [[1, "li-container"], [1, "main-div", "menu-item", 3, "ngStyle", "title", "click"], ["class", "icon-container", 4, "ngIf"], [1, "fas", "fa-chevron-down"], ["class", "item-children", 4, "ngIf"], [1, "icon-container"], ["class", "sf-icons-menu", 3, "collapsed", 4, "ngIf"], ["class", "sf-menu-logo-bg", 3, "ngStyle", 4, "ngIf"], [1, "sf-icons-menu"], [3, "ngClass"], [1, "sf-menu-logo-bg", 3, "ngStyle"], [1, "item-children"], ["sf-menu-item", "", 3, "show-arrow", "active-item", "item", "level", "itemType", 4, "ngFor", "ngForOf"], ["sf-menu-item", "", 3, "item", "level", "itemType"]], template: function MenuItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0)(1, "div", 1);
        ɵngcc0.ɵɵlistener("click", function MenuItemComponent_Template_div_click_1_listener($event) { return ctx.click($event); });
        ɵngcc0.ɵɵtemplate(2, MenuItemComponent_div_2_Template, 3, 2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "span");
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(5, "em", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(6, MenuItemComponent_ul_6_Template, 2, 1, "ul", 4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("color", ctx.text);
        ɵngcc0.ɵɵclassProp("show-arrow", ctx.item.items)("active-item", ctx.item.active)("collapsed", ctx.collapsed);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵclassProp("expanded-item", ctx.item.expanded && ctx.item.items);
        ɵngcc0.ɵɵpropertyInterpolate("title", ctx.menuConfig.tooltips || ctx.collapsed ? ctx.item.text : "");
        ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction2(15, _c5, ctx.menuConfig == null ? null : ctx.menuConfig.borderColor, 10 * ctx.level + "px"));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.item.iconClass || ctx.item.iconPath);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.item.text);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.item.items);
    } }, directives: [ɵngcc2.NgStyle, ɵngcc2.NgIf, ɵngcc2.NgClass, ɵngcc2.NgForOf, MenuItemComponent], styles: ["[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%]{display:inline-block;width:100%;padding:0;overflow:hidden;cursor:pointer}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%]   .icon-container[_ngcontent-%COMP%]{width:24px}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > div.main-div[_ngcontent-%COMP%]{margin-left:2px}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > .menu-item[_ngcontent-%COMP%]{display:flex;align-items:center;justify-content:flex-start;padding:7px 10px;position:relative;min-height:38px}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > .menu-item[_ngcontent-%COMP%]   .sf-menu-logo-bg[_ngcontent-%COMP%]{width:1.6em;height:1.55em}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > .menu-item[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{width:calc(100% - 24px);min-width:100px;padding:0 10px;font-size:14px;line-height:18px;opacity:.71;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > .menu-item[_ngcontent-%COMP%]:hover{position:relative}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > .menu-item[_ngcontent-%COMP%]:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > .menu-item[_ngcontent-%COMP%]:hover   span[_ngcontent-%COMP%]{opacity:1}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > .menu-item[_ngcontent-%COMP%]   .fa-chevron-down[_ngcontent-%COMP%]{display:none;position:absolute;right:15px;top:16px;width:auto;height:auto;font-size:12px}[_nghost-%COMP%]   .li-container.show-arrow[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]   .fa-chevron-down[_ngcontent-%COMP%]{display:inline-block;transform:rotate(0deg)}[_nghost-%COMP%]   .li-container.show-arrow[_ngcontent-%COMP%] > div.expanded-item[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{opacity:1}[_nghost-%COMP%]   .li-container.show-arrow[_ngcontent-%COMP%] > div.expanded-item[_ngcontent-%COMP%]   em.fa-chevron-down[_ngcontent-%COMP%], [_nghost-%COMP%]   .li-container.show-arrow[_ngcontent-%COMP%] > div.expanded-item[_ngcontent-%COMP%]   i.fa-chevron-down[_ngcontent-%COMP%]{transform:rotate(180deg)}[_nghost-%COMP%]   .li-container.show-arrow[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]:hover{position:relative}[_nghost-%COMP%]   .li-container.show-arrow[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}[_nghost-%COMP%]   .li-container.show-arrow[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]:hover   span[_ngcontent-%COMP%]{opacity:1}[_nghost-%COMP%]   .li-container.active-item[_ngcontent-%COMP%] > div.main-div[_ngcontent-%COMP%]{border-left:3px solid;margin-left:0}[_nghost-%COMP%]   .li-container.active-item[_ngcontent-%COMP%] > div.main-div[_ngcontent-%COMP%]:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.2}[_nghost-%COMP%]   .li-container.active-item[_ngcontent-%COMP%] > div.main-div[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{opacity:1}[_nghost-%COMP%]   .li-container.collapsed[_ngcontent-%COMP%]   div[_ngcontent-%COMP%]{justify-content:center}[_nghost-%COMP%]   .li-container.collapsed[_ngcontent-%COMP%]   em[_ngcontent-%COMP%], [_nghost-%COMP%]   .li-container.collapsed[_ngcontent-%COMP%]   i[_ngcontent-%COMP%]{width:auto}[_nghost-%COMP%]   .li-container.collapsed[_ngcontent-%COMP%]   .item-children[_ngcontent-%COMP%], [_nghost-%COMP%]   .li-container.collapsed[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{display:none}[_nghost-%COMP%]   .sf-icons-menu[_ngcontent-%COMP%]{text-align:center}[_nghost-%COMP%]   .sf-icons-menu[_ngcontent-%COMP%]   em[_ngcontent-%COMP%], [_nghost-%COMP%]   .sf-icons-menu[_ngcontent-%COMP%]   i[_ngcontent-%COMP%]{font-size:16px}[_nghost-%COMP%]   ul[_ngcontent-%COMP%]{list-style:none;padding-left:0;width:100%}.expanded-item[_ngcontent-%COMP%] + .item-children[_ngcontent-%COMP%]{display:block}ul.item-children[_ngcontent-%COMP%]{display:none}.sf-icons-menu.collapsed[_ngcontent-%COMP%]   em[_ngcontent-%COMP%]{font-size:var(--fs-300)}.sf-menu-logo-bg[_ngcontent-%COMP%]{background:50%/contain no-repeat}"] });
MenuItemComponent.ctorParameters = () => [
    { type: SfMenuService }
];
MenuItemComponent.propDecorators = {
    item: [{ type: Input }],
    level: [{ type: Input }],
    itemType: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MenuItemComponent, [{
        type: Component,
        args: [{
                // tslint:disable-next-line: component-selector
                selector: '[sf-menu-item]',
                template: "<div\n  class=\"li-container\"\n  [class.show-arrow]=\"item.items\"\n  [class.active-item]=\"item.active\"\n  [class.collapsed]=\"collapsed\"\n  [style.color]=\"text\"\n>\n  <div\n    [class.expanded-item]=\"item.expanded && item.items\"\n    class=\"main-div menu-item\"\n    [ngStyle]=\"{ 'border-left-color': menuConfig?.borderColor, 'padding-left': 10 * level + 'px' }\"\n    title=\"{{ menuConfig.tooltips || collapsed ? item.text : '' }}\"\n    (click)=\"click($event)\"\n  >\n    <div class=\"icon-container\" *ngIf=\"item.iconClass || item.iconPath\">\n      <div class=\"sf-icons-menu\" *ngIf=\"item.iconClass\" [class.collapsed]=\"collapsed\">\n        <em [ngClass]=\"item.iconClass\"></em>\n      </div>\n      <div\n        class=\"sf-menu-logo-bg\"\n        *ngIf=\"!item.iconClass && item.iconPath\"\n        [ngStyle]=\"{\n          'background-image': 'url(' + item.iconPath + ')'\n        }\"\n      ></div>\n    </div>\n    <span>{{ item.text }}</span>\n    <em class=\"fas fa-chevron-down\"></em>\n  </div>\n  <ul *ngIf=\"item.items\" class=\"item-children\">\n    <li\n      sf-menu-item\n      *ngFor=\"let child of item.items\"\n      [class.show-arrow]=\"child.items\"\n      [class.active-item]=\"child.active\"\n      [item]=\"child\"\n      [level]=\"level + 1\"\n      [itemType]=\"itemType\"\n    ></li>\n  </ul>\n</div>\n",
                styles: [":host .li-container{display:inline-block;width:100%;padding:0;overflow:hidden;cursor:pointer}:host .li-container .icon-container{width:24px}:host .li-container>div.main-div{margin-left:2px}:host .li-container>.menu-item{display:flex;align-items:center;justify-content:flex-start;padding:7px 10px;position:relative;min-height:38px}:host .li-container>.menu-item .sf-menu-logo-bg{width:1.6em;height:1.55em}:host .li-container>.menu-item span{width:calc(100% - 24px);min-width:100px;padding:0 10px;font-size:14px;line-height:18px;opacity:.71;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}:host .li-container>.menu-item:hover{position:relative}:host .li-container>.menu-item:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}:host .li-container>.menu-item:hover span{opacity:1}:host .li-container>.menu-item .fa-chevron-down{display:none;position:absolute;right:15px;top:16px;width:auto;height:auto;font-size:12px}:host .li-container.show-arrow>div .fa-chevron-down{display:inline-block;transform:rotate(0deg)}:host .li-container.show-arrow>div.expanded-item span{opacity:1}:host .li-container.show-arrow>div.expanded-item em.fa-chevron-down,:host .li-container.show-arrow>div.expanded-item i.fa-chevron-down{transform:rotate(180deg)}:host .li-container.show-arrow>div:hover{position:relative}:host .li-container.show-arrow>div:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}:host .li-container.show-arrow>div:hover span{opacity:1}:host .li-container.active-item>div.main-div{border-left:3px solid;margin-left:0}:host .li-container.active-item>div.main-div:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.2}:host .li-container.active-item>div.main-div span{opacity:1}:host .li-container.collapsed div{justify-content:center}:host .li-container.collapsed em,:host .li-container.collapsed i{width:auto}:host .li-container.collapsed .item-children,:host .li-container.collapsed span{display:none}:host .sf-icons-menu{text-align:center}:host .sf-icons-menu em,:host .sf-icons-menu i{font-size:16px}:host ul{list-style:none;padding-left:0;width:100%}.expanded-item+.item-children{display:block}ul.item-children{display:none}.sf-icons-menu.collapsed em{font-size:var(--fs-300)}.sf-menu-logo-bg{background:50%/contain no-repeat}"]
            }]
    }], function () { return [{ type: SfMenuService }]; }, { level: [{
            type: Input
        }], item: [{
            type: Input
        }], itemType: [{
            type: Input
        }] }); })();

class MenuProductsComponent {
    constructor(svc) {
        this.svc = svc;
        this.collapsed = false;
        this.productsExpanded = false;
    }
    get clr() {
        return this.border;
    }
    ngOnInit() {
        this.border = this.svc.colors.gray800;
        this.chevron = this.svc.colors.gray900;
        this.text = this.svc.colors.gray800;
        this.productsBg = this.svc.colors.gray100;
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            this.menuConfig = cfg;
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
        });
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
    currentProductClicked() {
        this.productsExpanded = !this.productsExpanded;
    }
    productClicked(item, ev) {
        this.svc.events$.next({
            type: MenuEventType.ProductClick,
            value: item
        });
        if (window.innerWidth <= this.svc.SMALL_RESOLUTION && !this.collapsed) {
            this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
        }
        window.location.href = item.routeHref;
    }
}
MenuProductsComponent.ɵfac = function MenuProductsComponent_Factory(t) { return new (t || MenuProductsComponent)(ɵngcc0.ɵɵdirectiveInject(SfMenuService)); };
MenuProductsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: MenuProductsComponent, selectors: [["sf-menu-products"]], hostVars: 2, hostBindings: function MenuProductsComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("border-bottom-color", ctx.clr);
    } }, decls: 3, vars: 3, consts: [["class", "sf-menu-product", 3, "collapsed", "click", 4, "ngIf"], ["class", "sf-customer-title", 3, "color", 4, "ngIf"], ["class", "sf-menu-products-list", 3, "background-color", "expanded", 4, "ngIf"], [1, "sf-menu-product", 3, "click"], [3, "ngClass", 4, "ngIf"], ["class", "sf-menu-logo-bg", 3, "ngStyle", 4, "ngIf"], ["class", "fas fa-chevron-down", "data-testid", "menu-product-display-products-icon", 3, "color", "expanded", "collapsed", 4, "ngIf"], [3, "ngClass"], [1, "sf-menu-logo-bg", 3, "ngStyle"], ["data-testid", "menu-product-display-products-icon", 1, "fas", "fa-chevron-down"], [1, "sf-customer-title"], [4, "ngIf"], [1, "sf-menu-products-list"], ["class", "option", "data-testid", "menu-product-option-link", 3, "collapsed", "two-columns", "click", 4, "ngFor", "ngForOf"], ["data-testid", "menu-product-option-link", 1, "option", 3, "click"]], template: function MenuProductsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, MenuProductsComponent_div_0_Template, 4, 5, "div", 0);
        ɵngcc0.ɵɵtemplate(1, MenuProductsComponent_div_1_Template, 2, 3, "div", 1);
        ɵngcc0.ɵɵtemplate(2, MenuProductsComponent_div_2_Template, 2, 5, "div", 2);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.menuConfig.currentProduct);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.menuConfig.hideProductName && !ctx.collapsed);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.menuConfig.products && ctx.menuConfig.products.length > 0);
    } }, directives: [ɵngcc2.NgIf, ɵngcc2.NgClass, ɵngcc2.NgStyle, ɵngcc2.NgForOf], styles: ["[_nghost-%COMP%]{position:relative;width:calc(100% - 2px);height:auto;display:flex;flex-direction:column;flex-wrap:wrap;align-items:center;padding:10px 0 0}.sf-menu-product[_ngcontent-%COMP%]{width:76px;height:76px;flex:0 0 76px;padding:5px;position:relative;cursor:pointer;margin-bottom:10px}.sf-menu-product[_ngcontent-%COMP%]   em[_ngcontent-%COMP%], .sf-menu-product[_ngcontent-%COMP%]   i[_ngcontent-%COMP%], .sf-menu-product[_ngcontent-%COMP%]   img[_ngcontent-%COMP%]{-o-object-fit:contain;object-fit:contain;text-align:center;align-self:center;justify-self:center;height:auto;width:50%;width:100%;height:100%;background-size:cover}.sf-menu-product[_ngcontent-%COMP%]   .sf-menu-logo-bg[_ngcontent-%COMP%]{height:100%;width:100%}.sf-menu-product[_ngcontent-%COMP%]   .fa-chevron-down[_ngcontent-%COMP%]{position:absolute;width:auto;height:auto;top:calc(50% - 7px);right:-20px;transition:transform .15s ease-out;font-size:13px}.sf-menu-product[_ngcontent-%COMP%]   .fa-chevron-down.expanded[_ngcontent-%COMP%]{transform:rotate(180deg)}.sf-menu-product[_ngcontent-%COMP%]   .fa-chevron-down.collapsed[_ngcontent-%COMP%]{right:calc(50% - 5px);top:calc(100% - 2px);font-size:10px}.sf-menu-product.collapsed[_ngcontent-%COMP%]{flex:0 0 60px;margin:20px 0 10px;width:60px;height:60px}.sf-menu-products-list[_ngcontent-%COMP%]{height:auto;max-height:0;width:100%;overflow:hidden;display:flex;flex-wrap:wrap;justify-content:space-around;transition:max-height .24s ease-in-out;opacity:.7}.sf-menu-products-list.expanded[_ngcontent-%COMP%]{max-height:304px}.sf-menu-products-list[_ngcontent-%COMP%]   .option[_ngcontent-%COMP%]{width:62px;height:62px;padding:5px;cursor:pointer}.sf-menu-products-list[_ngcontent-%COMP%]   .option[_ngcontent-%COMP%]:hover   .sf-menu-logo-bg[_ngcontent-%COMP%]{filter:brightness(1.2)}.sf-menu-products-list[_ngcontent-%COMP%]   .option.collapsed[_ngcontent-%COMP%]{flex:0 0 50px;height:55.8px}.sf-menu-products-list[_ngcontent-%COMP%]   .option.two-columns[_ngcontent-%COMP%]{width:80.6px}.sf-menu-products-list[_ngcontent-%COMP%]   .option.two-columns[_ngcontent-%COMP%]:before{content:\" \";height:1px;width:86px;position:absolute;bottom:-1px;left:14px}.sf-menu-products-list[_ngcontent-%COMP%]   .option[_ngcontent-%COMP%]   em[_ngcontent-%COMP%], .sf-menu-products-list[_ngcontent-%COMP%]   .option[_ngcontent-%COMP%]   i[_ngcontent-%COMP%], .sf-menu-products-list[_ngcontent-%COMP%]   .option[_ngcontent-%COMP%]   img[_ngcontent-%COMP%]{-o-object-fit:contain;object-fit:contain;text-align:center;align-self:center;justify-self:center;height:auto;width:50%;width:100%;height:100%;transition:background-color .12s,transform}.sf-menu-products-list[_ngcontent-%COMP%]   .option[_ngcontent-%COMP%]   .sf-menu-logo-bg[_ngcontent-%COMP%]{width:100%;height:100%}.sf-menu-products-list[_ngcontent-%COMP%]   .option[_ngcontent-%COMP%]:hover   .logo-bg[_ngcontent-%COMP%]{background-color:#3a3b3e;transform:scale(1.15)}.sf-menu-products-list[_ngcontent-%COMP%]   .option[_ngcontent-%COMP%]:after{content:\" \";background-color:#35353b;height:46px;width:1px;position:absolute;top:15px;right:-1px}.sf-customer-title[_ngcontent-%COMP%]{margin:10px 0}.sf-customer-title[_ngcontent-%COMP%]   h4[_ngcontent-%COMP%]{margin:0;font-size:1em;opacity:.6;text-align:center}.sf-menu-logo-bg[_ngcontent-%COMP%]{background:50%/contain no-repeat}"] });
MenuProductsComponent.ctorParameters = () => [
    { type: SfMenuService }
];
MenuProductsComponent.propDecorators = {
    clr: [{ type: HostBinding, args: ['style.border-bottom-color',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MenuProductsComponent, [{
        type: Component,
        args: [{
                selector: 'sf-menu-products',
                template: "<div class=\"sf-menu-product\" *ngIf=\"menuConfig.currentProduct\" (click)=\"currentProductClicked()\" [class.collapsed]=\"collapsed\">\n  <em [ngClass]=\"menuConfig.currentProduct.iconClass\" *ngIf=\"menuConfig.currentProduct.iconClass\"></em>\n  <div\n    class=\"sf-menu-logo-bg\"\n    *ngIf=\"!menuConfig.currentProduct.iconClass && menuConfig.currentProduct.iconPath\"\n    [ngStyle]=\"{\n      'background-image': 'url(' + menuConfig.currentProduct.iconPath + ')'\n    }\"\n  ></div>\n  <em\n    class=\"fas fa-chevron-down\"\n    *ngIf=\"menuConfig.products && menuConfig.products.length > 0\"\n    [style.color]=\"chevron\"\n    [class.expanded]=\"productsExpanded\"\n    [class.collapsed]=\"collapsed\"\n    data-testid=\"menu-product-display-products-icon\"\n  ></em>\n</div>\n\n<div class=\"sf-customer-title\" *ngIf=\"!menuConfig.hideProductName && !collapsed\" [style.color]=\"text\">\n  <h4 *ngIf=\"menuConfig.currentProduct?.text\">\n    {{ menuConfig.currentProduct.text }}\n  </h4>\n</div>\n\n<div\n  class=\"sf-menu-products-list\"\n  [style.background-color]=\"productsBg\"\n  [class.expanded]=\"productsExpanded\"\n  *ngIf=\"menuConfig.products && menuConfig.products.length > 0\"\n>\n  <a\n    class=\"option\"\n    [class.collapsed]=\"collapsed\"\n    [class.two-columns]=\"menuConfig.products.length > 3\"\n    *ngFor=\"let option of menuConfig.products\"\n    (click)=\"productClicked(option, $event)\"\n    [attr.href]=\"option.routeHref\"\n    data-testid=\"menu-product-option-link\"\n  >\n    <em [ngClass]=\"option.iconClass\" *ngIf=\"option.iconClass\"></em>\n    <div\n      class=\"sf-menu-logo-bg\"\n      *ngIf=\"!option.iconClass && option.iconPath\"\n      [ngStyle]=\"{ 'background-image': 'url(' + option.iconPath + ')' }\"\n    ></div>\n  </a>\n</div>\n",
                styles: [":host{position:relative;width:calc(100% - 2px);height:auto;display:flex;flex-direction:column;flex-wrap:wrap;align-items:center;padding:10px 0 0}.sf-menu-product{width:76px;height:76px;flex:0 0 76px;padding:5px;position:relative;cursor:pointer;margin-bottom:10px}.sf-menu-product em,.sf-menu-product i,.sf-menu-product img{-o-object-fit:contain;object-fit:contain;text-align:center;align-self:center;justify-self:center;height:auto;width:50%;width:100%;height:100%;background-size:cover}.sf-menu-product .sf-menu-logo-bg{height:100%;width:100%}.sf-menu-product .fa-chevron-down{position:absolute;width:auto;height:auto;top:calc(50% - 7px);right:-20px;transition:transform .15s ease-out;font-size:13px}.sf-menu-product .fa-chevron-down.expanded{transform:rotate(180deg)}.sf-menu-product .fa-chevron-down.collapsed{right:calc(50% - 5px);top:calc(100% - 2px);font-size:10px}.sf-menu-product.collapsed{flex:0 0 60px;margin:20px 0 10px;width:60px;height:60px}.sf-menu-products-list{height:auto;max-height:0;width:100%;overflow:hidden;display:flex;flex-wrap:wrap;justify-content:space-around;transition:max-height .24s ease-in-out;opacity:.7}.sf-menu-products-list.expanded{max-height:304px}.sf-menu-products-list .option{width:62px;height:62px;padding:5px;cursor:pointer}.sf-menu-products-list .option:hover .sf-menu-logo-bg{filter:brightness(1.2)}.sf-menu-products-list .option.collapsed{flex:0 0 50px;height:55.8px}.sf-menu-products-list .option.two-columns{width:80.6px}.sf-menu-products-list .option.two-columns:before{content:\" \";height:1px;width:86px;position:absolute;bottom:-1px;left:14px}.sf-menu-products-list .option em,.sf-menu-products-list .option i,.sf-menu-products-list .option img{-o-object-fit:contain;object-fit:contain;text-align:center;align-self:center;justify-self:center;height:auto;width:50%;width:100%;height:100%;transition:background-color .12s,transform}.sf-menu-products-list .option .sf-menu-logo-bg{width:100%;height:100%}.sf-menu-products-list .option:hover .logo-bg{background-color:#3a3b3e;transform:scale(1.15)}.sf-menu-products-list .option:after{content:\" \";background-color:#35353b;height:46px;width:1px;position:absolute;top:15px;right:-1px}.sf-customer-title{margin:10px 0}.sf-customer-title h4{margin:0;font-size:1em;opacity:.6;text-align:center}.sf-menu-logo-bg{background:50%/contain no-repeat}"]
            }]
    }], function () { return [{ type: SfMenuService }]; }, { clr: [{
            type: HostBinding,
            args: ['style.border-bottom-color']
        }] }); })();

class MenuRoutesComponent {
    constructor(svc) {
        this.svc = svc;
        this.collapsed = false;
    }
    ngOnInit() {
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            this.menuConfig = cfg;
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
        });
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
    routeClicked(item, evt) {
        this.svc.routeClicked(item, evt);
    }
    showText(index) {
        return index.indexOf('_') > -1;
    }
}
MenuRoutesComponent.ɵfac = function MenuRoutesComponent_Factory(t) { return new (t || MenuRoutesComponent)(ɵngcc0.ɵɵdirectiveInject(SfMenuService)); };
MenuRoutesComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: MenuRoutesComponent, selectors: [["sf-menu-routes"]], decls: 2, vars: 2, consts: [[4, "ngIf"], [3, "items", "vertical", 4, "ngIf"], ["sf-menu-item", "", "data-testid", "menu-route-item", 3, "show-arrow", "active-item", "item", "level", "itemType", 4, "ngFor", "ngForOf"], ["sf-menu-item", "", "data-testid", "menu-route-item", 3, "item", "level", "itemType"], [3, "items", "vertical"], ["kendoMenuItemLinkTemplate", ""], [1, "sf-menu-item-wrapper", 3, "ngStyle", "title", "click"], [1, "icon-container"], [3, "ngClass", 4, "ngIf"], ["class", "sf-menu-logo-bg", 3, "ngStyle", 4, "ngIf"], ["class", "fas fa-chevron-right", 4, "ngIf"], [3, "ngClass"], [1, "sf-menu-logo-bg", 3, "ngStyle"], [1, "fas", "fa-chevron-right"]], template: function MenuRoutesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, MenuRoutesComponent_ul_0_Template, 2, 2, "ul", 0);
        ɵngcc0.ɵɵtemplate(1, MenuRoutesComponent_kendo_menu_1_Template, 2, 2, "kendo-menu", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.collapsed);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", (ctx.menuConfig == null ? null : ctx.menuConfig.routes) && ctx.collapsed);
    } }, directives: [ɵngcc2.NgIf, ɵngcc2.NgForOf, MenuItemComponent, ɵngcc4.MenuComponent, ɵngcc4.ItemLinkTemplateDirective, ɵngcc2.NgStyle, ɵngcc2.NgClass], styles: ["[_nghost-%COMP%]{display:flex;flex-direction:column;width:100%;height:auto;align-items:flex-start;justify-content:flex-start;overflow:hidden;position:relative}[_nghost-%COMP%]:before{content:\" \";width:100%;height:1px;background-color:#fff;opacity:.15}[_nghost-%COMP%]   ul[_ngcontent-%COMP%]{list-style:none;padding-left:0}[_nghost-%COMP%]   ul[_ngcontent-%COMP%], kendo-menu[_ngcontent-%COMP%]{width:100%}.sf-menu-item-wrapper[_ngcontent-%COMP%]{width:100%;height:44px;max-height:75px;overflow:hidden;background-color:transparent;display:flex;justify-content:center;align-items:center}.sf-menu-item-wrapper[_ngcontent-%COMP%]   img[_ngcontent-%COMP%]{width:2em;height:auto;-o-object-fit:contain;object-fit:contain}.sf-menu-item-wrapper[_ngcontent-%COMP%]   .sf-menu-logo-bg[_ngcontent-%COMP%]{background:50%/contain no-repeat;width:1.6em;height:1.55em}.sf-menu-item-wrapper[_ngcontent-%COMP%]   em[_ngcontent-%COMP%], .sf-menu-item-wrapper[_ngcontent-%COMP%]   i[_ngcontent-%COMP%]{-o-object-fit:contain;object-fit:contain;color:#d5d7db}.sf-menu-item-wrapper[_ngcontent-%COMP%]   em.fa-chevron-right[_ngcontent-%COMP%], .sf-menu-item-wrapper[_ngcontent-%COMP%]   i.fa-chevron-right[_ngcontent-%COMP%]{font-size:12px;position:absolute;right:8px}.sf-menu-item-wrapper[_ngcontent-%COMP%]:hover{position:relative}.sf-menu-item-wrapper[_ngcontent-%COMP%]:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}.sf-menu-item-wrapper[_ngcontent-%COMP%]:hover   span[_ngcontent-%COMP%]{opacity:1}.sf-menu-item-wrapper.active-item[_ngcontent-%COMP%]{position:relative;border-left:2px solid #3298aa}.sf-menu-item-wrapper.active-item[_ngcontent-%COMP%]:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}.sf-menu-item-wrapper.active-item[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{opacity:1}.sf-menu-item-wrapper.active-item[_ngcontent-%COMP%]   em[_ngcontent-%COMP%], .sf-menu-item-wrapper.active-item[_ngcontent-%COMP%]   i[_ngcontent-%COMP%]{color:#fff}.sf-menu-item-wrapper[_ngcontent-%COMP%]   .icon-container.collapsed[_ngcontent-%COMP%]   em[_ngcontent-%COMP%]{font-size:var(--fs-300)}  .k-menu-popup .sf-menu-item-wrapper{padding:0 15px;justify-content:flex-start;background-color:#2f3037;width:188px}  .k-menu-popup .sf-menu-item-wrapper .icon-container{width:24px;font-size:16px}  .k-menu-popup .sf-menu-item-wrapper span{width:calc(100% - 24px);min-width:100px;color:#fefefe;padding:0 10px;font-size:14px;line-height:18px;opacity:.6;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}  .k-menu-popup .sf-menu-item-wrapper:hover{background-color:#000;color:#fefefe}  .k-menu-popup .sf-menu-item-wrapper.active-item{position:relative;background-color:#464853;border-left:2px solid #3298aa}  .k-menu-popup .sf-menu-item-wrapper.active-item:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}  .k-menu-popup .sf-menu-item-wrapper.active-item span{opacity:1}  .k-menu-popup .sf-menu-item-wrapper.active-item em,   .k-menu-popup .sf-menu-item-wrapper.active-item i{color:#fff}"], data: { animation: [
            trigger('appear-fast', [transition(':enter', [style({ opacity: 0 }), animate('80ms 120ms ease-in', style({ opacity: 1 }))])]),
        ] } });
MenuRoutesComponent.ctorParameters = () => [
    { type: SfMenuService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MenuRoutesComponent, [{
        type: Component,
        args: [{
                selector: 'sf-menu-routes',
                template: "<ul *ngIf=\"!collapsed\" [@appear-fast]>\n  <li\n    sf-menu-item\n    *ngFor=\"let item of menuConfig.routes\"\n    [class.show-arrow]=\"item.items\"\n    [class.active-item]=\"item.active\"\n    [item]=\"item\"\n    [level]=\"1\"\n    [itemType]=\"2\"\n    data-testid=\"menu-route-item\"\n  ></li>\n</ul>\n<kendo-menu *ngIf=\"menuConfig?.routes && collapsed\" [items]=\"menuConfig.routes\" [vertical]=\"true\">\n  <ng-template kendoMenuItemLinkTemplate let-item=\"item\" let-index=\"index\">\n    <div\n      class=\"sf-menu-item-wrapper\"\n      (click)=\"routeClicked(item, $event)\"\n      [class.active-item]=\"item.active\"\n      [ngStyle]=\"{ 'border-left-color': menuConfig?.borderColor }\"\n      title=\"{{ item.items ? undefined : item.text }}\"\n      [@appear-fast]\n    >\n      <div class=\"icon-container\" [class.collapsed]=\"collapsed\">\n        <em [ngClass]=\"item.iconClass\" *ngIf=\"item.iconClass\"></em>\n        <div\n          class=\"sf-menu-logo-bg\"\n          *ngIf=\"!item.iconClass && item.iconPath\"\n          [ngStyle]=\"{\n            'background-image': 'url(' + item.iconPath + ')'\n          }\"\n        ></div>\n      </div>\n      <span *ngIf=\"showText(index)\">{{ item.text }}</span>\n      <em class=\"fas fa-chevron-right\" *ngIf=\"item.items\"></em>\n    </div>\n  </ng-template>\n</kendo-menu>\n",
                animations: [
                    trigger('appear-fast', [transition(':enter', [style({ opacity: 0 }), animate('80ms 120ms ease-in', style({ opacity: 1 }))])]),
                ],
                styles: [":host{display:flex;flex-direction:column;width:100%;height:auto;align-items:flex-start;justify-content:flex-start;overflow:hidden;position:relative}:host:before{content:\" \";width:100%;height:1px;background-color:#fff;opacity:.15}:host ul{list-style:none;padding-left:0}:host ul,kendo-menu{width:100%}.sf-menu-item-wrapper{width:100%;height:44px;max-height:75px;overflow:hidden;background-color:transparent;display:flex;justify-content:center;align-items:center}.sf-menu-item-wrapper img{width:2em;height:auto;-o-object-fit:contain;object-fit:contain}.sf-menu-item-wrapper .sf-menu-logo-bg{background:50%/contain no-repeat;width:1.6em;height:1.55em}.sf-menu-item-wrapper em,.sf-menu-item-wrapper i{-o-object-fit:contain;object-fit:contain;color:#d5d7db}.sf-menu-item-wrapper em.fa-chevron-right,.sf-menu-item-wrapper i.fa-chevron-right{font-size:12px;position:absolute;right:8px}.sf-menu-item-wrapper:hover{position:relative}.sf-menu-item-wrapper:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}.sf-menu-item-wrapper:hover span{opacity:1}.sf-menu-item-wrapper.active-item{position:relative;border-left:2px solid #3298aa}.sf-menu-item-wrapper.active-item:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}.sf-menu-item-wrapper.active-item span{opacity:1}.sf-menu-item-wrapper.active-item em,.sf-menu-item-wrapper.active-item i{color:#fff}.sf-menu-item-wrapper .icon-container.collapsed em{font-size:var(--fs-300)}::ng-deep .k-menu-popup .sf-menu-item-wrapper{padding:0 15px;justify-content:flex-start;background-color:#2f3037;width:188px}::ng-deep .k-menu-popup .sf-menu-item-wrapper .icon-container{width:24px;font-size:16px}::ng-deep .k-menu-popup .sf-menu-item-wrapper span{width:calc(100% - 24px);min-width:100px;color:#fefefe;padding:0 10px;font-size:14px;line-height:18px;opacity:.6;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}::ng-deep .k-menu-popup .sf-menu-item-wrapper:hover{background-color:#000;color:#fefefe}::ng-deep .k-menu-popup .sf-menu-item-wrapper.active-item{position:relative;background-color:#464853;border-left:2px solid #3298aa}::ng-deep .k-menu-popup .sf-menu-item-wrapper.active-item:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}::ng-deep .k-menu-popup .sf-menu-item-wrapper.active-item span{opacity:1}::ng-deep .k-menu-popup .sf-menu-item-wrapper.active-item em,::ng-deep .k-menu-popup .sf-menu-item-wrapper.active-item i{color:#fff}"]
            }]
    }], function () { return [{ type: SfMenuService }]; }, null); })();

class MenuShortcutsComponent {
    constructor(svc) {
        this.svc = svc;
        this.collapsed = false;
    }
    ngOnInit() {
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            this.menuConfig = cfg;
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
        });
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
}
MenuShortcutsComponent.ɵfac = function MenuShortcutsComponent_Factory(t) { return new (t || MenuShortcutsComponent)(ɵngcc0.ɵɵdirectiveInject(SfMenuService)); };
MenuShortcutsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: MenuShortcutsComponent, selectors: [["sf-menu-shortcuts"]], decls: 2, vars: 1, consts: [["sf-menu-item", "", 3, "item", "level", "itemType", 4, "ngFor", "ngForOf"], ["sf-menu-item", "", 3, "item", "level", "itemType"]], template: function MenuShortcutsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "ul");
        ɵngcc0.ɵɵtemplate(1, MenuShortcutsComponent_li_1_Template, 1, 3, "li", 0);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.menuConfig.shortcuts);
    } }, directives: [ɵngcc2.NgForOf, MenuItemComponent], styles: ["[_nghost-%COMP%]{display:flex;flex-direction:column;width:100%;height:auto;align-items:flex-start;justify-content:flex-start;overflow:hidden;position:relative}[_nghost-%COMP%]:before{content:\" \";width:100%;height:1px;background-color:#fff;opacity:.15}[_nghost-%COMP%]   ul[_ngcontent-%COMP%]{list-style:none;padding-left:0;width:100%}"], data: { animation: [
            trigger('appear-fast', [transition(':enter', [style({ opacity: 0 }), animate('80ms 120ms ease-in', style({ opacity: 1 }))])]),
        ] } });
MenuShortcutsComponent.ctorParameters = () => [
    { type: SfMenuService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MenuShortcutsComponent, [{
        type: Component,
        args: [{
                selector: 'sf-menu-shortcuts',
                template: "<ul>\n  <li\n    sf-menu-item\n    *ngFor=\"let item of menuConfig.shortcuts\"\n    [item]=\"item\"\n    [level]=\"1\"\n    [itemType]=\"1\"\n  >\n  </li>\n</ul>\n",
                animations: [
                    trigger('appear-fast', [transition(':enter', [style({ opacity: 0 }), animate('80ms 120ms ease-in', style({ opacity: 1 }))])]),
                ],
                styles: [":host{display:flex;flex-direction:column;width:100%;height:auto;align-items:flex-start;justify-content:flex-start;overflow:hidden;position:relative}:host:before{content:\" \";width:100%;height:1px;background-color:#fff;opacity:.15}:host ul{list-style:none;padding-left:0;width:100%}"]
            }]
    }], function () { return [{ type: SfMenuService }]; }, null); })();

class MenuToggleBtnComponent {
    constructor(svc) {
        this.svc = svc;
        this.collapsed = false;
    }
    ngOnInit() {
        this.color = this.svc.colors.gray700;
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            this.menuConfig = cfg;
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
        });
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
    toggleMenu() {
        this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
    }
}
MenuToggleBtnComponent.ɵfac = function MenuToggleBtnComponent_Factory(t) { return new (t || MenuToggleBtnComponent)(ɵngcc0.ɵɵdirectiveInject(SfMenuService)); };
MenuToggleBtnComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: MenuToggleBtnComponent, selectors: [["sf-menu-toggle-btn"]], decls: 4, vars: 4, consts: [[1, "sf-menu-toggle-btn", 3, "click"], [1, "fas", "fa-chevron-left"]], template: function MenuToggleBtnComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "button", 0);
        ɵngcc0.ɵɵlistener("click", function MenuToggleBtnComponent_Template_button_click_0_listener() { return ctx.toggleMenu(); });
        ɵngcc0.ɵɵelementStart(1, "div");
        ɵngcc0.ɵɵelement(2, "em", 1)(3, "em", 1);
        ɵngcc0.ɵɵelementEnd()();
    } if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("color", ctx.color);
        ɵngcc0.ɵɵclassProp("collapsed", ctx.collapsed);
    } }, styles: [".sf-menu-toggle-btn[_ngcontent-%COMP%]{display:none;position:absolute;z-index:10000;top:8px;right:8px;background-color:transparent;border:none;width:1.8rem;height:1.8rem;transition:transform .15s ease-in-out;cursor:pointer}@media screen and (min-width:768px){.sf-menu-toggle-btn[_ngcontent-%COMP%]{display:block}}.sf-menu-toggle-btn[_ngcontent-%COMP%]   div[_ngcontent-%COMP%]{display:flex;justify-content:center;align-items:center}.sf-menu-toggle-btn[_ngcontent-%COMP%]   .fas[_ngcontent-%COMP%]{font-size:16px;opacity:.64}.sf-menu-toggle-btn[_ngcontent-%COMP%]   .fas[_ngcontent-%COMP%]:nth-child(2){margin-left:-4px}.sf-menu-toggle-btn.collapsed[_ngcontent-%COMP%]   .fas[_ngcontent-%COMP%]{transform:rotate(180deg)}.sf-menu-toggle-btn[_ngcontent-%COMP%]:hover   .fas[_ngcontent-%COMP%]{opacity:1}"] });
MenuToggleBtnComponent.ctorParameters = () => [
    { type: SfMenuService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MenuToggleBtnComponent, [{
        type: Component,
        args: [{
                selector: 'sf-menu-toggle-btn',
                template: "<button class=\"sf-menu-toggle-btn\" [class.collapsed]=\"collapsed\" (click)=\"toggleMenu()\" [style.color]=\"color\">\n  <div>\n    <em class=\"fas fa-chevron-left\"></em>\n    <em class=\"fas fa-chevron-left\"></em>\n  </div>\n</button>\n",
                styles: [".sf-menu-toggle-btn{display:none;position:absolute;z-index:10000;top:8px;right:8px;background-color:transparent;border:none;width:1.8rem;height:1.8rem;transition:transform .15s ease-in-out;cursor:pointer}@media screen and (min-width:768px){.sf-menu-toggle-btn{display:block}}.sf-menu-toggle-btn div{display:flex;justify-content:center;align-items:center}.sf-menu-toggle-btn .fas{font-size:16px;opacity:.64}.sf-menu-toggle-btn .fas:nth-child(2){margin-left:-4px}.sf-menu-toggle-btn.collapsed .fas{transform:rotate(180deg)}.sf-menu-toggle-btn:hover .fas{opacity:1}"]
            }]
    }], function () { return [{ type: SfMenuService }]; }, null); })();

class MenuUserComponent {
    constructor(svc) {
        this.svc = svc;
        this.tenants = [];
        this.collapsed = false;
        this.multipleTenantMenu = false;
        this.anchorAlign = { horizontal: 'right', vertical: 'top' };
        this.popupAlign = { horizontal: 'left', vertical: 'top' };
        this.userName = 'Username';
        this.darkMode = false;
    }
    ngOnInit() {
        this.color = this.svc.colors.gray900;
        this.tenantBg = this.svc.colors.gray200;
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            var _a;
            this.menuConfig = cfg;
            this.userActions = {
                userName: { id: 'username', text: this.menuConfig.userName, iconClass: 'fas fa-user-circle' },
                tenant: { id: 'tenant', text: this.menuConfig.fleetName, iconClass: 'fas fa-building' },
                help: { id: 'help', text: this.menuConfig.helpText, iconClass: 'fas fa-question-circle' },
                logout: { id: 'logout', text: this.menuConfig.logoutText, iconClass: 'fas fa-sign-out-alt' }
            };
            this.tenants.length = 0;
            if (((_a = cfg.multipleTenants) === null || _a === void 0 ? void 0 : _a.length) > 1) {
                for (const t of cfg.multipleTenants.sort((a, b) => a.localeCompare(b))) {
                    this.tenants.push({ id: t, active: t === cfg.fleetName });
                }
            }
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
        });
        this.svc.currentTheme.subscribe((theme) => (this.darkMode = theme === 'dark'));
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
    get disableUserNameClick() {
        return this.menuConfig.disableUserNameClick && !this.collapsed;
    }
    changeTenant(tenant) {
        this.svc.events$.next({
            type: MenuEventType.ChangeTenantClick,
            value: { text: tenant.id }
        });
        this.multipleTenantMenu = false;
    }
    showMultipleTenantMenu() {
        if (this.collapsed) {
            this.svc.menuCollapse$.next(false);
        }
        this.svc.events$.next({
            type: MenuEventType.OpenTenantListClick,
            value: null
        });
        this.multipleTenantMenu = true;
    }
    hideMultipleTenantMenu() {
        this.multipleTenantMenu = false;
    }
    userOptionClicked(item) {
        if (!this.disableUserNameClick) {
            if (this.collapsed) {
                this.svc.menuCollapse$.next(false);
                return;
            }
            this.svc.events$.next({
                type: MenuEventType.UserActionClick,
                value: item
            });
            if (window.innerWidth <= this.svc.SMALL_RESOLUTION && !this.collapsed) {
                this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
            }
        }
    }
    setDarkMode() {
        console.log('set dark mode');
        this.svc.setTheme(this.darkMode ? 'dark' : 'light');
    }
    themeIconClicked() {
        console.log('icon clicked');
        if (this.collapsed) {
            this.svc.menuCollapse$.next(false);
        }
    }
    onSsoManageAccount(event, config) {
        event.stopPropagation();
        if (!config.disable) {
            window.open(config.url);
        }
    }
    sortMultipleTenant(tenants) {
        if (tenants) {
            return tenants.sort((a, b) => a.localeCompare(b));
        }
        else {
            return [];
        }
    }
}
MenuUserComponent.ɵfac = function MenuUserComponent_Factory(t) { return new (t || MenuUserComponent)(ɵngcc0.ɵɵdirectiveInject(SfMenuService)); };
MenuUserComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: MenuUserComponent, selectors: [["sf-menu-user"]], decls: 13, vars: 15, consts: [[1, "ul-container"], [1, "li-container", "user-li", 3, "title", "click"], [1, "main-div", "menu-item"], [1, "icon-container"], [1, "sf-icons-menu"], [1, "fas", "fa-user-circle"], [4, "ngIf"], ["class", "sso-settings", 3, "click", 4, "ngIf"], [3, "click", 4, "ngIf"], ["sf-menu-item", "", "data-testid", "menu-user-help", 3, "show-arrow", "active-item", "item", "itemType", 4, "ngIf"], ["sf-menu-item", "", "data-testid", "menu-user-logout", 3, "show-arrow", "active-item", "item", "itemType", 4, "ngIf"], ["data-testid", "menu-user-name", 1, "sf-user-text"], ["data-testid", "menu-user-fleet-name-single", 1, "sf-user-fleet-name", 3, "click"], [1, "sso-settings", 3, "click"], ["kendoTooltip", "", 1, "sf-icons-menu", 3, "showOn", "title"], ["data-testid", "menu-products-sso-settings-icon", 1, "icon-manage-account"], ["anchor", ""], [1, "li-container"], [1, "main-div", "menu-item", "sf-fleet-list", 3, "ngStyle", "title", "click"], [1, "fas", "fa-building"], ["class", "fas fa-chevron-right", "data-testid", "menu-user-show-tenant-list", 4, "ngIf"], [3, "anchor", "anchorAlign", "popupAlign", "clickOutside", 4, "ngIf"], ["class", "sf-fleet-title", 4, "ngIf"], [1, "sf-fleet-title"], ["data-testid", "menu-user-show-tenant-list", 1, "fas", "fa-chevron-right"], [3, "anchor", "anchorAlign", "popupAlign", "clickOutside"], [1, "tenant-picker"], ["class", "tenant", "data-testid", "menu-user-tenants-list", 3, "active", "click", 4, "ngFor", "ngForOf"], ["data-testid", "menu-user-tenants-list", 1, "tenant", 3, "click"], [3, "click"], [1, "main-div", "menu-item", "theme-selector-container", 3, "title"], [1, "fas", "fa-moon"], [3, "ngModel", "valueChange", "ngModelChange", 4, "ngIf"], [3, "ngModel", "valueChange", "ngModelChange"], ["sf-menu-item", "", "data-testid", "menu-user-help", 3, "item", "itemType"], ["sf-menu-item", "", "data-testid", "menu-user-logout", 3, "item", "itemType"]], template: function MenuUserComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "ul", 0)(1, "li")(2, "div", 1);
        ɵngcc0.ɵɵlistener("click", function MenuUserComponent_Template_div_click_2_listener() { return ctx.userOptionClicked(ctx.userActions.userName); });
        ɵngcc0.ɵɵelementStart(3, "div", 2)(4, "div", 3)(5, "div", 4);
        ɵngcc0.ɵɵelement(6, "em", 5);
        ɵngcc0.ɵɵelementEnd()();
        ɵngcc0.ɵɵtemplate(7, MenuUserComponent_div_7_Template, 5, 2, "div", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(8, MenuUserComponent_div_8_Template, 3, 4, "div", 7);
        ɵngcc0.ɵɵelementEnd()();
        ɵngcc0.ɵɵtemplate(9, MenuUserComponent_li_9_Template, 10, 13, "li", 6);
        ɵngcc0.ɵɵtemplate(10, MenuUserComponent_li_10_Template, 8, 5, "li", 8);
        ɵngcc0.ɵɵtemplate(11, MenuUserComponent_li_11_Template, 1, 6, "li", 9);
        ɵngcc0.ɵɵtemplate(12, MenuUserComponent_li_12_Template, 1, 6, "li", 10);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("color", ctx.color);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵclassProp("collapsed", ctx.collapsed);
        ɵngcc0.ɵɵpropertyInterpolate("title", ctx.menuConfig.tooltips || ctx.collapsed ? ctx.menuConfig.userName : "");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵclassProp("disable-user-click", ctx.disableUserNameClick)("sso-manage-account", ctx.menuConfig.ssoManageAccount && ctx.menuConfig.ssoManageAccount.url);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.menuConfig.userName && !ctx.collapsed);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", (ctx.menuConfig.ssoManageAccount == null ? null : ctx.menuConfig.ssoManageAccount.url) && !ctx.collapsed);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", (ctx.menuConfig == null ? null : ctx.menuConfig.multipleTenants == null ? null : ctx.menuConfig.multipleTenants.length) > 1);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !(ctx.menuConfig == null ? null : ctx.menuConfig.denyChangeTheme) && (ctx.menuConfig == null ? null : ctx.menuConfig.themeTitle));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.userActions.help.text);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.userActions.logout.text);
    } }, directives: [ɵngcc2.NgIf, ɵngcc3.TooltipDirective, ɵngcc2.NgStyle, ɵngcc5.PopupComponent, ClickOutsideDirective, ɵngcc2.NgForOf, ɵngcc6.SwitchComponent, ɵngcc7.NgControlStatus, ɵngcc7.NgModel, MenuItemComponent], styles: ["[_nghost-%COMP%]{display:flex;flex-direction:column;width:100%;height:auto;align-items:flex-start;justify-content:flex-start;overflow:hidden;position:relative}[_nghost-%COMP%]:before{content:\" \";width:100%;height:1px;background-color:#fff;opacity:.15}[_nghost-%COMP%]   ul[_ngcontent-%COMP%]{list-style:none;padding-left:0;width:100%}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%]{display:inline-block;width:100%;padding:0;overflow:hidden;cursor:pointer}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%]   .icon-container[_ngcontent-%COMP%]{width:24px}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > div.main-div[_ngcontent-%COMP%]{margin-left:2px}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > .menu-item[_ngcontent-%COMP%]{display:flex;align-items:center;justify-content:flex-start;padding:7px 10px;position:relative;min-height:38px}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > .menu-item[_ngcontent-%COMP%]   .sf-menu-logo-bg[_ngcontent-%COMP%]{width:1.6em;height:1.55em}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > .menu-item[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{width:calc(100% - 24px);min-width:100px;padding:0 10px;font-size:14px;line-height:18px;opacity:.71;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > .menu-item[_ngcontent-%COMP%]:hover{position:relative}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > .menu-item[_ngcontent-%COMP%]:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > .menu-item[_ngcontent-%COMP%]:hover   span[_ngcontent-%COMP%]{opacity:1}[_nghost-%COMP%]   .li-container[_ngcontent-%COMP%] > .menu-item[_ngcontent-%COMP%]   .fa-chevron-down[_ngcontent-%COMP%]{display:none;position:absolute;right:15px;top:16px;width:auto;height:auto;font-size:12px}[_nghost-%COMP%]   .li-container.show-arrow[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]   .fa-chevron-down[_ngcontent-%COMP%]{display:inline-block;transform:rotate(0deg)}[_nghost-%COMP%]   .li-container.show-arrow[_ngcontent-%COMP%] > div.expanded-item[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{opacity:1}[_nghost-%COMP%]   .li-container.show-arrow[_ngcontent-%COMP%] > div.expanded-item[_ngcontent-%COMP%]   em.fa-chevron-down[_ngcontent-%COMP%], [_nghost-%COMP%]   .li-container.show-arrow[_ngcontent-%COMP%] > div.expanded-item[_ngcontent-%COMP%]   i.fa-chevron-down[_ngcontent-%COMP%]{transform:rotate(180deg)}[_nghost-%COMP%]   .li-container.show-arrow[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]:hover{position:relative}[_nghost-%COMP%]   .li-container.show-arrow[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}[_nghost-%COMP%]   .li-container.show-arrow[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]:hover   span[_ngcontent-%COMP%]{opacity:1}[_nghost-%COMP%]   .li-container.active-item[_ngcontent-%COMP%] > div.main-div[_ngcontent-%COMP%]{border-left:3px solid;margin-left:0}[_nghost-%COMP%]   .li-container.active-item[_ngcontent-%COMP%] > div.main-div[_ngcontent-%COMP%]:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.2}[_nghost-%COMP%]   .li-container.active-item[_ngcontent-%COMP%] > div.main-div[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{opacity:1}[_nghost-%COMP%]   .li-container.collapsed[_ngcontent-%COMP%]   div[_ngcontent-%COMP%]{justify-content:center}[_nghost-%COMP%]   .li-container.collapsed[_ngcontent-%COMP%]   em[_ngcontent-%COMP%], [_nghost-%COMP%]   .li-container.collapsed[_ngcontent-%COMP%]   i[_ngcontent-%COMP%]{width:auto}[_nghost-%COMP%]   .li-container.collapsed[_ngcontent-%COMP%]   .item-children[_ngcontent-%COMP%], [_nghost-%COMP%]   .li-container.collapsed[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{display:none}[_nghost-%COMP%]   .sf-icons-menu[_ngcontent-%COMP%]{text-align:center}[_nghost-%COMP%]   .sf-icons-menu[_ngcontent-%COMP%]   em[_ngcontent-%COMP%], [_nghost-%COMP%]   .sf-icons-menu[_ngcontent-%COMP%]   i[_ngcontent-%COMP%]{font-size:16px}.user-li[_ngcontent-%COMP%]{display:flex!important}.user-li[_ngcontent-%COMP%]   .menu-item[_ngcontent-%COMP%]{flex:1}.user-li[_ngcontent-%COMP%]   .menu-item.sso-manage-account[_ngcontent-%COMP%]{flex-basis:75%}.user-li[_ngcontent-%COMP%]   .menu-item.disable-user-click[_ngcontent-%COMP%]{cursor:default}.user-li[_ngcontent-%COMP%]   .menu-item.disable-user-click[_ngcontent-%COMP%]:hover{background-color:transparent}.user-li[_ngcontent-%COMP%]   .menu-item.disable-user-click[_ngcontent-%COMP%]:hover   .sf-user-fleet-name[_ngcontent-%COMP%], .user-li[_ngcontent-%COMP%]   .menu-item.disable-user-click[_ngcontent-%COMP%]:hover   .sf-user-text[_ngcontent-%COMP%]{opacity:.7}.user-li[_ngcontent-%COMP%]   .menu-item.disable-user-click[_ngcontent-%COMP%]:hover:before{display:none}.user-li[_ngcontent-%COMP%]   .sso-settings[_ngcontent-%COMP%]{display:flex;justify-content:flex-end;flex:1;align-items:center;padding-right:15px;cursor:default}.user-li[_ngcontent-%COMP%]   .sso-settings[_ngcontent-%COMP%]   .sf-icons-menu[_ngcontent-%COMP%]{cursor:pointer}.user-li[_ngcontent-%COMP%]   .sso-settings[_ngcontent-%COMP%]   .sf-icons-menu.disable[_ngcontent-%COMP%]{opacity:.5;cursor:not-allowed}.user-li[_ngcontent-%COMP%]   .sso-settings[_ngcontent-%COMP%]   .sf-icons-menu[_ngcontent-%COMP%]:hover{position:relative}.user-li[_ngcontent-%COMP%]   .sso-settings[_ngcontent-%COMP%]   .sf-icons-menu[_ngcontent-%COMP%]:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}.user-li[_ngcontent-%COMP%]   .sso-settings[_ngcontent-%COMP%]   .sf-icons-menu[_ngcontent-%COMP%]:hover   span[_ngcontent-%COMP%]{opacity:1}.sf-user-fleet-name[_ngcontent-%COMP%]{display:block;font-size:.8em!important;margin-top:.2em}.sf-fleet-list[_ngcontent-%COMP%]   .sf-fleet-title[_ngcontent-%COMP%]{font-size:.8em}.theme-selector-container[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{width:calc(100% - 84px)!important}.theme-selector-container[_ngcontent-%COMP%]   .k-switch[_ngcontent-%COMP%]{margin-right:0}.ul-container[_ngcontent-%COMP%]{margin:5px 0 0}.tenant-picker[_ngcontent-%COMP%]{padding:.5em 0;max-height:14em;max-width:14em;min-width:12em;overflow-y:auto;margin-left:-1px}.tenant-picker[_ngcontent-%COMP%]   .tenant[_ngcontent-%COMP%]{padding:.7em 1em;position:relative}.tenant-picker[_ngcontent-%COMP%]   .tenant[_ngcontent-%COMP%]:hover:before{opacity:.1}.tenant-picker[_ngcontent-%COMP%]   .tenant.active[_ngcontent-%COMP%]:before, .tenant-picker[_ngcontent-%COMP%]   .tenant[_ngcontent-%COMP%]:hover:before{content:\"\";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff}.tenant-picker[_ngcontent-%COMP%]   .tenant.active[_ngcontent-%COMP%]:before{opacity:.2}.tenant-picker[_ngcontent-%COMP%]::-webkit-scrollbar{width:6px;background-color:transparent}.tenant-picker[_ngcontent-%COMP%]::-webkit-scrollbar-thumb{border-radius:10px;background-color:var(--clr-gray-500)}  .sf-menu-container .k-popup{background-color:transparent;border-color:transparent}"] });
MenuUserComponent.ctorParameters = () => [
    { type: SfMenuService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MenuUserComponent, [{
        type: Component,
        args: [{
                selector: 'sf-menu-user',
                template: "<ul class=\"ul-container\" [style.color]=\"color\">\n  <li>\n    <div\n      class=\"li-container user-li\"\n      (click)=\"userOptionClicked(userActions.userName)\"\n      title=\"{{ menuConfig.tooltips || collapsed ? menuConfig.userName : '' }}\"\n      [class.collapsed]=\"collapsed\"\n    >\n      <div\n        class=\"main-div menu-item\"\n        [class.disable-user-click]=\"disableUserNameClick\"\n        [class.sso-manage-account]=\"menuConfig.ssoManageAccount && menuConfig.ssoManageAccount.url\"\n      >\n        <div class=\"icon-container\">\n          <div class=\"sf-icons-menu\">\n            <em class=\"fas fa-user-circle\"></em>\n          </div>\n        </div>\n        <div *ngIf=\"menuConfig.userName && !collapsed\">\n          <span class=\"sf-user-text\" data-testid=\"menu-user-name\">{{ menuConfig.userName }} </span>\n          <span class=\"sf-user-fleet-name\" (click)=\"showMultipleTenantMenu()\" data-testid=\"menu-user-fleet-name-single\">\n            {{ menuConfig?.multipleTenants?.length >= 2 ? '' : menuConfig.fleetName }}\n          </span>\n        </div>\n      </div>\n      <div\n\n        class=\"sso-settings\"\n        *ngIf=\"menuConfig.ssoManageAccount?.url && !collapsed\"\n        (click)=\"onSsoManageAccount($event, menuConfig.ssoManageAccount)\"\n      >\n        <div kendoTooltip showOn=\"{{ !menuConfig.tooltips ? 'hover' : 'none' }}\"\n          class=\"sf-icons-menu\" [class.disable]=\"menuConfig.ssoManageAccount.disable\"\n          title=\"{{ menuConfig.ssoManageAccount.title }}\">\n          <em class=\"icon-manage-account\"\n            data-testid=\"menu-products-sso-settings-icon\"></em>\n        </div>\n      </div>\n    </div>\n  </li>\n\n  <li *ngIf=\"menuConfig?.multipleTenants?.length > 1\" #anchor>\n    <div class=\"li-container\" [class.show-arrow]=\"menuConfig?.multipleTenants\" [class.collapsed]=\"collapsed\">\n      <div\n        [class.expanded-item]=\"multipleTenantMenu\"\n        class=\"main-div menu-item sf-fleet-list\"\n        [ngStyle]=\"{ 'border-left-color': menuConfig?.borderColor }\"\n        title=\"{{ menuConfig.tooltips || collapsed ? menuConfig.fleetName : '' }}\"\n        (click)=\"showMultipleTenantMenu()\"\n      >\n        <div class=\"icon-container\">\n          <div class=\"sf-icons-menu\">\n            <em class=\"fas fa-building\"></em>\n          </div>\n        </div>\n        <span *ngIf=\"!collapsed\">\n          <div class=\"sf-fleet-title\" *ngIf=\"menuConfig.fleetTitle\">\n            {{ menuConfig.fleetTitle }}\n          </div>\n          <div>\n            {{ menuConfig.fleetName }}\n          </div>\n        </span>\n        <em *ngIf=\"!collapsed\" class=\"fas fa-chevron-right\"\n          data-testid=\"menu-user-show-tenant-list\"></em>\n      </div>\n    </div>\n\n    <kendo-popup\n      [anchor]=\"anchor\"\n      [anchorAlign]=\"anchorAlign\"\n      [popupAlign]=\"popupAlign\"\n      *ngIf=\"multipleTenantMenu\"\n      (clickOutside)=\"hideMultipleTenantMenu()\"\n    >\n      <div class=\"tenant-picker\" [style.color]=\"color\" [style.background-color]=\"tenantBg\">\n        <div class=\"tenant\" *ngFor=\"let tenant of tenants\" [class.active]=\"tenant.active\" (click)=\"changeTenant(tenant)\"\n          data-testid=\"menu-user-tenants-list\">\n          {{ tenant.id }}\n        </div>\n      </div>\n    </kendo-popup>\n  </li>\n\n  <li *ngIf=\"!menuConfig?.denyChangeTheme && menuConfig?.themeTitle\" (click)=\"themeIconClicked()\">\n    <div class=\"li-container\" [class.collapsed]=\"collapsed\">\n      <div class=\"main-div menu-item theme-selector-container\" title=\"{{ menuConfig.tooltips || collapsed ? menuConfig.themeTitle : '' }}\">\n        <div class=\"icon-container\">\n          <div class=\"sf-icons-menu\">\n            <em class=\"fas fa-moon\"></em>\n          </div>\n        </div>\n        <span *ngIf=\"!collapsed\">{{ menuConfig?.themeTitle }}</span>\n        <kendo-switch *ngIf=\"!collapsed\" (valueChange)=\"setDarkMode()\" [(ngModel)]=\"darkMode\"></kendo-switch>\n      </div>\n    </div>\n  </li>\n  <li\n    sf-menu-item\n    *ngIf=\"userActions.help.text\"\n    [class.show-arrow]=\"userActions.help.items\"\n    [class.active-item]=\"userActions.help.active\"\n    [item]=\"userActions.help\"\n    [itemType]=\"3\"\n    data-testid=\"menu-user-help\"\n  ></li>\n  <li\n    sf-menu-item\n    *ngIf=\"userActions.logout.text\"\n    [class.show-arrow]=\"userActions.logout.items\"\n    [class.active-item]=\"userActions.logout.active\"\n    [item]=\"userActions.logout\"\n    [itemType]=\"3\"\n    data-testid=\"menu-user-logout\"\n  ></li>\n</ul>\n",
                styles: [":host{display:flex;flex-direction:column;width:100%;height:auto;align-items:flex-start;justify-content:flex-start;overflow:hidden;position:relative}:host:before{content:\" \";width:100%;height:1px;background-color:#fff;opacity:.15}:host ul{list-style:none;padding-left:0;width:100%}:host .li-container{display:inline-block;width:100%;padding:0;overflow:hidden;cursor:pointer}:host .li-container .icon-container{width:24px}:host .li-container>div.main-div{margin-left:2px}:host .li-container>.menu-item{display:flex;align-items:center;justify-content:flex-start;padding:7px 10px;position:relative;min-height:38px}:host .li-container>.menu-item .sf-menu-logo-bg{width:1.6em;height:1.55em}:host .li-container>.menu-item span{width:calc(100% - 24px);min-width:100px;padding:0 10px;font-size:14px;line-height:18px;opacity:.71;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}:host .li-container>.menu-item:hover{position:relative}:host .li-container>.menu-item:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}:host .li-container>.menu-item:hover span{opacity:1}:host .li-container>.menu-item .fa-chevron-down{display:none;position:absolute;right:15px;top:16px;width:auto;height:auto;font-size:12px}:host .li-container.show-arrow>div .fa-chevron-down{display:inline-block;transform:rotate(0deg)}:host .li-container.show-arrow>div.expanded-item span{opacity:1}:host .li-container.show-arrow>div.expanded-item em.fa-chevron-down,:host .li-container.show-arrow>div.expanded-item i.fa-chevron-down{transform:rotate(180deg)}:host .li-container.show-arrow>div:hover{position:relative}:host .li-container.show-arrow>div:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}:host .li-container.show-arrow>div:hover span{opacity:1}:host .li-container.active-item>div.main-div{border-left:3px solid;margin-left:0}:host .li-container.active-item>div.main-div:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.2}:host .li-container.active-item>div.main-div span{opacity:1}:host .li-container.collapsed div{justify-content:center}:host .li-container.collapsed em,:host .li-container.collapsed i{width:auto}:host .li-container.collapsed .item-children,:host .li-container.collapsed span{display:none}:host .sf-icons-menu{text-align:center}:host .sf-icons-menu em,:host .sf-icons-menu i{font-size:16px}.user-li{display:flex!important}.user-li .menu-item{flex:1}.user-li .menu-item.sso-manage-account{flex-basis:75%}.user-li .menu-item.disable-user-click{cursor:default}.user-li .menu-item.disable-user-click:hover{background-color:transparent}.user-li .menu-item.disable-user-click:hover .sf-user-fleet-name,.user-li .menu-item.disable-user-click:hover .sf-user-text{opacity:.7}.user-li .menu-item.disable-user-click:hover:before{display:none}.user-li .sso-settings{display:flex;justify-content:flex-end;flex:1;align-items:center;padding-right:15px;cursor:default}.user-li .sso-settings .sf-icons-menu{cursor:pointer}.user-li .sso-settings .sf-icons-menu.disable{opacity:.5;cursor:not-allowed}.user-li .sso-settings .sf-icons-menu:hover{position:relative}.user-li .sso-settings .sf-icons-menu:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}.user-li .sso-settings .sf-icons-menu:hover span{opacity:1}.sf-user-fleet-name{display:block;font-size:.8em!important;margin-top:.2em}.sf-fleet-list .sf-fleet-title{font-size:.8em}.theme-selector-container span{width:calc(100% - 84px)!important}.theme-selector-container .k-switch{margin-right:0}.ul-container{margin:5px 0 0}.tenant-picker{padding:.5em 0;max-height:14em;max-width:14em;min-width:12em;overflow-y:auto;margin-left:-1px}.tenant-picker .tenant{padding:.7em 1em;position:relative}.tenant-picker .tenant:hover:before{opacity:.1}.tenant-picker .tenant.active:before,.tenant-picker .tenant:hover:before{content:\"\";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff}.tenant-picker .tenant.active:before{opacity:.2}.tenant-picker::-webkit-scrollbar{width:6px;background-color:transparent}.tenant-picker::-webkit-scrollbar-thumb{border-radius:10px;background-color:var(--clr-gray-500)}::ng-deep .sf-menu-container .k-popup{background-color:transparent;border-color:transparent}"]
            }]
    }], function () { return [{ type: SfMenuService }]; }, null); })();

class TopMenuComponent {
    constructor(svc) {
        this.svc = svc;
        this.collapsed = false;
    }
    get bgColor() {
        return this.backgroundColor;
    }
    ngOnInit() {
        this.activeLink = this.svc.colors.gray300;
        this.iconColor = this.svc.colors.gray800;
        this.backgroundColor = this.svc.colors.gray200;
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            this.menuConfig = cfg;
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
        });
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
    shortcutClicked(shortcut) {
        this.svc.events$.next({
            type: MenuEventType.ShortcutClick,
            value: shortcut
        });
        if (window.innerWidth <= this.svc.SMALL_RESOLUTION && !this.collapsed) {
            this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
        }
    }
    toggleMenu() {
        this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
    }
}
TopMenuComponent.ɵfac = function TopMenuComponent_Factory(t) { return new (t || TopMenuComponent)(ɵngcc0.ɵɵdirectiveInject(SfMenuService)); };
TopMenuComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: TopMenuComponent, selectors: [["sf-top-menu"]], hostVars: 2, hostBindings: function TopMenuComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("background-color", ctx.bgColor);
    } }, decls: 8, vars: 6, consts: [[1, "sf-top-menu-left"], [1, "sf-top-menu-toggle-bars", "fa", "fa-2x", "fa-bars", 3, "click"], [1, "sf-top-menu-middle"], ["class", "sf-menu-logo-bg", 3, "ngStyle", 4, "ngIf"], ["class", "sf-menu-top-logo", 3, "ngClass", 4, "ngIf"], ["class", "sf-menu-top-logo", "xmlns", "http://www.w3.org/2000/svg", "viewBox", "0 0 1182 263", "fill", "#1B4C86", 4, "ngIf"], [1, "sf-top-menu-right"], [4, "ngFor", "ngForOf"], [1, "sf-menu-logo-bg", 3, "ngStyle"], [1, "sf-menu-top-logo", 3, "ngClass"], ["xmlns", "http://www.w3.org/2000/svg", "viewBox", "0 0 1182 263", "fill", "#1B4C86", 1, "sf-menu-top-logo"], ["d", "M55.2,145.2 C74.8,147.2 94.4,150.1 94.4,162.9 C94.4,179.3 70.4,183.8 49.4,183.8 C22,183.8 2.1,178 0.4,160 L11.7,160 C13.7,174 28.8,177.4 49.5,177.4 C65.7,177.4 82.5,174.7 82.5,164.1 C82.5,153.9 62.8,152.7 43.2,150.7 C17.5,148.2 3.7,143.7 3.7,132.9 C3.7,118.3 25.4,113.5 46.3,113.5 C68.3,113.5 90.6,118.4 91.3,134.3 L80.1,134.3 C78.6,122.9 63.6,119.9 46.2,119.9 C28.9,119.9 15.7,124.1 15.7,131.7 C15.7,140.4 28.9,142.5 55.2,145.2 Z M128.3,155.4 L176,155.4 L152.3,122.1 L128.3,155.4 Z M146,115.2 L159,115.2 L208,182.1 L195,182.1 L180.6,161.8 L123.8,161.8 L109.1,182.1 L97,182.1 L146,115.2 Z M217,182.1 L217,115.2 L297.9,115.2 L297.9,121.6 L228.2,121.6 L228.2,144.3 L294.4,144.3 L294.4,150.6 L228.2,150.6 L228.2,182.1 L217,182.1 Z M322.4,150.6 L322.4,175.7 L398.3,175.7 L398.3,182.1 L311.2,182.1 L311.2,115.2 L397.1,115.2 L397.1,121.6 L322.4,121.6 L322.4,144.2 L394.4,144.2 L394.4,150.6 L322.4,150.6 Z M707.1,182.1 L707.1,115.2 L788,115.2 L788,121.6 L718.3,121.6 L718.3,144.3 L784.5,144.3 L784.5,150.6 L718.3,150.6 L718.3,182.1 L707.1,182.1 Z M810.7,115.2 L810.7,175.7 L879.1,175.7 L879.1,182.1 L799.5,182.1 L799.5,115.2 L810.7,115.2 Z M902.1,150.6 L902.1,175.7 L978,175.7 L978,182.1 L890.9,182.1 L890.9,115.2 L976.8,115.2 L976.8,121.6 L902.1,121.6 L902.1,144.2 L974.1,144.2 L974.1,150.6 L902.1,150.6 Z M1005.7,150.6 L1005.7,175.7 L1081.5,175.7 L1081.5,182.1 L994.5,182.1 L994.5,115.2 L1080.3,115.2 L1080.3,121.6 L1005.7,121.6 L1005.7,144.2 L1077.6,144.2 L1077.6,150.6 L1005.7,150.6 Z M1085.9,115.2 L1181.8,115.2 L1181.8,121.6 L1139.4,121.6 L1139.4,182.1 L1128.2,182.1 L1128.2,121.6 L1085.9,121.6 L1085.9,115.2 Z M645.7,172.8 C668.1,124.1 664.4,42.1 664.3,38.6 L664.1,34.3 L659.8,34.3 C595.3,34.3 555.2,9.6 554.8,9.3 L552.4,7.8 L550,9.3 C549.7,9.6 509.7,34.3 445.2,34.3 L440.9,34.3 L440.7,38.6 C440.5,42 436.8,124.1 459.2,172.8 C476.4,210 508.1,237.8 550.9,253.4 L552.4,254 L553.9,253.4 C596.8,237.9 628.5,210 645.7,172.8 Z M433.1,28.5 C507.3,28.5 552.3,0 552.3,0 C552.3,0 597.5,28.5 671.6,28.5 C671.6,28.5 676.1,121.7 651.4,175.3 C634.1,212.6 602.2,244.2 552.2,262.3 C502.4,244.2 470.5,212.6 453.2,175.3 C428.5,121.7 433.1,28.5 433.1,28.5 Z M629.9,76.8 C630.1,70.1 630.1,64.5 630,60.5 L523.9,60.5 C486.9,60.5 479.1,77.1 479.1,91.1 C479.1,101.6 484.9,111 496.3,118.9 L523.5,137.5 C529.9,141.8 532.8,144.4 534.1,145.8 C536.2,147.9 537.2,150 537.2,151.8 C537.2,154.8 537.2,161.2 523.6,161.2 L487,161 C490.3,166.8 494.3,172.4 498.8,177.6 L520.3,177.6 C532.7,177.6 544.6,175.9 551.9,172.5 L551.9,211.1 C562.6,207.2 569.7,202.4 578,196.9 L578,122.7 L625.3,122.7 C626.4,117.1 627.2,111.4 627.9,105.8 L578,105.8 L578,81 C578,78.8 578.4,77.7 578.9,77.4 C578.9,77.3 580.2,76.8 585.1,76.8 L629.9,76.8 Z M552.3,18.4 C561.9,23.8 598.9,42.4 655.4,43.3 C655.8,60.8 656.2,128.5 637.5,169.1 C621.5,203.6 592.1,229.7 552.3,244.5 C512.6,229.6 483.2,203.6 467.3,169.1 C448.6,128.5 449,60.8 449.4,43.3 C505.9,42.4 542.8,23.8 552.3,18.4 Z M507.6,86.1 C507.6,83.1 507.6,76.8 521.2,76.8 L551.9,76.8 L551.9,120 C551.4,119.7 551,119.4 550.5,119.1 L521.1,100.9 C516,97.4 512.5,94.5 510.5,92.4 C508.5,90.3 507.6,88.2 507.6,86.1 Z"], [3, "ngClass", "background-color", "color", "click", 4, "ngIf"], ["class", "img-wrapper", 3, "background-color", 4, "ngIf"], [3, "ngClass", "click"], [1, "img-wrapper"], [1, "sf-menu-logo-bg", 3, "ngStyle", "click"]], template: function TopMenuComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0)(1, "em", 1);
        ɵngcc0.ɵɵlistener("click", function TopMenuComponent_Template_em_click_1_listener() { return ctx.toggleMenu(); });
        ɵngcc0.ɵɵelementEnd()();
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵtemplate(3, TopMenuComponent_div_3_Template, 1, 3, "div", 3);
        ɵngcc0.ɵɵtemplate(4, TopMenuComponent_em_4_Template, 1, 1, "em", 4);
        ɵngcc0.ɵɵtemplate(5, TopMenuComponent__svg_svg_5_Template, 2, 2, "svg", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 6);
        ɵngcc0.ɵɵtemplate(7, TopMenuComponent_ng_container_7_Template, 3, 2, "ng-container", 7);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵstyleProp("color", ctx.iconColor);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.menuConfig.logoUrl);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.menuConfig.logoUrl && ctx.menuConfig.logoClass);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.menuConfig.logoUrl && !ctx.menuConfig.logoClass);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.menuConfig == null ? null : ctx.menuConfig.shortcuts);
    } }, directives: [ɵngcc2.NgIf, ɵngcc2.NgStyle, ɵngcc2.NgClass, ɵngcc2.NgForOf], styles: ["[_nghost-%COMP%]{display:block;display:flex;flex-direction:row;justify-content:space-between;align-items:center;height:70px;border-bottom:4px solid transparent;max-width:100vw}@media screen and (min-width:768px){[_nghost-%COMP%]{display:none}}[_nghost-%COMP%]   .sf-menu-logo-bg[_ngcontent-%COMP%]{background:50%/contain no-repeat;width:8em;height:4em}[_nghost-%COMP%]   .sf-top-menu-middle[_ngcontent-%COMP%]{flex:1 0 4em;height:100%;padding:5px 0;display:flex}[_nghost-%COMP%]   .sf-top-menu-left[_ngcontent-%COMP%], [_nghost-%COMP%]   .sf-top-menu-right[_ngcontent-%COMP%]{height:100%;flex:0 1 100%;width:8em;padding:5px 10px;display:flex;align-items:center;overflow:hidden}[_nghost-%COMP%]   .sf-top-menu-left[_ngcontent-%COMP%] > *[_ngcontent-%COMP%], [_nghost-%COMP%]   .sf-top-menu-right[_ngcontent-%COMP%] > *[_ngcontent-%COMP%]{padding:0 10px;border-radius:4px}[_nghost-%COMP%]   .sf-top-menu-left[_ngcontent-%COMP%]   em[_ngcontent-%COMP%], [_nghost-%COMP%]   .sf-top-menu-left[_ngcontent-%COMP%]   i[_ngcontent-%COMP%], [_nghost-%COMP%]   .sf-top-menu-right[_ngcontent-%COMP%]   em[_ngcontent-%COMP%], [_nghost-%COMP%]   .sf-top-menu-right[_ngcontent-%COMP%]   i[_ngcontent-%COMP%]{font-size:1.6em;height:100%;display:flex;align-items:center}[_nghost-%COMP%]   .sf-top-menu-left[_ngcontent-%COMP%] > div[_ngcontent-%COMP%], [_nghost-%COMP%]   .sf-top-menu-right[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]{height:100%}[_nghost-%COMP%]   .sf-top-menu-left[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]   div[_ngcontent-%COMP%], [_nghost-%COMP%]   .sf-top-menu-right[_ngcontent-%COMP%] > div[_ngcontent-%COMP%]   div[_ngcontent-%COMP%]{width:1.6em}[_nghost-%COMP%]   .sf-top-menu-right[_ngcontent-%COMP%]{justify-content:flex-end}@media screen and (max-width:380px){[_nghost-%COMP%]{height:48px}[_nghost-%COMP%]   .sf-menu-logo-bg[_ngcontent-%COMP%]{height:2.4em;width:4.8em}[_nghost-%COMP%]   .sf-top-menu-left[_ngcontent-%COMP%], [_nghost-%COMP%]   .sf-top-menu-right[_ngcontent-%COMP%]{font-size:.8em}}.sf-menu-top-logo[_ngcontent-%COMP%]{-o-object-fit:contain;object-fit:contain;text-align:center;align-self:center;justify-self:center;height:auto;width:50%;max-height:calc(70px * .7);width:120px;filter:brightness(1) grayscale(1)}"], data: { animation: [
            trigger('appear-fast', [transition(':enter', [style({ opacity: 0 }), animate('80ms 120ms ease-in', style({ opacity: 1 }))])])
        ] } });
TopMenuComponent.ctorParameters = () => [
    { type: SfMenuService }
];
TopMenuComponent.propDecorators = {
    bgColor: [{ type: HostBinding, args: ['style.background-color',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TopMenuComponent, [{
        type: Component,
        args: [{
                selector: 'sf-top-menu',
                template: "<div class=\"sf-top-menu-left\">\n  <em class=\"sf-top-menu-toggle-bars fa fa-2x fa-bars\" (click)=\"toggleMenu()\" [style.color]=\"iconColor\"></em>\n</div>\n\n<div class=\"sf-top-menu-middle\">\n  <div class=\"sf-menu-logo-bg\" *ngIf=\"menuConfig.logoUrl\" [ngStyle]=\"{ 'background-image': 'url(' + menuConfig.logoUrl + ')' }\"></div>\n  <em [ngClass]=\"menuConfig.logoClass\" class=\"sf-menu-top-logo\" *ngIf=\"!menuConfig.logoUrl && menuConfig.logoClass\"></em>\n  <svg\n    class=\"sf-menu-top-logo\"\n    xmlns=\"http://www.w3.org/2000/svg\"\n    viewBox=\"0 0 1182 263\"\n    fill=\"#1B4C86\"\n    *ngIf=\"!menuConfig.logoUrl && !menuConfig.logoClass\"\n    [@appear-fast]\n  >\n    <path\n      [attr.fill]=\"iconColor\"\n      d=\"M55.2,145.2 C74.8,147.2 94.4,150.1 94.4,162.9 C94.4,179.3 70.4,183.8 49.4,183.8 C22,183.8 2.1,178 0.4,160 L11.7,160 C13.7,174 28.8,177.4 49.5,177.4 C65.7,177.4 82.5,174.7 82.5,164.1 C82.5,153.9 62.8,152.7 43.2,150.7 C17.5,148.2 3.7,143.7 3.7,132.9 C3.7,118.3 25.4,113.5 46.3,113.5 C68.3,113.5 90.6,118.4 91.3,134.3 L80.1,134.3 C78.6,122.9 63.6,119.9 46.2,119.9 C28.9,119.9 15.7,124.1 15.7,131.7 C15.7,140.4 28.9,142.5 55.2,145.2 Z M128.3,155.4 L176,155.4 L152.3,122.1 L128.3,155.4 Z M146,115.2 L159,115.2 L208,182.1 L195,182.1 L180.6,161.8 L123.8,161.8 L109.1,182.1 L97,182.1 L146,115.2 Z M217,182.1 L217,115.2 L297.9,115.2 L297.9,121.6 L228.2,121.6 L228.2,144.3 L294.4,144.3 L294.4,150.6 L228.2,150.6 L228.2,182.1 L217,182.1 Z M322.4,150.6 L322.4,175.7 L398.3,175.7 L398.3,182.1 L311.2,182.1 L311.2,115.2 L397.1,115.2 L397.1,121.6 L322.4,121.6 L322.4,144.2 L394.4,144.2 L394.4,150.6 L322.4,150.6 Z M707.1,182.1 L707.1,115.2 L788,115.2 L788,121.6 L718.3,121.6 L718.3,144.3 L784.5,144.3 L784.5,150.6 L718.3,150.6 L718.3,182.1 L707.1,182.1 Z M810.7,115.2 L810.7,175.7 L879.1,175.7 L879.1,182.1 L799.5,182.1 L799.5,115.2 L810.7,115.2 Z M902.1,150.6 L902.1,175.7 L978,175.7 L978,182.1 L890.9,182.1 L890.9,115.2 L976.8,115.2 L976.8,121.6 L902.1,121.6 L902.1,144.2 L974.1,144.2 L974.1,150.6 L902.1,150.6 Z M1005.7,150.6 L1005.7,175.7 L1081.5,175.7 L1081.5,182.1 L994.5,182.1 L994.5,115.2 L1080.3,115.2 L1080.3,121.6 L1005.7,121.6 L1005.7,144.2 L1077.6,144.2 L1077.6,150.6 L1005.7,150.6 Z M1085.9,115.2 L1181.8,115.2 L1181.8,121.6 L1139.4,121.6 L1139.4,182.1 L1128.2,182.1 L1128.2,121.6 L1085.9,121.6 L1085.9,115.2 Z M645.7,172.8 C668.1,124.1 664.4,42.1 664.3,38.6 L664.1,34.3 L659.8,34.3 C595.3,34.3 555.2,9.6 554.8,9.3 L552.4,7.8 L550,9.3 C549.7,9.6 509.7,34.3 445.2,34.3 L440.9,34.3 L440.7,38.6 C440.5,42 436.8,124.1 459.2,172.8 C476.4,210 508.1,237.8 550.9,253.4 L552.4,254 L553.9,253.4 C596.8,237.9 628.5,210 645.7,172.8 Z M433.1,28.5 C507.3,28.5 552.3,0 552.3,0 C552.3,0 597.5,28.5 671.6,28.5 C671.6,28.5 676.1,121.7 651.4,175.3 C634.1,212.6 602.2,244.2 552.2,262.3 C502.4,244.2 470.5,212.6 453.2,175.3 C428.5,121.7 433.1,28.5 433.1,28.5 Z M629.9,76.8 C630.1,70.1 630.1,64.5 630,60.5 L523.9,60.5 C486.9,60.5 479.1,77.1 479.1,91.1 C479.1,101.6 484.9,111 496.3,118.9 L523.5,137.5 C529.9,141.8 532.8,144.4 534.1,145.8 C536.2,147.9 537.2,150 537.2,151.8 C537.2,154.8 537.2,161.2 523.6,161.2 L487,161 C490.3,166.8 494.3,172.4 498.8,177.6 L520.3,177.6 C532.7,177.6 544.6,175.9 551.9,172.5 L551.9,211.1 C562.6,207.2 569.7,202.4 578,196.9 L578,122.7 L625.3,122.7 C626.4,117.1 627.2,111.4 627.9,105.8 L578,105.8 L578,81 C578,78.8 578.4,77.7 578.9,77.4 C578.9,77.3 580.2,76.8 585.1,76.8 L629.9,76.8 Z M552.3,18.4 C561.9,23.8 598.9,42.4 655.4,43.3 C655.8,60.8 656.2,128.5 637.5,169.1 C621.5,203.6 592.1,229.7 552.3,244.5 C512.6,229.6 483.2,203.6 467.3,169.1 C448.6,128.5 449,60.8 449.4,43.3 C505.9,42.4 542.8,23.8 552.3,18.4 Z M507.6,86.1 C507.6,83.1 507.6,76.8 521.2,76.8 L551.9,76.8 L551.9,120 C551.4,119.7 551,119.4 550.5,119.1 L521.1,100.9 C516,97.4 512.5,94.5 510.5,92.4 C508.5,90.3 507.6,88.2 507.6,86.1 Z\"\n    />\n  </svg>\n</div>\n\n<div class=\"sf-top-menu-right\">\n  <ng-container *ngFor=\"let shortcut of menuConfig?.shortcuts\">\n    <em\n      *ngIf=\"shortcut.iconClass\"\n      [ngClass]=\"shortcut.iconClass\"\n      (click)=\"shortcutClicked(shortcut)\"\n      [style.background-color]=\"shortcut.active ? activeLink : ''\"\n      [style.color]=\"iconColor\"\n    ></em>\n    <div *ngIf=\"!shortcut.iconClass && shortcut.iconPath\" class=\"img-wrapper\" [style.background-color]=\"shortcut.active ? activeLink : ''\">\n      <div\n        class=\"sf-menu-logo-bg\"\n        [ngStyle]=\"{ 'background-image': 'url(' + shortcut.iconPath + ')' }\"\n        (click)=\"shortcutClicked(shortcut)\"\n      ></div>\n    </div>\n  </ng-container>\n</div>\n",
                animations: [
                    trigger('appear-fast', [transition(':enter', [style({ opacity: 0 }), animate('80ms 120ms ease-in', style({ opacity: 1 }))])])
                ],
                styles: [":host{display:block;display:flex;flex-direction:row;justify-content:space-between;align-items:center;height:70px;border-bottom:4px solid transparent;max-width:100vw}@media screen and (min-width:768px){:host{display:none}}:host .sf-menu-logo-bg{background:50%/contain no-repeat;width:8em;height:4em}:host .sf-top-menu-middle{flex:1 0 4em;height:100%;padding:5px 0;display:flex}:host .sf-top-menu-left,:host .sf-top-menu-right{height:100%;flex:0 1 100%;width:8em;padding:5px 10px;display:flex;align-items:center;overflow:hidden}:host .sf-top-menu-left>*,:host .sf-top-menu-right>*{padding:0 10px;border-radius:4px}:host .sf-top-menu-left em,:host .sf-top-menu-left i,:host .sf-top-menu-right em,:host .sf-top-menu-right i{font-size:1.6em;height:100%;display:flex;align-items:center}:host .sf-top-menu-left>div,:host .sf-top-menu-right>div{height:100%}:host .sf-top-menu-left>div div,:host .sf-top-menu-right>div div{width:1.6em}:host .sf-top-menu-right{justify-content:flex-end}@media screen and (max-width:380px){:host{height:48px}:host .sf-menu-logo-bg{height:2.4em;width:4.8em}:host .sf-top-menu-left,:host .sf-top-menu-right{font-size:.8em}}.sf-menu-top-logo{-o-object-fit:contain;object-fit:contain;text-align:center;align-self:center;justify-self:center;height:auto;width:50%;max-height:calc(70px * .7);width:120px;filter:brightness(1) grayscale(1)}"]
            }]
    }], function () { return [{ type: SfMenuService }]; }, { bgColor: [{
            type: HostBinding,
            args: ['style.background-color']
        }] }); })();

class SfMenuModule {
}
SfMenuModule.ɵfac = function SfMenuModule_Factory(t) { return new (t || SfMenuModule)(); };
SfMenuModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SfMenuModule });
SfMenuModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [SfMenuService], imports: [[CommonModule, FormsModule, RouterModule, TooltipModule, MenusModule, InputsModule, PopupModule, BrowserAnimationsModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SfMenuModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    SfMenuComponent,
                    ClickOutsideDirective,
                    TopMenuComponent,
                    MenuFooterLogoComponent,
                    MenuUserComponent,
                    CustomerLogoComponent,
                    MenuProductsComponent,
                    MenuShortcutsComponent,
                    MenuToggleBtnComponent,
                    MenuRoutesComponent,
                    MenuItemComponent
                ],
                imports: [CommonModule, FormsModule, RouterModule, TooltipModule, MenusModule, InputsModule, PopupModule, BrowserAnimationsModule],
                providers: [SfMenuService],
                exports: [SfMenuComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SfMenuModule, { declarations: function () { return [SfMenuComponent, ClickOutsideDirective, TopMenuComponent, MenuFooterLogoComponent, MenuUserComponent, CustomerLogoComponent, MenuProductsComponent, MenuShortcutsComponent, MenuToggleBtnComponent, MenuRoutesComponent, MenuItemComponent]; }, imports: function () { return [CommonModule, FormsModule, RouterModule, TooltipModule, MenusModule, InputsModule, PopupModule, BrowserAnimationsModule]; }, exports: function () { return [SfMenuComponent]; } }); })();

/*
 * Public API Surface of sf-menu
 */

/**
 * Generated bundle index. Do not edit.
 */

export { MenuConfig, MenuEventType, MenuItem, SfMenuComponent, SfMenuModule, SfMenuService, SsoManageAccountConfig, ClickOutsideDirective as ɵa, TopMenuComponent as ɵb, MenuFooterLogoComponent as ɵc, MenuUserComponent as ɵd, CustomerLogoComponent as ɵe, MenuProductsComponent as ɵf, MenuShortcutsComponent as ɵg, MenuToggleBtnComponent as ɵh, MenuRoutesComponent as ɵi, MenuItemComponent as ɵj };
