// @progress/kendo-angular-grid/dist/es/filtering/operators/before-eq-filter-operator.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { FilterOperatorBase } from './filter-operator.base';
import { LocalizationService } from '@progress/kendo-angular-l10n';
/*
 * Represents the `LessOrEqualTo` (**Is before or equal to**) date filter operator.
 *
 * For more information and examples, refer to:
 * * [Setting the default filter operators]({% slug builtinfiltertemplate_grid %}#toc-setting-the-default-filter-operator)
 * * [Setting the order of the filter operators]({% slug builtinfiltertemplate_grid %}#toc-setting-the-order-of-the-filter-operators)
 */
var BeforeEqFilterOperatorComponent = /** @class */ (function (_super) {
    tslib_1.__extends(BeforeEqFilterOperatorComponent, _super);
    function BeforeEqFilterOperatorComponent(localization) {
        return _super.call(this, "before-eq", localization) || this;
    }
    BeforeEqFilterOperatorComponent_1 = BeforeEqFilterOperatorComponent;
    /**
     * @hidden
     */
    BeforeEqFilterOperatorComponent.prototype.toJSON = function () {
        return {
            text: this.text,
            value: "lte"
        };
    };
    var BeforeEqFilterOperatorComponent_1;
    BeforeEqFilterOperatorComponent = BeforeEqFilterOperatorComponent_1 = tslib_1.__decorate([
        Component({
            providers: [
                {
                    provide: FilterOperatorBase,
                    useExisting: forwardRef(function () { return BeforeEqFilterOperatorComponent_1; })
                }
            ],
            selector: 'kendo-filter-before-eq-operator',
            template: ""
        }),
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], BeforeEqFilterOperatorComponent);
    return BeforeEqFilterOperatorComponent;
}(FilterOperatorBase));
export { BeforeEqFilterOperatorComponent };
