// @progress/kendo-angular-layout/dist/es/tabstrip/localization/custom-messages.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { TabStripMessages } from './messages';
/**
 * Custom component messages override default component messages
 * ([see example]({% slug rtl_layout %})).
 */
var TabStripCustomMessagesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TabStripCustomMessagesComponent, _super);
    function TabStripCustomMessagesComponent(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    TabStripCustomMessagesComponent_1 = TabStripCustomMessagesComponent;
    Object.defineProperty(TabStripCustomMessagesComponent.prototype, "override", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    var TabStripCustomMessagesComponent_1;
    TabStripCustomMessagesComponent = TabStripCustomMessagesComponent_1 = tslib_1.__decorate([
        Component({
            providers: [
                {
                    provide: TabStripMessages,
                    useExisting: forwardRef(function () { return TabStripCustomMessagesComponent_1; })
                }
            ],
            selector: 'kendo-tabstrip-messages',
            template: ""
        }),
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], TabStripCustomMessagesComponent);
    return TabStripCustomMessagesComponent;
}(TabStripMessages));
export { TabStripCustomMessagesComponent };
