// @progress/kendo-angular-dropdowns/dist/npm/common/list.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var kendo_angular_common_1 = require("@progress/kendo-angular-common");
var list_item_directive_1 = require("./list-item.directive");
var item_template_directive_1 = require("./templates/item-template.directive");
var group_template_directive_1 = require("./templates/group-template.directive");
var fixed_group_template_directive_1 = require("./templates/fixed-group-template.directive");
var util_1 = require("./util");
var selection_service_1 = require("./selection/selection.service");
var rxjs_1 = require("rxjs");
var operators_1 = require("rxjs/operators");
var disabled_items_service_1 = require("./disabled-items/disabled-items.service");
var data_service_1 = require("./data.service");
/**
 * @hidden
 */
var ListComponent = /** @class */ (function () {
    /* tslint:disable:member-ordering */
    function ListComponent(dataService, wrapper, selectionService, disabledItemsService, cdr, zone, renderer) {
        var _this = this;
        this.dataService = dataService;
        this.wrapper = wrapper;
        this.selectionService = selectionService;
        this.disabledItemsService = disabledItemsService;
        this.cdr = cdr;
        this.zone = zone;
        this.renderer = renderer;
        this.selected = [];
        this.focused = -1;
        this.show = true;
        this.multipleSelection = false;
        this.type = 'list';
        this.checkboxes = { enabled: false };
        this.rounded = 'medium';
        this.onClick = new core_1.EventEmitter();
        this.pageChange = new core_1.EventEmitter();
        this.listResize = new core_1.EventEmitter();
        this.startFrom = 0;
        this.lastLoaded = 0;
        this.lastScrollTop = 0;
        this.scrollToFocused = false;
        this._size = 'medium';
        this.selectSubscription = rxjs_1.merge(this.selectionService.onSelect.pipe(operators_1.map(function (args) { return args.indices[0]; })), this.selectionService.onFocus)
            .pipe(
        // handle only the very last onSelect/onFocus emission
        operators_1.switchMap(function (event) { return _this.zone.onStable.pipe(operators_1.take(1), operators_1.map(function () { return event; })); }))
            .subscribe(this.scrollToItem.bind(this));
        this.prepareClasses();
    }
    Object.defineProperty(ListComponent.prototype, "data", {
        get: function () {
            return this._data;
        },
        set: function (data) {
            this._data = data[0] && data[0].header ? data.slice(0) : data;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListComponent.prototype, "size", {
        get: function () {
            return this._size;
        },
        set: function (size) {
            if (this.type === 'list') {
                this.renderer.removeClass(this.wrapper.nativeElement, util_1.getSizeClass('list', this.size));
                if (size) {
                    this.renderer.addClass(this.wrapper.nativeElement, util_1.getSizeClass('list', size));
                }
                this._size = size;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListComponent.prototype, "pageSize", {
        get: function () {
            if (this.virtual.pageSize) {
                return this.virtual.pageSize;
            }
            var size = Math.round(this.height / this.virtual.itemHeight);
            return size;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListComponent.prototype, "scrollHeight", {
        get: function () {
            return (this.dataService.grouped ? this.virtual.total - 1 : this.virtual.total) * this.virtual.itemHeight;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListComponent.prototype, "overflowY", {
        get: function () {
            if (util_1.isPresent(this.virtual)) {
                var overflow = this.hasVirtualScrollbar() ? 'scroll' : 'hidden';
                return overflow;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListComponent.prototype, "checkboxClasses", {
        /**
         * @hidden
         */
        get: function () {
            return (this.size ? util_1.getSizeClass('checkbox', this.size) : '') + " " + (this.rounded ? util_1.getRoundedClass(this.rounded) : '');
        },
        enumerable: true,
        configurable: true
    });
    ListComponent.prototype.ngOnChanges = function (changes) {
        if (kendo_angular_common_1.isChanged('data', changes, false)) {
            if (this.lastLoaded <= 0) {
                this.lastLoaded = this.data.length - 1;
                this.scrollToFocused = !changes.data.isFirstChange();
            }
            this.setOverflow();
        }
        if (kendo_angular_common_1.isChanged('virtual', changes, false)) {
            this.setOverflow();
        }
        if (kendo_angular_common_1.isChanged('type', changes, false)) {
            this.prepareClasses();
        }
    };
    ListComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        this.setComponentClasses();
        this.zone.runOutsideAngular(function () {
            _this.scrollSubscription = rxjs_1.fromEvent(_this.content.nativeElement, "scroll").pipe(operators_1.auditTime(100), operators_1.tap(_this.prefetchData.bind(_this)), operators_1.tap(_this.findCurrentGroup.bind(_this))).subscribe(function () {
                _this.lastScrollTop = _this.content.nativeElement.scrollTop;
            });
        });
        this.setOverflow();
    };
    ListComponent.prototype.ngAfterViewChecked = function () {
        if (this.virtual) {
            this.positionItems();
        }
        if (this.items && this.scrollToFocused) {
            this.scrollToFocused = false;
            var scrollTarget = this.items.length && this.selectionService.focused === -1 ? 0 : this.selectionService.focused;
            this.scrollToItem(scrollTarget);
        }
        if (this.dataService.grouped) {
            this.findCurrentGroup();
        }
    };
    ListComponent.prototype.ngOnDestroy = function () {
        this.selectSubscription.unsubscribe();
        if (this.scrollSubscription) {
            this.scrollSubscription.unsubscribe();
        }
    };
    ListComponent.prototype.onCheckedChange = function (e, index) {
        var isChecked = e.target['checked'];
        if (isChecked && !this.selectionService.isSelected(index)) {
            this.selectionService.add(index);
        }
        if (!isChecked && this.selectionService.isSelected(index)) {
            this.selectionService.unselect(index);
        }
    };
    ListComponent.prototype.prepareClasses = function () {
        if (this.type === 'list') {
            this.listContentClass = 'k-list-content';
            this.listClass = 'k-list-ul';
            this.listItemClass = 'k-list-item';
            this.listVirtualClass = 'k-virtual-list';
            this.listGroupStickyHeaderClass = 'k-list-group-sticky-header';
            this.listGroupStickyHeaderTextClass = 'k-list-header-text';
            this.listGroupItemClass = 'k-list-group-item';
            this.listGroupItemTextClass = 'k-list-item-text';
        }
        else {
            this.listContentClass = 'k-table-body k-table-scroller';
            this.listClass = 'k-table k-table-list';
            this.listItemClass = 'k-table-row';
            this.listVirtualClass = 'k-virtual-table';
            this.listGroupStickyHeaderClass = 'k-table-group-sticky-header';
            this.listGroupStickyHeaderTextClass = 'k-table-th';
            this.listGroupItemClass = 'k-table-group-row';
            this.listGroupItemTextClass = 'k-table-th';
        }
    };
    ListComponent.prototype.isChecked = function (index) {
        var normalizedIndex = this.virtual ? index + this.virtual.skip : index;
        return this.selectionService.isSelected(normalizedIndex);
    };
    ListComponent.prototype.firstVisibleItem = function () {
        var _this = this;
        var content = this.content.nativeElement;
        var rect = content.getBoundingClientRect();
        // IE9 hack
        var disabled = Array.prototype.slice.call(content.querySelectorAll(".k-disabled"));
        // This is a workaround for finding elements with pointer-events: none;
        disabled.forEach(function (el) { return _this.renderer.setStyle(el, "pointer-events", "auto"); });
        var item = document.elementFromPoint(rect.left + 1, rect.top + 1);
        disabled.forEach(function (el) { return _this.renderer.setStyle(el, "pointer-events", "none"); });
        // return the closest `li` item to cover the custom template scenario
        return util_1.closest(item, 'li');
    };
    ListComponent.prototype.findCurrentGroup = function () {
        if (!this.dataService.grouped) {
            this.currentGroup = undefined;
            return;
        }
        var item = this.firstVisibleItem();
        if (item) {
            var index = void 0;
            if (item.getAttribute("role") === "group") {
                index = parseInt(item.getAttribute("group-index"), 10);
                this.currentGroup = this.dataService.groupAt(index).value;
            }
            else {
                index = parseInt(item.getAttribute("index"), 10);
                this.currentGroup = this.dataService.itemGroup(this.dataService.itemAt(index));
            }
        }
        else {
            this.currentGroup = undefined;
        }
        this.cdr.detectChanges();
    };
    ListComponent.prototype.prefetchData = function () {
        if (!this.virtual) {
            return;
        }
        var visibleItems = Math.trunc(this.content.nativeElement.clientHeight / this.virtual.itemHeight);
        var offsetY = this.content.nativeElement.scrollTop;
        var start = Math.trunc(offsetY / this.virtual.itemHeight);
        var down = offsetY > this.lastScrollTop;
        var nextPage = (start + visibleItems >= this.lastLoaded) && this.lastLoaded < this.virtual.total - 1;
        var leftOver = this.pageSize - (this.lastLoaded - this.startFrom);
        var prevPage = this.lastLoaded - this.pageSize + visibleItems >= start - leftOver;
        if (down && nextPage) {
            this.changePage(start);
        }
        if (!down && prevPage) {
            this.changePage(start - this.pageSize + visibleItems + 1);
        }
    };
    ListComponent.prototype.changePage = function (start) {
        var _this = this;
        this.zone.run(function () {
            var end = _this.pageSize + start;
            if (end > _this.virtual.total) {
                start--;
                end = _this.virtual.total;
            }
            if (start < 0) {
                start = 0;
            }
            _this.startFrom = start;
            _this.lastLoaded = end;
            _this.pageChange.emit({ skip: start, take: _this.pageSize });
        });
    };
    ListComponent.prototype.index = function (groupIndex, itemIndex) {
        return groupIndex > 0 ? (this.dataService.groupIndices[groupIndex - 1] + itemIndex) : itemIndex;
    };
    ListComponent.prototype.getText = function (dataItem) {
        return util_1.getter(dataItem, this.textField);
    };
    ListComponent.prototype.getValue = function (dataItem) {
        return util_1.getter(dataItem, this.valueField);
    };
    ListComponent.prototype.isDisabled = function (index) {
        if (util_1.isPresent(this.virtual)) {
            index += this.virtual.skip;
        }
        return this.disabledItemsService.isIndexDisabled(index);
    };
    ListComponent.prototype.isAltRow = function (index) {
        return this.type === 'dropdowngrid' && index % 2 !== 0;
    };
    ListComponent.prototype.scrollToItem = function (index) {
        var flatIndex = index;
        if (this.dataService.grouped) {
            // takes into account the group header items
            flatIndex = this.dataService.flatIndex(index);
            /* The first group header item is not rendered in the list (see template), so subtract 1 when calulating the flat index.
               With virtualization enabled, the first group header could be in a previous page, in which case don't subtract anything. */
            var groupHeaderOffset = this.firstGroupHeaderInTargetedPage(flatIndex) ? -1 : 0;
            flatIndex += groupHeaderOffset;
        }
        if (this.virtual && flatIndex > -1) {
            this.scrollToIndex(flatIndex);
            return;
        }
        var items = this.items.toArray();
        if (util_1.isPresent(items[flatIndex]) && flatIndex !== -1) {
            this.scroll(items[flatIndex].element);
        }
    };
    ListComponent.prototype.scrollToIndex = function (index) {
        var content = this.content.nativeElement;
        var contentScrollTop = content.scrollTop;
        var itemOffsetTop = index * this.virtual.itemHeight;
        var itemOffsetHeight = this.virtual.itemHeight;
        var contentOffsetHeight = content.clientHeight;
        var bottomDistance = itemOffsetTop + itemOffsetHeight;
        if (contentScrollTop > itemOffsetTop) {
            contentScrollTop = itemOffsetTop;
        }
        else if (bottomDistance > (contentScrollTop + contentOffsetHeight)) {
            contentScrollTop = (bottomDistance - contentOffsetHeight);
        }
        content.scrollTop = contentScrollTop;
    };
    ListComponent.prototype.scroll = function (item) {
        if (!item) {
            return;
        }
        var nativeElement = item.nativeElement;
        var content = this.content.nativeElement, itemOffsetTop = nativeElement.offsetTop, itemOffsetHeight = nativeElement.offsetHeight, contentScrollTop = content.scrollTop, contentOffsetHeight = content.clientHeight, bottomDistance = itemOffsetTop + itemOffsetHeight;
        if (contentScrollTop > itemOffsetTop) {
            contentScrollTop = itemOffsetTop;
        }
        else if (bottomDistance > (contentScrollTop + contentOffsetHeight)) {
            contentScrollTop = (bottomDistance - contentOffsetHeight);
        }
        content.scrollTop = contentScrollTop;
    };
    /**
     * Indicates whether a scrollbar is currently rendered in the list.
     */
    ListComponent.prototype.hasScrollbar = function () {
        if (!(util_1.isPresent(this.items) && this.items.length && util_1.isPresent(this.list) && util_1.isPresent(this.content))) {
            return false;
        }
        var hasVirtualScroll = util_1.isPresent(this.virtual) && this.hasVirtualScrollbar();
        return hasVirtualScroll || this.list.nativeElement.scrollHeight > this.content.nativeElement.offsetHeight;
    };
    /**
     * Sets the list's content overflow (hides/shows scrollbar)
     */
    ListComponent.prototype.setOverflow = function () {
        if (this.virtual) {
            var overflow = this.hasVirtualScrollbar() ? 'scroll' : 'hidden';
            this.renderer.setStyle(this.content.nativeElement, 'overflow-y', overflow);
        }
    };
    /**
     * Indicates whether the scrollbar should be visible in virtual mode.
     */
    ListComponent.prototype.hasVirtualScrollbar = function () {
        var contentOffsetHeight = this.content.nativeElement.offsetHeight;
        var virtualOffsetHeight = this.virtualContainer && this.virtualContainer.nativeElement.offsetHeight;
        return this.virtualContainer && virtualOffsetHeight > contentOffsetHeight;
    };
    ListComponent.prototype.positionItems = function () {
        var _this = this;
        this.items.forEach(function (item, index) {
            var offsetY = (index + _this.startFrom) * _this.virtual.itemHeight;
            _this.renderer.setStyle(item.element.nativeElement, "transform", "translateY(" + offsetY + "px");
        });
    };
    /**
     * Indicates whether the first group header from the data set is in the targeted virtual page.
     */
    ListComponent.prototype.firstGroupHeaderInTargetedPage = function (itemIndex) {
        if (!util_1.isPresent(this.virtual)) {
            return true;
        }
        return this.virtual.skip === 0 && (this.virtual.pageSize > itemIndex);
    };
    ListComponent.prototype.setComponentClasses = function () {
        if (this.type === 'list') {
            this.renderer.addClass(this.wrapper.nativeElement, 'k-list');
            if (this.size) {
                this.renderer.addClass(this.wrapper.nativeElement, util_1.getSizeClass('list', this.size));
            }
        }
        if (util_1.isPresent(this.virtual)) {
            this.renderer.addClass(this.wrapper.nativeElement, this.listVirtualClass);
        }
    };
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Array)
    ], ListComponent.prototype, "selected", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Number)
    ], ListComponent.prototype, "focused", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], ListComponent.prototype, "textField", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], ListComponent.prototype, "valueField", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Number)
    ], ListComponent.prototype, "height", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", item_template_directive_1.ItemTemplateDirective)
    ], ListComponent.prototype, "template", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", group_template_directive_1.GroupTemplateDirective)
    ], ListComponent.prototype, "groupTemplate", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", fixed_group_template_directive_1.FixedGroupTemplateDirective)
    ], ListComponent.prototype, "fixedGroupTemplate", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], ListComponent.prototype, "show", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], ListComponent.prototype, "id", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], ListComponent.prototype, "optionPrefix", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], ListComponent.prototype, "multipleSelection", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], ListComponent.prototype, "virtual", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], ListComponent.prototype, "type", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], ListComponent.prototype, "checkboxes", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Array),
        tslib_1.__metadata("design:paramtypes", [Array])
    ], ListComponent.prototype, "data", null);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], ListComponent.prototype, "size", null);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], ListComponent.prototype, "rounded", void 0);
    tslib_1.__decorate([
        core_1.Output(),
        tslib_1.__metadata("design:type", core_1.EventEmitter)
    ], ListComponent.prototype, "onClick", void 0);
    tslib_1.__decorate([
        core_1.Output(),
        tslib_1.__metadata("design:type", core_1.EventEmitter)
    ], ListComponent.prototype, "pageChange", void 0);
    tslib_1.__decorate([
        core_1.Output(),
        tslib_1.__metadata("design:type", core_1.EventEmitter)
    ], ListComponent.prototype, "listResize", void 0);
    tslib_1.__decorate([
        core_1.ViewChildren(list_item_directive_1.ListItemDirective),
        tslib_1.__metadata("design:type", core_1.QueryList)
    ], ListComponent.prototype, "items", void 0);
    tslib_1.__decorate([
        core_1.ViewChild('content', { static: true }),
        tslib_1.__metadata("design:type", core_1.ElementRef)
    ], ListComponent.prototype, "content", void 0);
    tslib_1.__decorate([
        core_1.ViewChild('list', { static: true }),
        tslib_1.__metadata("design:type", core_1.ElementRef)
    ], ListComponent.prototype, "list", void 0);
    tslib_1.__decorate([
        core_1.ViewChild('virtualContainer', { static: false }),
        tslib_1.__metadata("design:type", core_1.ElementRef)
    ], ListComponent.prototype, "virtualContainer", void 0);
    ListComponent = tslib_1.__decorate([
        core_1.Component({
            selector: 'kendo-list',
            template: "\n    <div *ngIf=\"dataService.grouped\"\n        [class]=\"listGroupStickyHeaderClass\"\n        [ngStyle]=\"{\n            'height.px': virtual?.itemHeight,\n            'minHeight.px' : virtual?.itemHeight,\n            'boxSizing' : virtual ? 'border-box' : 'inherit'}\"\n        >\n        <ng-template *ngIf=\"fixedGroupTemplate\"\n            [templateContext]=\"{\n                templateRef: fixedGroupTemplate.templateRef,\n                $implicit: currentGroup\n            }\">\n        </ng-template>\n        <ng-template [ngIf]=\"!fixedGroupTemplate\"><span [class]=\"listGroupStickyHeaderTextClass\">{{ currentGroup }}</span></ng-template>\n    </div>\n    <div #content\n        [class]=\"listContentClass\"\n        [style.overscrollBehavior]=\"'none'\"\n        [style.maxHeight.px]=\"height\"\n        unselectable=\"on\">\n    <ul #list\n        role=\"listbox\"\n        [class]=\"listClass\"\n        [attr.id]=\"id\"\n        [attr.aria-hidden]=\"!show\">\n         <ng-template *ngIf=\"!dataService.grouped && show\" ngFor let-dataItem let-itemIndex=\"index\" [ngForOf]=\"data\">\n            <li\n                role=\"option\"\n                kendoDropDownsSelectable\n                [checkboxes]=\"checkboxes\"\n                [height]=\"virtual?.itemHeight\"\n                [index]=\"itemIndex + startFrom\"\n                [multipleSelection]=\"multipleSelection\"\n                [attr.id]=\"optionPrefix + '-' + getValue(dataItem)\"\n                [attr.tabIndex]=\"-1\"\n                [class]=\"listItemClass\"\n                [ngClass]=\"{\n                    'k-disabled': isDisabled(itemIndex),\n                    'k-table-alt-row': isAltRow(itemIndex)\n                }\"\n            >\n                <input\n                    *ngIf=\"checkboxes.enabled\"\n                    type=\"checkbox\"\n                    class=\"k-checkbox\"\n                    [ngClass]=\"checkboxClasses\"\n                    (change)=\"onCheckedChange($event, itemIndex)\"\n                    [checked]=\"isChecked(itemIndex)\"\n                />\n                <ng-template *ngIf=\"template\"\n                    [templateContext]=\"{\n                        templateRef: template.templateRef,\n                        $implicit: dataItem\n                    }\">\n                </ng-template>\n                <ng-template [ngIf]=\"!template\"><span class=\"k-list-item-text\">{{ getText(dataItem) }}</span></ng-template>\n            </li>\n         </ng-template>\n         <ng-template *ngIf=\"dataService.grouped\" ngFor let-dataItem let-itemIndex=\"index\" [ngForOf]=\"data\">\n            <li\n                *ngIf=\"dataItem.header && dataItem.index > 0\"\n                role=\"group\"\n                [class]=\"listGroupItemClass\"\n                [class.k-table-alt-row]=\"isAltRow(itemIndex - 1)\"\n                [ngStyle]=\"{\n                    'height.px': virtual?.itemHeight,\n                    'minHeight.px' : virtual?.itemHeight,\n                    'boxSizing' : virtual ? 'border-box' : 'inherit'}\"\n                [attr.group-index]=\"dataItem.index\"\n                [attr.id]=\"optionPrefix + '-' + getValue(dataItem.value)\"\n                [attr.tabIndex]=\"-1\">\n                    <span [class]=\"listGroupItemTextClass\">\n                        <ng-template *ngIf=\"groupTemplate\"\n                            [templateContext]=\"{\n                                templateRef: groupTemplate.templateRef,\n                                $implicit: dataItem.value\n                        }\">\n                        </ng-template>\n                        <ng-template [ngIf]=\"!groupTemplate\">{{ dataItem.value }}</ng-template>\n                    </span>\n            </li>\n            <li\n                *ngIf=\"!dataItem.header\"\n                role=\"option\"\n                kendoDropDownsSelectable\n                [height]=\"virtual?.itemHeight\"\n                [index]=\"dataItem.offsetIndex\"\n                [multipleSelection]=\"multipleSelection\"\n                [attr.absolute-index]=\"dataItem.index\"\n                [attr.id]=\"optionPrefix + '-' + getValue(dataItem.value)\"\n                [attr.tabIndex]=\"-1\"\n                [class]=\"listItemClass\"\n                [ngClass]=\"{\n                    'k-disabled': isDisabled(dataItem.offsetIndex),\n                    'k-table-alt-row': isAltRow(itemIndex - 1)\n                }\"\n            >\n                <ng-template *ngIf=\"template\"\n                    [templateContext]=\"{\n                        templateRef: template.templateRef,\n                        $implicit: dataItem.value\n                    }\">\n                </ng-template>\n                <ng-template [ngIf]=\"!template\"><span class=\"k-list-item-text\">{{ getText(dataItem.value) }}</span></ng-template>\n            </li>\n        </ng-template>\n        <kendo-resize-sensor\n            *ngIf=\"!virtual\"\n            (resize)=\"listResize.emit()\"\n        >\n        </kendo-resize-sensor>\n    </ul>\n    <div *ngIf=\"virtual\" #virtualContainer class=\"k-height-container\" role=\"presentation\">\n        <div [style.height.px]=\"scrollHeight\">\n            <kendo-resize-sensor (resize)=\"listResize.emit()\"></kendo-resize-sensor>\n        </div>\n    </div>\n    </div>\n  "
        }),
        tslib_1.__metadata("design:paramtypes", [data_service_1.DataService,
            core_1.ElementRef,
            selection_service_1.SelectionService,
            disabled_items_service_1.DisabledItemsService,
            core_1.ChangeDetectorRef,
            core_1.NgZone,
            core_1.Renderer2])
    ], ListComponent);
    return ListComponent;
}());
exports.ListComponent = ListComponent;
