// @safe-fleet/sf-menu/esm2015/lib/sf-menu.service.js
import { BehaviorSubject, Subject } from 'rxjs';
import { Injectable } from '@angular/core';
import { NavigationEnd, Router } from '@angular/router';
import { MenuConfig } from './models/menu-config';
import { MenuEventType } from './models/menu-events';
import { filter } from 'rxjs/operators';
export class SfMenuService {
    constructor(router) {
        this.router = router;
        this.SMALL_RESOLUTION = 768;
        this.menuCollapse$ = new BehaviorSubject(false);
        this.events$ = new Subject();
        this.menuConfig$ = new BehaviorSubject(undefined);
        // store nodes to expand on navigation end
        this.nodesToExpand = [];
        this.themes = ['light', 'dark'];
        // theming
        this.setBaseColors();
        this.currentTheme = new BehaviorSubject(this.themes[0]);
        this.storedTheme = localStorage.getItem('theme');
        if (!this.storedTheme) {
            this.storedTheme = window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches ? this.themes[1] : this.themes[0];
        }
        this.setMenuConfig(new MenuConfig());
        this.router.events.pipe(filter((e) => e instanceof NavigationEnd)).subscribe((e) => {
            var _a, _b;
            this.nodesToExpand.length = 0;
            (_b = (_a = this.menuConfig) === null || _a === void 0 ? void 0 : _a.routes) === null || _b === void 0 ? void 0 : _b.forEach((r) => this.setCurrentRouteActive(r, e.urlAfterRedirects));
        });
    }
    get colors() {
        return this.menuConfig.menuIgnoresTheme || this.menuConfig.denyChangeTheme ? this.noThemeColors : this.themeColors;
    }
    setMenuConfig(cfg) {
        this.menuConfig = cfg || new MenuConfig();
        this.menuConfig$.next(this.menuConfig);
        this.setTheme(this.storedTheme);
    }
    setBaseColors() {
        const notheme = {};
        const theme = {};
        const root = getComputedStyle(document.querySelector(':root'));
        ['blue', 'gray', 'green', 'red', 'orange', 'purple', 'teal', 'yellow'].forEach((col) => {
            for (let i = 1; i < 10; i++) {
                notheme[`${col}${i * 100}`] = root.getPropertyValue(`--clr-${col}-${1000 - i * 100}`);
                theme[`${col}${i * 100}`] = `var(--clr-${col}-${i * 100})`;
            }
        });
        this.noThemeColors = notheme;
        this.themeColors = theme;
    }
    routeClicked(item, evt) {
        evt.stopPropagation();
        this.events$.next({
            type: MenuEventType.RouteClick,
            value: item
        });
        if (item.items) {
            item.expanded = !item.expanded;
            return;
        }
        if (item.routePath) {
            if (!this.menuConfig.routeClickPreventCollapse) {
                this.collapseAllItems(this.menuConfig.routes);
            }
            this.router.navigate([item.routePath]);
            return;
        }
        if (item.routeHref) {
            if (!this.menuConfig.routeClickPreventCollapse) {
                this.collapseAllItems(this.menuConfig.routes);
            }
            if (item.newTab) {
                window.open(item.routeHref);
                return;
            }
            if (item.newWindow) {
                let top;
                let left;
                let width;
                let height;
                if (item.newWindowWidth > 200 && item.newWindowHeight > 100) {
                    width = item.newWindowWidth;
                    height = item.newWindowHeight;
                    left = window.screenLeft + window.outerWidth / 2 - width / 2;
                    top = window.screenTop + window.outerHeight / 2 - height / 2;
                }
                else {
                    height = window.outerHeight;
                    width = window.outerWidth;
                    left = window.screenLeft;
                    top = window.screenTop;
                }
                window.open(item.routeHref, item.id, `top=${top},left=${left},height=${height},width=${width},resizable=1,scrollbars=1`);
            }
        }
        if (window.innerWidth <= this.SMALL_RESOLUTION && !this.menuCollapse$.value) {
            this.menuCollapse$.next(!this.menuCollapse$.value);
        }
    }
    setCurrentRouteActive(route, url) {
        var _a, _b;
        route.active = !((_a = route.items) === null || _a === void 0 ? void 0 : _a.length) && `/${route.routePath}` === url;
        if (route.active) {
            this.nodesToExpand.forEach((n) => (n.expanded = true));
        }
        this.nodesToExpand.push(route);
        (_b = route.items) === null || _b === void 0 ? void 0 : _b.forEach((ch) => this.setCurrentRouteActive(ch, url));
        this.nodesToExpand.pop();
    }
    // theming
    setTheme(theme) {
        if (this.menuConfig.denyChangeTheme || this.themes.indexOf(theme) < 0) {
            theme = this.themes[0];
        }
        this.currentTheme.next(theme);
        localStorage.setItem('theme', theme);
        this.themes.forEach((t) => document.querySelector('html').classList.toggle(t, t === theme));
    }
    collapseAllItems(routes) {
        if (routes) {
            routes.forEach((route) => {
                route.expanded = false;
                if (route.items) {
                    this.collapseAllItems(route.items);
                }
            });
        }
    }
}
SfMenuService.decorators = [
    { type: Injectable }
];
SfMenuService.ctorParameters = () => [
    { type: Router }
];
