// @progress/kendo-angular-grid/dist/es2015/rendering/common/spacer.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, HostBinding, Input } from '@angular/core';
import { isPresent } from '../../utils';
/**
 * Represents the Kendo UI GridSpacer component for Angular.
 * Used to give additional white space between the Pager inner elements,
 * and provides a way for customizing the spacer width.
 * It can also be used in any flex container within the Grid.
 */
let GridSpacerComponent = class GridSpacerComponent {
    /**
     * Represents the Kendo UI GridSpacer component for Angular.
     * Used to give additional white space between the Pager inner elements,
     * and provides a way for customizing the spacer width.
     * It can also be used in any flex container within the Grid.
     */
    constructor() {
        this.hostClass = true;
    }
    get sizedClass() {
        return isPresent(this.width);
    }
    get flexBasisStyle() {
        return this.width;
    }
};
tslib_1.__decorate([
    HostBinding('class.k-spacer'),
    tslib_1.__metadata("design:type", Boolean)
], GridSpacerComponent.prototype, "hostClass", void 0);
tslib_1.__decorate([
    HostBinding('class.k-spacer-sized'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], GridSpacerComponent.prototype, "sizedClass", null);
tslib_1.__decorate([
    HostBinding('style.flexBasis'),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [])
], GridSpacerComponent.prototype, "flexBasisStyle", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], GridSpacerComponent.prototype, "width", void 0);
GridSpacerComponent = tslib_1.__decorate([
    Component({
        selector: 'kendo-grid-spacer, kendo-pager-spacer',
        template: ``
    })
], GridSpacerComponent);
export { GridSpacerComponent };
