// @progress/kendo-angular-dateinputs/dist/es2015/dateinput/localization/dateinput-localized-messages.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
var DateInputLocalizedMessagesDirective_1;
import { Directive, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { DateInputMessages } from './messages';
/**
 * @hidden
 */
let DateInputLocalizedMessagesDirective = DateInputLocalizedMessagesDirective_1 = class DateInputLocalizedMessagesDirective extends DateInputMessages {
    constructor(service) {
        super();
        this.service = service;
    }
};
DateInputLocalizedMessagesDirective = DateInputLocalizedMessagesDirective_1 = tslib_1.__decorate([
    Directive({
        providers: [
            {
                provide: DateInputMessages,
                useExisting: forwardRef(() => DateInputLocalizedMessagesDirective_1) // tslint:disable-line:no-forward-ref
            }
        ],
        selector: '[kendoDateInputLocalizedMessages]'
    }),
    tslib_1.__metadata("design:paramtypes", [LocalizationService])
], DateInputLocalizedMessagesDirective);
export { DateInputLocalizedMessagesDirective };
