// @progress/kendo-angular-tooltip/dist/npm/popover/popover.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
var rxjs_1 = require("rxjs");
var kendo_licensing_1 = require("@progress/kendo-licensing");
var package_metadata_1 = require("../package-metadata");
var constants_1 = require("../constants");
var title_template_directive_1 = require("./template-directives/title-template.directive");
var body_template_directive_1 = require("./template-directives/body-template.directive");
var actions_template_directive_1 = require("./template-directives/actions-template.directive");
/**
 * Represents the [Kendo UI Popover component for Angular]({% slug overview_popover %}).
 * Used to display additional information that is related to a target element.
 *
 * @example
 * ```ts-no-run
 * <kendo-popover>
 *    <ng-template kendoPopoverTitleTemplate>Foo Title</ng-template>
 *    <ng-template kendoPopoverBodyTemplate>Foo Body</ng-template>
 *    <ng-template kendoPopoverActionsTemplate>Foo Actions</ng-template>
 * </kendo-popover>
 * ```
 */
var PopoverComponent = /** @class */ (function () {
    function PopoverComponent(localization) {
        this.localization = localization;
        /**
         * Specifies the position of the Popover in relation to its anchor element. [See example]({% slug positioning_popover %})
         *
         * The possible options are:
         * `top`
         * `bottom`
         * `right` (Default)
         * `left`
         */
        this.position = 'right';
        /**
         * Determines whether a callout will be rendered along the Popover. [See example]({% slug callout_popover %})
         *
         * @default true
         */
        this.callout = true;
        /**
         * Enables and configures the Popover animation. [See example]({% slug animations_popover %})
         *
         * The possible options are:
         *
         * * `boolean`&mdash;Enables the default animation
         * * `PopoverAnimation`&mdash;A configuration object which allows setting the `direction`, `duration` and `type` of the animation.
         *
         * @default false
         */
        this.animation = false;
        /**
         * @hidden
         * Determines the visibility of the Popover.
         */
        this.visible = false;
        /**
         * Fires before the Popover is about to be shown ([see example]({% slug events_popover %})).
         * The event is preventable. If canceled, the Popover will not be displayed. [See example]({% slug events_popover %})
         */
        this.show = new core_1.EventEmitter();
        /**
         * Fires after the Popover has been shown and the animation has ended. [See example]({% slug events_popover %})
         */
        this.shown = new core_1.EventEmitter();
        /**
         * Fires when the Popover is about to be hidden ([see example]({% slug events_popover %})).
         * The event is preventable. If canceled, the Popover will remain visible.
         */
        this.hide = new core_1.EventEmitter();
        /**
         * Fires after the Popover has been hidden and the animation has ended. [See example]({% slug events_popover %})
         */
        this.hidden = new core_1.EventEmitter();
        this._offset = 6;
        this._width = 'auto';
        this._height = 'auto';
        this.subs = new rxjs_1.Subscription();
        /**
         * @hidden
         */
        this._templateData = function () { return null; };
        kendo_licensing_1.validatePackage(package_metadata_1.packageMetadata);
    }
    Object.defineProperty(PopoverComponent.prototype, "offset", {
        get: function () {
            var calloutBuffer = 14;
            return this.callout
                ? calloutBuffer + this._offset
                : this._offset;
        },
        /**
         * Specifies the distance from the Popover to its anchor element in pixels.
         *
         * @default `6`
         */
        set: function (value) {
            this._offset = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PopoverComponent.prototype, "width", {
        get: function () {
            return this._width;
        },
        /**
         * Determines the width of the popover. Numeric values are treated as pixels.
         * @default 'auto'
         */
        set: function (value) {
            this._width = typeof value === 'number' ? value + "px" : value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PopoverComponent.prototype, "height", {
        get: function () {
            return this._height;
        },
        /**
         * Determines the height of the popover. Numeric values are treated as pixels.
         * @default 'auto'
         */
        set: function (value) {
            this._height = typeof value === 'number' ? value + "px" : value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PopoverComponent.prototype, "templateData", {
        get: function () {
            return this._templateData;
        },
        /**
         * Defines a callback function which returns custom data passed to the Popover templates.
         * It exposes the current anchor element as an argument. [See example]({% slug templates_popover %}#toc-popoverdatacallback)
         */
        set: function (fn) {
            if (core_1.isDevMode && typeof fn !== 'function') {
                throw new Error(constants_1.ERRORS.templateData + " " + JSON.stringify(fn) + ".");
            }
            this._templateData = fn;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PopoverComponent.prototype, "isHidden", {
        /**
         * @hidden
         */
        get: function () {
            return !this.visible;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PopoverComponent.prototype, "hasAttributeHidden", {
        /**
         * @hidden
         */
        get: function () {
            return !this.visible;
        },
        enumerable: true,
        configurable: true
    });
    PopoverComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.subs.add(this.localization.changes.subscribe(function (_a) {
            var rtl = _a.rtl;
            _this.direction = rtl ? 'rtl' : 'ltr';
        }));
    };
    PopoverComponent.prototype.ngOnDestroy = function () {
        this.subs.unsubscribe();
    };
    /**
     * @hidden
     */
    PopoverComponent.prototype.getCalloutPosition = function () {
        switch (this.position) {
            case 'top': return { 'k-callout-s': true };
            case 'bottom': return { 'k-callout-n': true };
            case 'left': return { 'k-callout-e': true };
            case 'right': return { 'k-callout-w': true };
            default: return { 'k-callout-s': true };
        }
    };
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], PopoverComponent.prototype, "position", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Number),
        tslib_1.__metadata("design:paramtypes", [Number])
    ], PopoverComponent.prototype, "offset", null);
    tslib_1.__decorate([
        core_1.HostBinding('style.width'),
        core_1.Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], PopoverComponent.prototype, "width", null);
    tslib_1.__decorate([
        core_1.HostBinding('style.height'),
        core_1.Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], PopoverComponent.prototype, "height", null);
    tslib_1.__decorate([
        core_1.HostBinding('attr.dir'),
        tslib_1.__metadata("design:type", String)
    ], PopoverComponent.prototype, "direction", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], PopoverComponent.prototype, "title", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], PopoverComponent.prototype, "subtitle", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], PopoverComponent.prototype, "body", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], PopoverComponent.prototype, "callout", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], PopoverComponent.prototype, "animation", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Function])
    ], PopoverComponent.prototype, "templateData", null);
    tslib_1.__decorate([
        core_1.HostBinding('class.k-hidden'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [])
    ], PopoverComponent.prototype, "isHidden", null);
    tslib_1.__decorate([
        core_1.HostBinding('attr.aria-hidden'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [])
    ], PopoverComponent.prototype, "hasAttributeHidden", null);
    tslib_1.__decorate([
        core_1.Output(),
        tslib_1.__metadata("design:type", core_1.EventEmitter)
    ], PopoverComponent.prototype, "show", void 0);
    tslib_1.__decorate([
        core_1.Output(),
        tslib_1.__metadata("design:type", core_1.EventEmitter)
    ], PopoverComponent.prototype, "shown", void 0);
    tslib_1.__decorate([
        core_1.Output(),
        tslib_1.__metadata("design:type", core_1.EventEmitter)
    ], PopoverComponent.prototype, "hide", void 0);
    tslib_1.__decorate([
        core_1.Output(),
        tslib_1.__metadata("design:type", core_1.EventEmitter)
    ], PopoverComponent.prototype, "hidden", void 0);
    tslib_1.__decorate([
        core_1.ContentChild(title_template_directive_1.PopoverTitleTemplateDirective, { static: false }),
        tslib_1.__metadata("design:type", title_template_directive_1.PopoverTitleTemplateDirective)
    ], PopoverComponent.prototype, "titleTemplate", void 0);
    tslib_1.__decorate([
        core_1.ContentChild(body_template_directive_1.PopoverBodyTemplateDirective, { static: false }),
        tslib_1.__metadata("design:type", body_template_directive_1.PopoverBodyTemplateDirective)
    ], PopoverComponent.prototype, "bodyTemplate", void 0);
    tslib_1.__decorate([
        core_1.ContentChild(actions_template_directive_1.PopoverActionsTemplateDirective, { static: false }),
        tslib_1.__metadata("design:type", actions_template_directive_1.PopoverActionsTemplateDirective)
    ], PopoverComponent.prototype, "actionsTemplate", void 0);
    PopoverComponent = tslib_1.__decorate([
        core_1.Component({
            selector: 'kendo-popover',
            providers: [
                kendo_angular_l10n_1.LocalizationService,
                {
                    provide: kendo_angular_l10n_1.L10N_PREFIX,
                    useValue: 'kendo.popover'
                }
            ],
            template: "\n    <div *ngIf=\"visible\" role=\"tooltip\" class=\"k-popover\" [ngStyle]=\"{'width.px': width, 'height.px': height}\">\n        <div class=\"k-popover-callout\" [ngClass]=\"getCalloutPosition()\" *ngIf=\"callout\"></div>\n\n        <div *ngIf=\"titleTemplate || title\" class=\"k-popover-header\">\n            <ng-template *ngIf=\"titleTemplate\"\n                [ngTemplateOutlet]=\"titleTemplate?.templateRef\"\n                [ngTemplateOutletContext]=\"{ $implicit: anchor, data: contextData }\">\n            </ng-template>\n            <ng-container *ngIf=\"title && !titleTemplate\">\n                {{ title }}\n            </ng-container>\n        </div>\n\n        <div *ngIf=\"bodyTemplate || body\" class=\"k-popover-body\">\n            <ng-template *ngIf=\"bodyTemplate\"\n                [ngTemplateOutlet]=\"bodyTemplate?.templateRef\"\n                [ngTemplateOutletContext]=\"{ $implicit: anchor, data: contextData }\">\n            </ng-template>\n            <ng-container *ngIf=\"body && !bodyTemplate\">\n                {{ body }}\n            </ng-container>\n        </div>\n\n        <div *ngIf=\"actionsTemplate\" class=\"k-popover-actions k-actions k-hstack k-justify-content-between\">\n            <ng-template *ngIf=\"actionsTemplate\"\n                [ngTemplateOutlet]=\"actionsTemplate?.templateRef\"\n                [ngTemplateOutletContext]=\"{ $implicit: anchor, data: contextData }\">\n            </ng-template>\n        </div>\n    </div>\n    "
        }),
        tslib_1.__metadata("design:paramtypes", [kendo_angular_l10n_1.LocalizationService])
    ], PopoverComponent);
    return PopoverComponent;
}());
exports.PopoverComponent = PopoverComponent;
