// @progress/kendo-angular-listview/dist/npm/utils.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var kendo_angular_common_1 = require("@progress/kendo-angular-common");
var LISTVIEW_ITEM_SELECTOR = '.k-listview-item';
/**
 * @hidden
 */
exports.isPresent = function (item) { return item !== null && item !== undefined; };
/**
 * @hidden
 */
exports.isObject = function (item) { return exports.isPresent(item) && typeof item === 'object'; };
/**
 * @hidden
 *
 * Polyfill for `Element.matches`.
 * https://developer.mozilla.org/en-US/docs/Web/API/Element/matches
 */
exports.match = function (element, selector) {
    var matcher = element.matches || element.msMatchesSelector || element.webkitMatchesSelector;
    if (!exports.isPresent(matcher)) {
        return false;
    }
    return matcher.call(element, selector);
};
/**
 * @hidden
 *
 * Checks if a target element has the `.k-listview-item` CSS class.
 */
exports.isListViewItem = function (element) {
    if (!exports.isPresent(element)) {
        return false;
    }
    return exports.match(element, LISTVIEW_ITEM_SELECTOR);
};
/**
 * @hidden
 *
 * Extracts and parses to a number the `data-kendo-listview-item-index` attribute value from the targeted element.
 */
exports.getListItemIndex = function (item) {
    if (!exports.isPresent(item)) {
        return null;
    }
    return Number(item.getAttribute('data-kendo-listview-item-index'));
};
/**
 * @hidden
 *
 * Gets the new focus target from a blur event.
 * Queries both event.relatedTarget and document.activeElement for compatibility with IE.
 */
exports.relatedTarget = function (event) {
    if (!exports.isPresent(event.relatedTarget) || !kendo_angular_common_1.isDocumentAvailable()) {
        return null;
    }
    return event.relatedTarget || document.activeElement;
};
/**
 * @hidden
 *
 * If the given contender number is not defined or lower than the specified min - returns min, if its above the specified max - returns max.
 * If the number is in the given bounds, it is returned.
 */
exports.fitIntoRange = function (contender, min, max) {
    if (!exports.isPresent(contender) || contender <= min) {
        return min;
    }
    else if (contender >= max) {
        return max;
    }
    else {
        return contender;
    }
};
/**
 * @hidden
 */
exports.closestWithMatch = function (element, selector) {
    var parent = element;
    while (parent !== null && parent.nodeType === 1) {
        if (exports.match(parent, selector)) {
            return parent;
        }
        parent = parent.parentElement || parent.parentNode;
    }
    return null;
};
/**
 * @hidden
 *
 * Extracts and parses to a number the `data-kendo-listview-item-index` attribute value from the targeted element.
 */
exports.getClosestListItemIndex = function (element) {
    if (!exports.isPresent(element)) {
        return null;
    }
    var closestListViewItem = exports.closestWithMatch(element, LISTVIEW_ITEM_SELECTOR);
    return exports.getListItemIndex(closestListViewItem);
};
