// @progress/kendo-angular-menu/dist/npm/context-menu/context-menu-target-container.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var context_menu_target_service_1 = require("./context-menu-target.service");
/**
 * Specifies a container for the [targets]({% slug api_menu_contextmenutargetdirective %}) of the ContextMenu.
 */
var ContextMenuTargetContainerDirective = /** @class */ (function () {
    /**
     * @hidden
     */
    function ContextMenuTargetContainerDirective(elementRef, targetService) {
        this.targetService = targetService;
        if (elementRef) {
            this.element = elementRef.nativeElement;
        }
    }
    ContextMenuTargetContainerDirective = tslib_1.__decorate([
        core_1.Directive({
            selector: '[kendoContextMenuTargetContainer]',
            exportAs: 'kendoContextMenuTargetContainer',
            providers: [context_menu_target_service_1.ContextMenuTargetService]
        }),
        tslib_1.__metadata("design:paramtypes", [core_1.ElementRef, context_menu_target_service_1.ContextMenuTargetService])
    ], ContextMenuTargetContainerDirective);
    return ContextMenuTargetContainerDirective;
}());
exports.ContextMenuTargetContainerDirective = ContextMenuTargetContainerDirective;
