// @progress/kendo-angular-listview/dist/npm/navigation/listview-navigable-item.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var navigation_service_1 = require("./navigation.service");
var utils_1 = require("../utils");
/**
 * @hidden
 */
var ListViewNavigableItemDirective = /** @class */ (function () {
    function ListViewNavigableItemDirective(hostElement, renderer, navigationService) {
        this.hostElement = hostElement;
        this.renderer = renderer;
        this.navigationService = navigationService;
    }
    ListViewNavigableItemDirective.prototype.ngOnChanges = function () {
        this.updateNavigationState();
    };
    ListViewNavigableItemDirective.prototype.ngOnInit = function () {
        this.navigationSubscription = this.navigationService.changes
            .subscribe(this.updateNavigationState.bind(this));
    };
    ListViewNavigableItemDirective.prototype.ngOnDestroy = function () {
        if (utils_1.isPresent(this.navigationSubscription)) {
            this.navigationSubscription.unsubscribe();
        }
    };
    ListViewNavigableItemDirective.prototype.updateNavigationState = function () {
        this.updateTabIndex();
        this.updateFocusedState();
    };
    ListViewNavigableItemDirective.prototype.updateFocusedState = function () {
        var shouldFocus = this.navigationService.isActive(this.index) && this.navigationService.isFocused;
        var focusedCssClass = 'k-state-focused';
        if (shouldFocus) {
            this.renderer.addClass(this.hostElement.nativeElement, focusedCssClass);
            this.hostElement.nativeElement.focus();
        }
        else {
            this.renderer.removeClass(this.hostElement.nativeElement, focusedCssClass);
        }
    };
    ListViewNavigableItemDirective.prototype.updateTabIndex = function () {
        if (!this.navigationService.isEnabled) {
            this.renderer.removeAttribute(this.hostElement.nativeElement, 'tabindex');
        }
        else if (this.navigationService.isActive(this.index)) {
            this.renderer.setAttribute(this.hostElement.nativeElement, 'tabindex', '0');
        }
        else {
            this.renderer.setAttribute(this.hostElement.nativeElement, 'tabindex', '-1');
        }
    };
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Number)
    ], ListViewNavigableItemDirective.prototype, "index", void 0);
    ListViewNavigableItemDirective = tslib_1.__decorate([
        core_1.Directive({
            selector: '[kendoListViewNavigableItem]'
        }),
        tslib_1.__metadata("design:paramtypes", [core_1.ElementRef,
            core_1.Renderer2,
            navigation_service_1.NavigationService])
    ], ListViewNavigableItemDirective);
    return ListViewNavigableItemDirective;
}());
exports.ListViewNavigableItemDirective = ListViewNavigableItemDirective;
