// @progress/kendo-angular-dropdowns/dist/es/common/filter-input.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, NgZone } from '@angular/core';
/**
 * @hidden
 */
var FilterInputDirective = /** @class */ (function () {
    function FilterInputDirective(element, zone) {
        this.element = element;
        this.zone = zone;
    }
    FilterInputDirective.prototype.ngOnChanges = function () {
        var _this = this;
        if (this.focused) {
            this.nextTick(function () { return _this.element.nativeElement.focus(); });
        }
    };
    FilterInputDirective.prototype.nextTick = function (fn) {
        this.zone.runOutsideAngular(function () { return setTimeout(fn); });
    };
    tslib_1.__decorate([
        Input('filterInput'),
        tslib_1.__metadata("design:type", Boolean)
    ], FilterInputDirective.prototype, "focused", void 0);
    FilterInputDirective = tslib_1.__decorate([
        Directive({
            selector: '[filterInput]' // tslint:disable-line
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef,
            NgZone])
    ], FilterInputDirective);
    return FilterInputDirective;
}());
export { FilterInputDirective };
