// @progress/kendo-angular-dateinputs/dist/es2015/calendar/localization/calendar-localized-messages.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
var CalendarLocalizedMessagesDirective_1;
import { Directive, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { CalendarMessages } from './calendar-messages';
/**
 * @hidden
 */
let CalendarLocalizedMessagesDirective = CalendarLocalizedMessagesDirective_1 = class CalendarLocalizedMessagesDirective extends CalendarMessages {
    constructor(service) {
        super();
        this.service = service;
    }
};
CalendarLocalizedMessagesDirective = CalendarLocalizedMessagesDirective_1 = tslib_1.__decorate([
    Directive({
        providers: [
            {
                provide: CalendarMessages,
                useExisting: forwardRef(() => CalendarLocalizedMessagesDirective_1) // tslint:disable-line:no-forward-ref
            }
        ],
        selector: '[kendoCalendarLocalizedMessages]'
    }),
    tslib_1.__metadata("design:paramtypes", [LocalizationService])
], CalendarLocalizedMessagesDirective);
export { CalendarLocalizedMessagesDirective };
