// @progress/kendo-angular-layout/dist/es2015/stepper/localization/localized-messages.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
var LocalizedStepperMessagesDirective_1;
import { Directive, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { StepperMessages } from './messages';
/**
 * @hidden
 */
let LocalizedStepperMessagesDirective = LocalizedStepperMessagesDirective_1 = class LocalizedStepperMessagesDirective extends StepperMessages {
    constructor(service) {
        super();
        this.service = service;
    }
};
LocalizedStepperMessagesDirective = LocalizedStepperMessagesDirective_1 = tslib_1.__decorate([
    Directive({
        providers: [
            {
                provide: StepperMessages,
                useExisting: forwardRef(() => LocalizedStepperMessagesDirective_1)
            }
        ],
        selector: `
      [kendoStepperLocalizedMessages]
    `
    }),
    tslib_1.__metadata("design:paramtypes", [LocalizationService])
], LocalizedStepperMessagesDirective);
export { LocalizedStepperMessagesDirective };
