// @safe-fleet/sf-security-menu/bundles/safe-fleet-sf-security-menu.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('@safe-fleet/sf-menu'), require('@progress/kendo-angular-menu'), require('@angular/common/http'), require('rxjs/operators'), require('@progress/kendo-angular-popup'), require('@angular/common'), require('@progress/kendo-angular-buttons')) :
    typeof define === 'function' && define.amd ? define('@safe-fleet/sf-security-menu', ['exports', '@angular/core', 'rxjs', '@safe-fleet/sf-menu', '@progress/kendo-angular-menu', '@angular/common/http', 'rxjs/operators', '@progress/kendo-angular-popup', '@angular/common', '@progress/kendo-angular-buttons'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["safe-fleet"] = global["safe-fleet"] || {}, global["safe-fleet"]["sf-security-menu"] = {}), global.ng.core, global.rxjs, global["sf-menu"], global.kendoAngularMenu, global.ng.common.http, global.rxjs.operators, global.kendoAngularPopup, global.ng.common, global.kendoAngularButtons));
})(this, (function (exports, i0, rxjs, i1, kendoAngularMenu, i1$1, operators, kendoAngularPopup, common, kendoAngularButtons) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace$1 = /*#__PURE__*/_interopNamespace(i1);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1$1);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || Array.prototype.slice.call(from));
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var IMenuConfig = /** @class */ (function (_super) {
        __extends(IMenuConfig, _super);
        function IMenuConfig() {
            return _super.call(this) || this;
        }
        return IMenuConfig;
    }(i1.MenuConfig));
    var IUnauthorizedPopup = /** @class */ (function () {
        function IUnauthorizedPopup() {
        }
        return IUnauthorizedPopup;
    }());
    var IMenuItem = /** @class */ (function (_super) {
        __extends(IMenuItem, _super);
        function IMenuItem() {
            return _super.call(this) || this;
        }
        return IMenuItem;
    }(i1.MenuItem));
    var IEvent = /** @class */ (function (_super) {
        __extends(IEvent, _super);
        function IEvent() {
            return _super.call(this) || this;
        }
        return IEvent;
    }(kendoAngularMenu.MenuEvent));
    var EventTypeEnum = Object.assign({}, i1.MenuEventType);

    var SecurityService = /** @class */ (function () {
        function SecurityService(httpClient) {
            this.httpClient = httpClient;
        }
        SecurityService.prototype.setSecurityInformation = function (securityEndpoint, settingsEndpoint, headers) {
            this.securityHeaders = headers;
            this.securityEndpoint = securityEndpoint;
            this.settingsEndpoint = settingsEndpoint;
        };
        SecurityService.prototype.getApplications = function () {
            var url = this.securityEndpoint + "/applications";
            return this.httpClient.get(url, {
                headers: this.securityHeaders,
            });
        };
        SecurityService.prototype.getPermissions = function () {
            return this.permissions;
        };
        SecurityService.prototype.getInfo = function () {
            return this.applicationInfo;
        };
        SecurityService.prototype.getSecurityPermissions = function (resourceName) {
            var _this = this;
            var url = this.securityEndpoint + "/permissions?resourceName=" + resourceName;
            return this.httpClient
                .get(url, {
                headers: this.securityHeaders,
            })
                .pipe(operators.map(function (x) { return (_this.permissions = x); }), operators.catchError(function (error) {
                throw error;
            }));
        };
        SecurityService.prototype.getApplicationInfo = function () {
            var _this = this;
            var url = this.settingsEndpoint + "/info";
            return this.httpClient
                .get(url, {
                headers: this.securityHeaders,
            })
                .pipe(operators.map(function (x) { return (_this.applicationInfo = x); }), operators.catchError(function (error) {
                throw error;
            }));
        };
        return SecurityService;
    }());
    SecurityService.ɵprov = i0__namespace.ɵɵdefineInjectable({ factory: function SecurityService_Factory() { return new SecurityService(i0__namespace.ɵɵinject(i1__namespace.HttpClient)); }, token: SecurityService, providedIn: "root" });
    SecurityService.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    SecurityService.ctorParameters = function () { return [
        { type: i1$1.HttpClient }
    ]; };

    var SfSecurityMenuService = /** @class */ (function () {
        function SfSecurityMenuService(sfMenuService, securityService) {
            this.sfMenuService = sfMenuService;
            this.securityService = securityService;
            this.infoMessageVisibilityChange = new rxjs.BehaviorSubject(false);
            this.events$ = this.sfMenuService.events$;
        }
        SfSecurityMenuService.prototype.configureMenu = function (config, securityInfo) {
            var _this = this;
            this.securityService.setSecurityInformation(securityInfo.securityEndpoint, securityInfo.settingsEndpoint, securityInfo.headers);
            this.refreshMenu(config);
            this.onMenuChanges();
            this.infoMessageVisibilityChange.next(false);
            this.unauthorizedPopup = this.getUnauthorizedPopup(config.unauthorizedPopup);
            var permissionsObservable = this.securityService.getSecurityPermissions(config.currentProduct.id).pipe(operators.catchError(function (error) {
                if (error.status === 403 && config.multipleTenants && config.unauthorizedPopup) {
                    _this.infoMessageVisibilityChange.next(true);
                }
                throw error;
            }));
            var applicationsObservable = this.securityService.getApplications().pipe(operators.map(function (applications) { return _this.mapConfiguration(config, applications, securityInfo); }), operators.catchError(function (error) {
                _this.refreshMenu(config);
                throw error;
            }));
            var applicationInfoObservable = this.securityService.getApplicationInfo();
            return applicationsObservable.pipe(operators.mergeMap(function () { return permissionsObservable; }), operators.mergeMap(function () { return applicationInfoObservable; }), operators.catchError(function () { return rxjs.of(''); }));
        };
        SfSecurityMenuService.prototype.refreshMenu = function (config) {
            this.sfMenuService.setMenuConfig(config);
        };
        SfSecurityMenuService.prototype.getPermissions = function () {
            return this.securityService.getPermissions();
        };
        SfSecurityMenuService.prototype.getSettingsInfo = function () {
            return this.securityService.getInfo();
        };
        SfSecurityMenuService.prototype.mapConfiguration = function (config, applications, securityInfo) {
            var _this = this;
            config.products = [];
            var menuItems = [];
            applications.forEach(function (app) {
                menuItems.push(_this.mapMenuApplication(app));
            });
            var selectedProduct = menuItems.find(function (app) { return app.text === securityInfo.selectedApplication; });
            if (!selectedProduct || selectedProduct === null) {
                config.products = menuItems;
                this.refreshMenu(config);
                return;
            }
            config.currentProduct = selectedProduct;
            config.products = menuItems.filter(function (item) { return item.id !== config.currentProduct.id; });
            this.refreshMenu(config);
        };
        SfSecurityMenuService.prototype.mapMenuApplication = function (app) {
            return {
                text: app.applicationName,
                routeHref: app.applicationUrl,
                iconPath: app.applicationIconUrl,
                id: app.applicationName,
                newWindow: false,
                active: false,
                expanded: false,
                newTab: false,
            };
        };
        SfSecurityMenuService.prototype.getUnauthorizedPopup = function (popupConfig) {
            return {
                popupButtonText: popupConfig ? popupConfig.popupButtonText : '',
                popupMessageActionText: popupConfig ? popupConfig.popupMessageActionText : '',
                popupMessageText: popupConfig ? popupConfig.popupMessageText : '',
            };
        };
        SfSecurityMenuService.prototype.reloadPermissions = function (applicationName) {
            return this.securityService.getSecurityPermissions(applicationName);
        };
        SfSecurityMenuService.prototype.onMenuChanges = function () {
            var _this = this;
            return this.sfMenuService.events$.subscribe(function (menuEvent) {
                if (menuEvent.type === EventTypeEnum.OpenTenantListClick) {
                    _this.infoMessageVisibilityChange.next(false);
                }
            });
        };
        return SfSecurityMenuService;
    }());
    SfSecurityMenuService.ɵprov = i0__namespace.ɵɵdefineInjectable({ factory: function SfSecurityMenuService_Factory() { return new SfSecurityMenuService(i0__namespace.ɵɵinject(i1__namespace$1.SfMenuService), i0__namespace.ɵɵinject(SecurityService)); }, token: SfSecurityMenuService, providedIn: "root" });
    SfSecurityMenuService.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    SfSecurityMenuService.ctorParameters = function () { return [
        { type: i1.SfMenuService },
        { type: SecurityService }
    ]; };

    var SfSecurityMenuComponent = /** @class */ (function () {
        function SfSecurityMenuComponent(sfSecMenuService, sfMenuService) {
            this.sfSecMenuService = sfSecMenuService;
            this.sfMenuService = sfMenuService;
            this.highlightClass = 'highlight-org-menu';
            this.paddingWidth = 10;
            this.popupAlign = { horizontal: 'left', vertical: 'center' };
            this.delayTime = 300;
            this.offsetPopup = { left: -1000, top: 0 };
            this.showOrganizationMsgPopup = false;
        }
        SfSecurityMenuComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.sfSecMenuService.infoMessageVisibilityChange.subscribe(function (showInfoMessage) {
                if (showInfoMessage) {
                    _this.unauthorizedPopup = _this.sfSecMenuService.unauthorizedPopup;
                    _this.onMenuCollapseOrExpand();
                }
                else {
                    _this.closeShowOrganizationMsgPopup();
                }
            });
        };
        SfSecurityMenuComponent.prototype.onMenuCollapseOrExpand = function () {
            var _this = this;
            this.showOrganizationMsgPopup = true;
            this.sfMenuService.menuCollapse$.pipe(operators.debounceTime(this.delayTime)).subscribe(function () {
                _this.getOrganizationPosition();
            });
        };
        SfSecurityMenuComponent.prototype.onResize = function () {
            this.getOrganizationPosition();
        };
        SfSecurityMenuComponent.prototype.getOrganizationPosition = function () {
            if (this.showOrganizationMsgPopup) {
                var elem = document.querySelector('.sf-fleet-list');
                this.organizationItemMenu = elem && elem.parentElement ? elem.parentElement : null;
                if (this.organizationItemMenu) {
                    var bounding = this.organizationItemMenu.getBoundingClientRect();
                    var top = bounding.top + bounding.height / 2;
                    var left = bounding.left + bounding.width + this.paddingWidth;
                    this.offsetPopup = { left: left, top: top };
                    this.organizationItemMenu.classList.add(this.highlightClass);
                }
                else {
                    this.showOrganizationMsgPopup = false;
                }
            }
        };
        SfSecurityMenuComponent.prototype.closeShowOrganizationMsgPopup = function () {
            this.showOrganizationMsgPopup = false;
            if (this.organizationItemMenu) {
                this.organizationItemMenu.classList.remove(this.highlightClass);
            }
        };
        return SfSecurityMenuComponent;
    }());
    SfSecurityMenuComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'sf-security-menu',
                    template: "<sf-menu>\n  <ng-content></ng-content>\n</sf-menu>\n\n<kendo-popup [popupAlign]=\"popupAlign\" *ngIf=\"unauthorizedPopup && showOrganizationMsgPopup\" [offset]=\"offsetPopup\">\n  <div class=\"content-popup\">\n    <div class=\"message\">{{unauthorizedPopup.popupMessageText}}</div>\n    <div class=\"message action-bold\">{{unauthorizedPopup.popupMessageActionText}}</div>\n    <div>\n      <button type=\"button\" class=\"primary-size\" [primary]=\"true\" kendoButton (click)=\"closeShowOrganizationMsgPopup()\">\n        {{unauthorizedPopup.popupButtonText}}\n      </button>\n    </div>\n  </div>\n</kendo-popup>\n",
                    styles: [":host{width:100%;height:100%}:host ::ng-deep .highlight-org-menu{background-color:#00000080}:host ::ng-deep .k-popup{border-radius:5px}:host ::ng-deep .k-popup .content-popup{width:210px;padding:25px 15px;text-align:center;background-color:var(--clr-gray-500);color:#fff;border-radius:5px}:host ::ng-deep .k-popup .content-popup .message{margin:10px}:host ::ng-deep .k-popup .content-popup .action-bold{font-weight:600}:host ::ng-deep .k-popup .content-popup:before{content:\"\";position:absolute;top:75px;left:-38px;border-left:10px solid transparent;border-right:10px solid var(--clr-gray-500);border-bottom:10px solid transparent;border-top:10px solid transparent;width:20px}\n"]
                },] }
    ];
    SfSecurityMenuComponent.ctorParameters = function () { return [
        { type: SfSecurityMenuService },
        { type: i1.SfMenuService }
    ]; };
    SfSecurityMenuComponent.propDecorators = {
        onResize: [{ type: i0.HostListener, args: ['window:resize', ['$event'],] }]
    };

    var SfSecurityMenuModule = /** @class */ (function () {
        function SfSecurityMenuModule() {
        }
        return SfSecurityMenuModule;
    }());
    SfSecurityMenuModule.decorators = [
        { type: i0.NgModule, args: [{
                    declarations: [SfSecurityMenuComponent],
                    imports: [common.CommonModule, i1.SfMenuModule, i1$1.HttpClientModule, kendoAngularPopup.PopupModule, kendoAngularButtons.ButtonsModule],
                    exports: [SfSecurityMenuComponent],
                },] }
    ];

    /*
     * Public API Surface of sf-security-menu
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.EventTypeEnum = EventTypeEnum;
    exports.IEvent = IEvent;
    exports.IMenuConfig = IMenuConfig;
    exports.IMenuItem = IMenuItem;
    exports.IUnauthorizedPopup = IUnauthorizedPopup;
    exports.SfSecurityMenuComponent = SfSecurityMenuComponent;
    exports.SfSecurityMenuModule = SfSecurityMenuModule;
    exports.SfSecurityMenuService = SfSecurityMenuService;
    exports["ɵa"] = SecurityService;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
