// @progress/kendo-angular-menu/dist/es2015/menu-base.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Input, ContentChildren, QueryList } from '@angular/core';
import { MenuItemComponent } from './menu-item.component';
import { ItemTemplateDirective } from './templates/item-template.directive';
import { ItemLinkTemplateDirective } from './templates/item-link-template.directive';
/**
 * @hidden
 */
export class MenuBase {
    constructor() {
        /**
         * Specifies if the Menu will be vertical ([see example]({% slug vertical_menu %})).
         */
        this.vertical = false;
        /**
         * Specifies that the root items can be opened only on click
         * ([see example]({% slug openclose_menu %}#toc-opening-on-click)).
         */
        this.openOnClick = false;
        /**
         * Specifies the delay in milliseconds before the Menu items are opened or closed on item hover
         * or leave ([see example]({% slug openclose_menu %}#toc-delay-on-hover)). Used to avoid the accidental
         * opening or closing of the items.
         */
        this.hoverDelay = 100;
        /**
         * Sets the Menu animation.
         */
        this.animate = true;
    }
    /**
     * @hidden
     */
    get rootItems() {
        return this.items || (this.children ? this.children.toArray() : []);
    }
}
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Array)
], MenuBase.prototype, "items", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], MenuBase.prototype, "vertical", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], MenuBase.prototype, "openOnClick", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Number)
], MenuBase.prototype, "hoverDelay", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], MenuBase.prototype, "animate", void 0);
tslib_1.__decorate([
    ContentChildren(ItemTemplateDirective),
    tslib_1.__metadata("design:type", QueryList)
], MenuBase.prototype, "itemTemplate", void 0);
tslib_1.__decorate([
    ContentChildren(ItemLinkTemplateDirective),
    tslib_1.__metadata("design:type", QueryList)
], MenuBase.prototype, "itemLinkTemplate", void 0);
tslib_1.__decorate([
    ContentChildren(MenuItemComponent),
    tslib_1.__metadata("design:type", QueryList)
], MenuBase.prototype, "children", void 0);
