// @safe-fleet/sf-menu/esm2015/lib/components/menu-shortcuts/menu-shortcuts.component.js
import { animate, style, transition, trigger } from '@angular/animations';
import { Component } from '@angular/core';
import { SfMenuService } from '../../sf-menu.service';
export class MenuShortcutsComponent {
    constructor(svc) {
        this.svc = svc;
        this.collapsed = false;
    }
    ngOnInit() {
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            this.menuConfig = cfg;
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
        });
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
}
MenuShortcutsComponent.decorators = [
    { type: Component, args: [{
                selector: 'sf-menu-shortcuts',
                template: "<ul>\n  <li\n    sf-menu-item\n    *ngFor=\"let item of menuConfig.shortcuts\"\n    [item]=\"item\"\n    [level]=\"1\"\n    [itemType]=\"1\"\n  >\n  </li>\n</ul>\n",
                animations: [
                    trigger('appear-fast', [transition(':enter', [style({ opacity: 0 }), animate('80ms 120ms ease-in', style({ opacity: 1 }))])]),
                ],
                styles: [":host{display:flex;flex-direction:column;width:100%;height:auto;align-items:flex-start;justify-content:flex-start;overflow:hidden;position:relative}:host:before{content:\" \";width:100%;height:1px;background-color:#fff;opacity:.15}:host ul{list-style:none;padding-left:0;width:100%}"]
            },] }
];
MenuShortcutsComponent.ctorParameters = () => [
    { type: SfMenuService }
];
