// @progress/kendo-angular-inputs/dist/npm/sliders-common/slider-model.base.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var sliders_util_1 = require("../sliders-common/sliders-util");
var math_1 = require("../common/math");
/**
 * @hidden
 */
var SliderModelBase = /** @class */ (function () {
    function SliderModelBase(props, wrapper, track, renderer) {
        this.props = props;
        this.wrapper = wrapper;
        this.track = track;
        this.renderer = renderer;
        this.props = props;
        this.wrapper = wrapper;
        this.track = track;
        this.tickSizes = this.getTickSizes();
    }
    SliderModelBase.prototype.resizeTrack = function () {
        var orientation = this.props.vertical ? 'height' : 'width';
        var altOrientation = this.props.vertical ? 'width' : 'height';
        var trackWidth = this.trackWidth();
        this.track.parentElement.style[orientation] = trackWidth + "px";
        this.track.parentElement.style[altOrientation] = '';
    };
    SliderModelBase.prototype.resizeTicks = function (ticksContainer, ticks) {
        var _this = this;
        var dimension = this.props.vertical ? "height" : "width";
        ticks.slice().map(function (tick, index) { return tick.style[dimension] = _this.tickSizes[index] + "px"; });
        if (this.props.vertical) {
            this.adjustPadding(ticksContainer);
        }
    };
    SliderModelBase.prototype.resizeWrapper = function () {
        var dimension = this.props.vertical ? "height" : "width";
        var fixedTrackWidth = sliders_util_1.calculateFixedTrackSize(this.props);
        var wrapperParentEl = this.wrapper.parentElement;
        if (fixedTrackWidth) {
            wrapperParentEl.style[dimension] = "auto";
        }
    };
    SliderModelBase.prototype.trackWidth = function () {
        if (this.props.fixedTickWidth) {
            return sliders_util_1.calculateFixedTrackSize(this.props);
        }
        return this.elementSize(this.track.parentElement);
    };
    SliderModelBase.prototype.getTickSizes = function () {
        var _a = this.props, min = _a.min, max = _a.max, smallStep = _a.smallStep;
        var count = sliders_util_1.calculateTicksCount(min, max, smallStep);
        var trackSize = this.trackWidth();
        var distStep = trackSize / math_1.subtract(max, min);
        var result = [];
        var usedSpace = 0;
        var endPoint = 0;
        for (var i = 0; i < count; i++) {
            if (i === 0 || i === count - 1) {
                endPoint += (smallStep / 2) * distStep;
            }
            else {
                endPoint += smallStep * distStep;
            }
            // ensure that the sum of the tick sizes does not exceed the track width
            endPoint = +endPoint.toFixed(2) - 0.01;
            var size = Math.round(endPoint - usedSpace);
            result.push(size);
            usedSpace += size;
        }
        if (usedSpace >= trackSize) {
            result[result.length - 1] -= 1;
        }
        return result;
    };
    SliderModelBase.prototype.adjustPadding = function (ticksContainer) {
        var totalTickSize = this.tickSizes.reduce(function (prev, curr) { return prev + curr; }, 0);
        var trackWidth = this.trackWidth();
        var reminder = trackWidth - totalTickSize;
        if (reminder !== 0) {
            var padding = reminder + this.elementOffset(this.track);
            ticksContainer.style.paddingTop = padding + "px";
        }
    };
    SliderModelBase.prototype.elementOffset = function (element) {
        var vertical = this.props.vertical;
        var style = getComputedStyle(element);
        return parseInt(vertical ? style.bottom : style.left, 10);
    };
    SliderModelBase.prototype.elementSize = function (element) {
        var vertical = this.props.vertical;
        return vertical ? element.clientHeight : element.clientWidth;
    };
    return SliderModelBase;
}());
exports.SliderModelBase = SliderModelBase;
