// @progress/kendo-angular-pager/dist/npm/pager/pager-prev-buttons.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
// tslint:disable:no-access-missing-member
var core_1 = require("@angular/core");
var pager_context_service_1 = require("./pager-context.service");
var pager_element_component_1 = require("./pager-element.component");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
/**
 * Displays buttons for navigating to the first and to the previous page ([see example]({% slug pager_settings %})).
 */
var PagerPrevButtonsComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PagerPrevButtonsComponent, _super);
    function PagerPrevButtonsComponent(localization, pagerContext, cd) {
        return _super.call(this, localization, pagerContext, cd) || this;
    }
    Object.defineProperty(PagerPrevButtonsComponent.prototype, "disabled", {
        /**
         * @hidden
         *
         * @readonly
         * @type {boolean}
         * @memberOf PagerPrevButtonsComponent
         */
        get: function () {
            return this.currentPage === 1 || !this.total;
        },
        enumerable: true,
        configurable: true
    });
    PagerPrevButtonsComponent.prototype.onChanges = function (_a) {
        var total = _a.total, skip = _a.skip, pageSize = _a.pageSize;
        this.total = total;
        this.skip = skip;
        this.pageSize = pageSize;
        this.cd.markForCheck();
    };
    PagerPrevButtonsComponent = tslib_1.__decorate([
        core_1.Component({
            changeDetection: core_1.ChangeDetectionStrategy.OnPush,
            selector: 'kendo-datapager-prev-buttons',
            template: "\n        <button\n            type=\"button\"\n            class=\"k-link k-pager-nav k-pager-first\"\n            [class.k-disabled]=\"disabled\"\n            [title]=\"textFor('firstPage')\"\n            [attr.aria-label]=\"textFor('firstPage')\"\n            [attr.aria-disabled]=\"disabled ? true : undefined\"\n            (click)=\"currentPage !== 1 ? changePage(0) : false\">\n            <span class=\"k-icon k-i-seek-w\" role=\"note\" [attr.aria-label]=\"textFor('firstPage')\"></span>\n        </button>\n        <button\n            type=\"button\"\n            class=\"k-link k-pager-nav\"\n            [class.k-disabled]=\"disabled\"\n            [title]=\"textFor('previousPage')\"\n            [attr.aria-label]=\"textFor('previousPage')\"\n            [attr.aria-disabled]=\"disabled ? true : undefined\"\n            (click)=\"currentPage !== 1 ? changePage(currentPage-2) : false\">\n            <span class=\"k-icon k-i-arrow-w\" role=\"note\" [attr.aria-label]=\"textFor('previousPage')\"></span>\n        </button>\n    "
        }),
        tslib_1.__metadata("design:paramtypes", [kendo_angular_l10n_1.LocalizationService,
            pager_context_service_1.PagerContextService,
            core_1.ChangeDetectorRef])
    ], PagerPrevButtonsComponent);
    return PagerPrevButtonsComponent;
}(pager_element_component_1.PagerElementComponent));
exports.PagerPrevButtonsComponent = PagerPrevButtonsComponent;
