// @progress/kendo-angular-pager/dist/npm/pager/pager.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var kendo_angular_inputs_1 = require("@progress/kendo-angular-inputs");
var kendo_angular_dropdowns_1 = require("@progress/kendo-angular-dropdowns");
var pager_component_1 = require("./pager.component");
var pager_prev_buttons_component_1 = require("./pager-prev-buttons.component");
var pager_next_buttons_component_1 = require("./pager-next-buttons.component");
var pager_numeric_buttons_component_1 = require("./pager-numeric-buttons.component");
var pager_input_component_1 = require("./pager-input.component");
var pager_info_component_1 = require("./pager-info.component");
var pager_page_sizes_component_1 = require("./pager-page-sizes.component");
var pager_template_directive_1 = require("./pager-template.directive");
var kendo_angular_common_1 = require("@progress/kendo-angular-common");
var localized_messages_directive_1 = require("./localization/localized-messages.directive");
var custom_messages_component_1 = require("./localization/custom-messages.component");
var kendo_angular_common_2 = require("@progress/kendo-angular-common");
var importedModules = [
    common_1.CommonModule,
    kendo_angular_inputs_1.NumericTextBoxModule,
    kendo_angular_dropdowns_1.DropDownListModule,
    kendo_angular_common_1.EventsModule,
    kendo_angular_common_2.ResizeSensorModule
];
var INTERNAL_COMPONENTS = [
    pager_component_1.PagerComponent,
    pager_prev_buttons_component_1.PagerPrevButtonsComponent,
    pager_next_buttons_component_1.PagerNextButtonsComponent,
    pager_numeric_buttons_component_1.PagerNumericButtonsComponent,
    pager_input_component_1.PagerInputComponent,
    pager_info_component_1.PagerInfoComponent,
    pager_page_sizes_component_1.PagerPageSizesComponent,
    pager_template_directive_1.PagerTemplateDirective,
    localized_messages_directive_1.LocalizedMessagesDirective,
    custom_messages_component_1.CustomMessagesComponent
];
var providers = [
    {
        provide: kendo_angular_l10n_1.L10N_PREFIX,
        useValue: 'kendo.pager'
    }
];
/**
 * @hidden
 */
var PagerModule = /** @class */ (function () {
    function PagerModule() {
    }
    PagerModule.exports = function () {
        return [
            pager_component_1.PagerComponent,
            pager_prev_buttons_component_1.PagerPrevButtonsComponent,
            pager_next_buttons_component_1.PagerNextButtonsComponent,
            pager_numeric_buttons_component_1.PagerNumericButtonsComponent,
            pager_input_component_1.PagerInputComponent,
            pager_info_component_1.PagerInfoComponent,
            pager_page_sizes_component_1.PagerPageSizesComponent,
            pager_template_directive_1.PagerTemplateDirective
        ];
    };
    PagerModule = tslib_1.__decorate([
        core_1.NgModule({
            declarations: [INTERNAL_COMPONENTS],
            exports: [INTERNAL_COMPONENTS],
            imports: importedModules.slice(),
            providers: [providers]
        })
    ], PagerModule);
    return PagerModule;
}());
exports.PagerModule = PagerModule;
