// angular-auth-oidc-client/esm2020/lib/interceptor/closest-matching-route.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../config/provider/config.provider";
export class ClosestMatchingRouteService {
    constructor(configProvider) {
        this.configProvider = configProvider;
    }
    getConfigIdForClosestMatchingRoute(route) {
        const allConfiguredRoutes = this.getAllConfiguredRoutes();
        for (const routesWithConfig of allConfiguredRoutes) {
            const allRoutesForConfig = routesWithConfig.routes;
            for (const configuredRoute of allRoutesForConfig) {
                if (route.startsWith(configuredRoute)) {
                    return {
                        matchingRoute: configuredRoute,
                        matchingConfigId: routesWithConfig.configId,
                    };
                }
            }
        }
        return {
            matchingRoute: null,
            matchingConfigId: null,
        };
    }
    getAllConfiguredRoutes() {
        const allConfigurations = this.configProvider.getAllConfigurations();
        return allConfigurations.map((x) => ({ routes: x.secureRoutes, configId: x.configId }));
    }
}
ClosestMatchingRouteService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: ClosestMatchingRouteService, deps: [{ token: i1.ConfigurationProvider }], target: i0.ɵɵFactoryTarget.Injectable });
ClosestMatchingRouteService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: ClosestMatchingRouteService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: ClosestMatchingRouteService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.ConfigurationProvider }]; } });
