// angular-auth-oidc-client/esm2020/lib/auth-state/auth-state.service.mjs
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { distinctUntilChanged } from 'rxjs/operators';
import { EventTypes } from '../public-events/event-types';
import * as i0 from "@angular/core";
import * as i1 from "../storage/storage-persistence.service";
import * as i2 from "../logging/logger.service";
import * as i3 from "../public-events/public-events.service";
import * as i4 from "../config/provider/config.provider";
import * as i5 from "../validation/token-validation.service";
const DEFAULT_AUTHRESULT = { isAuthenticated: false, allConfigsAuthenticated: [] };
export class AuthStateService {
    constructor(storagePersistenceService, loggerService, publicEventsService, configurationProvider, tokenValidationService) {
        this.storagePersistenceService = storagePersistenceService;
        this.loggerService = loggerService;
        this.publicEventsService = publicEventsService;
        this.configurationProvider = configurationProvider;
        this.tokenValidationService = tokenValidationService;
        this.authenticatedInternal$ = new BehaviorSubject(DEFAULT_AUTHRESULT);
    }
    get authenticated$() {
        return this.authenticatedInternal$.asObservable().pipe(distinctUntilChanged());
    }
    setAuthenticatedAndFireEvent() {
        const result = this.composeAuthenticatedResult();
        this.authenticatedInternal$.next(result);
    }
    setUnauthenticatedAndFireEvent(configIdToReset) {
        this.storagePersistenceService.resetAuthStateInStorage(configIdToReset);
        const result = this.composeUnAuthenticatedResult();
        this.authenticatedInternal$.next(result);
    }
    updateAndPublishAuthState(authenticationResult) {
        this.publicEventsService.fireEvent(EventTypes.NewAuthenticationResult, authenticationResult);
    }
    setAuthorizationData(accessToken, authResult, configId) {
        this.loggerService.logDebug(configId, `storing the accessToken '${accessToken}'`);
        this.storagePersistenceService.write('authzData', accessToken, configId);
        this.persistAccessTokenExpirationTime(authResult, configId);
        this.setAuthenticatedAndFireEvent();
    }
    getAccessToken(configId) {
        if (!this.isAuthenticated(configId)) {
            return null;
        }
        const token = this.storagePersistenceService.getAccessToken(configId);
        return this.decodeURIComponentSafely(token);
    }
    getIdToken(configId) {
        if (!this.isAuthenticated(configId)) {
            return null;
        }
        const token = this.storagePersistenceService.getIdToken(configId);
        return this.decodeURIComponentSafely(token);
    }
    getRefreshToken(configId) {
        if (!this.isAuthenticated(configId)) {
            return null;
        }
        const token = this.storagePersistenceService.getRefreshToken(configId);
        return this.decodeURIComponentSafely(token);
    }
    getAuthenticationResult(configId) {
        if (!this.isAuthenticated(configId)) {
            return null;
        }
        return this.storagePersistenceService.getAuthenticationResult(configId);
    }
    areAuthStorageTokensValid(configId) {
        if (!this.isAuthenticated(configId)) {
            return false;
        }
        if (this.hasIdTokenExpiredAndRenewCheckIsEnabled(configId)) {
            this.loggerService.logDebug(configId, 'persisted idToken is expired');
            return false;
        }
        if (this.hasAccessTokenExpiredIfExpiryExists(configId)) {
            this.loggerService.logDebug(configId, 'persisted accessToken is expired');
            return false;
        }
        this.loggerService.logDebug(configId, 'persisted idToken and accessToken are valid');
        return true;
    }
    hasIdTokenExpiredAndRenewCheckIsEnabled(configId) {
        const { renewTimeBeforeTokenExpiresInSeconds, enableIdTokenExpiredValidationInRenew } = this.configurationProvider.getOpenIDConfiguration(configId);
        if (!enableIdTokenExpiredValidationInRenew) {
            return false;
        }
        const tokenToCheck = this.storagePersistenceService.getIdToken(configId);
        const idTokenExpired = this.tokenValidationService.hasIdTokenExpired(tokenToCheck, configId, renewTimeBeforeTokenExpiresInSeconds);
        if (idTokenExpired) {
            this.publicEventsService.fireEvent(EventTypes.IdTokenExpired, idTokenExpired);
        }
        return idTokenExpired;
    }
    hasAccessTokenExpiredIfExpiryExists(configId) {
        const { renewTimeBeforeTokenExpiresInSeconds } = this.configurationProvider.getOpenIDConfiguration(configId);
        const accessTokenExpiresIn = this.storagePersistenceService.read('access_token_expires_at', configId);
        const accessTokenHasNotExpired = this.tokenValidationService.validateAccessTokenNotExpired(accessTokenExpiresIn, configId, renewTimeBeforeTokenExpiresInSeconds);
        const hasExpired = !accessTokenHasNotExpired;
        if (hasExpired) {
            this.publicEventsService.fireEvent(EventTypes.TokenExpired, hasExpired);
        }
        return hasExpired;
    }
    isAuthenticated(configId) {
        return !!this.storagePersistenceService.getAccessToken(configId) && !!this.storagePersistenceService.getIdToken(configId);
    }
    decodeURIComponentSafely(token) {
        if (token) {
            return decodeURIComponent(token);
        }
        else {
            return '';
        }
    }
    persistAccessTokenExpirationTime(authResult, configId) {
        if (authResult?.expires_in) {
            const accessTokenExpiryTime = new Date(new Date().toUTCString()).valueOf() + authResult.expires_in * 1000;
            this.storagePersistenceService.write('access_token_expires_at', accessTokenExpiryTime, configId);
        }
    }
    composeAuthenticatedResult() {
        if (!this.configurationProvider.hasManyConfigs()) {
            const { configId } = this.configurationProvider.getOpenIDConfiguration();
            return { isAuthenticated: true, allConfigsAuthenticated: [{ configId, isAuthenticated: true }] };
        }
        return this.checkAllConfigsIfTheyAreAuthenticated();
    }
    composeUnAuthenticatedResult() {
        if (!this.configurationProvider.hasManyConfigs()) {
            const { configId } = this.configurationProvider.getOpenIDConfiguration();
            return { isAuthenticated: false, allConfigsAuthenticated: [{ configId, isAuthenticated: false }] };
        }
        return this.checkAllConfigsIfTheyAreAuthenticated();
    }
    checkAllConfigsIfTheyAreAuthenticated() {
        const configs = this.configurationProvider.getAllConfigurations();
        const allConfigsAuthenticated = configs.map(({ configId }) => ({
            configId,
            isAuthenticated: this.isAuthenticated(configId),
        }));
        const isAuthenticated = allConfigsAuthenticated.every((x) => !!x.isAuthenticated);
        return { allConfigsAuthenticated, isAuthenticated };
    }
}
AuthStateService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: AuthStateService, deps: [{ token: i1.StoragePersistenceService }, { token: i2.LoggerService }, { token: i3.PublicEventsService }, { token: i4.ConfigurationProvider }, { token: i5.TokenValidationService }], target: i0.ɵɵFactoryTarget.Injectable });
AuthStateService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: AuthStateService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: AuthStateService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.StoragePersistenceService }, { type: i2.LoggerService }, { type: i3.PublicEventsService }, { type: i4.ConfigurationProvider }, { type: i5.TokenValidationService }]; } });
