// @progress/kendo-angular-grid/dist/es/selection/pair-set.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
/**
 * @hidden
 *
 * Quick look-up structure for combinations of keys.
 * Similar to the native JS Set, however, working with a couple of keys instead of with a single key.
 * Supports both primitive keys and object keys (compared by reference).
 */
var PairSet = /** @class */ (function () {
    function PairSet(items, keyXField, keyYField) {
        var _this = this;
        /**
         * Holds a set of Y keys for each defined X key.
         * Each X key creates a map which holds a set of Y keys.
         *
         * Map { 1 => Set { 1, 2, 3 } } // pairs: [1, 1], [1, 2], [1, 3]
         */
        this.keysX = new Map();
        /**
         * Count the each added or deleted key manually to avoid iterating over all items when calling `this.size`.
         */
        this.totalKeysCount = 0;
        if (items && keyXField && keyYField) {
            items.forEach(function (item) { return _this.add(item[keyXField], item[keyYField]); });
        }
    }
    Object.defineProperty(PairSet.prototype, "size", {
        /**
         * Gets the total number of X/Y key pairs.
         */
        get: function () {
            return this.totalKeysCount;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Adds a couple of items identified as a combination.
     */
    PairSet.prototype.add = function (keyX, keyY) {
        if (!this.keysX.has(keyX)) {
            this.keysX.set(keyX, new Set());
        }
        if (!this.has(keyX, keyY)) {
            this.keysX.get(keyX).add(keyY);
            this.totalKeysCount += 1;
        }
    };
    /**
     * Adds a combination of a couple of items identified together.
     */
    PairSet.prototype.delete = function (keyX, keyY) {
        if (this.has(keyX, keyY)) {
            this.keysX.get(keyX).delete(keyY);
            this.totalKeysCount -= 1;
        }
    };
    /**
     * Checks whether the defined combination is stored.
     */
    PairSet.prototype.has = function (keyX, keyY) {
        return this.keysX.has(keyX) && this.keysX.get(keyX).has(keyY);
    };
    /**
     * Clears all key combinations.
     */
    PairSet.prototype.clear = function () {
        this.keysX.clear();
        this.totalKeysCount = 0;
    };
    /**
     * Converts the persisted data structure to an array of objects,
     * using the provided field names for the object props.
     */
    PairSet.prototype.toArray = function (keyXField, keyYField) {
        return Array.from(this.keysX).reduce(function (pairs, pair) {
            // Array.from(mapInstance) returns an array of arrays [[itemKey1, columnKeysSet1], [itemKey2, columnKeysSet2]]
            var keyX = pair[0], keysY = pair[1];
            Array.from(keysY).forEach(function (keyY) {
                var _a;
                return pairs.push((_a = {}, _a[keyXField] = keyX, _a[keyYField] = keyY, _a));
            });
            return pairs;
        }, []);
    };
    return PairSet;
}());
export { PairSet };
