// @progress/kendo-angular-grid/dist/es/selection/selectall-checkbox.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { detectIE } from './../utils';
import { Directive, HostBinding, Renderer2, ElementRef, Input, Output, EventEmitter, NgZone } from '@angular/core';
import { SelectionService } from './selection.service';
import { isPresent } from '../utils';
import { hasObservers } from '@progress/kendo-angular-common';
/**
 * Represents the select-all checkbox feature of the Grid ([see example]({% slug selection_grid %}#toc-select-all-feature)).
 */
var SelectAllCheckboxDirective = /** @class */ (function () {
    function SelectAllCheckboxDirective(selectionService, el, renderer, ngZone) {
        var _this = this;
        this.selectionService = selectionService;
        this.el = el;
        this.renderer = renderer;
        this.ngZone = ngZone;
        /**
         * Fires when the user clicks the `kendoGridSelectAllCheckbox` select-all checkbox
         * ([see example]({% slug selection_grid %}#toc-select-all-feature)).
         */
        this.selectAllChange = new EventEmitter();
        this.type = 'checkbox';
        this.stateSet = false;
        this.ngZone.runOutsideAngular(function () {
            _this.destroyClick = _this.renderer.listen(_this.el.nativeElement, 'click', _this.onClick.bind(_this));
        });
    }
    SelectAllCheckboxDirective.prototype.ngAfterContentChecked = function () {
        this.setState();
    };
    SelectAllCheckboxDirective.prototype.ngOnChanges = function () {
        this.stateSet = true;
    };
    SelectAllCheckboxDirective.prototype.ngOnDestroy = function () {
        if (this.destroyClick) {
            this.destroyClick();
        }
    };
    /**
     * @hidden
     */
    SelectAllCheckboxDirective.prototype.onClick = function () {
        var _this = this;
        // yields consistent cross-browser behavior when clicking an indeterminate checkbox
        var undefinedCheckedStateInIE = detectIE() && this.selectionService.selectAllState === undefined;
        var isChecked = undefinedCheckedStateInIE ? true : this.el.nativeElement.checked;
        var options = this.selectionService.options;
        var enabledAndMultiple = options.enabled && options.mode === 'multiple';
        var shouldEmitSelectAll = hasObservers(this.selectAllChange);
        if (enabledAndMultiple || shouldEmitSelectAll) {
            this.ngZone.run(function () {
                if (enabledAndMultiple) {
                    _this.selectionService.updateAll(isChecked);
                }
                if (shouldEmitSelectAll) {
                    _this.selectAllChange.emit(isChecked ? 'checked' : 'unchecked');
                }
            });
        }
    };
    /**
     * @hidden
     */
    SelectAllCheckboxDirective.prototype.setState = function () {
        var state = this.stateSet ? this.stateToBool() : this.selectionService.selectAllState;
        var elem = this.el.nativeElement;
        this.renderer.setProperty(elem, 'indeterminate', !isPresent(state));
        this.renderer.setProperty(elem, 'checked', isPresent(state) ? state : false);
    };
    /**
     * @hidden
     */
    SelectAllCheckboxDirective.prototype.stateToBool = function () {
        switch (this.state) {
            case 'checked':
                return true;
            case 'unchecked':
                return false;
            default:
                return undefined;
        }
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], SelectAllCheckboxDirective.prototype, "state", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], SelectAllCheckboxDirective.prototype, "selectAllChange", void 0);
    tslib_1.__decorate([
        HostBinding('attr.type'),
        tslib_1.__metadata("design:type", String)
    ], SelectAllCheckboxDirective.prototype, "type", void 0);
    SelectAllCheckboxDirective = tslib_1.__decorate([
        Directive({
            selector: '[kendoGridSelectAllCheckbox]'
        }),
        tslib_1.__metadata("design:paramtypes", [SelectionService,
            ElementRef,
            Renderer2,
            NgZone])
    ], SelectAllCheckboxDirective);
    return SelectAllCheckboxDirective;
}());
export { SelectAllCheckboxDirective };
