// @progress/kendo-angular-dateinputs/dist/es2015/timepicker/localization/timepicker-localized-messages.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
var TimePickerLocalizedMessagesDirective_1;
import { Directive, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { TimePickerMessages } from './messages';
/**
 * @hidden
 */
let TimePickerLocalizedMessagesDirective = TimePickerLocalizedMessagesDirective_1 = class TimePickerLocalizedMessagesDirective extends TimePickerMessages {
    constructor(service) {
        super();
        this.service = service;
    }
};
TimePickerLocalizedMessagesDirective = TimePickerLocalizedMessagesDirective_1 = tslib_1.__decorate([
    Directive({
        providers: [
            {
                provide: TimePickerMessages,
                useExisting: forwardRef(() => TimePickerLocalizedMessagesDirective_1) // tslint:disable-line:no-forward-ref
            }
        ],
        selector: '[kendoTimePickerLocalizedMessages]'
    }),
    tslib_1.__metadata("design:paramtypes", [LocalizationService])
], TimePickerLocalizedMessagesDirective);
export { TimePickerLocalizedMessagesDirective };
