// @safe-fleet/sf-menu/esm2015/lib/components/menu-item/menu-item.component.js
import { Component, Input } from '@angular/core';
import { MenuEventType } from '../../models/menu-events';
import { MenuItem } from '../../models/menu-item';
import { SfMenuService } from '../../sf-menu.service';
export class MenuItemComponent {
    constructor(svc) {
        this.svc = svc;
        this.collapsed = false;
        this.level = 1;
    }
    ngOnInit() {
        this.text = this.svc.colors.gray900;
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            this.menuConfig = cfg;
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
        });
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
    click(evt) {
        switch (this.itemType) {
            case MenuEventType.ShortcutClick:
                this.shortcutClicked();
                break;
            case MenuEventType.UserActionClick:
                this.userOptionClicked();
                break;
            case MenuEventType.RouteClick:
                this.svc.routeClicked(this.item, evt);
                break;
            default:
                console.error('event not found');
        }
    }
    shortcutClicked() {
        this.svc.events$.next({
            type: MenuEventType.ShortcutClick,
            value: this.item
        });
        if (window.innerWidth <= this.svc.SMALL_RESOLUTION && !this.collapsed) {
            this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
        }
    }
    userOptionClicked() {
        this.svc.events$.next({
            type: MenuEventType.UserActionClick,
            value: this.item
        });
        if (window.innerWidth <= this.svc.SMALL_RESOLUTION && !this.collapsed) {
            this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
        }
    }
}
MenuItemComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line: component-selector
                selector: '[sf-menu-item]',
                template: "<div\n  class=\"li-container\"\n  [class.show-arrow]=\"item.items\"\n  [class.active-item]=\"item.active\"\n  [class.collapsed]=\"collapsed\"\n  [style.color]=\"text\"\n>\n  <div\n    [class.expanded-item]=\"item.expanded && item.items\"\n    class=\"main-div menu-item\"\n    [ngStyle]=\"{ 'border-left-color': menuConfig?.borderColor, 'padding-left': 10 * level + 'px' }\"\n    title=\"{{ menuConfig.tooltips || collapsed ? item.text : '' }}\"\n    (click)=\"click($event)\"\n  >\n    <div class=\"icon-container\" *ngIf=\"item.iconClass || item.iconPath\">\n      <div class=\"sf-icons-menu\" *ngIf=\"item.iconClass\" [class.collapsed]=\"collapsed\">\n        <em [ngClass]=\"item.iconClass\"></em>\n      </div>\n      <div\n        class=\"sf-menu-logo-bg\"\n        *ngIf=\"!item.iconClass && item.iconPath\"\n        [ngStyle]=\"{\n          'background-image': 'url(' + item.iconPath + ')'\n        }\"\n      ></div>\n    </div>\n    <span>{{ item.text }}</span>\n    <em class=\"fas fa-chevron-down\"></em>\n  </div>\n  <ul *ngIf=\"item.items\" class=\"item-children\">\n    <li\n      sf-menu-item\n      *ngFor=\"let child of item.items\"\n      [class.show-arrow]=\"child.items\"\n      [class.active-item]=\"child.active\"\n      [item]=\"child\"\n      [level]=\"level + 1\"\n      [itemType]=\"itemType\"\n    ></li>\n  </ul>\n</div>\n",
                styles: [":host .li-container{display:inline-block;width:100%;padding:0;overflow:hidden;cursor:pointer}:host .li-container .icon-container{width:24px}:host .li-container>div.main-div{margin-left:2px}:host .li-container>.menu-item{display:flex;align-items:center;justify-content:flex-start;padding:7px 10px;position:relative;min-height:38px}:host .li-container>.menu-item .sf-menu-logo-bg{width:1.6em;height:1.55em}:host .li-container>.menu-item span{width:calc(100% - 24px);min-width:100px;padding:0 10px;font-size:14px;line-height:18px;opacity:.71;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}:host .li-container>.menu-item:hover{position:relative}:host .li-container>.menu-item:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}:host .li-container>.menu-item:hover span{opacity:1}:host .li-container>.menu-item .fa-chevron-down{display:none;position:absolute;right:15px;top:16px;width:auto;height:auto;font-size:12px}:host .li-container.show-arrow>div .fa-chevron-down{display:inline-block;transform:rotate(0deg)}:host .li-container.show-arrow>div.expanded-item span{opacity:1}:host .li-container.show-arrow>div.expanded-item em.fa-chevron-down,:host .li-container.show-arrow>div.expanded-item i.fa-chevron-down{transform:rotate(180deg)}:host .li-container.show-arrow>div:hover{position:relative}:host .li-container.show-arrow>div:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}:host .li-container.show-arrow>div:hover span{opacity:1}:host .li-container.active-item>div.main-div{border-left:3px solid;margin-left:0}:host .li-container.active-item>div.main-div:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.2}:host .li-container.active-item>div.main-div span{opacity:1}:host .li-container.collapsed div{justify-content:center}:host .li-container.collapsed em,:host .li-container.collapsed i{width:auto}:host .li-container.collapsed .item-children,:host .li-container.collapsed span{display:none}:host .sf-icons-menu{text-align:center}:host .sf-icons-menu em,:host .sf-icons-menu i{font-size:16px}:host ul{list-style:none;padding-left:0;width:100%}.expanded-item+.item-children{display:block}ul.item-children{display:none}.sf-icons-menu.collapsed em{font-size:var(--fs-300)}.sf-menu-logo-bg{background:50%/contain no-repeat}"]
            },] }
];
MenuItemComponent.ctorParameters = () => [
    { type: SfMenuService }
];
MenuItemComponent.propDecorators = {
    item: [{ type: Input }],
    level: [{ type: Input }],
    itemType: [{ type: Input }]
};
