// @safe-fleet/sf-menu/esm2015/lib/components/customer-logo/customer-logo.component.js
import { Component, HostBinding } from '@angular/core';
import { SfMenuService } from '../../sf-menu.service';
export class CustomerLogoComponent {
    constructor(svc) {
        this.svc = svc;
        this.collapsed = false;
    }
    get clr() {
        return this.background;
    }
    ngOnInit() {
        this.background = this.svc.colors.gray100;
        this.configSubs$ = this.svc.menuConfig$.subscribe((cfg) => {
            this.menuConfig = cfg;
        });
        this.collapsedSubs$ = this.svc.menuCollapse$.subscribe((value) => {
            this.collapsed = value;
        });
    }
    ngOnDestroy() {
        this.configSubs$.unsubscribe();
        this.collapsedSubs$.unsubscribe();
    }
}
CustomerLogoComponent.decorators = [
    { type: Component, args: [{
                selector: 'sf-customer-logo',
                template: "<div\n  data-testid=\"menu-logo-bg\"\n  class=\"sf-menu-logo-bg\"\n  *ngIf=\"menuConfig.logoUrl && !menuConfig.logoClass\"\n  [ngStyle]=\"{\n    'background-image': 'url(' + menuConfig.logoUrl + ')'\n  }\"\n></div>\n<em [ngClass]=\"menuConfig.logoClass\" *ngIf=\"menuConfig.logoClass && !menuConfig.logoUrl\"></em>\n",
                styles: [":host{display:flex;align-items:center;justify-content:center;width:100%;flex-direction:column;height:auto;padding:5px;position:relative}:host .sf-menu-logo-bg{background:50%/contain no-repeat;height:2.0625rem;width:4.8125rem}:host.collapsed .sf-menu-logo-bg{width:80%}"]
            },] }
];
CustomerLogoComponent.ctorParameters = () => [
    { type: SfMenuService }
];
CustomerLogoComponent.propDecorators = {
    clr: [{ type: HostBinding, args: ['style.background-color',] }]
};
