// @progress/kendo-angular-grid/dist/es2015/filtering/cell/filter-cell-wrapper.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, Input, HostBinding } from '@angular/core';
import { FilterService } from '../filter.service';
import { isPresent, isNullOrEmptyString } from '../../utils';
import { FilterInputWrapperComponent } from "../filter-input-wrapper.component";
const EMPTY_FILTER_OPERATORS = ['isnull', 'isnotnull', 'isempty', 'isnotempty'];
/**
 * @hidden
 */
let FilterCellWrapperComponent = class FilterCellWrapperComponent extends FilterInputWrapperComponent {
    constructor(filterService) {
        super(filterService);
        this.showOperators = true;
    }
    get hostClasses() {
        return true;
    }
    get overrideBaseClasses() {
        return false;
    }
    get showButton() {
        const filter = this.currentFilter;
        return isPresent(filter) && (!isNullOrEmptyString(filter.value) ||
            EMPTY_FILTER_OPERATORS.indexOf(String(filter.operator)) >= 0);
    }
    filterChange(filter) {
        this.applyFilter(filter);
    }
};
tslib_1.__decorate([
    HostBinding('class.k-filtercell-wrapper'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], FilterCellWrapperComponent.prototype, "hostClasses", null);
tslib_1.__decorate([
    HostBinding('class.k-filtercell'),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [])
], FilterCellWrapperComponent.prototype, "overrideBaseClasses", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], FilterCellWrapperComponent.prototype, "showOperators", void 0);
FilterCellWrapperComponent = tslib_1.__decorate([
    Component({
        selector: 'kendo-grid-filter-wrapper-cell',
        template: `
        <ng-content></ng-content>
        <kendo-grid-filter-cell-operators
            [showOperators]="showOperators"
            [operators]="operators"
            (clear)="onClear()"
            [showButton]="showButton"
            [(value)]="currentOperator">
        </kendo-grid-filter-cell-operators>
    `
    }),
    tslib_1.__metadata("design:paramtypes", [FilterService])
], FilterCellWrapperComponent);
export { FilterCellWrapperComponent };
