// @progress/kendo-angular-layout/dist/npm/tabstrip/models/scrollable-settings.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var constants_1 = require("../constants");
/**
 * @hidden
 */
var normalizeSettings = function (_a) {
    var _b = _a.enabled, enabled = _b === void 0 ? true : _b, _c = _a.scrollButtons, scrollButtons = _c === void 0 ? 'auto' : _c, _d = _a.mouseScroll, mouseScroll = _d === void 0 ? true : _d, _e = _a.buttonScrollSpeed, buttonScrollSpeed = _e === void 0 ? constants_1.BUTTON_SCROLL_SPEED : _e, _f = _a.mouseScrollSpeed, mouseScrollSpeed = _f === void 0 ? constants_1.MOUSE_SCROLL_SPEED : _f, prevButtonIcon = _a.prevButtonIcon, nextButtonIcon = _a.nextButtonIcon;
    return ({
        enabled: enabled,
        scrollButtons: scrollButtons,
        mouseScroll: mouseScroll,
        buttonScrollSpeed: buttonScrollSpeed,
        mouseScrollSpeed: mouseScrollSpeed,
        prevButtonIcon: prevButtonIcon,
        nextButtonIcon: nextButtonIcon
    });
};
var ɵ0 = normalizeSettings;
exports.ɵ0 = ɵ0;
/**
 * @hidden
 */
exports.normalizeScrollableSettings = function (settings) {
    return normalizeSettings(settings === false ? { enabled: false } : settings);
};
