// @progress/kendo-angular-buttons/dist/es/splitbutton/localization/custom-messages.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { Messages } from './messages';
/**
 * Custom component messages override default component messages
 * ([see example]({% slug rtl_buttons %}).
 */
var SplitButtonCustomMessagesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(SplitButtonCustomMessagesComponent, _super);
    function SplitButtonCustomMessagesComponent(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    SplitButtonCustomMessagesComponent_1 = SplitButtonCustomMessagesComponent;
    Object.defineProperty(SplitButtonCustomMessagesComponent.prototype, "override", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    var SplitButtonCustomMessagesComponent_1;
    SplitButtonCustomMessagesComponent = SplitButtonCustomMessagesComponent_1 = tslib_1.__decorate([
        Component({
            providers: [
                {
                    provide: Messages,
                    useExisting: forwardRef(function () { return SplitButtonCustomMessagesComponent_1; })
                }
            ],
            selector: 'kendo-splitbutton-messages',
            template: ""
        }),
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], SplitButtonCustomMessagesComponent);
    return SplitButtonCustomMessagesComponent;
}(Messages));
export { SplitButtonCustomMessagesComponent };
