// @progress/kendo-angular-listview/dist/npm/editing/edit.service.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var forms_1 = require("@angular/forms");
var utils_1 = require("../utils");
var rxjs_1 = require("rxjs");
var operators_1 = require("rxjs/operators");
/**
 * @hidden
 */
var isEqual = function (index) { return function (item) { return item.index === index; }; };
var ɵ0 = isEqual;
exports.ɵ0 = ɵ0;
/**
 * @hidden
 */
var isNotEqual = function (index) { return function (item) { return item.index !== index; }; };
var ɵ1 = isNotEqual;
exports.ɵ1 = ɵ1;
/**
 * @hidden
 */
var isNewItem = function (index) { return index === -1 || index === undefined; };
var ɵ2 = isNewItem;
exports.ɵ2 = ɵ2;
/**
 * @hidden
 */
var EditService = /** @class */ (function () {
    function EditService(ngZone) {
        var _this = this;
        this.ngZone = ngZone;
        this.changes = new core_1.EventEmitter();
        this.editedIndices = [];
        this.changedSource = new rxjs_1.Subject();
        this.changed = this.changedSource.asObservable().pipe(operators_1.switchMap(function () { return _this.ngZone.onStable.asObservable().pipe(operators_1.take(1)); }));
    }
    EditService.prototype.editItem = function (index, group) {
        if (group === void 0) { group = undefined; }
        this.editedIndices.push({ index: index, group: group });
        this.onChanged();
    };
    EditService.prototype.addItem = function (group) {
        this.newItem = { group: group };
        this.onChanged();
    };
    EditService.prototype.isEditing = function () {
        return this.editedIndices.length > 0;
    };
    Object.defineProperty(EditService.prototype, "hasNewItem", {
        get: function () {
            return utils_1.isPresent(this.newItem);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(EditService.prototype, "newDataItem", {
        get: function () {
            if (this.hasNewItem) {
                return this.newItem.group.value;
            }
            return {};
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(EditService.prototype, "newItemGroup", {
        get: function () {
            if (this.hasNewItem) {
                return this.newItem.group;
            }
            return new forms_1.FormGroup({});
        },
        enumerable: true,
        configurable: true
    });
    EditService.prototype.editGroup = function (index) {
        return this.context(index).group;
    };
    EditService.prototype.close = function (index) {
        if (isNewItem(index)) {
            this.newItem = undefined;
            return;
        }
        this.editedIndices = this.editedIndices.filter(isNotEqual(index));
        this.onChanged();
    };
    EditService.prototype.context = function (index) {
        if (isNewItem(index)) {
            return this.newItem;
        }
        return this.findByIndex(index);
    };
    EditService.prototype.isEdited = function (index) {
        if (isNewItem(index) && utils_1.isPresent(this.newItem)) {
            return true;
        }
        return utils_1.isPresent(this.findByIndex(index));
    };
    EditService.prototype.hasEdited = function (index) {
        return utils_1.isPresent(this.context(index));
    };
    EditService.prototype.beginEdit = function (itemIndex) {
        this.changes.emit({ action: 'edit', itemIndex: itemIndex });
    };
    EditService.prototype.beginAdd = function () {
        this.changes.emit({ action: 'add' });
    };
    EditService.prototype.endEdit = function (itemIndex) {
        var formGroup = this.context(itemIndex).group;
        this.changes.emit({ action: 'cancel', itemIndex: itemIndex, formGroup: formGroup, isNew: isNewItem(itemIndex) });
    };
    EditService.prototype.save = function (itemIndex) {
        var formGroup = this.context(itemIndex).group;
        this.changes.emit({ action: 'save', itemIndex: itemIndex, formGroup: formGroup, isNew: isNewItem(itemIndex) });
    };
    EditService.prototype.remove = function (itemIndex) {
        this.changes.emit({ action: 'remove', itemIndex: itemIndex });
    };
    EditService.prototype.findByIndex = function (index) {
        return this.editedIndices.find(isEqual(index));
    };
    EditService.prototype.onChanged = function () {
        var _this = this;
        this.ngZone.runOutsideAngular(function () {
            _this.changedSource.next();
        });
    };
    EditService = tslib_1.__decorate([
        core_1.Injectable(),
        tslib_1.__metadata("design:paramtypes", [core_1.NgZone])
    ], EditService);
    return EditService;
}());
exports.EditService = EditService;
