// @progress/kendo-angular-label/dist/npm/main.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var label_directive_1 = require("./label.directive");
exports.LabelDirective = label_directive_1.LabelDirective;
var label_module_1 = require("./label.module");
exports.LabelModule = label_module_1.LabelModule;
var floating_label_module_1 = require("./floating-label/floating-label.module");
exports.FloatingLabelModule = floating_label_module_1.FloatingLabelModule;
var floating_label_component_1 = require("./floating-label/floating-label.component");
exports.FloatingLabelComponent = floating_label_component_1.FloatingLabelComponent;
var label_component_1 = require("./label/label.component");
exports.LabelComponent = label_component_1.LabelComponent;
var custom_messages_component_1 = require("./localization/custom-messages.component");
exports.CustomMessagesComponent = custom_messages_component_1.CustomMessagesComponent;
