// @progress/kendo-angular-inputs/dist/es/colorpicker/color-contrast-svg.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, HostBinding, Input } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { isPresent } from '../common/utils';
import { AA_RATIO, STEP_COUNT, AAA_RATIO } from './constants';
import { bezierCommand, getContrastFromTwoRGBAs, getRGBA, getColorFromHSV, svgPath, controlPoint, line } from './utils';
/**
 * @hidden
 */
var ColorContrastSvgComponent = /** @class */ (function () {
    function ColorContrastSvgComponent() {
        this.hostClass = true;
    }
    ColorContrastSvgComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        this.metrics = this.wrapper.getBoundingClientRect();
        this.oldA = this.hsva.value.a;
        this.oldH = this.hsva.value.h;
        this.hsva.subscribe(function (value) {
            if (value.h !== _this.oldH || value.a !== _this.oldA) {
                _this.oldH = value.h;
                _this.oldA = value.a;
                _this.setPaths();
            }
        });
    };
    ColorContrastSvgComponent.prototype.ngOnChanges = function (changes) {
        if (isPresent(changes.backgroundColor) && this.metrics) {
            this.setPaths();
        }
    };
    ColorContrastSvgComponent.prototype.setPaths = function () {
        var bezierCommandCalc = bezierCommand(controlPoint(line));
        this.paths = [svgPath(this.getPaths(AA_RATIO, STEP_COUNT), bezierCommandCalc),
            svgPath(this.getPaths(AA_RATIO, STEP_COUNT, true), bezierCommandCalc),
            svgPath(this.getPaths(AAA_RATIO, STEP_COUNT), bezierCommandCalc),
            svgPath(this.getPaths(AAA_RATIO, STEP_COUNT, true), bezierCommandCalc)];
    };
    ColorContrastSvgComponent.prototype.findValue = function (contrast, saturation, low, high, comparer) {
        var mid = (low + high) / 2;
        var hsva = Object.assign({}, this.hsva.value, { s: saturation / this.metrics.width, v: 1 - mid / this.metrics.height });
        var currentContrast = getContrastFromTwoRGBAs(getRGBA(getColorFromHSV(hsva)), getRGBA(this.backgroundColor || ''));
        if (low + 0.5 > high) {
            if (currentContrast < contrast + 1 && currentContrast > contrast - 1) {
                return mid;
            }
            else {
                return null;
            }
        }
        if (comparer(currentContrast, contrast)) {
            return this.findValue(contrast, saturation, low, high - (high - low) / 2, comparer);
        }
        return this.findValue(contrast, saturation, low + (high - low) / 2, high, comparer);
    };
    ColorContrastSvgComponent.prototype.getPaths = function (contrast, stepCount, reversed) {
        if (reversed === void 0) { reversed = false; }
        var points = [];
        for (var i = 0; i <= this.metrics.width; i += this.metrics.width / stepCount) {
            var value = this.findValue(contrast, i, 0, this.metrics.height, reversed ? (function (a, b) { return a < b; }) : (function (a, b) { return a > b; }));
            if (value !== null) {
                points.push([i, value]);
            }
        }
        return points;
    };
    tslib_1.__decorate([
        HostBinding('class.k-color-contrast-svg'),
        tslib_1.__metadata("design:type", Boolean)
    ], ColorContrastSvgComponent.prototype, "hostClass", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], ColorContrastSvgComponent.prototype, "wrapper", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", BehaviorSubject)
    ], ColorContrastSvgComponent.prototype, "hsva", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], ColorContrastSvgComponent.prototype, "backgroundColor", void 0);
    ColorContrastSvgComponent = tslib_1.__decorate([
        Component({
            selector: '[kendoColorContrastSvg]',
            template: "\n        <svg:path *ngFor=\"let path of paths\" [attr.d]=\"path\" fill=\"none\" stroke=\"white\" stroke-width=\"1\"></svg:path>\n    "
        })
    ], ColorContrastSvgComponent);
    return ColorContrastSvgComponent;
}());
export { ColorContrastSvgComponent };
