// @progress/kendo-angular-inputs/dist/es/colorpicker/flatcolorpicker-actions.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, HostBinding, EventEmitter, Output, ElementRef, ViewChild } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
/**
 * @hidden
 */
var FlatColorPickerActionButtonsComponent = /** @class */ (function () {
    function FlatColorPickerActionButtonsComponent(localizationService) {
        this.localizationService = localizationService;
        this.hostClasses = true;
        this.actionButtonClick = new EventEmitter();
    }
    FlatColorPickerActionButtonsComponent.prototype.getText = function (text) {
        return this.localizationService.get(text);
    };
    FlatColorPickerActionButtonsComponent.prototype.onActionButtonClick = function (type, ev) {
        var args = {
            target: type,
            originalEvent: ev
        };
        this.actionButtonClick.emit(args);
    };
    tslib_1.__decorate([
        HostBinding('class.k-coloreditor-footer'),
        HostBinding('class.k-actions'),
        HostBinding('class.k-hstack'),
        tslib_1.__metadata("design:type", Boolean)
    ], FlatColorPickerActionButtonsComponent.prototype, "hostClasses", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], FlatColorPickerActionButtonsComponent.prototype, "actionButtonClick", void 0);
    tslib_1.__decorate([
        ViewChild('last', { read: ElementRef, static: false }),
        tslib_1.__metadata("design:type", ElementRef)
    ], FlatColorPickerActionButtonsComponent.prototype, "lastButton", void 0);
    FlatColorPickerActionButtonsComponent = tslib_1.__decorate([
        Component({
            selector: '[kendoFlatColorPickerActionButtons]',
            template: "\n        <button\n            class=\"k-coloreditor-cancel k-button k-button-md k-button-solid k-button-solid-base\"\n            [attr.title]=\"getText('cancelButton')\"\n            (click)=\"onActionButtonClick('cancel', $event)\"\n            type=\"button\"\n        >{{getText('cancelButton')}}</button>\n        <button #last\n            class=\"k-coloreditor-apply k-button k-button-md k-button-solid k-button-solid-primary\"\n            [attr.title]=\"getText('applyButton')\"\n            (click)=\"onActionButtonClick('apply', $event)\"\n            type=\"button\"\n        >{{getText('applyButton')}}</button>\n    "
        }),
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], FlatColorPickerActionButtonsComponent);
    return FlatColorPickerActionButtonsComponent;
}());
export { FlatColorPickerActionButtonsComponent };
