// @safe-fleet/sf-menu/bundles/safe-fleet-sf-menu.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('rxjs'), require('@angular/core'), require('@angular/router'), require('rxjs/operators'), require('@angular/animations'), require('@progress/kendo-angular-tooltip'), require('@angular/common'), require('@angular/forms'), require('@angular/platform-browser/animations'), require('@progress/kendo-angular-inputs'), require('@progress/kendo-angular-menu'), require('@progress/kendo-angular-popup')) :
    typeof define === 'function' && define.amd ? define('@safe-fleet/sf-menu', ['exports', 'rxjs', '@angular/core', '@angular/router', 'rxjs/operators', '@angular/animations', '@progress/kendo-angular-tooltip', '@angular/common', '@angular/forms', '@angular/platform-browser/animations', '@progress/kendo-angular-inputs', '@progress/kendo-angular-menu', '@progress/kendo-angular-popup'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['safe-fleet'] = global['safe-fleet'] || {}, global['safe-fleet']['sf-menu'] = {}), global.rxjs, global.ng.core, global.ng.router, global.rxjs.operators, global.ng.animations, global.kendoAngularTooltip, global.ng.common, global.ng.forms, global.ng.platformBrowser.animations, global.kendoAngularInputs, global.kendoAngularMenu, global.kendoAngularPopup));
}(this, (function (exports, rxjs, core, router, operators, animations, kendoAngularTooltip, common, forms, animations$1, kendoAngularInputs, kendoAngularMenu, kendoAngularPopup) { 'use strict';

    var MenuConfig = /** @class */ (function () {
        function MenuConfig() {
            this.helpText = 'Help';
            this.logoutText = 'Logout';
            // optional menu
            this.hideMenu = false;
            this.tooltips = false;
            this.denyChangeTheme = false;
            this.menuIgnoresTheme = true;
            this.hideProductName = false;
            this.routeClickPreventCollapse = false;
            this.disableUserNameClick = false;
        }
        return MenuConfig;
    }());
    var SsoManageAccountConfig = /** @class */ (function () {
        function SsoManageAccountConfig() {
            this.disable = false;
        }
        return SsoManageAccountConfig;
    }());

    (function (MenuEventType) {
        MenuEventType[MenuEventType["ProductClick"] = 0] = "ProductClick";
        MenuEventType[MenuEventType["ShortcutClick"] = 1] = "ShortcutClick";
        MenuEventType[MenuEventType["RouteClick"] = 2] = "RouteClick";
        MenuEventType[MenuEventType["UserActionClick"] = 3] = "UserActionClick";
        MenuEventType[MenuEventType["ChangeTenantClick"] = 4] = "ChangeTenantClick";
        MenuEventType[MenuEventType["OpenTenantListClick"] = 5] = "OpenTenantListClick";
    })(exports.MenuEventType || (exports.MenuEventType = {}));

    var SfMenuService = /** @class */ (function () {
        function SfMenuService(router$1) {
            var _this = this;
            this.router = router$1;
            this.SMALL_RESOLUTION = 768;
            this.menuCollapse$ = new rxjs.BehaviorSubject(false);
            this.events$ = new rxjs.Subject();
            this.menuConfig$ = new rxjs.BehaviorSubject(undefined);
            // store nodes to expand on navigation end
            this.nodesToExpand = [];
            this.themes = ['light', 'dark'];
            // theming
            this.setBaseColors();
            this.currentTheme = new rxjs.BehaviorSubject(this.themes[0]);
            this.storedTheme = localStorage.getItem('theme');
            if (!this.storedTheme) {
                this.storedTheme = window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches ? this.themes[1] : this.themes[0];
            }
            this.setMenuConfig(new MenuConfig());
            this.router.events.pipe(operators.filter(function (e) { return e instanceof router.NavigationEnd; })).subscribe(function (e) {
                var _a, _b;
                _this.nodesToExpand.length = 0;
                (_b = (_a = _this.menuConfig) === null || _a === void 0 ? void 0 : _a.routes) === null || _b === void 0 ? void 0 : _b.forEach(function (r) { return _this.setCurrentRouteActive(r, e.urlAfterRedirects); });
            });
        }
        Object.defineProperty(SfMenuService.prototype, "colors", {
            get: function () {
                return this.menuConfig.menuIgnoresTheme || this.menuConfig.denyChangeTheme ? this.noThemeColors : this.themeColors;
            },
            enumerable: false,
            configurable: true
        });
        SfMenuService.prototype.setMenuConfig = function (cfg) {
            this.menuConfig = cfg || new MenuConfig();
            this.menuConfig$.next(this.menuConfig);
            this.setTheme(this.storedTheme);
        };
        SfMenuService.prototype.setBaseColors = function () {
            var notheme = {};
            var theme = {};
            var root = getComputedStyle(document.querySelector(':root'));
            ['blue', 'gray', 'green', 'red', 'orange', 'purple', 'teal', 'yellow'].forEach(function (col) {
                for (var i = 1; i < 10; i++) {
                    notheme["" + col + i * 100] = root.getPropertyValue("--clr-" + col + "-" + (1000 - i * 100));
                    theme["" + col + i * 100] = "var(--clr-" + col + "-" + i * 100 + ")";
                }
            });
            this.noThemeColors = notheme;
            this.themeColors = theme;
        };
        SfMenuService.prototype.routeClicked = function (item, evt) {
            evt.stopPropagation();
            this.events$.next({
                type: exports.MenuEventType.RouteClick,
                value: item
            });
            if (item.items) {
                item.expanded = !item.expanded;
                return;
            }
            if (item.routePath) {
                if (!this.menuConfig.routeClickPreventCollapse) {
                    this.collapseAllItems(this.menuConfig.routes);
                }
                this.router.navigate([item.routePath]);
                return;
            }
            if (item.routeHref) {
                if (!this.menuConfig.routeClickPreventCollapse) {
                    this.collapseAllItems(this.menuConfig.routes);
                }
                if (item.newTab) {
                    window.open(item.routeHref);
                    return;
                }
                if (item.newWindow) {
                    var top = void 0;
                    var left = void 0;
                    var width = void 0;
                    var height = void 0;
                    if (item.newWindowWidth > 200 && item.newWindowHeight > 100) {
                        width = item.newWindowWidth;
                        height = item.newWindowHeight;
                        left = window.screenLeft + window.outerWidth / 2 - width / 2;
                        top = window.screenTop + window.outerHeight / 2 - height / 2;
                    }
                    else {
                        height = window.outerHeight;
                        width = window.outerWidth;
                        left = window.screenLeft;
                        top = window.screenTop;
                    }
                    window.open(item.routeHref, item.id, "top=" + top + ",left=" + left + ",height=" + height + ",width=" + width + ",resizable=1,scrollbars=1");
                }
            }
            if (window.innerWidth <= this.SMALL_RESOLUTION && !this.menuCollapse$.value) {
                this.menuCollapse$.next(!this.menuCollapse$.value);
            }
        };
        SfMenuService.prototype.setCurrentRouteActive = function (route, url) {
            var _this = this;
            var _a, _b;
            route.active = !((_a = route.items) === null || _a === void 0 ? void 0 : _a.length) && "/" + route.routePath === url;
            if (route.active) {
                this.nodesToExpand.forEach(function (n) { return (n.expanded = true); });
            }
            this.nodesToExpand.push(route);
            (_b = route.items) === null || _b === void 0 ? void 0 : _b.forEach(function (ch) { return _this.setCurrentRouteActive(ch, url); });
            this.nodesToExpand.pop();
        };
        // theming
        SfMenuService.prototype.setTheme = function (theme) {
            if (this.menuConfig.denyChangeTheme || this.themes.indexOf(theme) < 0) {
                theme = this.themes[0];
            }
            this.currentTheme.next(theme);
            localStorage.setItem('theme', theme);
            this.themes.forEach(function (t) { return document.querySelector('html').classList.toggle(t, t === theme); });
        };
        SfMenuService.prototype.collapseAllItems = function (routes) {
            var _this = this;
            if (routes) {
                routes.forEach(function (route) {
                    route.expanded = false;
                    if (route.items) {
                        _this.collapseAllItems(route.items);
                    }
                });
            }
        };
        return SfMenuService;
    }());
    SfMenuService.decorators = [
        { type: core.Injectable }
    ];
    SfMenuService.ctorParameters = function () { return [
        { type: router.Router }
    ]; };

    var SfMenuComponent = /** @class */ (function () {
        function SfMenuComponent(svc) {
            this.svc = svc;
            this.collapsed = false;
        }
        Object.defineProperty(SfMenuComponent.prototype, "config", {
            set: function (value) {
                this.svc.setMenuConfig(value);
            },
            enumerable: false,
            configurable: true
        });
        SfMenuComponent.prototype.ngOnInit = function () {
            var _this = this;
            var clr = this.svc.colors;
            this.gradient = "linear-gradient(180deg, " + clr.gray200 + " 0%, " + clr.gray100 + " 100%)";
            this.color = clr.gray900;
            this.configSubs$ = this.svc.menuConfig$.subscribe(function (cfg) {
                _this.menuConfig = cfg;
            });
            this.collapsedSubs$ = this.svc.menuCollapse$.subscribe(function (value) {
                _this.collapsed = value;
                if (_this.tooltipDir) {
                    _this.tooltipDir.hide();
                }
            });
        };
        SfMenuComponent.prototype.ngOnDestroy = function () {
            this.configSubs$.unsubscribe();
            this.collapsedSubs$.unsubscribe();
        };
        SfMenuComponent.prototype.toggleMenu = function () {
            this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
        };
        return SfMenuComponent;
    }());
    SfMenuComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'sf-menu',
                    template: "<div class=\"sf-menu-container\" *ngIf=\"!menuConfig.hideMenu\">\n  <sf-top-menu [ngStyle]=\"{ 'border-bottom-color': menuConfig?.borderColor }\"></sf-top-menu>\n  <div\n    class=\"sf-menu-content\"\n    [class.collapsed]=\"collapsed\"\n    [ngStyle]=\"{ 'border-top-color': menuConfig?.borderColor }\"\n    [style.background]=\"gradient\"\n    kendoTooltip\n    showOn=\"{{ menuConfig.tooltips || collapsed ? 'hover' : 'none' }}\"\n    position=\"bottom\"\n  >\n    <sf-customer-logo [class.collapsed]=\"collapsed\" *ngIf=\"menuConfig.logoUrl || menuConfig.logoClass\"></sf-customer-logo>\n    <sf-menu-products [class.collapsed]=\"collapsed\"></sf-menu-products>\n    <div class=\"fm-menu-scrollable-content\">\n      <div class=\"scroll\">\n        <sf-menu-shortcuts *ngIf=\"menuConfig?.shortcuts && menuConfig.shortcuts.length > 0\"></sf-menu-shortcuts>\n        <sf-menu-routes *ngIf=\"menuConfig?.routes\"></sf-menu-routes>\n      </div>\n    </div>\n    <sf-menu-user></sf-menu-user>\n    <sf-menu-footer-logo></sf-menu-footer-logo>\n  </div>\n  <sf-menu-toggle-btn></sf-menu-toggle-btn>\n  <div class=\"sf-menu-backdrop\" *ngIf=\"!collapsed\" (click)=\"toggleMenu()\" [@appear]></div>\n</div>\n<div class=\"application-content\">\n  <ng-content></ng-content>\n</div>\n",
                    animations: [
                        animations.trigger('appear', [
                            animations.transition(':enter', [animations.style({ opacity: 0 }), animations.animate(180, animations.style({ opacity: 1 }))]),
                            animations.transition(':leave', [animations.style({ opacity: 1 }), animations.animate(180, animations.style({ opacity: 0 }))])
                        ])
                    ],
                    styles: [":host{width:100%;height:100%;display:grid;grid-template-areas:\"menu\" \"content\";grid-template-rows:auto 1fr;grid-template-columns:1fr}@media screen and (min-width:768px){:host{grid-template-areas:\"menu content\";grid-template-rows:1fr;grid-template-columns:auto 1fr}}.sf-menu-container{grid-area:menu;position:relative}.sf-menu-content{height:auto;max-height:calc(100% - 70px);width:250px;border-bottom:none;transition:transform .12s ease-out,width .12s ease-out;overflow-y:auto;position:fixed;z-index:10000}.sf-menu-content::-webkit-scrollbar{width:6px;background-color:transparent}.sf-menu-content::-webkit-scrollbar-thumb{border-radius:10px;background-color:var(--clr-gray-400)}.sf-menu-content.collapsed{transform:translateX(-100%)}@media screen and (min-width:768px){.sf-menu-content{display:flex;flex-direction:column;justify-content:space-between;border-top:5px solid #00a2fe;position:relative;height:100%;max-height:100%;width:250px;overflow:hidden}.sf-menu-content.collapsed{transform:translateX(0);width:75px}}@media screen and (max-width:380px){.sf-menu-content{max-height:calc(100% - 48px)}}.sf-menu-backdrop{position:fixed;width:100%;height:100%;background-color:rgba(0,0,0,.71);z-index:9999}.application-content{grid-area:content;overflow:auto;height:100%;position:relative}.sf-menu-logo-bg{background:50%/contain no-repeat}.fm-menu-scrollable-content{display:flex;flex-direction:column;width:100%;height:auto;align-items:flex-start;justify-content:flex-start;overflow:hidden;position:relative;flex:1 0 120px;padding:0}.fm-menu-scrollable-content:before{content:\" \";width:100%;height:1px;background-color:#fff;opacity:.15}.fm-menu-scrollable-content ul{list-style:none;padding-left:0;width:100%}.fm-menu-scrollable-content :before{display:none}.fm-menu-scrollable-content .scroll{overflow-y:auto;overflow-x:hidden;width:100%}.fm-menu-scrollable-content .scroll::-webkit-scrollbar{width:6px;background-color:transparent}.fm-menu-scrollable-content .scroll::-webkit-scrollbar-thumb{border-radius:10px;background-color:#5b616a}@media screen and (min-width:768px){.fm-menu-scrollable-content .scroll{max-height:65vh}}"]
                },] }
    ];
    SfMenuComponent.ctorParameters = function () { return [
        { type: SfMenuService }
    ]; };
    SfMenuComponent.propDecorators = {
        borderColor: [{ type: core.Input }],
        config: [{ type: core.Input, args: ['config',] }],
        tooltipDir: [{ type: core.ViewChild, args: [kendoAngularTooltip.TooltipDirective,] }]
    };

    var ClickOutsideDirective = /** @class */ (function () {
        function ClickOutsideDirective(elementRef) {
            var _this = this;
            this.elementRef = elementRef;
            this.clickOutside = new core.EventEmitter();
            this.ready = false;
            setTimeout(function () { return (_this.ready = true); }, 200);
        }
        ClickOutsideDirective.prototype.onClick = function (event, targetElement) {
            if (!this.ready || !this.elementRef || !targetElement) {
                return;
            }
            var clickedInside = this.elementRef.nativeElement.contains(targetElement);
            if (!clickedInside) {
                this.clickOutside.emit(event);
            }
        };
        return ClickOutsideDirective;
    }());
    ClickOutsideDirective.decorators = [
        { type: core.Directive, args: [{
                    selector: '[clickOutside]',
                },] }
    ];
    ClickOutsideDirective.ctorParameters = function () { return [
        { type: core.ElementRef }
    ]; };
    ClickOutsideDirective.propDecorators = {
        clickOutside: [{ type: core.Output }],
        onClick: [{ type: core.HostListener, args: ['document:click', ['$event', '$event.target'],] }]
    };

    var CustomerLogoComponent = /** @class */ (function () {
        function CustomerLogoComponent(svc) {
            this.svc = svc;
            this.collapsed = false;
        }
        Object.defineProperty(CustomerLogoComponent.prototype, "clr", {
            get: function () {
                return this.background;
            },
            enumerable: false,
            configurable: true
        });
        CustomerLogoComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.background = this.svc.colors.gray100;
            this.configSubs$ = this.svc.menuConfig$.subscribe(function (cfg) {
                _this.menuConfig = cfg;
            });
            this.collapsedSubs$ = this.svc.menuCollapse$.subscribe(function (value) {
                _this.collapsed = value;
            });
        };
        CustomerLogoComponent.prototype.ngOnDestroy = function () {
            this.configSubs$.unsubscribe();
            this.collapsedSubs$.unsubscribe();
        };
        return CustomerLogoComponent;
    }());
    CustomerLogoComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'sf-customer-logo',
                    template: "<div\n  data-testid=\"menu-logo-bg\"\n  class=\"sf-menu-logo-bg\"\n  *ngIf=\"menuConfig.logoUrl && !menuConfig.logoClass\"\n  [ngStyle]=\"{\n    'background-image': 'url(' + menuConfig.logoUrl + ')'\n  }\"\n></div>\n<em [ngClass]=\"menuConfig.logoClass\" *ngIf=\"menuConfig.logoClass && !menuConfig.logoUrl\"></em>\n",
                    styles: [":host{display:flex;align-items:center;justify-content:center;width:100%;flex-direction:column;height:auto;padding:5px;position:relative}:host .sf-menu-logo-bg{background:50%/contain no-repeat;height:2.0625rem;width:4.8125rem}:host.collapsed .sf-menu-logo-bg{width:80%}"]
                },] }
    ];
    CustomerLogoComponent.ctorParameters = function () { return [
        { type: SfMenuService }
    ]; };
    CustomerLogoComponent.propDecorators = {
        clr: [{ type: core.HostBinding, args: ['style.background-color',] }]
    };

    var MenuFooterLogoComponent = /** @class */ (function () {
        function MenuFooterLogoComponent(svc) {
            this.svc = svc;
            this.collapsed = false;
        }
        MenuFooterLogoComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.color = this.svc.colors.gray900;
            this.text = this.svc.colors.gray800;
            this.configSubs$ = this.svc.menuConfig$.subscribe(function (cfg) {
                _this.menuConfig = cfg;
            });
            this.collapsedSubs$ = this.svc.menuCollapse$.subscribe(function (value) {
                _this.collapsed = value;
            });
        };
        MenuFooterLogoComponent.prototype.ngOnDestroy = function () {
            this.configSubs$.unsubscribe();
            this.collapsedSubs$.unsubscribe();
        };
        return MenuFooterLogoComponent;
    }());
    MenuFooterLogoComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'sf-menu-footer-logo',
                    template: "<svg class=\"sf-menu-logo\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 1182 263\" [attr.fill]=\"color\" *ngIf=\"!collapsed\" [@appear-fast]>\n  <path\n    d=\"M55.2,145.2 C74.8,147.2 94.4,150.1 94.4,162.9 C94.4,179.3 70.4,183.8 49.4,183.8 C22,183.8 2.1,178 0.4,160 L11.7,160 C13.7,174 28.8,177.4 49.5,177.4 C65.7,177.4 82.5,174.7 82.5,164.1 C82.5,153.9 62.8,152.7 43.2,150.7 C17.5,148.2 3.7,143.7 3.7,132.9 C3.7,118.3 25.4,113.5 46.3,113.5 C68.3,113.5 90.6,118.4 91.3,134.3 L80.1,134.3 C78.6,122.9 63.6,119.9 46.2,119.9 C28.9,119.9 15.7,124.1 15.7,131.7 C15.7,140.4 28.9,142.5 55.2,145.2 Z M128.3,155.4 L176,155.4 L152.3,122.1 L128.3,155.4 Z M146,115.2 L159,115.2 L208,182.1 L195,182.1 L180.6,161.8 L123.8,161.8 L109.1,182.1 L97,182.1 L146,115.2 Z M217,182.1 L217,115.2 L297.9,115.2 L297.9,121.6 L228.2,121.6 L228.2,144.3 L294.4,144.3 L294.4,150.6 L228.2,150.6 L228.2,182.1 L217,182.1 Z M322.4,150.6 L322.4,175.7 L398.3,175.7 L398.3,182.1 L311.2,182.1 L311.2,115.2 L397.1,115.2 L397.1,121.6 L322.4,121.6 L322.4,144.2 L394.4,144.2 L394.4,150.6 L322.4,150.6 Z M707.1,182.1 L707.1,115.2 L788,115.2 L788,121.6 L718.3,121.6 L718.3,144.3 L784.5,144.3 L784.5,150.6 L718.3,150.6 L718.3,182.1 L707.1,182.1 Z M810.7,115.2 L810.7,175.7 L879.1,175.7 L879.1,182.1 L799.5,182.1 L799.5,115.2 L810.7,115.2 Z M902.1,150.6 L902.1,175.7 L978,175.7 L978,182.1 L890.9,182.1 L890.9,115.2 L976.8,115.2 L976.8,121.6 L902.1,121.6 L902.1,144.2 L974.1,144.2 L974.1,150.6 L902.1,150.6 Z M1005.7,150.6 L1005.7,175.7 L1081.5,175.7 L1081.5,182.1 L994.5,182.1 L994.5,115.2 L1080.3,115.2 L1080.3,121.6 L1005.7,121.6 L1005.7,144.2 L1077.6,144.2 L1077.6,150.6 L1005.7,150.6 Z M1085.9,115.2 L1181.8,115.2 L1181.8,121.6 L1139.4,121.6 L1139.4,182.1 L1128.2,182.1 L1128.2,121.6 L1085.9,121.6 L1085.9,115.2 Z M645.7,172.8 C668.1,124.1 664.4,42.1 664.3,38.6 L664.1,34.3 L659.8,34.3 C595.3,34.3 555.2,9.6 554.8,9.3 L552.4,7.8 L550,9.3 C549.7,9.6 509.7,34.3 445.2,34.3 L440.9,34.3 L440.7,38.6 C440.5,42 436.8,124.1 459.2,172.8 C476.4,210 508.1,237.8 550.9,253.4 L552.4,254 L553.9,253.4 C596.8,237.9 628.5,210 645.7,172.8 Z M433.1,28.5 C507.3,28.5 552.3,0 552.3,0 C552.3,0 597.5,28.5 671.6,28.5 C671.6,28.5 676.1,121.7 651.4,175.3 C634.1,212.6 602.2,244.2 552.2,262.3 C502.4,244.2 470.5,212.6 453.2,175.3 C428.5,121.7 433.1,28.5 433.1,28.5 Z M629.9,76.8 C630.1,70.1 630.1,64.5 630,60.5 L523.9,60.5 C486.9,60.5 479.1,77.1 479.1,91.1 C479.1,101.6 484.9,111 496.3,118.9 L523.5,137.5 C529.9,141.8 532.8,144.4 534.1,145.8 C536.2,147.9 537.2,150 537.2,151.8 C537.2,154.8 537.2,161.2 523.6,161.2 L487,161 C490.3,166.8 494.3,172.4 498.8,177.6 L520.3,177.6 C532.7,177.6 544.6,175.9 551.9,172.5 L551.9,211.1 C562.6,207.2 569.7,202.4 578,196.9 L578,122.7 L625.3,122.7 C626.4,117.1 627.2,111.4 627.9,105.8 L578,105.8 L578,81 C578,78.8 578.4,77.7 578.9,77.4 C578.9,77.3 580.2,76.8 585.1,76.8 L629.9,76.8 Z M552.3,18.4 C561.9,23.8 598.9,42.4 655.4,43.3 C655.8,60.8 656.2,128.5 637.5,169.1 C621.5,203.6 592.1,229.7 552.3,244.5 C512.6,229.6 483.2,203.6 467.3,169.1 C448.6,128.5 449,60.8 449.4,43.3 C505.9,42.4 542.8,23.8 552.3,18.4 Z M507.6,86.1 C507.6,83.1 507.6,76.8 521.2,76.8 L551.9,76.8 L551.9,120 C551.4,119.7 551,119.4 550.5,119.1 L521.1,100.9 C516,97.4 512.5,94.5 510.5,92.4 C508.5,90.3 507.6,88.2 507.6,86.1 Z\"\n  />\n</svg>\n\n<svg class=\"sf-menu-logo\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 256 256\" [attr.fill]=\"color\" *ngIf=\"collapsed\" [@appear-fast]>\n  <path\n    d=\"M219.012551,168.99085 C240.789143,121.646207 237.192116,41.9283263 237.0949,38.5257339 L236.900466,34.345406 L232.720138,34.345406 C170.01522,34.345406 131.031232,10.332825 130.642364,10.0411742 L128.309158,8.58292032 L125.975952,10.0411742 C125.684301,10.332825 86.7975301,34.345406 24.0926121,34.345406 L19.9122842,34.345406 L19.7178504,38.5257339 C19.5234165,41.8311094 15.9263902,121.646207 37.7029819,168.99085 C54.4242934,205.155547 85.2420593,232.181853 126.850904,247.347693 L128.309158,247.930995 L129.767412,247.347693 C171.473474,232.27907 202.29124,205.155547 219.012551,168.99085 Z M128.211941,1 C128.211941,1 172.153992,28.7068242 244.191735,28.7068242 C244.191735,28.7068242 248.566497,119.313 224.553916,171.421273 C207.735388,207.683187 176.723188,238.403736 128.114724,256 C79.7006944,238.403736 48.6884947,207.683187 31.8699663,171.421273 C7.85738524,119.313 12.3293639,28.7068242 12.3293639,28.7068242 C84.4643239,28.7068242 128.211941,1 128.211941,1 Z M203.652277,75.6626001 C203.84671,69.149066 203.84671,63.704918 203.749494,59.8162409 L100.602334,59.8162409 C64.6320707,59.8162409 57.0491504,75.9542509 57.0491504,89.5646207 C57.0491504,99.772398 62.6877322,108.910789 73.7704619,116.590926 L100.213466,134.673275 C106.435349,138.853603 109.25464,141.381243 110.51846,142.74228 C112.560016,144.783835 113.532185,146.825391 113.532185,148.575295 C113.532185,151.491803 113.532185,157.713687 100.310683,157.713687 L64.7292876,157.519253 C67.9374462,163.157835 71.8261233,168.601982 76.200885,173.657263 L97.1025244,173.657263 C109.157423,173.657263 120.726238,172.004575 127.823073,168.699199 L127.823073,206.224933 C138.225285,202.433473 145.127686,197.767061 153.196691,192.42013 L153.196691,120.28517 L199.180298,120.28517 C200.249684,114.841022 201.02742,109.299657 201.707938,103.855509 L153.196691,103.855509 L153.196691,79.745711 C153.196691,77.6069386 153.585559,76.5375524 154.071644,76.2459016 C154.071644,76.1486847 155.335464,75.6626001 160.099093,75.6626001 L203.652277,75.6626001 Z M128.211941,18.8879146 C137.544766,24.1376287 173.515029,42.2199771 228.442593,43.0949295 C228.831461,60.1078917 229.220328,125.923751 211.040763,165.393824 C195.486055,198.933664 166.904278,224.307282 128.211941,238.695387 C89.616821,224.210065 61.0350444,198.933664 45.577553,165.393824 C27.3979876,125.923751 27.7868553,60.1078917 28.175723,43.0949295 C83.1032869,42.2199771 118.976333,24.1376287 128.211941,18.8879146 Z M84.7559746,84.7037743 C84.7559746,81.7872665 84.7559746,75.6626001 97.9774767,75.6626001 L127.823073,75.6626001 L127.823073,117.660313 C127.336989,117.368662 126.948121,117.077011 126.462036,116.78536 L97.8802598,99.0918795 C92.9221965,95.6892871 89.5196041,92.8699962 87.5752655,90.8284407 C85.630927,88.7868852 84.7559746,86.7453298 84.7559746,84.7037743 Z\"\n  />\n</svg>\n\n<div class=\"sf-version-text\" *ngIf=\"!collapsed && menuConfig.versionText\" [style.color]=\"text\">\n  {{ menuConfig.versionText }}\n</div>\n",
                    animations: [
                        animations.trigger('appear-fast', [animations.transition(':enter', [animations.style({ opacity: 0 }), animations.animate('80ms 120ms ease-in', animations.style({ opacity: 1 }))])])
                    ],
                    styles: [":host{display:flex;flex-direction:column;justify-content:center;align-items:center;padding:15px 5px}:host .sf-menu-logo{-o-object-fit:contain;object-fit:contain;text-align:center;align-self:center;justify-self:center;height:auto;width:50%;max-width:90%;height:2em}:host .sf-menu-logo-bg{width:100%;height:4em}@media screen and (min-width:768px){:host .sf-menu-logo-bg{width:65%}}:host .sf-version-text{font-size:.7em;padding:10px 10px 0;text-align:center}@media screen and (min-width:768px){:host.collapsed .sf-menu-logo{max-width:100%;width:100%}}"]
                },] }
    ];
    MenuFooterLogoComponent.ctorParameters = function () { return [
        { type: SfMenuService }
    ]; };

    var MenuItemComponent = /** @class */ (function () {
        function MenuItemComponent(svc) {
            this.svc = svc;
            this.collapsed = false;
            this.level = 1;
        }
        MenuItemComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.text = this.svc.colors.gray900;
            this.configSubs$ = this.svc.menuConfig$.subscribe(function (cfg) {
                _this.menuConfig = cfg;
            });
            this.collapsedSubs$ = this.svc.menuCollapse$.subscribe(function (value) {
                _this.collapsed = value;
            });
        };
        MenuItemComponent.prototype.ngOnDestroy = function () {
            this.configSubs$.unsubscribe();
            this.collapsedSubs$.unsubscribe();
        };
        MenuItemComponent.prototype.click = function (evt) {
            switch (this.itemType) {
                case exports.MenuEventType.ShortcutClick:
                    this.shortcutClicked();
                    break;
                case exports.MenuEventType.UserActionClick:
                    this.userOptionClicked();
                    break;
                case exports.MenuEventType.RouteClick:
                    this.svc.routeClicked(this.item, evt);
                    break;
                default:
                    console.error('event not found');
            }
        };
        MenuItemComponent.prototype.shortcutClicked = function () {
            this.svc.events$.next({
                type: exports.MenuEventType.ShortcutClick,
                value: this.item
            });
            if (window.innerWidth <= this.svc.SMALL_RESOLUTION && !this.collapsed) {
                this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
            }
        };
        MenuItemComponent.prototype.userOptionClicked = function () {
            this.svc.events$.next({
                type: exports.MenuEventType.UserActionClick,
                value: this.item
            });
            if (window.innerWidth <= this.svc.SMALL_RESOLUTION && !this.collapsed) {
                this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
            }
        };
        return MenuItemComponent;
    }());
    MenuItemComponent.decorators = [
        { type: core.Component, args: [{
                    // tslint:disable-next-line: component-selector
                    selector: '[sf-menu-item]',
                    template: "<div\n  class=\"li-container\"\n  [class.show-arrow]=\"item.items\"\n  [class.active-item]=\"item.active\"\n  [class.collapsed]=\"collapsed\"\n  [style.color]=\"text\"\n>\n  <div\n    [class.expanded-item]=\"item.expanded && item.items\"\n    class=\"main-div menu-item\"\n    [ngStyle]=\"{ 'border-left-color': menuConfig?.borderColor, 'padding-left': 10 * level + 'px' }\"\n    title=\"{{ menuConfig.tooltips || collapsed ? item.text : '' }}\"\n    (click)=\"click($event)\"\n  >\n    <div class=\"icon-container\" *ngIf=\"item.iconClass || item.iconPath\">\n      <div class=\"sf-icons-menu\" *ngIf=\"item.iconClass\" [class.collapsed]=\"collapsed\">\n        <em [ngClass]=\"item.iconClass\"></em>\n      </div>\n      <div\n        class=\"sf-menu-logo-bg\"\n        *ngIf=\"!item.iconClass && item.iconPath\"\n        [ngStyle]=\"{\n          'background-image': 'url(' + item.iconPath + ')'\n        }\"\n      ></div>\n    </div>\n    <span>{{ item.text }}</span>\n    <em class=\"fas fa-chevron-down\"></em>\n  </div>\n  <ul *ngIf=\"item.items\" class=\"item-children\">\n    <li\n      sf-menu-item\n      *ngFor=\"let child of item.items\"\n      [class.show-arrow]=\"child.items\"\n      [class.active-item]=\"child.active\"\n      [item]=\"child\"\n      [level]=\"level + 1\"\n      [itemType]=\"itemType\"\n    ></li>\n  </ul>\n</div>\n",
                    styles: [":host .li-container{display:inline-block;width:100%;padding:0;overflow:hidden;cursor:pointer}:host .li-container .icon-container{width:24px}:host .li-container>div.main-div{margin-left:2px}:host .li-container>.menu-item{display:flex;align-items:center;justify-content:flex-start;padding:7px 10px;position:relative;min-height:38px}:host .li-container>.menu-item .sf-menu-logo-bg{width:1.6em;height:1.55em}:host .li-container>.menu-item span{width:calc(100% - 24px);min-width:100px;padding:0 10px;font-size:14px;line-height:18px;opacity:.71;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}:host .li-container>.menu-item:hover{position:relative}:host .li-container>.menu-item:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}:host .li-container>.menu-item:hover span{opacity:1}:host .li-container>.menu-item .fa-chevron-down{display:none;position:absolute;right:15px;top:16px;width:auto;height:auto;font-size:12px}:host .li-container.show-arrow>div .fa-chevron-down{display:inline-block;transform:rotate(0deg)}:host .li-container.show-arrow>div.expanded-item span{opacity:1}:host .li-container.show-arrow>div.expanded-item em.fa-chevron-down,:host .li-container.show-arrow>div.expanded-item i.fa-chevron-down{transform:rotate(180deg)}:host .li-container.show-arrow>div:hover{position:relative}:host .li-container.show-arrow>div:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}:host .li-container.show-arrow>div:hover span{opacity:1}:host .li-container.active-item>div.main-div{border-left:3px solid;margin-left:0}:host .li-container.active-item>div.main-div:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.2}:host .li-container.active-item>div.main-div span{opacity:1}:host .li-container.collapsed div{justify-content:center}:host .li-container.collapsed em,:host .li-container.collapsed i{width:auto}:host .li-container.collapsed .item-children,:host .li-container.collapsed span{display:none}:host .sf-icons-menu{text-align:center}:host .sf-icons-menu em,:host .sf-icons-menu i{font-size:16px}:host ul{list-style:none;padding-left:0;width:100%}.expanded-item+.item-children{display:block}ul.item-children{display:none}.sf-icons-menu.collapsed em{font-size:var(--fs-300)}.sf-menu-logo-bg{background:50%/contain no-repeat}"]
                },] }
    ];
    MenuItemComponent.ctorParameters = function () { return [
        { type: SfMenuService }
    ]; };
    MenuItemComponent.propDecorators = {
        item: [{ type: core.Input }],
        level: [{ type: core.Input }],
        itemType: [{ type: core.Input }]
    };

    var MenuProductsComponent = /** @class */ (function () {
        function MenuProductsComponent(svc) {
            this.svc = svc;
            this.collapsed = false;
            this.productsExpanded = false;
        }
        Object.defineProperty(MenuProductsComponent.prototype, "clr", {
            get: function () {
                return this.border;
            },
            enumerable: false,
            configurable: true
        });
        MenuProductsComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.border = this.svc.colors.gray800;
            this.chevron = this.svc.colors.gray900;
            this.text = this.svc.colors.gray800;
            this.productsBg = this.svc.colors.gray100;
            this.configSubs$ = this.svc.menuConfig$.subscribe(function (cfg) {
                _this.menuConfig = cfg;
            });
            this.collapsedSubs$ = this.svc.menuCollapse$.subscribe(function (value) {
                _this.collapsed = value;
            });
        };
        MenuProductsComponent.prototype.ngOnDestroy = function () {
            this.configSubs$.unsubscribe();
            this.collapsedSubs$.unsubscribe();
        };
        MenuProductsComponent.prototype.currentProductClicked = function () {
            this.productsExpanded = !this.productsExpanded;
        };
        MenuProductsComponent.prototype.productClicked = function (item, ev) {
            this.svc.events$.next({
                type: exports.MenuEventType.ProductClick,
                value: item
            });
            if (window.innerWidth <= this.svc.SMALL_RESOLUTION && !this.collapsed) {
                this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
            }
            window.location.href = item.routeHref;
        };
        return MenuProductsComponent;
    }());
    MenuProductsComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'sf-menu-products',
                    template: "<div class=\"sf-menu-product\" *ngIf=\"menuConfig.currentProduct\" (click)=\"currentProductClicked()\" [class.collapsed]=\"collapsed\">\n  <em [ngClass]=\"menuConfig.currentProduct.iconClass\" *ngIf=\"menuConfig.currentProduct.iconClass\"></em>\n  <div\n    class=\"sf-menu-logo-bg\"\n    *ngIf=\"!menuConfig.currentProduct.iconClass && menuConfig.currentProduct.iconPath\"\n    [ngStyle]=\"{\n      'background-image': 'url(' + menuConfig.currentProduct.iconPath + ')'\n    }\"\n  ></div>\n  <em\n    class=\"fas fa-chevron-down\"\n    *ngIf=\"menuConfig.products && menuConfig.products.length > 0\"\n    [style.color]=\"chevron\"\n    [class.expanded]=\"productsExpanded\"\n    [class.collapsed]=\"collapsed\"\n    data-testid=\"menu-product-display-products-icon\"\n  ></em>\n</div>\n\n<div class=\"sf-customer-title\" *ngIf=\"!menuConfig.hideProductName && !collapsed\" [style.color]=\"text\">\n  <h4 *ngIf=\"menuConfig.currentProduct?.text\">\n    {{ menuConfig.currentProduct.text }}\n  </h4>\n</div>\n\n<div\n  class=\"sf-menu-products-list\"\n  [style.background-color]=\"productsBg\"\n  [class.expanded]=\"productsExpanded\"\n  *ngIf=\"menuConfig.products && menuConfig.products.length > 0\"\n>\n  <a\n    class=\"option\"\n    [class.collapsed]=\"collapsed\"\n    [class.two-columns]=\"menuConfig.products.length > 3\"\n    *ngFor=\"let option of menuConfig.products\"\n    (click)=\"productClicked(option, $event)\"\n    [attr.href]=\"option.routeHref\"\n    data-testid=\"menu-product-option-link\"\n  >\n    <em [ngClass]=\"option.iconClass\" *ngIf=\"option.iconClass\"></em>\n    <div\n      class=\"sf-menu-logo-bg\"\n      *ngIf=\"!option.iconClass && option.iconPath\"\n      [ngStyle]=\"{ 'background-image': 'url(' + option.iconPath + ')' }\"\n    ></div>\n  </a>\n</div>\n",
                    styles: [":host{position:relative;width:calc(100% - 2px);height:auto;display:flex;flex-direction:column;flex-wrap:wrap;align-items:center;padding:10px 0 0}.sf-menu-product{width:76px;height:76px;flex:0 0 76px;padding:5px;position:relative;cursor:pointer;margin-bottom:10px}.sf-menu-product em,.sf-menu-product i,.sf-menu-product img{-o-object-fit:contain;object-fit:contain;text-align:center;align-self:center;justify-self:center;height:auto;width:50%;width:100%;height:100%;background-size:cover}.sf-menu-product .sf-menu-logo-bg{height:100%;width:100%}.sf-menu-product .fa-chevron-down{position:absolute;width:auto;height:auto;top:calc(50% - 7px);right:-20px;transition:transform .15s ease-out;font-size:13px}.sf-menu-product .fa-chevron-down.expanded{transform:rotate(180deg)}.sf-menu-product .fa-chevron-down.collapsed{right:calc(50% - 5px);top:calc(100% - 2px);font-size:10px}.sf-menu-product.collapsed{flex:0 0 60px;margin:20px 0 10px;width:60px;height:60px}.sf-menu-products-list{height:auto;max-height:0;width:100%;overflow:hidden;display:flex;flex-wrap:wrap;justify-content:space-around;transition:max-height .24s ease-in-out;opacity:.7}.sf-menu-products-list.expanded{max-height:304px}.sf-menu-products-list .option{width:62px;height:62px;padding:5px;cursor:pointer}.sf-menu-products-list .option:hover .sf-menu-logo-bg{filter:brightness(1.2)}.sf-menu-products-list .option.collapsed{flex:0 0 50px;height:55.8px}.sf-menu-products-list .option.two-columns{width:80.6px}.sf-menu-products-list .option.two-columns:before{content:\" \";height:1px;width:86px;position:absolute;bottom:-1px;left:14px}.sf-menu-products-list .option em,.sf-menu-products-list .option i,.sf-menu-products-list .option img{-o-object-fit:contain;object-fit:contain;text-align:center;align-self:center;justify-self:center;height:auto;width:50%;width:100%;height:100%;transition:background-color .12s,transform}.sf-menu-products-list .option .sf-menu-logo-bg{width:100%;height:100%}.sf-menu-products-list .option:hover .logo-bg{background-color:#3a3b3e;transform:scale(1.15)}.sf-menu-products-list .option:after{content:\" \";background-color:#35353b;height:46px;width:1px;position:absolute;top:15px;right:-1px}.sf-customer-title{margin:10px 0}.sf-customer-title h4{margin:0;font-size:1em;opacity:.6;text-align:center}.sf-menu-logo-bg{background:50%/contain no-repeat}"]
                },] }
    ];
    MenuProductsComponent.ctorParameters = function () { return [
        { type: SfMenuService }
    ]; };
    MenuProductsComponent.propDecorators = {
        clr: [{ type: core.HostBinding, args: ['style.border-bottom-color',] }]
    };

    var MenuRoutesComponent = /** @class */ (function () {
        function MenuRoutesComponent(svc) {
            this.svc = svc;
            this.collapsed = false;
        }
        MenuRoutesComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.configSubs$ = this.svc.menuConfig$.subscribe(function (cfg) {
                _this.menuConfig = cfg;
            });
            this.collapsedSubs$ = this.svc.menuCollapse$.subscribe(function (value) {
                _this.collapsed = value;
            });
        };
        MenuRoutesComponent.prototype.ngOnDestroy = function () {
            this.configSubs$.unsubscribe();
            this.collapsedSubs$.unsubscribe();
        };
        MenuRoutesComponent.prototype.routeClicked = function (item, evt) {
            this.svc.routeClicked(item, evt);
        };
        MenuRoutesComponent.prototype.showText = function (index) {
            return index.indexOf('_') > -1;
        };
        return MenuRoutesComponent;
    }());
    MenuRoutesComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'sf-menu-routes',
                    template: "<ul *ngIf=\"!collapsed\" [@appear-fast]>\n  <li\n    sf-menu-item\n    *ngFor=\"let item of menuConfig.routes\"\n    [class.show-arrow]=\"item.items\"\n    [class.active-item]=\"item.active\"\n    [item]=\"item\"\n    [level]=\"1\"\n    [itemType]=\"2\"\n    data-testid=\"menu-route-item\"\n  ></li>\n</ul>\n<kendo-menu *ngIf=\"menuConfig?.routes && collapsed\" [items]=\"menuConfig.routes\" [vertical]=\"true\">\n  <ng-template kendoMenuItemLinkTemplate let-item=\"item\" let-index=\"index\">\n    <div\n      class=\"sf-menu-item-wrapper\"\n      (click)=\"routeClicked(item, $event)\"\n      [class.active-item]=\"item.active\"\n      [ngStyle]=\"{ 'border-left-color': menuConfig?.borderColor }\"\n      title=\"{{ item.items ? undefined : item.text }}\"\n      [@appear-fast]\n    >\n      <div class=\"icon-container\" [class.collapsed]=\"collapsed\">\n        <em [ngClass]=\"item.iconClass\" *ngIf=\"item.iconClass\"></em>\n        <div\n          class=\"sf-menu-logo-bg\"\n          *ngIf=\"!item.iconClass && item.iconPath\"\n          [ngStyle]=\"{\n            'background-image': 'url(' + item.iconPath + ')'\n          }\"\n        ></div>\n      </div>\n      <span *ngIf=\"showText(index)\">{{ item.text }}</span>\n      <em class=\"fas fa-chevron-right\" *ngIf=\"item.items\"></em>\n    </div>\n  </ng-template>\n</kendo-menu>\n",
                    animations: [
                        animations.trigger('appear-fast', [animations.transition(':enter', [animations.style({ opacity: 0 }), animations.animate('80ms 120ms ease-in', animations.style({ opacity: 1 }))])]),
                    ],
                    styles: [":host{display:flex;flex-direction:column;width:100%;height:auto;align-items:flex-start;justify-content:flex-start;overflow:hidden;position:relative}:host:before{content:\" \";width:100%;height:1px;background-color:#fff;opacity:.15}:host ul{list-style:none;padding-left:0}:host ul,kendo-menu{width:100%}.sf-menu-item-wrapper{width:100%;height:44px;max-height:75px;overflow:hidden;background-color:transparent;display:flex;justify-content:center;align-items:center}.sf-menu-item-wrapper img{width:2em;height:auto;-o-object-fit:contain;object-fit:contain}.sf-menu-item-wrapper .sf-menu-logo-bg{background:50%/contain no-repeat;width:1.6em;height:1.55em}.sf-menu-item-wrapper em,.sf-menu-item-wrapper i{-o-object-fit:contain;object-fit:contain;color:#d5d7db}.sf-menu-item-wrapper em.fa-chevron-right,.sf-menu-item-wrapper i.fa-chevron-right{font-size:12px;position:absolute;right:8px}.sf-menu-item-wrapper:hover{position:relative}.sf-menu-item-wrapper:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}.sf-menu-item-wrapper:hover span{opacity:1}.sf-menu-item-wrapper.active-item{position:relative;border-left:2px solid #3298aa}.sf-menu-item-wrapper.active-item:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}.sf-menu-item-wrapper.active-item span{opacity:1}.sf-menu-item-wrapper.active-item em,.sf-menu-item-wrapper.active-item i{color:#fff}.sf-menu-item-wrapper .icon-container.collapsed em{font-size:var(--fs-300)}::ng-deep .k-menu-popup .sf-menu-item-wrapper{padding:0 15px;justify-content:flex-start;background-color:#2f3037;width:188px}::ng-deep .k-menu-popup .sf-menu-item-wrapper .icon-container{width:24px;font-size:16px}::ng-deep .k-menu-popup .sf-menu-item-wrapper span{width:calc(100% - 24px);min-width:100px;color:#fefefe;padding:0 10px;font-size:14px;line-height:18px;opacity:.6;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}::ng-deep .k-menu-popup .sf-menu-item-wrapper:hover{background-color:#000;color:#fefefe}::ng-deep .k-menu-popup .sf-menu-item-wrapper.active-item{position:relative;background-color:#464853;border-left:2px solid #3298aa}::ng-deep .k-menu-popup .sf-menu-item-wrapper.active-item:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}::ng-deep .k-menu-popup .sf-menu-item-wrapper.active-item span{opacity:1}::ng-deep .k-menu-popup .sf-menu-item-wrapper.active-item em,::ng-deep .k-menu-popup .sf-menu-item-wrapper.active-item i{color:#fff}"]
                },] }
    ];
    MenuRoutesComponent.ctorParameters = function () { return [
        { type: SfMenuService }
    ]; };

    var MenuShortcutsComponent = /** @class */ (function () {
        function MenuShortcutsComponent(svc) {
            this.svc = svc;
            this.collapsed = false;
        }
        MenuShortcutsComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.configSubs$ = this.svc.menuConfig$.subscribe(function (cfg) {
                _this.menuConfig = cfg;
            });
            this.collapsedSubs$ = this.svc.menuCollapse$.subscribe(function (value) {
                _this.collapsed = value;
            });
        };
        MenuShortcutsComponent.prototype.ngOnDestroy = function () {
            this.configSubs$.unsubscribe();
            this.collapsedSubs$.unsubscribe();
        };
        return MenuShortcutsComponent;
    }());
    MenuShortcutsComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'sf-menu-shortcuts',
                    template: "<ul>\n  <li\n    sf-menu-item\n    *ngFor=\"let item of menuConfig.shortcuts\"\n    [item]=\"item\"\n    [level]=\"1\"\n    [itemType]=\"1\"\n  >\n  </li>\n</ul>\n",
                    animations: [
                        animations.trigger('appear-fast', [animations.transition(':enter', [animations.style({ opacity: 0 }), animations.animate('80ms 120ms ease-in', animations.style({ opacity: 1 }))])]),
                    ],
                    styles: [":host{display:flex;flex-direction:column;width:100%;height:auto;align-items:flex-start;justify-content:flex-start;overflow:hidden;position:relative}:host:before{content:\" \";width:100%;height:1px;background-color:#fff;opacity:.15}:host ul{list-style:none;padding-left:0;width:100%}"]
                },] }
    ];
    MenuShortcutsComponent.ctorParameters = function () { return [
        { type: SfMenuService }
    ]; };

    var MenuToggleBtnComponent = /** @class */ (function () {
        function MenuToggleBtnComponent(svc) {
            this.svc = svc;
            this.collapsed = false;
        }
        MenuToggleBtnComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.color = this.svc.colors.gray700;
            this.configSubs$ = this.svc.menuConfig$.subscribe(function (cfg) {
                _this.menuConfig = cfg;
            });
            this.collapsedSubs$ = this.svc.menuCollapse$.subscribe(function (value) {
                _this.collapsed = value;
            });
        };
        MenuToggleBtnComponent.prototype.ngOnDestroy = function () {
            this.configSubs$.unsubscribe();
            this.collapsedSubs$.unsubscribe();
        };
        MenuToggleBtnComponent.prototype.toggleMenu = function () {
            this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
        };
        return MenuToggleBtnComponent;
    }());
    MenuToggleBtnComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'sf-menu-toggle-btn',
                    template: "<button class=\"sf-menu-toggle-btn\" [class.collapsed]=\"collapsed\" (click)=\"toggleMenu()\" [style.color]=\"color\">\n  <div>\n    <em class=\"fas fa-chevron-left\"></em>\n    <em class=\"fas fa-chevron-left\"></em>\n  </div>\n</button>\n",
                    styles: [".sf-menu-toggle-btn{display:none;position:absolute;z-index:10000;top:8px;right:8px;background-color:transparent;border:none;width:1.8rem;height:1.8rem;transition:transform .15s ease-in-out;cursor:pointer}@media screen and (min-width:768px){.sf-menu-toggle-btn{display:block}}.sf-menu-toggle-btn div{display:flex;justify-content:center;align-items:center}.sf-menu-toggle-btn .fas{font-size:16px;opacity:.64}.sf-menu-toggle-btn .fas:nth-child(2){margin-left:-4px}.sf-menu-toggle-btn.collapsed .fas{transform:rotate(180deg)}.sf-menu-toggle-btn:hover .fas{opacity:1}"]
                },] }
    ];
    MenuToggleBtnComponent.ctorParameters = function () { return [
        { type: SfMenuService }
    ]; };

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from) {
        for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
            to[j] = from[i];
        return to;
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var MenuUserComponent = /** @class */ (function () {
        function MenuUserComponent(svc) {
            this.svc = svc;
            this.tenants = [];
            this.collapsed = false;
            this.multipleTenantMenu = false;
            this.anchorAlign = { horizontal: 'right', vertical: 'top' };
            this.popupAlign = { horizontal: 'left', vertical: 'top' };
            this.userName = 'Username';
            this.darkMode = false;
        }
        MenuUserComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.color = this.svc.colors.gray900;
            this.tenantBg = this.svc.colors.gray200;
            this.configSubs$ = this.svc.menuConfig$.subscribe(function (cfg) {
                var e_1, _b;
                var _a;
                _this.menuConfig = cfg;
                _this.userActions = {
                    userName: { id: 'username', text: _this.menuConfig.userName, iconClass: 'fas fa-user-circle' },
                    tenant: { id: 'tenant', text: _this.menuConfig.fleetName, iconClass: 'fas fa-building' },
                    help: { id: 'help', text: _this.menuConfig.helpText, iconClass: 'fas fa-question-circle' },
                    logout: { id: 'logout', text: _this.menuConfig.logoutText, iconClass: 'fas fa-sign-out-alt' }
                };
                _this.tenants.length = 0;
                if (((_a = cfg.multipleTenants) === null || _a === void 0 ? void 0 : _a.length) > 1) {
                    try {
                        for (var _c = __values(cfg.multipleTenants.sort(function (a, b) { return a.localeCompare(b); })), _d = _c.next(); !_d.done; _d = _c.next()) {
                            var t = _d.value;
                            _this.tenants.push({ id: t, active: t === cfg.fleetName });
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                }
            });
            this.collapsedSubs$ = this.svc.menuCollapse$.subscribe(function (value) {
                _this.collapsed = value;
            });
            this.svc.currentTheme.subscribe(function (theme) { return (_this.darkMode = theme === 'dark'); });
        };
        MenuUserComponent.prototype.ngOnDestroy = function () {
            this.configSubs$.unsubscribe();
            this.collapsedSubs$.unsubscribe();
        };
        Object.defineProperty(MenuUserComponent.prototype, "disableUserNameClick", {
            get: function () {
                return this.menuConfig.disableUserNameClick && !this.collapsed;
            },
            enumerable: false,
            configurable: true
        });
        MenuUserComponent.prototype.changeTenant = function (tenant) {
            this.svc.events$.next({
                type: exports.MenuEventType.ChangeTenantClick,
                value: { text: tenant.id }
            });
            this.multipleTenantMenu = false;
        };
        MenuUserComponent.prototype.showMultipleTenantMenu = function () {
            if (this.collapsed) {
                this.svc.menuCollapse$.next(false);
            }
            this.svc.events$.next({
                type: exports.MenuEventType.OpenTenantListClick,
                value: null
            });
            this.multipleTenantMenu = true;
        };
        MenuUserComponent.prototype.hideMultipleTenantMenu = function () {
            this.multipleTenantMenu = false;
        };
        MenuUserComponent.prototype.userOptionClicked = function (item) {
            if (!this.disableUserNameClick) {
                if (this.collapsed) {
                    this.svc.menuCollapse$.next(false);
                    return;
                }
                this.svc.events$.next({
                    type: exports.MenuEventType.UserActionClick,
                    value: item
                });
                if (window.innerWidth <= this.svc.SMALL_RESOLUTION && !this.collapsed) {
                    this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
                }
            }
        };
        MenuUserComponent.prototype.setDarkMode = function () {
            console.log('set dark mode');
            this.svc.setTheme(this.darkMode ? 'dark' : 'light');
        };
        MenuUserComponent.prototype.themeIconClicked = function () {
            console.log('icon clicked');
            if (this.collapsed) {
                this.svc.menuCollapse$.next(false);
            }
        };
        MenuUserComponent.prototype.onSsoManageAccount = function (event, config) {
            event.stopPropagation();
            if (!config.disable) {
                window.open(config.url);
            }
        };
        MenuUserComponent.prototype.sortMultipleTenant = function (tenants) {
            if (tenants) {
                return tenants.sort(function (a, b) { return a.localeCompare(b); });
            }
            else {
                return [];
            }
        };
        return MenuUserComponent;
    }());
    MenuUserComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'sf-menu-user',
                    template: "<ul class=\"ul-container\" [style.color]=\"color\">\n  <li>\n    <div\n      class=\"li-container user-li\"\n      (click)=\"userOptionClicked(userActions.userName)\"\n      title=\"{{ menuConfig.tooltips || collapsed ? menuConfig.userName : '' }}\"\n      [class.collapsed]=\"collapsed\"\n    >\n      <div\n        class=\"main-div menu-item\"\n        [class.disable-user-click]=\"disableUserNameClick\"\n        [class.sso-manage-account]=\"menuConfig.ssoManageAccount && menuConfig.ssoManageAccount.url\"\n      >\n        <div class=\"icon-container\">\n          <div class=\"sf-icons-menu\">\n            <em class=\"fas fa-user-circle\"></em>\n          </div>\n        </div>\n        <div *ngIf=\"menuConfig.userName && !collapsed\">\n          <span class=\"sf-user-text\" data-testid=\"menu-user-name\">{{ menuConfig.userName }} </span>\n          <span class=\"sf-user-fleet-name\" (click)=\"showMultipleTenantMenu()\" data-testid=\"menu-user-fleet-name-single\">\n            {{ menuConfig?.multipleTenants?.length >= 2 ? '' : menuConfig.fleetName }}\n          </span>\n        </div>\n      </div>\n      <div\n\n        class=\"sso-settings\"\n        *ngIf=\"menuConfig.ssoManageAccount?.url && !collapsed\"\n        (click)=\"onSsoManageAccount($event, menuConfig.ssoManageAccount)\"\n      >\n        <div kendoTooltip showOn=\"{{ !menuConfig.tooltips ? 'hover' : 'none' }}\"\n          class=\"sf-icons-menu\" [class.disable]=\"menuConfig.ssoManageAccount.disable\"\n          title=\"{{ menuConfig.ssoManageAccount.title }}\">\n          <em class=\"icon-manage-account\"\n            data-testid=\"menu-products-sso-settings-icon\"></em>\n        </div>\n      </div>\n    </div>\n  </li>\n\n  <li *ngIf=\"menuConfig?.multipleTenants?.length > 1\" #anchor>\n    <div class=\"li-container\" [class.show-arrow]=\"menuConfig?.multipleTenants\" [class.collapsed]=\"collapsed\">\n      <div\n        [class.expanded-item]=\"multipleTenantMenu\"\n        class=\"main-div menu-item sf-fleet-list\"\n        [ngStyle]=\"{ 'border-left-color': menuConfig?.borderColor }\"\n        title=\"{{ menuConfig.tooltips || collapsed ? menuConfig.fleetName : '' }}\"\n        (click)=\"showMultipleTenantMenu()\"\n      >\n        <div class=\"icon-container\">\n          <div class=\"sf-icons-menu\">\n            <em class=\"fas fa-building\"></em>\n          </div>\n        </div>\n        <span *ngIf=\"!collapsed\">\n          <div class=\"sf-fleet-title\" *ngIf=\"menuConfig.fleetTitle\">\n            {{ menuConfig.fleetTitle }}\n          </div>\n          <div>\n            {{ menuConfig.fleetName }}\n          </div>\n        </span>\n        <em *ngIf=\"!collapsed\" class=\"fas fa-chevron-right\"\n          data-testid=\"menu-user-show-tenant-list\"></em>\n      </div>\n    </div>\n\n    <kendo-popup\n      [anchor]=\"anchor\"\n      [anchorAlign]=\"anchorAlign\"\n      [popupAlign]=\"popupAlign\"\n      *ngIf=\"multipleTenantMenu\"\n      (clickOutside)=\"hideMultipleTenantMenu()\"\n    >\n      <div class=\"tenant-picker\" [style.color]=\"color\" [style.background-color]=\"tenantBg\">\n        <div class=\"tenant\" *ngFor=\"let tenant of tenants\" [class.active]=\"tenant.active\" (click)=\"changeTenant(tenant)\"\n          data-testid=\"menu-user-tenants-list\">\n          {{ tenant.id }}\n        </div>\n      </div>\n    </kendo-popup>\n  </li>\n\n  <li *ngIf=\"!menuConfig?.denyChangeTheme && menuConfig?.themeTitle\" (click)=\"themeIconClicked()\">\n    <div class=\"li-container\" [class.collapsed]=\"collapsed\">\n      <div class=\"main-div menu-item theme-selector-container\" title=\"{{ menuConfig.tooltips || collapsed ? menuConfig.themeTitle : '' }}\">\n        <div class=\"icon-container\">\n          <div class=\"sf-icons-menu\">\n            <em class=\"fas fa-moon\"></em>\n          </div>\n        </div>\n        <span *ngIf=\"!collapsed\">{{ menuConfig?.themeTitle }}</span>\n        <kendo-switch *ngIf=\"!collapsed\" (valueChange)=\"setDarkMode()\" [(ngModel)]=\"darkMode\"></kendo-switch>\n      </div>\n    </div>\n  </li>\n  <li\n    sf-menu-item\n    *ngIf=\"userActions.help.text\"\n    [class.show-arrow]=\"userActions.help.items\"\n    [class.active-item]=\"userActions.help.active\"\n    [item]=\"userActions.help\"\n    [itemType]=\"3\"\n    data-testid=\"menu-user-help\"\n  ></li>\n  <li\n    sf-menu-item\n    *ngIf=\"userActions.logout.text\"\n    [class.show-arrow]=\"userActions.logout.items\"\n    [class.active-item]=\"userActions.logout.active\"\n    [item]=\"userActions.logout\"\n    [itemType]=\"3\"\n    data-testid=\"menu-user-logout\"\n  ></li>\n</ul>\n",
                    styles: [":host{display:flex;flex-direction:column;width:100%;height:auto;align-items:flex-start;justify-content:flex-start;overflow:hidden;position:relative}:host:before{content:\" \";width:100%;height:1px;background-color:#fff;opacity:.15}:host ul{list-style:none;padding-left:0;width:100%}:host .li-container{display:inline-block;width:100%;padding:0;overflow:hidden;cursor:pointer}:host .li-container .icon-container{width:24px}:host .li-container>div.main-div{margin-left:2px}:host .li-container>.menu-item{display:flex;align-items:center;justify-content:flex-start;padding:7px 10px;position:relative;min-height:38px}:host .li-container>.menu-item .sf-menu-logo-bg{width:1.6em;height:1.55em}:host .li-container>.menu-item span{width:calc(100% - 24px);min-width:100px;padding:0 10px;font-size:14px;line-height:18px;opacity:.71;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}:host .li-container>.menu-item:hover{position:relative}:host .li-container>.menu-item:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}:host .li-container>.menu-item:hover span{opacity:1}:host .li-container>.menu-item .fa-chevron-down{display:none;position:absolute;right:15px;top:16px;width:auto;height:auto;font-size:12px}:host .li-container.show-arrow>div .fa-chevron-down{display:inline-block;transform:rotate(0deg)}:host .li-container.show-arrow>div.expanded-item span{opacity:1}:host .li-container.show-arrow>div.expanded-item em.fa-chevron-down,:host .li-container.show-arrow>div.expanded-item i.fa-chevron-down{transform:rotate(180deg)}:host .li-container.show-arrow>div:hover{position:relative}:host .li-container.show-arrow>div:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}:host .li-container.show-arrow>div:hover span{opacity:1}:host .li-container.active-item>div.main-div{border-left:3px solid;margin-left:0}:host .li-container.active-item>div.main-div:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.2}:host .li-container.active-item>div.main-div span{opacity:1}:host .li-container.collapsed div{justify-content:center}:host .li-container.collapsed em,:host .li-container.collapsed i{width:auto}:host .li-container.collapsed .item-children,:host .li-container.collapsed span{display:none}:host .sf-icons-menu{text-align:center}:host .sf-icons-menu em,:host .sf-icons-menu i{font-size:16px}.user-li{display:flex!important}.user-li .menu-item{flex:1}.user-li .menu-item.sso-manage-account{flex-basis:75%}.user-li .menu-item.disable-user-click{cursor:default}.user-li .menu-item.disable-user-click:hover{background-color:transparent}.user-li .menu-item.disable-user-click:hover .sf-user-fleet-name,.user-li .menu-item.disable-user-click:hover .sf-user-text{opacity:.7}.user-li .menu-item.disable-user-click:hover:before{display:none}.user-li .sso-settings{display:flex;justify-content:flex-end;flex:1;align-items:center;padding-right:15px;cursor:default}.user-li .sso-settings .sf-icons-menu{cursor:pointer}.user-li .sso-settings .sf-icons-menu.disable{opacity:.5;cursor:not-allowed}.user-li .sso-settings .sf-icons-menu:hover{position:relative}.user-li .sso-settings .sf-icons-menu:hover:before{content:\" \";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff;opacity:.1}.user-li .sso-settings .sf-icons-menu:hover span{opacity:1}.sf-user-fleet-name{display:block;font-size:.8em!important;margin-top:.2em}.sf-fleet-list .sf-fleet-title{font-size:.8em}.theme-selector-container span{width:calc(100% - 84px)!important}.theme-selector-container .k-switch{margin-right:0}.ul-container{margin:5px 0 0}.tenant-picker{padding:.5em 0;max-height:14em;max-width:14em;min-width:12em;overflow-y:auto;margin-left:-1px}.tenant-picker .tenant{padding:.7em 1em;position:relative}.tenant-picker .tenant:hover:before{opacity:.1}.tenant-picker .tenant.active:before,.tenant-picker .tenant:hover:before{content:\"\";position:absolute;top:0;left:0;width:100%;height:100%;background-color:#fff}.tenant-picker .tenant.active:before{opacity:.2}.tenant-picker::-webkit-scrollbar{width:6px;background-color:transparent}.tenant-picker::-webkit-scrollbar-thumb{border-radius:10px;background-color:var(--clr-gray-500)}::ng-deep .sf-menu-container .k-popup{background-color:transparent;border-color:transparent}"]
                },] }
    ];
    MenuUserComponent.ctorParameters = function () { return [
        { type: SfMenuService }
    ]; };

    var TopMenuComponent = /** @class */ (function () {
        function TopMenuComponent(svc) {
            this.svc = svc;
            this.collapsed = false;
        }
        Object.defineProperty(TopMenuComponent.prototype, "bgColor", {
            get: function () {
                return this.backgroundColor;
            },
            enumerable: false,
            configurable: true
        });
        TopMenuComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.activeLink = this.svc.colors.gray300;
            this.iconColor = this.svc.colors.gray800;
            this.backgroundColor = this.svc.colors.gray200;
            this.configSubs$ = this.svc.menuConfig$.subscribe(function (cfg) {
                _this.menuConfig = cfg;
            });
            this.collapsedSubs$ = this.svc.menuCollapse$.subscribe(function (value) {
                _this.collapsed = value;
            });
        };
        TopMenuComponent.prototype.ngOnDestroy = function () {
            this.configSubs$.unsubscribe();
            this.collapsedSubs$.unsubscribe();
        };
        TopMenuComponent.prototype.shortcutClicked = function (shortcut) {
            this.svc.events$.next({
                type: exports.MenuEventType.ShortcutClick,
                value: shortcut
            });
            if (window.innerWidth <= this.svc.SMALL_RESOLUTION && !this.collapsed) {
                this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
            }
        };
        TopMenuComponent.prototype.toggleMenu = function () {
            this.svc.menuCollapse$.next(!this.svc.menuCollapse$.value);
        };
        return TopMenuComponent;
    }());
    TopMenuComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'sf-top-menu',
                    template: "<div class=\"sf-top-menu-left\">\n  <em class=\"sf-top-menu-toggle-bars fa fa-2x fa-bars\" (click)=\"toggleMenu()\" [style.color]=\"iconColor\"></em>\n</div>\n\n<div class=\"sf-top-menu-middle\">\n  <div class=\"sf-menu-logo-bg\" *ngIf=\"menuConfig.logoUrl\" [ngStyle]=\"{ 'background-image': 'url(' + menuConfig.logoUrl + ')' }\"></div>\n  <em [ngClass]=\"menuConfig.logoClass\" class=\"sf-menu-top-logo\" *ngIf=\"!menuConfig.logoUrl && menuConfig.logoClass\"></em>\n  <svg\n    class=\"sf-menu-top-logo\"\n    xmlns=\"http://www.w3.org/2000/svg\"\n    viewBox=\"0 0 1182 263\"\n    fill=\"#1B4C86\"\n    *ngIf=\"!menuConfig.logoUrl && !menuConfig.logoClass\"\n    [@appear-fast]\n  >\n    <path\n      [attr.fill]=\"iconColor\"\n      d=\"M55.2,145.2 C74.8,147.2 94.4,150.1 94.4,162.9 C94.4,179.3 70.4,183.8 49.4,183.8 C22,183.8 2.1,178 0.4,160 L11.7,160 C13.7,174 28.8,177.4 49.5,177.4 C65.7,177.4 82.5,174.7 82.5,164.1 C82.5,153.9 62.8,152.7 43.2,150.7 C17.5,148.2 3.7,143.7 3.7,132.9 C3.7,118.3 25.4,113.5 46.3,113.5 C68.3,113.5 90.6,118.4 91.3,134.3 L80.1,134.3 C78.6,122.9 63.6,119.9 46.2,119.9 C28.9,119.9 15.7,124.1 15.7,131.7 C15.7,140.4 28.9,142.5 55.2,145.2 Z M128.3,155.4 L176,155.4 L152.3,122.1 L128.3,155.4 Z M146,115.2 L159,115.2 L208,182.1 L195,182.1 L180.6,161.8 L123.8,161.8 L109.1,182.1 L97,182.1 L146,115.2 Z M217,182.1 L217,115.2 L297.9,115.2 L297.9,121.6 L228.2,121.6 L228.2,144.3 L294.4,144.3 L294.4,150.6 L228.2,150.6 L228.2,182.1 L217,182.1 Z M322.4,150.6 L322.4,175.7 L398.3,175.7 L398.3,182.1 L311.2,182.1 L311.2,115.2 L397.1,115.2 L397.1,121.6 L322.4,121.6 L322.4,144.2 L394.4,144.2 L394.4,150.6 L322.4,150.6 Z M707.1,182.1 L707.1,115.2 L788,115.2 L788,121.6 L718.3,121.6 L718.3,144.3 L784.5,144.3 L784.5,150.6 L718.3,150.6 L718.3,182.1 L707.1,182.1 Z M810.7,115.2 L810.7,175.7 L879.1,175.7 L879.1,182.1 L799.5,182.1 L799.5,115.2 L810.7,115.2 Z M902.1,150.6 L902.1,175.7 L978,175.7 L978,182.1 L890.9,182.1 L890.9,115.2 L976.8,115.2 L976.8,121.6 L902.1,121.6 L902.1,144.2 L974.1,144.2 L974.1,150.6 L902.1,150.6 Z M1005.7,150.6 L1005.7,175.7 L1081.5,175.7 L1081.5,182.1 L994.5,182.1 L994.5,115.2 L1080.3,115.2 L1080.3,121.6 L1005.7,121.6 L1005.7,144.2 L1077.6,144.2 L1077.6,150.6 L1005.7,150.6 Z M1085.9,115.2 L1181.8,115.2 L1181.8,121.6 L1139.4,121.6 L1139.4,182.1 L1128.2,182.1 L1128.2,121.6 L1085.9,121.6 L1085.9,115.2 Z M645.7,172.8 C668.1,124.1 664.4,42.1 664.3,38.6 L664.1,34.3 L659.8,34.3 C595.3,34.3 555.2,9.6 554.8,9.3 L552.4,7.8 L550,9.3 C549.7,9.6 509.7,34.3 445.2,34.3 L440.9,34.3 L440.7,38.6 C440.5,42 436.8,124.1 459.2,172.8 C476.4,210 508.1,237.8 550.9,253.4 L552.4,254 L553.9,253.4 C596.8,237.9 628.5,210 645.7,172.8 Z M433.1,28.5 C507.3,28.5 552.3,0 552.3,0 C552.3,0 597.5,28.5 671.6,28.5 C671.6,28.5 676.1,121.7 651.4,175.3 C634.1,212.6 602.2,244.2 552.2,262.3 C502.4,244.2 470.5,212.6 453.2,175.3 C428.5,121.7 433.1,28.5 433.1,28.5 Z M629.9,76.8 C630.1,70.1 630.1,64.5 630,60.5 L523.9,60.5 C486.9,60.5 479.1,77.1 479.1,91.1 C479.1,101.6 484.9,111 496.3,118.9 L523.5,137.5 C529.9,141.8 532.8,144.4 534.1,145.8 C536.2,147.9 537.2,150 537.2,151.8 C537.2,154.8 537.2,161.2 523.6,161.2 L487,161 C490.3,166.8 494.3,172.4 498.8,177.6 L520.3,177.6 C532.7,177.6 544.6,175.9 551.9,172.5 L551.9,211.1 C562.6,207.2 569.7,202.4 578,196.9 L578,122.7 L625.3,122.7 C626.4,117.1 627.2,111.4 627.9,105.8 L578,105.8 L578,81 C578,78.8 578.4,77.7 578.9,77.4 C578.9,77.3 580.2,76.8 585.1,76.8 L629.9,76.8 Z M552.3,18.4 C561.9,23.8 598.9,42.4 655.4,43.3 C655.8,60.8 656.2,128.5 637.5,169.1 C621.5,203.6 592.1,229.7 552.3,244.5 C512.6,229.6 483.2,203.6 467.3,169.1 C448.6,128.5 449,60.8 449.4,43.3 C505.9,42.4 542.8,23.8 552.3,18.4 Z M507.6,86.1 C507.6,83.1 507.6,76.8 521.2,76.8 L551.9,76.8 L551.9,120 C551.4,119.7 551,119.4 550.5,119.1 L521.1,100.9 C516,97.4 512.5,94.5 510.5,92.4 C508.5,90.3 507.6,88.2 507.6,86.1 Z\"\n    />\n  </svg>\n</div>\n\n<div class=\"sf-top-menu-right\">\n  <ng-container *ngFor=\"let shortcut of menuConfig?.shortcuts\">\n    <em\n      *ngIf=\"shortcut.iconClass\"\n      [ngClass]=\"shortcut.iconClass\"\n      (click)=\"shortcutClicked(shortcut)\"\n      [style.background-color]=\"shortcut.active ? activeLink : ''\"\n      [style.color]=\"iconColor\"\n    ></em>\n    <div *ngIf=\"!shortcut.iconClass && shortcut.iconPath\" class=\"img-wrapper\" [style.background-color]=\"shortcut.active ? activeLink : ''\">\n      <div\n        class=\"sf-menu-logo-bg\"\n        [ngStyle]=\"{ 'background-image': 'url(' + shortcut.iconPath + ')' }\"\n        (click)=\"shortcutClicked(shortcut)\"\n      ></div>\n    </div>\n  </ng-container>\n</div>\n",
                    animations: [
                        animations.trigger('appear-fast', [animations.transition(':enter', [animations.style({ opacity: 0 }), animations.animate('80ms 120ms ease-in', animations.style({ opacity: 1 }))])])
                    ],
                    styles: [":host{display:block;display:flex;flex-direction:row;justify-content:space-between;align-items:center;height:70px;border-bottom:4px solid transparent;max-width:100vw}@media screen and (min-width:768px){:host{display:none}}:host .sf-menu-logo-bg{background:50%/contain no-repeat;width:8em;height:4em}:host .sf-top-menu-middle{flex:1 0 4em;height:100%;padding:5px 0;display:flex}:host .sf-top-menu-left,:host .sf-top-menu-right{height:100%;flex:0 1 100%;width:8em;padding:5px 10px;display:flex;align-items:center;overflow:hidden}:host .sf-top-menu-left>*,:host .sf-top-menu-right>*{padding:0 10px;border-radius:4px}:host .sf-top-menu-left em,:host .sf-top-menu-left i,:host .sf-top-menu-right em,:host .sf-top-menu-right i{font-size:1.6em;height:100%;display:flex;align-items:center}:host .sf-top-menu-left>div,:host .sf-top-menu-right>div{height:100%}:host .sf-top-menu-left>div div,:host .sf-top-menu-right>div div{width:1.6em}:host .sf-top-menu-right{justify-content:flex-end}@media screen and (max-width:380px){:host{height:48px}:host .sf-menu-logo-bg{height:2.4em;width:4.8em}:host .sf-top-menu-left,:host .sf-top-menu-right{font-size:.8em}}.sf-menu-top-logo{-o-object-fit:contain;object-fit:contain;text-align:center;align-self:center;justify-self:center;height:auto;width:50%;max-height:calc(70px * .7);width:120px;filter:brightness(1) grayscale(1)}"]
                },] }
    ];
    TopMenuComponent.ctorParameters = function () { return [
        { type: SfMenuService }
    ]; };
    TopMenuComponent.propDecorators = {
        bgColor: [{ type: core.HostBinding, args: ['style.background-color',] }]
    };

    var SfMenuModule = /** @class */ (function () {
        function SfMenuModule() {
        }
        return SfMenuModule;
    }());
    SfMenuModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [
                        SfMenuComponent,
                        ClickOutsideDirective,
                        TopMenuComponent,
                        MenuFooterLogoComponent,
                        MenuUserComponent,
                        CustomerLogoComponent,
                        MenuProductsComponent,
                        MenuShortcutsComponent,
                        MenuToggleBtnComponent,
                        MenuRoutesComponent,
                        MenuItemComponent
                    ],
                    imports: [common.CommonModule, forms.FormsModule, router.RouterModule, kendoAngularTooltip.TooltipModule, kendoAngularMenu.MenusModule, kendoAngularInputs.InputsModule, kendoAngularPopup.PopupModule, animations$1.BrowserAnimationsModule],
                    providers: [SfMenuService],
                    exports: [SfMenuComponent]
                },] }
    ];

    var MenuItem = /** @class */ (function () {
        function MenuItem() {
            this.newTab = false;
            this.newWindow = false;
            this.expanded = false;
            this.active = false;
        }
        return MenuItem;
    }());

    /*
     * Public API Surface of sf-menu
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.MenuConfig = MenuConfig;
    exports.MenuItem = MenuItem;
    exports.SfMenuComponent = SfMenuComponent;
    exports.SfMenuModule = SfMenuModule;
    exports.SfMenuService = SfMenuService;
    exports.SsoManageAccountConfig = SsoManageAccountConfig;
    exports.a = ClickOutsideDirective;
    exports.b = TopMenuComponent;
    exports.c = MenuFooterLogoComponent;
    exports.d = MenuUserComponent;
    exports.e = CustomerLogoComponent;
    exports.f = MenuProductsComponent;
    exports.g = MenuShortcutsComponent;
    exports.h = MenuToggleBtnComponent;
    exports.i = MenuRoutesComponent;
    exports.j = MenuItemComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
