// @progress/kendo-angular-grid/dist/npm/filtering/filter-input.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var forms_1 = require("@angular/forms");
var operators_1 = require("rxjs/operators");
var rxjs_1 = require("rxjs");
var utils_1 = require("../utils");
var kendo_angular_common_1 = require("@progress/kendo-angular-common");
/**
 * @hidden
 */
var FilterInputDirective = /** @class */ (function () {
    function FilterInputDirective(valueAccessors, ngZone, element, renderer) {
        var _this = this;
        this.element = element;
        this.renderer = renderer;
        this.change = new core_1.EventEmitter();
        this.composing = false;
        this.filterDelay = 500;
        this.changeRequests = new rxjs_1.Subject();
        this.accessor = valueAccessors[0];
        ngZone.runOutsideAngular(function () {
            var unsubscribeStart = renderer.listen(element.nativeElement, 'compositionstart', function () { return _this.composing = true; });
            var unsubscribeEnd = renderer.listen(element.nativeElement, 'compositionend', function () { return _this.composing = false; });
            _this.unsubscribeEvents = function () {
                unsubscribeStart();
                unsubscribeEnd();
            };
        });
    }
    Object.defineProperty(FilterInputDirective.prototype, "value", {
        set: function (value) {
            this.accessor.writeValue(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FilterInputDirective.prototype, "disabled", {
        set: function (value) {
            this.accessor.setDisabledState(value);
        },
        enumerable: true,
        configurable: true
    });
    FilterInputDirective.prototype.ngAfterViewInit = function () {
        var _this = this;
        this.addAriaAttributes();
        this.accessor.registerOnChange(function (x) {
            return _this.filterDelay > 0 ?
                _this.changeRequests.next(x) :
                _this.change.emit(x);
        });
        this.subscribeChanges();
    };
    FilterInputDirective.prototype.ngOnChanges = function (changes) {
        if (utils_1.isChanged('filterDelay', changes)) {
            this.unsubscribeChanges();
            this.subscribeChanges();
        }
    };
    FilterInputDirective.prototype.ngOnDestroy = function () {
        this.unsubscribeChanges();
        this.unsubscribeEvents();
    };
    FilterInputDirective.prototype.subscribeChanges = function () {
        var _this = this;
        this.changeRequestsSubscription = this.changeRequests
            .pipe(operators_1.debounceTime(this.filterDelay), operators_1.filter(function () { return !_this.composing; }))
            .subscribe(function (x) { return _this.change.emit(x); });
    };
    FilterInputDirective.prototype.unsubscribeChanges = function () {
        if (this.changeRequestsSubscription) {
            this.changeRequestsSubscription.unsubscribe();
        }
    };
    FilterInputDirective.prototype.addAriaAttributes = function () {
        var ariaValue = this.columnLabel;
        if (this.kendoInput && this.kendoInput.focusableId && kendo_angular_common_1.isDocumentAvailable()) {
            var focusableElement = this.element.nativeElement.querySelector("#" + this.kendoInput.focusableId) ||
                this.element.nativeElement;
            this.renderer.setAttribute(focusableElement, 'aria-label', ariaValue);
        }
        else {
            this.renderer.setAttribute(this.element.nativeElement, 'aria-label', ariaValue);
        }
    };
    tslib_1.__decorate([
        core_1.ContentChild(kendo_angular_common_1.KendoInput, { static: true }),
        tslib_1.__metadata("design:type", Object)
    ], FilterInputDirective.prototype, "kendoInput", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Number)
    ], FilterInputDirective.prototype, "filterDelay", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], FilterInputDirective.prototype, "columnLabel", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], FilterInputDirective.prototype, "value", null);
    FilterInputDirective = tslib_1.__decorate([
        core_1.Directive({
            selector: '[kendoFilterInput]'
        }),
        tslib_1.__param(0, core_1.Self()), tslib_1.__param(0, core_1.Inject(forms_1.NG_VALUE_ACCESSOR)),
        tslib_1.__metadata("design:paramtypes", [Array, core_1.NgZone,
            core_1.ElementRef,
            core_1.Renderer2])
    ], FilterInputDirective);
    return FilterInputDirective;
}());
exports.FilterInputDirective = FilterInputDirective;
