// @progress/kendo-angular-inputs/dist/npm/colorpicker/localization/custom-messages.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
var messages_1 = require("./messages");
/**
 * Custom component messages override default component messages.
 */
var ColorPickerCustomMessagesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ColorPickerCustomMessagesComponent, _super);
    function ColorPickerCustomMessagesComponent(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    ColorPickerCustomMessagesComponent_1 = ColorPickerCustomMessagesComponent;
    Object.defineProperty(ColorPickerCustomMessagesComponent.prototype, "override", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    var ColorPickerCustomMessagesComponent_1;
    ColorPickerCustomMessagesComponent = ColorPickerCustomMessagesComponent_1 = tslib_1.__decorate([
        core_1.Component({
            providers: [
                {
                    provide: messages_1.ColorPickerMessages,
                    useExisting: core_1.forwardRef(function () { return ColorPickerCustomMessagesComponent_1; }) // tslint:disable-line:no-forward-ref
                }
            ],
            selector: 'kendo-colorpicker-messages, kendo-flatcolorpicker-messages, kendo-colorgradient-messages, kendo-colorpalette-messages',
            template: ""
        }),
        tslib_1.__metadata("design:paramtypes", [kendo_angular_l10n_1.LocalizationService])
    ], ColorPickerCustomMessagesComponent);
    return ColorPickerCustomMessagesComponent;
}(messages_1.ColorPickerMessages));
exports.ColorPickerCustomMessagesComponent = ColorPickerCustomMessagesComponent;
