// @progress/kendo-angular-tooltip/dist/es/popover.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { PopupModule } from "@progress/kendo-angular-popup";
import { PopoverAnchorDirective } from "./popover/anchor.directive";
import { PopoverContainerDirective } from "./popover/container.directive";
import { PopoverComponent } from "./popover/popover.component";
import { PopoverTitleTemplateDirective } from './popover/template-directives/title-template.directive';
import { PopoverBodyTemplateDirective } from './popover/template-directives/body-template.directive';
import { PopoverActionsTemplateDirective } from './popover/template-directives/actions-template.directive';
var DIRECTIVES = [
    PopoverActionsTemplateDirective,
    PopoverBodyTemplateDirective,
    PopoverTitleTemplateDirective,
    PopoverAnchorDirective,
    PopoverContainerDirective
];
/**
 * Represents the [NgModule]({{ site.data.urls.angular['ngmoduleapi'] }})
 * definition for the Popover component.
 *
 * @example
 *
 * ```ts-no-run
 * import { PopoverModule } from '@progress/kendo-angular-tooltip';
 *
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 * import { NgModule } from '@angular/core';
 *
 * import { AppComponent } from './app.component';
 *
 * _@NgModule({
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, PopoverModule],
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
var PopoverModule = /** @class */ (function () {
    function PopoverModule() {
    }
    PopoverModule = tslib_1.__decorate([
        NgModule({
            declarations: DIRECTIVES.concat([
                PopoverComponent
            ]),
            entryComponents: [PopoverComponent],
            exports: DIRECTIVES.concat([PopoverComponent]),
            imports: [
                CommonModule,
                PopupModule
            ]
        })
    ], PopoverModule);
    return PopoverModule;
}());
export { PopoverModule };
