// @progress/kendo-angular-menu/dist/es/rendering/link.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Directive, Input, HostBinding, isDevMode } from '@angular/core';
import { ItemsService } from '../services/items.service';
/**
 * Represents a directive that can be used in the [`linkTemplate`]({% slug api_menu_itemlinktemplatedirective %})
 * of the items to apply the default styling and behavior.
 */
var LinkDirective = /** @class */ (function () {
    function LinkDirective(itemsService) {
        this.itemsService = itemsService;
        this.hostClasses = true;
        this.role = 'presentation';
        this.tabindex = '-1';
    }
    Object.defineProperty(LinkDirective.prototype, "activeClass", {
        get: function () {
            return this.item.opened;
        },
        enumerable: true,
        configurable: true
    });
    LinkDirective.prototype.ngOnInit = function () {
        if (isDevMode() && !this.index) {
            throw new Error('The kendoMenuItemLink directive requires the item index to be set.');
        }
        this.item = this.itemsService.get(this.index) || {};
    };
    tslib_1.__decorate([
        Input('kendoMenuItemLink'),
        tslib_1.__metadata("design:type", String)
    ], LinkDirective.prototype, "index", void 0);
    tslib_1.__decorate([
        HostBinding('class.k-link'),
        HostBinding('class.k-menu-link'),
        tslib_1.__metadata("design:type", Boolean)
    ], LinkDirective.prototype, "hostClasses", void 0);
    tslib_1.__decorate([
        HostBinding('attr.role'),
        tslib_1.__metadata("design:type", String)
    ], LinkDirective.prototype, "role", void 0);
    tslib_1.__decorate([
        HostBinding('attr.tabindex'),
        tslib_1.__metadata("design:type", String)
    ], LinkDirective.prototype, "tabindex", void 0);
    tslib_1.__decorate([
        HostBinding('class.k-state-active'),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [])
    ], LinkDirective.prototype, "activeClass", null);
    LinkDirective = tslib_1.__decorate([
        Directive({
            selector: '[kendoMenuItemLink]'
        }),
        tslib_1.__metadata("design:paramtypes", [ItemsService])
    ], LinkDirective);
    return LinkDirective;
}());
export { LinkDirective };
