// @progress/kendo-angular-dateinputs/dist/es2015/datepicker/localization/datepicker-localized-messages.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
var DatePickerLocalizedMessagesDirective_1;
import { Directive, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { DatePickerMessages } from './messages';
/**
 * @hidden
 */
let DatePickerLocalizedMessagesDirective = DatePickerLocalizedMessagesDirective_1 = class DatePickerLocalizedMessagesDirective extends DatePickerMessages {
    constructor(service) {
        super();
        this.service = service;
    }
};
DatePickerLocalizedMessagesDirective = DatePickerLocalizedMessagesDirective_1 = tslib_1.__decorate([
    Directive({
        providers: [
            {
                provide: DatePickerMessages,
                useExisting: forwardRef(() => DatePickerLocalizedMessagesDirective_1) // tslint:disable-line:no-forward-ref
            }
        ],
        selector: '[kendoDatePickerLocalizedMessages]'
    }),
    tslib_1.__metadata("design:paramtypes", [LocalizationService])
], DatePickerLocalizedMessagesDirective);
export { DatePickerLocalizedMessagesDirective };
